// @taiga-ui/addon-preview/esm2015/components/preview/preview-action/preview-action.directive.js
import { Directive } from '@angular/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiPreviewActionDirective {
}
TuiPreviewActionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewActionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiPreviewActionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewActionDirective, selector: "[tuiPreviewAction]", providers: [
        tuiButtonOptionsProvider({
            appearance: 'preview-action',
            shape: 'rounded',
            size: 's',
        }),
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewActionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPreviewAction]',
                    providers: [
                        tuiButtonOptionsProvider({
                            appearance: 'preview-action',
                            shape: 'rounded',
                            size: 's',
                        }),
                    ],
                }]
        }] });
