// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-line-chart.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-charts/utils'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/line-chart', ['exports', '@angular/core', '@taiga-ui/addon-charts/utils', '@taiga-ui/cdk', '@taiga-ui/core', 'rxjs', 'rxjs/operators', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components["line-chart"] = {}), global.ng.core, global["taiga-ui"]["addon-charts"].utils, global.i1, global.i3, global.rxjs, global.rxjs.operators, global.ng.common, global.ngPolymorpheus));
})(this, (function (exports, i0, utils, i1, i3, i2, operators, i2$1, ngPolymorpheus) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i2__namespace$1 = /*#__PURE__*/_interopNamespace(i2$1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TUI_LINE_CHART_DEFAULT_OPTIONS = {
        dots: false,
        filled: false,
        smoothingFactor: 0,
    };
    var TUI_LINE_CHART_OPTIONS = i1.tuiCreateToken(TUI_LINE_CHART_DEFAULT_OPTIONS);
    function tuiLineChartOptionsProvider(options) {
        return i1.tuiProvideOptions(TUI_LINE_CHART_OPTIONS, options, TUI_LINE_CHART_DEFAULT_OPTIONS);
    }

    var TuiLineChartHintDirective = /** @class */ (function () {
        function TuiLineChartHintDirective(renderer, destroy$, zone, hovered$) {
            this.renderer = renderer;
            this.destroy$ = destroy$;
            this.zone = zone;
            this.hovered$ = hovered$;
            this.charts = i1.EMPTY_QUERY;
            this.chartsRef = i1.EMPTY_QUERY;
        }
        TuiLineChartHintDirective.prototype.ngAfterViewInit = function () {
            var _this = this;
            i2.combineLatest([tuiLineChartDrivers(this.charts), this.hovered$])
                .pipe(operators.filter(function (result) { return !result.some(Boolean); }), i1.tuiZonefree(this.zone), operators.takeUntil(this.destroy$))
                .subscribe(function () {
                _this.charts.forEach(function (chart) { return chart.onHovered(NaN); });
            });
        };
        // _chart is required by TuiLineDaysChartComponent that impersonates this directive
        TuiLineChartHintDirective.prototype.getContext = function (index, _chart) {
            return this.computeContext.apply(this, __spreadArray([], __read(this.charts.map(function (_a) {
                var value = _a.value;
                return value;
            }))))[index];
        };
        // _chart is required by TuiLineDaysChartComponent that impersonates this directive
        TuiLineChartHintDirective.prototype.raise = function (index, _chart) {
            var _this = this;
            var current = this.charts.map(function (chart) { return chart.value[index]; });
            var sorted = __spreadArray([], __read(current)).sort(function (a, b) { return a[1] - b[1]; });
            this.charts.forEach(function (chart) { return chart.onHovered(index); });
            this.chartsRef.forEach(function (_a, index) {
                var nativeElement = _a.nativeElement;
                return _this.renderer.setStyle(nativeElement, 'z-index', sorted.indexOf(current[index]));
            });
        };
        TuiLineChartHintDirective.prototype.computeContext = function () {
            var values = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                values[_i] = arguments[_i];
            }
            return (values[0] || []).map(function (_, index) { return values.map(function (value) { return value[index]; }); });
        };
        return TuiLineChartHintDirective;
    }());
    TuiLineChartHintDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartHintDirective, deps: [{ token: i0.Renderer2 }, { token: i1.TuiDestroyService, self: true }, { token: i0.NgZone }, { token: i1.TuiHoveredService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiLineChartHintDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineChartHintDirective, selector: "[tuiLineChartHint]", inputs: { hint: ["tuiLineChartHint", "hint"] }, providers: [i1.TuiDestroyService, i1.TuiHoveredService], queries: [{ propertyName: "charts", predicate: i0__namespace.forwardRef(function () { return TuiLineChartComponent; }) }, { propertyName: "chartsRef", predicate: i0__namespace.forwardRef(function () { return TuiLineChartComponent; }), read: i0.ElementRef }], ngImport: i0__namespace });
    __decorate([
        i1.tuiPure
    ], TuiLineChartHintDirective.prototype, "computeContext", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartHintDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiLineChartHint]',
                        providers: [i1.TuiDestroyService, i1.TuiHoveredService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.Renderer2, decorators: [{
                            type: i0.Inject,
                            args: [i0.Renderer2]
                        }] }, { type: i1__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiDestroyService]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiHoveredService]
                        }] }];
        }, propDecorators: { charts: [{
                    type: i0.ContentChildren,
                    args: [i0.forwardRef(function () { return TuiLineChartComponent; })]
                }], chartsRef: [{
                    type: i0.ContentChildren,
                    args: [i0.forwardRef(function () { return TuiLineChartComponent; }), { read: i0.ElementRef }]
                }], hint: [{
                    type: i0.Input,
                    args: ['tuiLineChartHint']
                }], computeContext: [] } });
    function tuiLineChartDrivers(charts) {
        return i2.combineLatest(charts.map(function (_a) {
            var drivers = _a.drivers;
            return i1.tuiQueryListChanges(drivers).pipe(operators.map(function (drivers) { return drivers.map(function (driver) { return driver.pipe(operators.startWith(false)); }); }));
        })).pipe(operators.map(function (all) { return all.reduce(function (acc, drivers) { return acc.concat(drivers); }, []); }), operators.switchMap(function (drivers) { return i2.combineLatest(drivers); }), operators.map(function (values) { return values.some(Boolean); }), operators.distinctUntilChanged());
    }

    var TuiLineChartComponent = /** @class */ (function () {
        function TuiLineChartComponent(idService, zone, hintDirective, hintOptions, options) {
            this.zone = zone;
            this.hintDirective = hintDirective;
            this.hintOptions = hintOptions;
            this.options = options;
            this.hover$ = new i2.Subject();
            this.drivers = i1.EMPTY_QUERY;
            this.x = 0;
            this.y = 0;
            this.width = 0;
            this.height = 0;
            this.smoothingFactor = this.options.smoothingFactor;
            this.xStringify = null;
            this.yStringify = null;
            this.filled = this.options.filled;
            this.dots = this.options.dots;
            this.value = [];
            this.autoIdString = idService.generate();
        }
        Object.defineProperty(TuiLineChartComponent.prototype, "valueSetter", {
            set: function (value) {
                this.value = value.filter(function (item) { return !item.some(Number.isNaN); });
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "hovered$", {
            get: function () {
                return this.hover$.pipe(operators.distinctUntilChanged(), i1.tuiZoneOptimized(this.zone));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "hintContent", {
            get: function () {
                var _a;
                return ((_a = this.hintOptions) === null || _a === void 0 ? void 0 : _a.content) || '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "fillId", {
            get: function () {
                return "tui-line-chart-" + this.autoIdString;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "fill", {
            get: function () {
                return this.filled ? "url(#" + this.fillId + ")" : 'none';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "viewBox", {
            get: function () {
                return this.x + " " + this.y + " " + this.width + " " + this.height;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "d", {
            get: function () {
                return this.getD(this.value, this.smoothingFactor);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "fillD", {
            get: function () {
                return this.value.length
                    ? this.d + "V " + this.y + " H " + this.value[0][0] + " V " + this.value[0][1]
                    : this.d;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "isFocusable", {
            get: function () {
                return !this.hintDirective && this.hasHints;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineChartComponent.prototype, "hasHints", {
            get: function () {
                var _a;
                return (!!this.xStringify ||
                    !!this.yStringify ||
                    !!((_a = this.hintDirective) === null || _a === void 0 ? void 0 : _a.hint) ||
                    !!this.hintContent);
            },
            enumerable: false,
            configurable: true
        });
        TuiLineChartComponent.prototype.onMouseLeave = function () {
            if (!this.hintDirective) {
                this.onHovered(NaN);
            }
        };
        TuiLineChartComponent.prototype.getX = function (index) {
            if (this.isSinglePoint) {
                return this.value[0][0] / 2;
            }
            return index
                ? (this.value[index - 1][0] + this.value[index][0]) / 2
                : 2 * this.value[0][0] - this.getX(1);
        };
        TuiLineChartComponent.prototype.getWidth = function (index) {
            return (100 * this.computeWidth(index)) / this.width;
        };
        TuiLineChartComponent.prototype.getHintId = function (index) {
            return this.autoIdString + "_" + index;
        };
        TuiLineChartComponent.prototype.getImplicit = function ($implicit) {
            var _a;
            return (((_a = this.hintDirective) === null || _a === void 0 ? void 0 : _a.getContext(this.value.indexOf($implicit), this)) ||
                $implicit);
        };
        TuiLineChartComponent.prototype.getHovered = function (hovered) {
            // This checks for NaN and null too since async pipe returns null before first item
            return i1.tuiIsPresent(hovered) && Number.isInteger(hovered)
                ? this.value[hovered]
                : null;
        };
        TuiLineChartComponent.prototype.getBottom = function (y) {
            return (100 * (y - this.y)) / this.height;
        };
        TuiLineChartComponent.prototype.getLeft = function (x) {
            return (100 * (x - this.x)) / this.width;
        };
        TuiLineChartComponent.prototype.getOffset = function (x) {
            return (100 * (this.value[x][0] - this.getX(x))) / this.computeWidth(x);
        };
        TuiLineChartComponent.prototype.onMouseEnter = function (index) {
            if (this.hintDirective) {
                this.hintDirective.raise(index, this);
            }
            else {
                this.onHovered(index);
            }
        };
        TuiLineChartComponent.prototype.onHovered = function (index) {
            this.hover$.next(index);
        };
        Object.defineProperty(TuiLineChartComponent.prototype, "isSinglePoint", {
            get: function () {
                return this.value.length === 1;
            },
            enumerable: false,
            configurable: true
        });
        TuiLineChartComponent.prototype.getD = function (value, smoothingFactor) {
            return value.reduce(function (d, point, index) { return index ? d + " " + utils.tuiDraw(value, index, smoothingFactor) : "M " + point; }, '');
        };
        TuiLineChartComponent.prototype.computeWidth = function (index) {
            return index === this.value.length - 1
                ? 2 * (this.value[index][0] - this.getX(index))
                : this.getX(index + 1) - this.getX(index);
        };
        return TuiLineChartComponent;
    }());
    TuiLineChartComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartComponent, deps: [{ token: i1.TuiIdService }, { token: i0.NgZone }, { token: TuiLineChartHintDirective, optional: true }, { token: i3.TuiHintOptionsDirective, optional: true }, { token: TUI_LINE_CHART_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiLineChartComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineChartComponent, selector: "tui-line-chart", inputs: { valueSetter: ["value", "valueSetter"], x: "x", y: "y", width: "width", height: "height", smoothingFactor: "smoothingFactor", xStringify: "xStringify", yStringify: "yStringify", filled: "filled", dots: "dots" }, host: { listeners: { "mouseleave": "onMouseLeave()" } }, viewQueries: [{ propertyName: "drivers", predicate: i3.TuiHintHoverDirective, descendants: true }], ngImport: i0__namespace, template: "<ng-container *tuiLet=\"hovered$ | async as hovered\">\n    <svg\n        focusable=\"false\"\n        height=\"100%\"\n        preserveAspectRatio=\"none\"\n        width=\"100%\"\n        xmlns=\"http://www.w3.org/2000/svg\"\n        class=\"t-svg\"\n        [attr.viewBox]=\"viewBox\"\n    >\n        <defs>\n            <linearGradient\n                x1=\"0\"\n                x2=\"0\"\n                y1=\"1\"\n                y2=\"0\"\n                [attr.id]=\"fillId\"\n            >\n                <stop\n                    offset=\"0%\"\n                    stop-color=\"currentColor\"\n                    stop-opacity=\"0.5\"\n                />\n                <stop\n                    offset=\"100%\"\n                    stop-color=\"currentColor\"\n                    stop-opacity=\"0\"\n                />\n            </linearGradient>\n        </defs>\n        <path\n            stroke=\"none\"\n            [attr.d]=\"fillD\"\n            [attr.fill]=\"fill\"\n        />\n        <path\n            fill=\"none\"\n            stroke=\"currentColor\"\n            stroke-width=\"2\"\n            vector-effect=\"non-scaling-stroke\"\n            [attr.d]=\"d\"\n        />\n    </svg>\n    <ng-container *ngIf=\"dots\">\n        <div\n            *ngFor=\"let point of value\"\n            class=\"t-dot\"\n            [style.bottom.%]=\"getBottom(point[1])\"\n            [style.left.%]=\"getLeft(point[0])\"\n        ></div>\n    </ng-container>\n    <ng-container *ngIf=\"hasHints\">\n        <ng-container *ngFor=\"let point of value; let index = index\">\n            <!-- TODO: Fix hint type -->\n            <div\n                *ngIf=\"value.length > 1 || dots\"\n                class=\"t-column\"\n                [class.t-column_hint_hovered]=\"drivers.get(index) | async\"\n                [class.t-column_hovered]=\"hovered === index\"\n                [style.left.%]=\"getLeft(getX(index))\"\n                [style.width.%]=\"getWidth(index)\"\n                [tuiHint]=\"$any(hintDirective?.hint || hintContent)\"\n                [tuiHintContext]=\"{$implicit: getImplicit(point), index: index}\"\n                [tuiHintDescribe]=\"isFocusable ? getHintId(index) : null\"\n                [tuiHintHost]=\"hintHost\"\n                (mouseenter)=\"onMouseEnter(index)\"\n            >\n                <div\n                    class=\"t-line t-line_vertical\"\n                    [style.left.%]=\"getOffset(index)\"\n                ></div>\n                <div\n                    #hintHost\n                    class=\"t-host\"\n                    [id]=\"getHintId(index)\"\n                    [style.bottom.%]=\"getBottom(point[1])\"\n                    [style.left.%]=\"getOffset(index)\"\n                    [tuiFocusable]=\"isFocusable\"\n                ></div>\n            </div>\n            <div\n                *ngIf=\"isFocusable\"\n                class=\"t-line t-line_horizontal\"\n                [style.bottom.%]=\"getBottom(point[1])\"\n            ></div>\n        </ng-container>\n    </ng-container>\n    <ng-container *ngIf=\"getHovered(hovered) as point\">\n        <div\n            *ngIf=\"xStringify\"\n            class=\"t-hint t-hint_x\"\n            [style.left.%]=\"getLeft(point[0])\"\n        >\n            {{ xStringify(point[0]) }}\n        </div>\n        <div\n            *ngIf=\"yStringify\"\n            class=\"t-hint t-hint_y\"\n            [style.bottom.%]=\"getBottom(point[1])\"\n        >\n            {{ yStringify(point[1]) }}\n        </div>\n    </ng-container>\n</ng-container>\n", styles: [":host{display:flex;width:100%;height:100%;pointer-events:none}.t-svg{transform:scaleY(-1)}.t-column{position:absolute;top:0;height:100%;pointer-events:auto}.t-dot{position:absolute;width:.375rem;height:.375rem;border-radius:100%;background:currentColor;margin:-.1875rem;box-shadow:0 0 0 2px #fff}.t-host{position:absolute;left:50%;width:.5rem;height:.5rem;border-radius:100%;opacity:0;background:#fff;margin:-.25rem;box-shadow:0 0 0 2px currentColor,0 .0625rem .1875rem .125rem #0000001a;outline:none;pointer-events:none}.t-host:focus,.t-column_hovered .t-host,.t-column:hover .t-host,.t-column_hint_hovered .t-host{opacity:1}.t-line{position:absolute;opacity:0;background:var(--tui-base-03)}.t-line_vertical{top:0;bottom:0;left:50%;width:1px}.t-line_horizontal{z-index:-1;width:100%;height:1px}:host[style^=\"z-index: 0\"] .t-column_hovered .t-line,:host:not([style]) .t-column:hover .t-line,:host:not([style]) .t-column_hint_hovered .t-line,:host[style^=\"z-index: 0\"] .t-column_hovered+.t-line,:host:not([style]) .t-column:hover+.t-line,:host:not([style]) .t-column_hint_hovered+.t-line{opacity:1}.t-hint{position:absolute;box-shadow:var(--tui-shadow);font:var(--tui-font-text-xs);height:1.25rem;line-height:1.25rem;margin-bottom:-.625rem;padding:0 .375rem;white-space:nowrap;color:var(--tui-base-09);background:var(--tui-base-01);transform:translate(-50%)}.t-hint_x{bottom:0}.t-hint_y{left:0}\n"], directives: [{ type: i1__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2__namespace$1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2__namespace$1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i3__namespace.TuiHintDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHint", "tuiHintContext", "tuiHintAppearance"] }, { type: i3__namespace.TuiHintDriverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)" }, { type: i3__namespace.TuiHintHoverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHintShowDelay", "tuiHintHideDelay"], exportAs: ["tuiHintHover"] }, { type: i3__namespace.TuiHintPositionDirective, selector: "[tuiHint]:not([tuiHintCustomPosition]):not(ng-container):not(ng-template)", inputs: ["tuiHintDirection"] }, { type: i3__namespace.TuiHintHostDirective, selector: "[tuiHint][tuiHintHost]", inputs: ["tuiHintHost"] }, { type: i3__namespace.TuiHintDescribeDirective, selector: "[tuiHintDescribe]", inputs: ["tuiHintDescribe"] }, { type: i1__namespace.TuiFocusableDirective, selector: "[tuiFocusable]", inputs: ["tuiFocusable"] }], pipes: { "async": i2__namespace$1.AsyncPipe }, viewProviders: [i3.tuiHintOptionsProvider({ direction: 'top', hideDelay: 0 })], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i1.tuiPure
    ], TuiLineChartComponent.prototype, "hovered$", null);
    __decorate([
        i1.tuiPure
    ], TuiLineChartComponent.prototype, "getD", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-line-chart',
                        templateUrl: './line-chart.template.html',
                        styleUrls: ['./line-chart.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        viewProviders: [i3.tuiHintOptionsProvider({ direction: 'top', hideDelay: 0 })],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiIdService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiIdService]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: TuiLineChartHintDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [TuiLineChartHintDirective]
                        }] }, { type: i3__namespace.TuiHintOptionsDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiHintOptionsDirective]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_LINE_CHART_OPTIONS]
                        }] }];
        }, propDecorators: { drivers: [{
                    type: i0.ViewChildren,
                    args: [i3.TuiHintHoverDirective]
                }], valueSetter: [{
                    type: i0.Input,
                    args: ['value']
                }], x: [{
                    type: i0.Input
                }], y: [{
                    type: i0.Input
                }], width: [{
                    type: i0.Input
                }], height: [{
                    type: i0.Input
                }], smoothingFactor: [{
                    type: i0.Input
                }], xStringify: [{
                    type: i0.Input
                }], yStringify: [{
                    type: i0.Input
                }], filled: [{
                    type: i0.Input
                }], dots: [{
                    type: i0.Input
                }], hovered$: [], onMouseLeave: [{
                    type: i0.HostListener,
                    args: ['mouseleave']
                }], getD: [] } });

    var TuiLineChartModule = /** @class */ (function () {
        function TuiLineChartModule() {
        }
        return TuiLineChartModule;
    }());
    TuiLineChartModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiLineChartModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartModule, declarations: [TuiLineChartComponent, TuiLineChartHintDirective], imports: [i2$1.CommonModule,
            ngPolymorpheus.PolymorpheusModule,
            i3.TuiHintModule,
            i1.TuiFocusableModule,
            i1.TuiLetModule], exports: [TuiLineChartComponent, TuiLineChartHintDirective] });
    TuiLineChartModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartModule, imports: [[
                i2$1.CommonModule,
                ngPolymorpheus.PolymorpheusModule,
                i3.TuiHintModule,
                i1.TuiFocusableModule,
                i1.TuiLetModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineChartModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i2$1.CommonModule,
                            ngPolymorpheus.PolymorpheusModule,
                            i3.TuiHintModule,
                            i1.TuiFocusableModule,
                            i1.TuiLetModule,
                        ],
                        declarations: [TuiLineChartComponent, TuiLineChartHintDirective],
                        exports: [TuiLineChartComponent, TuiLineChartHintDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_LINE_CHART_DEFAULT_OPTIONS = TUI_LINE_CHART_DEFAULT_OPTIONS;
    exports.TUI_LINE_CHART_OPTIONS = TUI_LINE_CHART_OPTIONS;
    exports.TuiLineChartComponent = TuiLineChartComponent;
    exports.TuiLineChartHintDirective = TuiLineChartHintDirective;
    exports.TuiLineChartModule = TuiLineChartModule;
    exports.tuiLineChartDrivers = tuiLineChartDrivers;
    exports.tuiLineChartOptionsProvider = tuiLineChartOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
