// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-legend-item.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/legend-item', ['exports', '@angular/core', '@taiga-ui/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components["legend-item"] = {}), global.ng.core, global.i1, global.ng.common));
})(this, (function (exports, i0, i1, common) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiLegendItemComponent = /** @class */ (function () {
        function TuiLegendItemComponent() {
            this.active = false;
            this.color = '';
            this.text = '';
            this.size = 'm';
            this.disabled = false;
        }
        Object.defineProperty(TuiLegendItemComponent.prototype, "computedColor", {
            get: function () {
                return this.color === 'var(--tui-primary)' ? null : this.color;
            },
            enumerable: false,
            configurable: true
        });
        return TuiLegendItemComponent;
    }());
    TuiLegendItemComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiLegendItemComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLegendItemComponent, selector: "tui-legend-item", inputs: { active: "active", color: "color", text: "text", size: "size", disabled: "disabled" }, host: { properties: { "attr.data-size": "this.size", "class._disabled": "this.disabled", "style.--tui-primary": "this.computedColor" } }, ngImport: i0__namespace, template: "<button\n    appearance=\"whiteblock\"\n    size=\"s\"\n    tuiButton\n    type=\"button\"\n    class=\"t-button\"\n    [pseudoHover]=\"active || null\"\n>\n    <div class=\"t-wrapper\">\n        <ng-content select=\"tui-primitive-checkbox\"></ng-content>\n        <span\n            class=\"t-dot\"\n            [style.backgroundColor]=\"color || 'var(--tui-primary)'\"\n        ></span>\n        <span class=\"t-text\">{{ text }}</span>\n        <ng-content></ng-content>\n    </div>\n</button>\n", styles: [":host{display:inline-block}.t-button{height:2rem;font-weight:bold}:host[data-size=m] .t-button{height:2.25rem;font:var(--tui-font-text-m);font-weight:bold}.t-wrapper{display:flex;align-items:center;pointer-events:none}:host._disabled .t-wrapper{color:var(--tui-text-03)}.t-dot{width:.5rem;height:.5rem;border-radius:100%;flex-shrink:0}:host._disabled .t-dot{background:var(--tui-base-04)}.t-dot:not(:first-child){display:none}.t-text{margin:0 .5rem;font-weight:normal}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-legend-item',
                        templateUrl: './legend-item.template.html',
                        styleUrls: ['./legend-item.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { active: [{
                    type: i0.Input
                }], color: [{
                    type: i0.Input
                }], text: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], disabled: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._disabled']
                }], computedColor: [{
                    type: i0.HostBinding,
                    args: ['style.--tui-primary']
                }] } });

    var TuiLegendItemModule = /** @class */ (function () {
        function TuiLegendItemModule() {
        }
        return TuiLegendItemModule;
    }());
    TuiLegendItemModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiLegendItemModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemModule, declarations: [TuiLegendItemComponent], imports: [common.CommonModule, i1.TuiButtonModule], exports: [TuiLegendItemComponent] });
    TuiLegendItemModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemModule, imports: [[common.CommonModule, i1.TuiButtonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLegendItemModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [common.CommonModule, i1.TuiButtonModule],
                        declarations: [TuiLegendItemComponent],
                        exports: [TuiLegendItemComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiLegendItemComponent = TuiLegendItemComponent;
    exports.TuiLegendItemModule = TuiLegendItemModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
