// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-pie-chart.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@ng-web-apis/common'), require('@taiga-ui/addon-charts/utils'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/pie-chart', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@ng-web-apis/common', '@taiga-ui/addon-charts/utils', 'rxjs', 'rxjs/operators', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components["pie-chart"] = {}), global.ng.core, global.i1$1, global.i3, global.common, global["taiga-ui"]["addon-charts"].utils, global.rxjs, global.rxjs.operators, global.ng.common, global.ngPolymorpheus));
})(this, (function (exports, i0, i1$1, i3, common, utils, i1, operators, common$1, ngPolymorpheus) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiPieChartDirective = /** @class */ (function () {
        function TuiPieChartDirective(_a, zone, destroy$, performance, animationFrame$, duration) {
            var nativeElement = _a.nativeElement;
            this.sector$ = new i1.BehaviorSubject([0, 0]);
            this.sector$
                .pipe(operators.pairwise(), operators.switchMap(function (_a) {
                var _b = __read(_a, 2), prev = _b[0], cur = _b[1];
                var now = performance.now();
                var startDelta = cur[0] - prev[0];
                var endDelta = cur[1] - prev[1];
                return animationFrame$.pipe(operators.map(function (timestamp) { return i1$1.tuiEaseInOutQuad(i1$1.tuiClamp((timestamp - now) / duration, 0, 1)); }), operators.takeWhile(function (progress) { return progress < 1; }, true), operators.map(function (progress) { return [
                    prev[0] + startDelta * progress,
                    cur[1] > 359 ? cur[1] : prev[1] + endDelta * progress,
                ]; }));
            }), i1$1.tuiZonefree(zone), operators.takeUntil(destroy$))
                .subscribe(function (_a) {
                var _b = __read(_a, 2), start = _b[0], end = _b[1];
                nativeElement.setAttribute('d', utils.tuiDescribeSector(start, end));
            });
        }
        Object.defineProperty(TuiPieChartDirective.prototype, "tuiPieChart", {
            set: function (sector) {
                this.sector$.next(sector);
            },
            enumerable: false,
            configurable: true
        });
        return TuiPieChartDirective;
    }());
    TuiPieChartDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartDirective, deps: [{ token: i0.ElementRef }, { token: i0.NgZone }, { token: i1$1.TuiDestroyService, self: true }, { token: common.PERFORMANCE }, { token: common.ANIMATION_FRAME }, { token: i3.TUI_ANIMATIONS_DURATION }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPieChartDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPieChartDirective, selector: "path[tuiPieChart]", inputs: { tuiPieChart: "tuiPieChart" }, providers: [i1$1.TuiDestroyService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'path[tuiPieChart]',
                        providers: [i1$1.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i1__namespace.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1$1.TuiDestroyService]
                        }] }, { type: Performance, decorators: [{
                            type: i0.Inject,
                            args: [common.PERFORMANCE]
                        }] }, { type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [common.ANIMATION_FRAME]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_ANIMATIONS_DURATION]
                        }] }];
        }, propDecorators: { tuiPieChart: [{
                    type: i0.Input
                }] } });

    var RADII = {
        xs: '50',
        s: '50',
        m: '77.8',
        l: '81.9',
        xl: '81.3',
    };
    var TRANSFORM = {
        xs: 1.15,
        s: 1.25,
        m: 1.11,
        l: 1.09,
        xl: 1.08,
    };
    var TuiPieChartComponent = /** @class */ (function () {
        function TuiPieChartComponent(idService, hintOptions) {
            this.hintOptions = hintOptions;
            this.value = [];
            this.size = 'm';
            this.masked = false;
            this.activeItemIndex = NaN;
            this.activeItemIndexChange = new i0.EventEmitter();
            this.autoIdString = idService.generate();
            if (this.hintOptions) {
                this.hintOptions.showDelay = 0;
                this.hintOptions.hideDelay = 0;
            }
        }
        Object.defineProperty(TuiPieChartComponent.prototype, "empty", {
            get: function () {
                return !this.getSum(this.value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPieChartComponent.prototype, "hintContent", {
            get: function () {
                var _a;
                return ((_a = this.hintOptions) === null || _a === void 0 ? void 0 : _a.content) || '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPieChartComponent.prototype, "maskId", {
            get: function () {
                return "tui-ring-chart-" + this.autoIdString;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPieChartComponent.prototype, "mask", {
            get: function () {
                return this.masked ? "url(#" + this.maskId + ")" : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPieChartComponent.prototype, "radius", {
            get: function () {
                return RADII[this.size];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPieChartComponent.prototype, "segments", {
            get: function () {
                return this.getSegments(this.value);
            },
            enumerable: false,
            configurable: true
        });
        TuiPieChartComponent.prototype.getTransform = function (index) {
            var transform = this.masked
                ? "scale(" + TRANSFORM[this.size] + ")"
                : "scale(" + TRANSFORM.xs + ")";
            return index === this.activeItemIndex ? transform : null;
        };
        TuiPieChartComponent.prototype.onHovered = function (hovered, index) {
            this.updateActiveItemIndex(hovered ? index : NaN);
        };
        TuiPieChartComponent.prototype.getColor = function (index) {
            return "var(--tui-chart-" + index + ")";
        };
        TuiPieChartComponent.prototype.getSum = function (value) {
            return i1$1.tuiSum.apply(void 0, __spreadArray([], __read(value)));
        };
        TuiPieChartComponent.prototype.getSegments = function (value) {
            var _this = this;
            return value
                .map(function (initial, i, array) { return array.reduce(function (sum, current, j) { return (j < i ? _this.getDeg(current) + sum : sum); }, _this.getDeg(initial)); })
                .map(function (angle, index, array) { return [
                array[index - 1] || 0,
                Math.min(angle, 359.9999),
            ]; });
        };
        TuiPieChartComponent.prototype.getDeg = function (value) {
            return 360 * (value / this.getSum(this.value));
        };
        TuiPieChartComponent.prototype.updateActiveItemIndex = function (index) {
            if (index === this.activeItemIndex) {
                return;
            }
            this.activeItemIndex = index;
            this.activeItemIndexChange.next(index);
        };
        return TuiPieChartComponent;
    }());
    TuiPieChartComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartComponent, deps: [{ token: i1$1.TuiIdService }, { token: i3.TuiHintOptionsDirective, optional: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPieChartComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPieChartComponent, selector: "tui-pie-chart", inputs: { value: "value", size: "size", masked: "masked", activeItemIndex: "activeItemIndex" }, outputs: { activeItemIndexChange: "activeItemIndexChange" }, host: { properties: { "attr.data-size": "this.size", "class._empty": "this.empty" } }, ngImport: i0__namespace, template: "<svg\n    focusable=\"false\"\n    height=\"100%\"\n    viewBox=\"-100 -100 200 200\"\n    width=\"100%\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    class=\"t-svg\"\n>\n    <defs>\n        <mask [attr.id]=\"maskId\">\n            <rect\n                fill=\"white\"\n                height=\"400\"\n                width=\"400\"\n                x=\"-200\"\n                y=\"-200\"\n            />\n            <circle\n                cx=\"0\"\n                cy=\"0\"\n                [attr.r]=\"radius\"\n            />\n        </mask>\n    </defs>\n    <g [style.mask]=\"mask\">\n        <circle\n            cx=\"0\"\n            cy=\"0\"\n            r=\"100\"\n            class=\"t-placeholder\"\n        />\n        <path\n            *tuiRepeatTimes=\"let index of segments.length\"\n            automation-id=\"tui-pie-chart__segment\"\n            d=\"\"\n            fill=\"currentColor\"\n            tuiHintPointer\n            class=\"t-segment\"\n            [attr.transform]=\"getTransform(index)\"\n            [style.color]=\"getColor(index)\"\n            [tuiHint]=\"hintContent\"\n            [tuiHintContext]=\"{$implicit: index}\"\n            [tuiPieChart]=\"segments[index]\"\n            (tuiHoveredChange)=\"onHovered($event, index)\"\n        />\n    </g>\n</svg>\n", styles: [":host{position:relative;display:block;width:var(--t-size);height:var(--t-size)}:host[data-size=xs]{--t-size: 2rem;pointer-events:none}:host[data-size=s]{--t-size: 4rem}:host[data-size=m]{--t-size: 9rem}:host[data-size=l]{--t-size: 11rem}:host[data-size=xl]{--t-size: 16rem}.t-svg{position:relative;overflow:visible;transform:rotate(-90deg)}.t-segment{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out}:host._empty .t-segment{display:none}.t-placeholder{fill:var(--tui-base-03)}\n"], directives: [{ type: i1__namespace$1.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }, { type: TuiPieChartDirective, selector: "path[tuiPieChart]", inputs: ["tuiPieChart"] }, { type: i3__namespace.TuiHintDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHint", "tuiHintContext", "tuiHintAppearance"] }, { type: i3__namespace.TuiHintDriverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)" }, { type: i3__namespace.TuiHintHoverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHintShowDelay", "tuiHintHideDelay"], exportAs: ["tuiHintHover"] }, { type: i3__namespace.TuiHintPositionDirective, selector: "[tuiHint]:not([tuiHintCustomPosition]):not(ng-container):not(ng-template)", inputs: ["tuiHintDirection"] }, { type: i3__namespace.TuiHintPointerDirective, selector: "[tuiHint][tuiHintPointer]" }, { type: i1__namespace$1.TuiHoveredDirective, selector: "[tuiHoveredChange]", outputs: ["tuiHoveredChange"] }], viewProviders: [
            i3.tuiHintOptionsProvider({ direction: 'top-right', appearance: 'onDark' }),
        ], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i1$1.tuiPure
    ], TuiPieChartComponent.prototype, "getSum", null);
    __decorate([
        i1$1.tuiPure
    ], TuiPieChartComponent.prototype, "getSegments", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-pie-chart',
                        templateUrl: './pie-chart.template.html',
                        styleUrls: ['./pie-chart.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        viewProviders: [
                            i3.tuiHintOptionsProvider({ direction: 'top-right', appearance: 'onDark' }),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace$1.TuiIdService, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TuiIdService]
                        }] }, { type: i3__namespace.TuiHintOptionsDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiHintOptionsDirective]
                        }] }];
        }, propDecorators: { value: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], masked: [{
                    type: i0.Input
                }], activeItemIndex: [{
                    type: i0.Input
                }], activeItemIndexChange: [{
                    type: i0.Output
                }], empty: [{
                    type: i0.HostBinding,
                    args: ['class._empty']
                }], getSum: [], getSegments: [] } });

    var TuiPieChartModule = /** @class */ (function () {
        function TuiPieChartModule() {
        }
        return TuiPieChartModule;
    }());
    TuiPieChartModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPieChartModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartModule, declarations: [TuiPieChartComponent, TuiPieChartDirective], imports: [common$1.CommonModule,
            i1$1.TuiRepeatTimesModule,
            i1$1.TuiHoveredModule,
            ngPolymorpheus.PolymorpheusModule,
            i3.TuiHintModule], exports: [TuiPieChartComponent] });
    TuiPieChartModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartModule, imports: [[
                common$1.CommonModule,
                i1$1.TuiRepeatTimesModule,
                i1$1.TuiHoveredModule,
                ngPolymorpheus.PolymorpheusModule,
                i3.TuiHintModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPieChartModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            common$1.CommonModule,
                            i1$1.TuiRepeatTimesModule,
                            i1$1.TuiHoveredModule,
                            ngPolymorpheus.PolymorpheusModule,
                            i3.TuiHintModule,
                        ],
                        declarations: [TuiPieChartComponent, TuiPieChartDirective],
                        exports: [TuiPieChartComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPieChartComponent = TuiPieChartComponent;
    exports.TuiPieChartDirective = TuiPieChartDirective;
    exports.TuiPieChartModule = TuiPieChartModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
