// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-axes.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-charts/constants'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/axes', ['exports', '@angular/core', '@taiga-ui/addon-charts/constants', '@taiga-ui/cdk', '@taiga-ui/core', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components.axes = {}), global.ng.core, global["taiga-ui"]["addon-charts"].constants, global.i2, global.core, global.ng.common, global.rxjs));
})(this, (function (exports, i0, constants, i2, core, i1, i3) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    var TuiAxesComponent = /** @class */ (function () {
        function TuiAxesComponent(mode$) {
            this.mode$ = mode$;
            this.axisX = 'solid';
            this.axisXLabels = [];
            this.axisY = 'solid';
            this.axisYInset = false;
            this.axisYLabels = [];
            this.axisYName = '';
            this.axisYSecondaryInset = false;
            this.axisYSecondaryLabels = [];
            this.axisYSecondaryName = '';
            this.horizontalLines = 0;
            this.horizontalLinesHandler = constants.TUI_ALWAYS_SOLID;
            this.verticalLines = 0;
            this.verticalLinesHandler = constants.TUI_ALWAYS_DASHED;
        }
        Object.defineProperty(TuiAxesComponent.prototype, "centeredXLabels", {
            get: function () {
                return this.axisY === 'none';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiAxesComponent.prototype, "hasXLabels", {
            get: function () {
                return !!this.axisXLabels.length;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiAxesComponent.prototype, "hasYLabels", {
            get: function () {
                return (!!this.axisYLabels.length && !this.axisYInset) || !!this.axisYName;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiAxesComponent.prototype, "hasYSecondaryLabels", {
            get: function () {
                return ((!!this.axisYSecondaryLabels.length && !this.axisYSecondaryInset) ||
                    !!this.axisYSecondaryName);
            },
            enumerable: false,
            configurable: true
        });
        TuiAxesComponent.prototype.fallback = function (label) {
            return label || i2.CHAR_NO_BREAK_SPACE;
        };
        return TuiAxesComponent;
    }());
    TuiAxesComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesComponent, deps: [{ token: core.TUI_MODE }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiAxesComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAxesComponent, selector: "tui-axes", inputs: { axisX: "axisX", axisXLabels: "axisXLabels", axisY: "axisY", axisYInset: "axisYInset", axisYLabels: "axisYLabels", axisYName: "axisYName", axisYSecondaryInset: "axisYSecondaryInset", axisYSecondaryLabels: "axisYSecondaryLabels", axisYSecondaryName: "axisYSecondaryName", horizontalLines: "horizontalLines", horizontalLinesHandler: "horizontalLinesHandler", verticalLines: "verticalLines", verticalLinesHandler: "verticalLinesHandler" }, host: { listeners: { "$.data-mode.attr": "mode$" }, properties: { "class._centered": "this.centeredXLabels" } }, providers: [core.MODE_PROVIDER], ngImport: i0__namespace, template: "<div\n    *ngIf=\"hasYLabels\"\n    class=\"t-side\"\n    [class.t-side_padding]=\"hasXLabels\"\n>\n    <div\n        *ngIf=\"axisYName\"\n        automation-id=\"tui-axex__axis-y-name\"\n        class=\"t-name t-name_primary\"\n    >\n        {{ axisYName }}\n    </div>\n    <div\n        *ngIf=\"!axisYInset\"\n        class=\"t-labels-y t-labels-y_primary\"\n    >\n        <div\n            *ngFor=\"let label of axisYLabels\"\n            automation-id=\"tui-axex__axis-y-label\"\n            class=\"t-label-y\"\n        >\n            {{ fallback(label) }}\n        </div>\n    </div>\n</div>\n<div class=\"t-wrapper\">\n    <div\n        class=\"t-grid\"\n        [style.borderBottomStyle]=\"axisX\"\n        [style.borderLeftStyle]=\"axisY\"\n    >\n        <div class=\"t-vertical\">\n            <div\n                *tuiRepeatTimes=\"let index of verticalLines\"\n                automation-id=\"tui-axex__vertical-line\"\n                class=\"t-line t-line_vertical\"\n                [style.borderRightStyle]=\"verticalLinesHandler(index, verticalLines)\"\n            ></div>\n        </div>\n        <div class=\"t-horizontal\">\n            <div\n                *tuiRepeatTimes=\"let index of horizontalLines\"\n                automation-id=\"tui-axex__horizontal-line\"\n                class=\"t-line\"\n                [style.borderTopStyle]=\"horizontalLinesHandler(index, horizontalLines)\"\n            ></div>\n        </div>\n        <div\n            *ngIf=\"axisYInset\"\n            class=\"t-labels-y t-labels-y_inset\"\n        >\n            <div\n                *ngFor=\"let label of axisYLabels\"\n                automation-id=\"tui-axex__axis-y-label\"\n                class=\"t-label-y\"\n            >\n                {{ fallback(label) }}\n            </div>\n        </div>\n        <div\n            *ngIf=\"axisYSecondaryInset\"\n            class=\"t-labels-y t-labels-y_inset t-labels-y_inset_secondary\"\n        >\n            <div\n                *ngFor=\"let label of axisYSecondaryLabels\"\n                automation-id=\"tui-axex__axis-y-secondary-label\"\n                class=\"t-label-y\"\n            >\n                {{ fallback(label) }}\n            </div>\n        </div>\n        <div class=\"t-content\">\n            <ng-content></ng-content>\n        </div>\n    </div>\n    <div\n        *ngIf=\"hasXLabels\"\n        class=\"t-labels-x\"\n    >\n        <div\n            *ngFor=\"let label of axisXLabels\"\n            automation-id=\"tui-axex__axis-x-label\"\n            class=\"t-label-x\"\n            [class.t-label-x_transparent]=\"label === null\"\n        >\n            {{ fallback(label) }}\n        </div>\n    </div>\n</div>\n<div\n    *ngIf=\"hasYSecondaryLabels\"\n    class=\"t-side\"\n    [class.t-side_padding]=\"hasXLabels\"\n>\n    <div\n        *ngIf=\"!axisYSecondaryInset\"\n        class=\"t-labels-y t-labels-y_secondary\"\n    >\n        <div\n            *ngFor=\"let label of axisYSecondaryLabels\"\n            automation-id=\"tui-axex__axis-y-secondary-label\"\n            class=\"t-label-y\"\n        >\n            {{ fallback(label) }}\n        </div>\n    </div>\n    <div\n        *ngIf=\"axisYSecondaryName\"\n        automation-id=\"tui-axex__axis-y-secondary-name\"\n        class=\"t-name\"\n    >\n        {{ axisYSecondaryName }}\n    </div>\n</div>\n", styles: [":host{display:flex}.t-wrapper{display:flex;flex:1;height:100%;flex-direction:column}.t-grid{position:relative;display:flex;flex:1;justify-content:space-around;align-items:flex-end;border-width:1px;border-color:var(--tui-base-03);isolation:isolate}:host[data-mode=onDark] .t-grid{border-color:#ffffff7a}:host[data-mode=onLight] .t-grid{border-color:#0000007a}.t-horizontal{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;flex-direction:column}.t-vertical{position:absolute;top:0;left:0;width:100%;height:100%;display:flex}:host._centered .t-vertical:after{content:\"\";display:block;flex:1 0 1px}.t-line{flex:2 0 1px;box-sizing:border-box;border-width:1px;border-color:var(--tui-base-03)}:host[data-mode=onDark] .t-line{border-color:#ffffff7a}:host[data-mode=onLight] .t-line{border-color:#0000007a}:host._centered .t-line_vertical:first-child{flex:1 0 1px;pointer-events:none}.t-side{display:flex;align-items:stretch}.t-side_padding{padding-bottom:2rem}.t-name{font:var(--tui-font-text-xs);writing-mode:tb;text-align:center;padding-left:.75rem;color:var(--tui-text-02)}.t-name_primary{transform:rotate(180deg)}:host[data-mode=onDark] .t-name{color:var(--tui-text-02-night)}.t-labels-y{display:flex;font:var(--tui-font-text-xs);flex-direction:column-reverse;justify-content:space-between;color:var(--tui-text-02)}:host[data-mode=onDark] .t-labels-y{color:var(--tui-text-02-night)}.t-labels-y_primary{text-align:right;padding-right:.75rem}.t-labels-y_secondary{padding-left:.75rem}.t-labels-y_transparent{border-color:transparent}.t-labels-y_inset{position:absolute;top:.5625rem;left:.25rem;bottom:-.75rem;pointer-events:none}.t-labels-y_inset_secondary{left:auto;right:.25rem;text-align:right}.t-labels-x{display:flex;font:var(--tui-font-text-xs);border-right:1px solid transparent;color:var(--tui-text-02)}.t-label-x{height:.4375rem;border-left:1px solid var(--tui-base-03);flex:1;margin-bottom:1.5625rem}:host[data-mode=onDark] .t-label-x{border-color:#ffffff7a}:host[data-mode=onLight] .t-label-x{border-color:#0000007a}.t-label-x:before{content:\"\";display:block;height:.5625rem}.t-label-x_transparent{border-color:transparent}:host._centered .t-label-x{height:2rem;text-align:center;border:none;margin:0}:host[new]:not(._centered) .t-label-x:last-child:not(:first-child){flex:.5;text-align:right;border-left:none}:host[new]:not(._centered) .t-label-x:nth-last-child(2){flex:.5}.t-label-y:first-child{margin-bottom:-.375rem}.t-label-y:last-child{margin-top:-.375rem}.t-content{position:absolute;top:0;left:-1px;right:0;bottom:-1px;display:flex;align-items:flex-end}\n"], directives: [{ type: i1__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2__namespace.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-axes',
                        templateUrl: './axes.template.html',
                        styleUrls: ['./axes.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [core.MODE_PROVIDER],
                        host: {
                            '($.data-mode.attr)': 'mode$',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [core.TUI_MODE]
                        }] }];
        }, propDecorators: { axisX: [{
                    type: i0.Input
                }], axisXLabels: [{
                    type: i0.Input
                }], axisY: [{
                    type: i0.Input
                }], axisYInset: [{
                    type: i0.Input
                }], axisYLabels: [{
                    type: i0.Input
                }], axisYName: [{
                    type: i0.Input
                }], axisYSecondaryInset: [{
                    type: i0.Input
                }], axisYSecondaryLabels: [{
                    type: i0.Input
                }], axisYSecondaryName: [{
                    type: i0.Input
                }], horizontalLines: [{
                    type: i0.Input
                }], horizontalLinesHandler: [{
                    type: i0.Input
                }], verticalLines: [{
                    type: i0.Input
                }], verticalLinesHandler: [{
                    type: i0.Input
                }], centeredXLabels: [{
                    type: i0.HostBinding,
                    args: ['class._centered']
                }] } });

    var TuiAxesModule = /** @class */ (function () {
        function TuiAxesModule() {
        }
        return TuiAxesModule;
    }());
    TuiAxesModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAxesModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesModule, declarations: [TuiAxesComponent], imports: [i1.CommonModule, i2.TuiRepeatTimesModule], exports: [TuiAxesComponent] });
    TuiAxesModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesModule, imports: [[i1.CommonModule, i2.TuiRepeatTimesModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAxesModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule, i2.TuiRepeatTimesModule],
                        declarations: [TuiAxesComponent],
                        exports: [TuiAxesComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiAxesComponent = TuiAxesComponent;
    exports.TuiAxesModule = TuiAxesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
