// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-ring-chart.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-charts/components/pie-chart'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/ring-chart', ['exports', '@angular/core', '@taiga-ui/addon-charts/components/pie-chart', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components["ring-chart"] = {}), global.ng.core, global["taiga-ui"]["addon-charts"].components["pie-chart"], global.ng.common, global.ngPolymorpheus));
})(this, (function (exports, i0, i1, common, ngPolymorpheus) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiRingChartComponent = /** @class */ (function () {
        function TuiRingChartComponent() {
            this.value = [];
            this.size = 'm';
            this.activeItemIndex = NaN;
            this.activeItemIndexChange = new i0.EventEmitter();
        }
        TuiRingChartComponent.prototype.onActiveItemIndexChange = function (index) {
            this.updateActiveItemIndex(index);
        };
        TuiRingChartComponent.prototype.updateActiveItemIndex = function (index) {
            if (index === this.activeItemIndex) {
                return;
            }
            this.activeItemIndex = index;
            this.activeItemIndexChange.next(index);
        };
        return TuiRingChartComponent;
    }());
    TuiRingChartComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiRingChartComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRingChartComponent, selector: "tui-ring-chart", inputs: { value: "value", size: "size", activeItemIndex: "activeItemIndex" }, outputs: { activeItemIndexChange: "activeItemIndexChange" }, host: { properties: { "attr.data-size": "this.size" } }, ngImport: i0__namespace, template: "<div class=\"t-content\">\n    <div class=\"t-wrapper\">\n        <ng-content></ng-content>\n    </div>\n</div>\n\n<tui-pie-chart\n    class=\"t-chart\"\n    [activeItemIndex]=\"activeItemIndex\"\n    [masked]=\"true\"\n    [size]=\"size\"\n    [value]=\"value\"\n    (activeItemIndexChange)=\"onActiveItemIndexChange($event)\"\n></tui-pie-chart>\n\n<div class=\"t-shield\"></div>\n", styles: [":host{position:relative;display:block;width:var(--t-size);height:var(--t-size);flex-shrink:0}:host[data-size=xs]{--t-size: 3rem}:host[data-size=s]{--t-size: 4rem}:host[data-size=m]{--t-size: 9rem}:host[data-size=l]{--t-size: 11rem}:host[data-size=xl]{--t-size: 16rem}.t-content{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;font:var(--tui-font-text-m);max-height:100%;flex-direction:column;justify-content:center;text-align:center;padding:1.5rem;border-radius:100%;box-sizing:border-box;overflow:hidden;word-break:break-word;white-space:pre-wrap;color:var(--tui-text-02)}:host[data-size=m] .t-content{font:var(--tui-font-text-xs)}.t-wrapper:first-line{color:var(--tui-text-01)}:host[data-size=l] .t-wrapper:first-line{font:var(--tui-font-text-l);font-weight:bold}:host[data-size=m] .t-wrapper:first-line{font:var(--tui-font-text-m);font-weight:bold}.t-shield{position:absolute;top:25%;left:25%;right:25%;bottom:25%;border-radius:100%}.t-chart{width:100%;height:100%}\n"], components: [{ type: i1__namespace.TuiPieChartComponent, selector: "tui-pie-chart", inputs: ["value", "size", "masked", "activeItemIndex"], outputs: ["activeItemIndexChange"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-ring-chart',
                        templateUrl: './ring-chart.template.html',
                        styleUrls: ['./ring-chart.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { value: [{
                    type: i0.Input
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], activeItemIndex: [{
                    type: i0.Input
                }], activeItemIndexChange: [{
                    type: i0.Output
                }] } });

    var TuiRingChartModule = /** @class */ (function () {
        function TuiRingChartModule() {
        }
        return TuiRingChartModule;
    }());
    TuiRingChartModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiRingChartModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartModule, declarations: [TuiRingChartComponent], imports: [common.CommonModule, ngPolymorpheus.PolymorpheusModule, i1.TuiPieChartModule], exports: [TuiRingChartComponent] });
    TuiRingChartModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartModule, imports: [[common.CommonModule, ngPolymorpheus.PolymorpheusModule, i1.TuiPieChartModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRingChartModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [common.CommonModule, ngPolymorpheus.PolymorpheusModule, i1.TuiPieChartModule],
                        declarations: [TuiRingChartComponent],
                        exports: [TuiRingChartComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiRingChartComponent = TuiRingChartComponent;
    exports.TuiRingChartModule = TuiRingChartModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
