// @taiga-ui/addon-charts/bundles/taiga-ui-addon-charts-components-line-days-chart.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-charts/components/line-chart'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/core'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-charts/components/line-days-chart', ['exports', '@angular/core', '@taiga-ui/addon-charts/components/line-chart', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/core', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-charts"] = global["taiga-ui"]["addon-charts"] || {}, global["taiga-ui"]["addon-charts"].components = global["taiga-ui"]["addon-charts"].components || {}, global["taiga-ui"]["addon-charts"].components["line-days-chart"] = {}), global.ng.core, global["taiga-ui"]["addon-charts"].components["line-chart"], global.i1, global.rxjs, global.rxjs.operators, global.ng.common, global.i3, global.i4));
})(this, (function (exports, i0, i1$1, i1, i2, operators, i2$1, i3, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i2__namespace$1 = /*#__PURE__*/_interopNamespace(i2$1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function find(value, current) {
        return value.find(function (_a) {
            var _b = __read(_a, 1), day = _b[0];
            return day.daySame(current);
        }) || [current, NaN];
    }
    // TODO: Consider extending TuiLineChartHintDirective
    var TuiLineDaysChartHintDirective = /** @class */ (function () {
        function TuiLineDaysChartHintDirective(destroy$, zone, hovered$) {
            this.destroy$ = destroy$;
            this.zone = zone;
            this.hovered$ = hovered$;
            this.charts = i1.EMPTY_QUERY;
        }
        TuiLineDaysChartHintDirective.prototype.ngAfterContentInit = function () {
            var _this = this;
            i2.combineLatest(__spreadArray(__spreadArray([], __read(this.charts.map(function (_a) {
                var charts = _a.charts;
                return i1$1.tuiLineChartDrivers(charts);
            }))), [
                this.hovered$,
            ]))
                .pipe(operators.filter(function (result) { return !result.some(Boolean); }), i1.tuiZonefree(this.zone), operators.takeUntil(this.destroy$))
                .subscribe(function () {
                _this.charts.forEach(function (chart) { return chart.onHovered(NaN); });
            });
        };
        TuiLineDaysChartHintDirective.prototype.getContext = function (day) {
            return this.getMap.apply(this, __spreadArray([], __read(this.charts.map(function (_a) {
                var value = _a.value;
                return value;
            })))).get(String(day)) || [];
        };
        TuiLineDaysChartHintDirective.prototype.raise = function (day) {
            var current = this.charts
                .map(function (_a) {
                var value = _a.value;
                return find(value, day);
            })
                .filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], value = _b[1];
                return !Number.isNaN(value);
            });
            var sorted = __spreadArray([], __read(current)).sort(function (a, b) { return a[1] - b[1]; });
            this.charts.forEach(function (chart, index) {
                chart.onHovered(day);
                chart.zIndex = Math.max(sorted.indexOf(current[index]), 0);
            });
        };
        TuiLineDaysChartHintDirective.prototype.getMap = function () {
            var values = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                values[_i] = arguments[_i];
            }
            return (values[0] || []).reduce(function (map, _a) {
                var _b = __read(_a, 1), day = _b[0];
                return map.set(String(day), values.map(function (value) { return find(value, day); }));
            }, new Map());
        };
        return TuiLineDaysChartHintDirective;
    }());
    TuiLineDaysChartHintDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartHintDirective, deps: [{ token: i1.TuiDestroyService, self: true }, { token: i0.NgZone }, { token: i1.TuiHoveredService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiLineDaysChartHintDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineDaysChartHintDirective, selector: "[tuiLineChartHint]", inputs: { hint: ["tuiLineChartHint", "hint"] }, providers: [i1.TuiDestroyService, i1.TuiHoveredService], queries: [{ propertyName: "charts", predicate: i0__namespace.forwardRef(function () { return TuiLineDaysChartComponent; }) }], ngImport: i0__namespace });
    __decorate([
        i1.tuiPure
    ], TuiLineDaysChartHintDirective.prototype, "getMap", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartHintDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiLineChartHint]',
                        providers: [i1.TuiDestroyService, i1.TuiHoveredService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiDestroyService]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiHoveredService]
                        }] }];
        }, propDecorators: { charts: [{
                    type: i0.ContentChildren,
                    args: [i0.forwardRef(function () { return TuiLineDaysChartComponent; })]
                }], hint: [{
                    type: i0.Input,
                    args: ['tuiLineChartHint']
                }], getMap: [] } });

    var DUMMY = [NaN, NaN];
    var TuiLineDaysChartComponent = /** @class */ (function () {
        function TuiLineDaysChartComponent(hintDirective, destroy$, zone, hovered$, options) {
            var _this = this;
            this.hintDirective = hintDirective;
            this.destroy$ = destroy$;
            this.zone = zone;
            this.hovered$ = hovered$;
            this.options = options;
            this.charts = i1.EMPTY_QUERY;
            this.y = 0;
            this.height = 0;
            this.smoothingFactor = this.options.smoothingFactor;
            this.xStringify = null;
            this.yStringify = null;
            this.dots = this.options.dots;
            this.zIndex = 0;
            this.value = [];
            this.daysStringify = function (index) { return _this.xStringify ? _this.xStringify(_this.getDay(index)) : ''; };
        }
        Object.defineProperty(TuiLineDaysChartComponent.prototype, "valueSetter", {
            set: function (value) {
                if (!value.length) {
                    this.value = [];
                    return;
                }
                var start = value[0][0];
                var mutable = __spreadArray([], __read(value));
                var length = i1.TuiDay.lengthBetween(start, value[value.length - 1][0]) + 1;
                this.value = Array.from({ length: length }, function (_, day) {
                    var currentDay = start.append({ day: day });
                    var shifted = currentDay.daySame(mutable[0][0]) ? mutable.shift() : null;
                    var currentValue = shifted ? shifted[1] : NaN;
                    return [currentDay, currentValue];
                });
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineDaysChartComponent.prototype, "months", {
            get: function () {
                return this.value.length ? this.breakMonths(this.value) : i1.EMPTY_ARRAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineDaysChartComponent.prototype, "firstWidth", {
            get: function () {
                return this.months.length * this.value[0][0].daysCount;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLineDaysChartComponent.prototype, "hint", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.hintDirective) === null || _a === void 0 ? void 0 : _a.hint) !== null && _b !== void 0 ? _b : this.hintContent;
            },
            enumerable: false,
            configurable: true
        });
        TuiLineDaysChartComponent.prototype.getHintContext = function (x, value) {
            return value[x - value[0][0].day + 1];
        };
        TuiLineDaysChartComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            i2.combineLatest([i1$1.tuiLineChartDrivers(this.charts), this.hovered$])
                .pipe(operators.filter(function (result) { return !result.some(Boolean); }), i1.tuiZonefree(this.zone), operators.takeUntil(this.destroy$))
                .subscribe(function () {
                _this.onHovered(NaN);
            });
        };
        TuiLineDaysChartComponent.prototype.getX = function (index) {
            var current = this.getDay(index);
            var months = i1.TuiMonth.lengthBetween(this.value[0][0], current);
            var offset = months * current.daysCount;
            return index - offset;
        };
        TuiLineDaysChartComponent.prototype.onHovered = function (day) {
            if (i1.tuiIsNumber(day)) {
                this.charts.forEach(function (chart) { return chart.onHovered(NaN); });
                return;
            }
            var index = i1.TuiMonth.lengthBetween(this.value[0][0], day);
            var x = i1.TuiDay.lengthBetween(this.value[0][0], day) + this.value[0][0].day - 1;
            var current = this.charts.get(index);
            this.charts.forEach(function (chart) {
                if (chart === current) {
                    current.onHovered(current.value.findIndex(function (point) { return point[0] === x; }));
                }
                else {
                    chart.onHovered(NaN);
                }
            });
        };
        TuiLineDaysChartComponent.prototype.raise = function (index, _c) {
            var value = _c.value;
            var x = value[index][0];
            var month = this.getDay(x);
            if (this.hintDirective) {
                this.hintDirective.raise(month);
            }
            else {
                this.onHovered(month);
            }
        };
        TuiLineDaysChartComponent.prototype.getWidth = function (index) {
            return this.getDay(index).daysCount * this.months.length;
        };
        TuiLineDaysChartComponent.prototype.getContext = function (index, _c) {
            var value = _c.value;
            var x = value[index][0];
            return this.hintDirective
                ? this.hintDirective.getContext(this.getDay(x))
                : this.getHintContext(x, this.value);
        };
        TuiLineDaysChartComponent.prototype.breakMonths = function (value) {
            var offset = value[0][0].day - 1;
            return Array.from({ length: i1.TuiMonth.lengthBetween(value[0][0], value[value.length - 1][0]) + 1 }, function (_, i) { return i + value[0][0].month + value[0][0].year * 12; })
                .map(function (absoluteMonth) { return value
                .map(function (_c, index) {
                var _d = __read(_c, 2), _e = _d[0], month = _e.month, year = _e.year, y = _d[1];
                return month + year * 12 === absoluteMonth ? [index + offset, y] : null;
            })
                .filter(i1.tuiIsPresent); })
                .map(function (month, index, array) { return index === array.length - 1
                ? month
                : __spreadArray(__spreadArray([], __read(month)), [
                    array[index + 1].find(function (day) { return !Number.isNaN(day[1]); }) || DUMMY,
                ]); });
        };
        TuiLineDaysChartComponent.prototype.getDay = function (index) {
            return this.value[index - this.value[0][0].day + 1][0];
        };
        return TuiLineDaysChartComponent;
    }());
    TuiLineDaysChartComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartComponent, deps: [{ token: TuiLineDaysChartHintDirective, optional: true }, { token: i1.TuiDestroyService, self: true }, { token: i0.NgZone }, { token: i1.TuiHoveredService }, { token: i1$1.TUI_LINE_CHART_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiLineDaysChartComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLineDaysChartComponent, selector: "tui-line-days-chart", inputs: { valueSetter: ["value", "valueSetter"], y: "y", height: "height", smoothingFactor: "smoothingFactor", hintContent: "hintContent", xStringify: "xStringify", yStringify: "yStringify", dots: "dots" }, host: { properties: { "style.zIndex": "this.zIndex" } }, providers: [
            i1.TuiDestroyService,
            i1.TuiHoveredService,
            {
                provide: i1$1.TuiLineChartHintDirective,
                useExisting: TuiLineDaysChartComponent,
            },
        ], viewQueries: [{ propertyName: "charts", predicate: i1$1.TuiLineChartComponent, descendants: true }], ngImport: i0__namespace, template: "<tui-line-chart\n    *ngFor=\"let month of months; let first = first\"\n    class=\"t-chart\"\n    [dots]=\"dots\"\n    [height]=\"height\"\n    [smoothingFactor]=\"smoothingFactor\"\n    [style.zIndex]=\"zIndex\"\n    [tuiHintContent]=\"hintContent ? hint : ''\"\n    [value]=\"month\"\n    [width]=\"first ? firstWidth : getWidth(month[0][0])\"\n    [x]=\"first ? 0 : getX(month[0][0])\"\n    [xStringify]=\"xStringify ? daysStringify : null\"\n    [y]=\"y\"\n    [yStringify]=\"yStringify\"\n></tui-line-chart>\n<ng-template\n    #hint\n    let-point\n>\n    <ng-container *polymorpheusOutlet=\"hintContent as text; context: {$implicit: getHintContext(point[0], value)}\">\n        {{ text }}\n    </ng-container>\n</ng-template>\n", styles: [":host{display:block}.t-chart{position:absolute;top:0;left:0;width:100%;height:100%}\n"], components: [{ type: i1__namespace$1.TuiLineChartComponent, selector: "tui-line-chart", inputs: ["value", "x", "y", "width", "height", "smoothingFactor", "xStringify", "yStringify", "filled", "dots"] }], directives: [{ type: i2__namespace$1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i3__namespace.TuiHintOptionsDirective, selector: "[tuiHintContent]", inputs: ["tuiHintContent", "tuiHintDirection", "tuiHintAppearance", "tuiHintShowDelay", "tuiHintHideDelay"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i1.tuiPure
    ], TuiLineDaysChartComponent.prototype, "getHintContext", null);
    __decorate([
        i1.tuiPure
    ], TuiLineDaysChartComponent.prototype, "breakMonths", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-line-days-chart',
                        templateUrl: './line-days-chart.template.html',
                        styleUrls: ['./line-days-chart.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i1.TuiDestroyService,
                            i1.TuiHoveredService,
                            {
                                provide: i1$1.TuiLineChartHintDirective,
                                useExisting: TuiLineDaysChartComponent,
                            },
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: TuiLineDaysChartHintDirective, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [TuiLineDaysChartHintDirective]
                        }] }, { type: i1__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiDestroyService]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiHoveredService]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TUI_LINE_CHART_OPTIONS]
                        }] }];
        }, propDecorators: { charts: [{
                    type: i0.ViewChildren,
                    args: [i1$1.TuiLineChartComponent]
                }], valueSetter: [{
                    type: i0.Input,
                    args: ['value']
                }], y: [{
                    type: i0.Input
                }], height: [{
                    type: i0.Input
                }], smoothingFactor: [{
                    type: i0.Input
                }], hintContent: [{
                    type: i0.Input
                }], xStringify: [{
                    type: i0.Input
                }], yStringify: [{
                    type: i0.Input
                }], dots: [{
                    type: i0.Input
                }], zIndex: [{
                    type: i0.HostBinding,
                    args: ['style.zIndex']
                }], getHintContext: [], breakMonths: [] } });

    var TuiLineDaysChartModule = /** @class */ (function () {
        function TuiLineDaysChartModule() {
        }
        return TuiLineDaysChartModule;
    }());
    TuiLineDaysChartModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiLineDaysChartModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartModule, declarations: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective], imports: [i2$1.CommonModule, i1$1.TuiLineChartModule, i4.PolymorpheusModule, i3.TuiHintModule], exports: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective] });
    TuiLineDaysChartModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartModule, imports: [[i2$1.CommonModule, i1$1.TuiLineChartModule, i4.PolymorpheusModule, i3.TuiHintModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLineDaysChartModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i2$1.CommonModule, i1$1.TuiLineChartModule, i4.PolymorpheusModule, i3.TuiHintModule],
                        declarations: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective],
                        exports: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiLineDaysChartComponent = TuiLineDaysChartComponent;
    exports.TuiLineDaysChartHintDirective = TuiLineDaysChartHintDirective;
    exports.TuiLineDaysChartModule = TuiLineDaysChartModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
