// @taiga-ui/core/esm2015/components/data-list/opt-group.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiOptGroupDirective {
}
TuiOptGroupDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiOptGroupDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiOptGroupDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiOptGroupDirective, selector: "tui-opt-group", inputs: { label: "label" }, host: { attributes: { "role": "group" }, properties: { "attr.data-label": "this.label" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiOptGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-opt-group',
                    host: {
                        role: 'group',
                    },
                }]
        }], propDecorators: { label: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-label']
            }] } });
