// @taiga-ui/addon-commerce/esm2015/components/money/money.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Inject, Input, } from '@angular/core';
import { CHAR_MINUS, CHAR_PLUS } from '@taiga-ui/cdk';
import { TUI_MONEY_OPTIONS } from './money.options';
import { tuiFormatSignSymbol } from './utils/format-sign-symbol';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
import * as i2 from "@angular/common";
import * as i3 from "./pipes/sign-symbol.pipe";
import * as i4 from "@taiga-ui/addon-commerce/pipes";
import * as i5 from "./pipes/integer-part.pipe";
import * as i6 from "./pipes/fraction-part.pipe";
export class TuiMoneyComponent {
    constructor(options) {
        this.options = options;
        this.value = NaN;
        this.decimal = this.options.decimal;
        this.currency = this.options.currency;
        this.currencyAlign = this.options.currencyAlign;
        this.sign = this.options.sign;
        this.colored = this.options.colored;
        this.precision = this.options.precision;
        this.singleColor = this.options.singleColor;
    }
    get signSymbol() {
        return tuiFormatSignSymbol(this.value, this.sign);
    }
    get red() {
        return (this.colored &&
            (this.signSymbol === CHAR_MINUS ||
                (this.value < 0 && this.sign !== 'force-positive')));
    }
    get green() {
        return (this.colored &&
            (this.signSymbol === CHAR_PLUS ||
                (this.value > 0 && this.sign !== 'force-negative')));
    }
    get alignRight() {
        return this.currencyAlign === 'right';
    }
    get inheritColor() {
        return this.singleColor || (this.value === 0 && this.colored);
    }
}
TuiMoneyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyComponent, deps: [{ token: TUI_MONEY_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiMoneyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMoneyComponent, selector: "tui-money", inputs: { value: "value", decimal: "decimal", currency: "currency", currencyAlign: "currencyAlign", sign: "sign", colored: "colored", precision: "precision", singleColor: "singleColor" }, host: { properties: { "class._red": "this.red", "class._green": "this.green", "class._inherit-color": "this.inheritColor" } }, ngImport: i0, template: "<span\n    automation-id=\"tui-money__sign\"\n    [textContent]=\"value | tuiSignSymbol: sign\"\n></span>\n\n<span\n    *tuiLet=\"currency | tuiCurrency as currencySymbol\"\n    class=\"t-container\"\n    [class.t-container_reverse]=\"!alignRight\"\n>\n    <span>\n        <span\n            automation-id=\"tui-money__integer-part\"\n            [textContent]=\"value | tuiIntegerPart: precision\"\n        ></span>\n        <span\n            automation-id=\"tui-money__fraction-part\"\n            class=\"t-lighter\"\n            [textContent]=\"value | tuiFractionPart: decimal : precision\"\n        ></span>\n    </span>\n    <span *ngIf=\"(alignRight && currencySymbol) || currencySymbol.length > 1\">&nbsp;</span>\n    <span\n        automation-id=\"tui-money__currency\"\n        class=\"t-lighter\"\n        [textContent]=\"currencySymbol\"\n    ></span>\n</span>\n", styles: [":host{white-space:nowrap}:host._red{color:var(--tui-negative)}:host._green{color:var(--tui-positive)}:host:not(._inherit-color) .t-lighter{opacity:var(--tui-disabled-opacity)}.t-container_reverse{display:inline-flex;flex-direction:row-reverse}\n"], directives: [{ type: i1.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "tuiSignSymbol": i3.TuiSignSymbolPipe, "tuiCurrency": i4.TuiCurrencyPipe, "tuiIntegerPart": i5.TuiIntegerPartPipe, "tuiFractionPart": i6.TuiFractionPartPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-money',
                    templateUrl: './money.template.html',
                    styleUrls: ['./money.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_MONEY_OPTIONS]
                }] }]; }, propDecorators: { value: [{
                type: Input
            }], decimal: [{
                type: Input
            }], currency: [{
                type: Input
            }], currencyAlign: [{
                type: Input
            }], sign: [{
                type: Input
            }], colored: [{
                type: Input
            }], precision: [{
                type: Input
            }], singleColor: [{
                type: Input
            }], red: [{
                type: HostBinding,
                args: ['class._red']
            }], green: [{
                type: HostBinding,
                args: ['class._green']
            }], inheritColor: [{
                type: HostBinding,
                args: ['class._inherit-color']
            }] } });
