// @taiga-ui/addon-commerce/esm2015/components/money/money.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiCurrencyPipeModule } from '@taiga-ui/addon-commerce/pipes';
import { TuiLetModule } from '@taiga-ui/cdk';
import { TuiMoneyComponent } from './money.component';
import { TuiFractionPartPipe } from './pipes/fraction-part.pipe';
import { TuiIntegerPartPipe } from './pipes/integer-part.pipe';
import { TuiSignSymbolPipe } from './pipes/sign-symbol.pipe';
import * as i0 from "@angular/core";
export class TuiMoneyModule {
}
TuiMoneyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMoneyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyModule, declarations: [TuiMoneyComponent,
        TuiFractionPartPipe,
        TuiIntegerPartPipe,
        TuiSignSymbolPipe], imports: [CommonModule, TuiCurrencyPipeModule, TuiLetModule], exports: [TuiMoneyComponent] });
TuiMoneyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyModule, imports: [[CommonModule, TuiCurrencyPipeModule, TuiLetModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMoneyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiCurrencyPipeModule, TuiLetModule],
                    declarations: [
                        TuiMoneyComponent,
                        TuiFractionPartPipe,
                        TuiIntegerPartPipe,
                        TuiSignSymbolPipe,
                    ],
                    exports: [TuiMoneyComponent],
                }]
        }] });
