// @taiga-ui/experimental/esm2015/directives/fade/fade.directive.js
import { Directive, ElementRef, HostBinding, Inject, Input, NgZone, Self, } from '@angular/core';
import { MUTATION_OBSERVER_INIT, MutationObserverService, } from '@ng-web-apis/mutation-observer';
import { TuiDestroyService, TuiDirectiveStylesService, TuiResizeService, tuiZonefree, } from '@taiga-ui/cdk';
import { fromEvent, merge } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TuiFadeComponent } from './fade.component';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
import * as i2 from "@taiga-ui/cdk";
const BUFFER = 1; // buffer for rounding issues
export class TuiFadeDirective {
    constructor(destroy$, resize$, mutate$, element, zone, directiveStyles) {
        // TODO: Remove when lh CSS units are supported: https://caniuse.com/mdn-css_types_length_lh
        this.lineHeight = null;
        this.size = '1.5em';
        this.offset = '0em';
        this.orientation = 'horizontal';
        const el = element.nativeElement;
        directiveStyles.addComponent(TuiFadeComponent);
        merge(resize$, mutate$, fromEvent(el, 'scroll'))
            .pipe(tuiZonefree(zone), takeUntil(destroy$))
            .subscribe(() => {
            el.classList.toggle('_start', !!el.scrollLeft || !!el.scrollTop);
            el.classList.toggle('_end', this.isEnd(el));
        });
    }
    isEnd(el) {
        if (this.orientation === 'vertical') {
            return Math.round(el.scrollTop) < el.scrollHeight - el.clientHeight - BUFFER;
        }
        return (Math.round(el.scrollLeft) < el.scrollWidth - el.clientWidth - BUFFER ||
            // horizontal multiline fade can kick in early due to hanging elements of fonts so using bigger buffer
            el.scrollHeight > el.clientHeight + 4 * BUFFER);
    }
}
TuiFadeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFadeDirective, deps: [{ token: TuiDestroyService, self: true }, { token: TuiResizeService }, { token: MutationObserverService }, { token: ElementRef }, { token: NgZone }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiFadeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiFadeDirective, selector: "[tuiFade]", inputs: { lineHeight: ["tuiFadeHeight", "lineHeight"], size: ["tuiFadeSize", "size"], offset: ["tuiFadeOffset", "offset"], orientation: ["tuiFade", "orientation"] }, host: { properties: { "style.line-height": "this.lineHeight", "style.--line-height": "this.lineHeight", "style.--fade-size": "this.size", "style.--fade-offset": "this.offset", "attr.data-orientation": "this.orientation" } }, providers: [
        TuiDestroyService,
        TuiResizeService,
        MutationObserverService,
        {
            provide: MUTATION_OBSERVER_INIT,
            useValue: { characterData: true, subtree: true },
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFadeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiFade]',
                    providers: [
                        TuiDestroyService,
                        TuiResizeService,
                        MutationObserverService,
                        {
                            provide: MUTATION_OBSERVER_INIT,
                            useValue: { characterData: true, subtree: true },
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [MutationObserverService]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: i2.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { lineHeight: [{
                type: Input,
                args: ['tuiFadeHeight']
            }, {
                type: HostBinding,
                args: ['style.line-height']
            }, {
                type: HostBinding,
                args: ['style.--line-height']
            }], size: [{
                type: Input,
                args: ['tuiFadeSize']
            }, {
                type: HostBinding,
                args: ['style.--fade-size']
            }], offset: [{
                type: Input,
                args: ['tuiFadeOffset']
            }, {
                type: HostBinding,
                args: ['style.--fade-offset']
            }], orientation: [{
                type: Input,
                args: ['tuiFade']
            }, {
                type: HostBinding,
                args: ['attr.data-orientation']
            }] } });
