// @taiga-ui/experimental/bundles/taiga-ui-experimental-pipes-amount.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-commerce'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('rxjs/operators'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/pipes/amount', ['exports', '@angular/core', '@taiga-ui/addon-commerce', '@taiga-ui/cdk', '@taiga-ui/core', 'rxjs/operators', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.pipes = global["taiga-ui"].experimental.pipes || {}, global["taiga-ui"].experimental.pipes.amount = {}), global.ng.core, global.addonCommerce, global.cdk, global.core, global.rxjs.operators, global.rxjs));
})(this, (function (exports, i0, addonCommerce, cdk, core, operators, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TUI_AMOUNT_DEFAULT_OPTIONS = {
        currency: null,
        currencyAlign: 'right',
        sign: 'negative-only',
    };
    var TUI_AMOUNT_OPTIONS = cdk.tuiCreateToken(TUI_AMOUNT_DEFAULT_OPTIONS);
    function tuiAmountOptionsProvider(options) {
        return cdk.tuiProvideOptions(TUI_AMOUNT_OPTIONS, options, TUI_AMOUNT_DEFAULT_OPTIONS);
    }

    var DEFAULT_DECIMAL_LIMIT = 2;
    var TuiAmountPipePipe = /** @class */ (function () {
        function TuiAmountPipePipe(options, format) {
            this.options = options;
            this.format = format;
        }
        TuiAmountPipePipe.prototype.transform = function (value, currency, currencyAlign) {
            var _this = this;
            if (currency === void 0) { currency = this.options.currency; }
            if (currencyAlign === void 0) { currencyAlign = this.options.currencyAlign; }
            return this.format.pipe(operators.map(function (format) {
                var sign = addonCommerce.tuiFormatSignSymbol(value, _this.options.sign);
                var currencySymbol = addonCommerce.tuiFormatCurrency(currency);
                var formatted = core.tuiFormatNumber(Math.abs(value), Object.assign(Object.assign({}, format), { decimalLimit: _this.getDecimalLimit(value, Number.isNaN(format.decimalLimit)
                        ? DEFAULT_DECIMAL_LIMIT
                        : format.decimalLimit, (format === null || format === void 0 ? void 0 : format.decimal) || 'not-zero') }));
                var space = (currencySymbol === null || currencySymbol === void 0 ? void 0 : currencySymbol.length) > 1 || currencyAlign === 'right'
                    ? cdk.CHAR_NO_BREAK_SPACE
                    : '';
                return currencyAlign === 'right'
                    ? "" + sign + formatted + space + currencySymbol
                    : "" + sign + currencySymbol + space + formatted;
            }));
        };
        TuiAmountPipePipe.prototype.getDecimalLimit = function (value, limit, decimal) {
            return decimal === 'always' || (decimal === 'not-zero' && value % 1) ? limit : 0;
        };
        return TuiAmountPipePipe;
    }());
    TuiAmountPipePipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipePipe, deps: [{ token: TUI_AMOUNT_OPTIONS }, { token: core.TUI_NUMBER_FORMAT_OBSERVABLE }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiAmountPipePipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipePipe, name: "tuiAmount" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipePipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiAmount',
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_AMOUNT_OPTIONS]
                        }] }, { type: i1__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [core.TUI_NUMBER_FORMAT_OBSERVABLE]
                        }] }];
        } });

    var TuiAmountPipeModule = /** @class */ (function () {
        function TuiAmountPipeModule() {
        }
        return TuiAmountPipeModule;
    }());
    TuiAmountPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAmountPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipeModule, declarations: [TuiAmountPipePipe], exports: [TuiAmountPipePipe] });
    TuiAmountPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAmountPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiAmountPipePipe],
                        exports: [TuiAmountPipePipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_AMOUNT_DEFAULT_OPTIONS = TUI_AMOUNT_DEFAULT_OPTIONS;
    exports.TUI_AMOUNT_OPTIONS = TUI_AMOUNT_OPTIONS;
    exports.TuiAmountPipeModule = TuiAmountPipeModule;
    exports.TuiAmountPipePipe = TuiAmountPipePipe;
    exports.tuiAmountOptionsProvider = tuiAmountOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
