// @taiga-ui/experimental/bundles/taiga-ui-experimental-pipes-auto-color.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/kit')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/pipes/auto-color', ['exports', '@angular/core', '@taiga-ui/kit'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.pipes = global["taiga-ui"].experimental.pipes || {}, global["taiga-ui"].experimental.pipes["auto-color"] = {}), global.ng.core, global.kit));
})(this, (function (exports, i0, kit) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiAutoColorPipe = /** @class */ (function () {
        function TuiAutoColorPipe() {
        }
        TuiAutoColorPipe.prototype.transform = function (text) {
            return kit.tuiStringHashToHsl(text);
        };
        return TuiAutoColorPipe;
    }());
    TuiAutoColorPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiAutoColorPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorPipe, name: "tuiAutoColor" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiAutoColor',
                    }]
            }] });

    var TuiAutoColorModule = /** @class */ (function () {
        function TuiAutoColorModule() {
        }
        return TuiAutoColorModule;
    }());
    TuiAutoColorModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAutoColorModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorModule, declarations: [TuiAutoColorPipe], exports: [TuiAutoColorPipe] });
    TuiAutoColorModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAutoColorModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiAutoColorPipe],
                        exports: [TuiAutoColorPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiAutoColorModule = TuiAutoColorModule;
    exports.TuiAutoColorPipe = TuiAutoColorPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
