// @taiga-ui/experimental/bundles/taiga-ui-experimental-pipes-fallback-src.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/pipes/fallback-src', ['exports', '@angular/core', '@taiga-ui/core', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.pipes = global["taiga-ui"].experimental.pipes || {}, global["taiga-ui"].experimental.pipes["fallback-src"] = {}), global.ng.core, global.core, global.rxjs, global.rxjs.operators));
})(this, (function (exports, i0, core, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiFallbackSrcPipe = /** @class */ (function () {
        function TuiFallbackSrcPipe(el) {
            this.el = el;
        }
        TuiFallbackSrcPipe.prototype.transform = function (src, fallback) {
            return rxjs.merge(rxjs.fromEvent(this.el.nativeElement, core.TUI_ICON_ERROR), rxjs.fromEvent(this.el.nativeElement, 'error', { capture: true })).pipe(operators.map(function () { return fallback; }), operators.startWith(src));
        };
        return TuiFallbackSrcPipe;
    }());
    TuiFallbackSrcPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcPipe, deps: [{ token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiFallbackSrcPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcPipe, name: "tuiFallbackSrc" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiFallbackSrc',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        } });

    var TuiFallbackSrcModule = /** @class */ (function () {
        function TuiFallbackSrcModule() {
        }
        return TuiFallbackSrcModule;
    }());
    TuiFallbackSrcModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFallbackSrcModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcModule, declarations: [TuiFallbackSrcPipe], exports: [TuiFallbackSrcPipe] });
    TuiFallbackSrcModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFallbackSrcModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFallbackSrcPipe],
                        exports: [TuiFallbackSrcPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiFallbackSrcModule = TuiFallbackSrcModule;
    exports.TuiFallbackSrcPipe = TuiFallbackSrcPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
