// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-button-group.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/button-group', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["button-group"] = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiButtonGroupComponent = /** @class */ (function () {
            function TuiButtonGroupComponent() {
            }
            return TuiButtonGroupComponent;
        }());
        TuiButtonGroupComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiButtonGroupComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonGroupComponent, selector: "ng-component", host: { classAttribute: "tui-button-group-styles" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiButtonGroup]{display:flex;border-radius:var(--tui-radius-xl);justify-content:center;overflow:hidden}[tuiButtonGroup]>button,[tuiButtonGroup]>a{transition-property:background;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;display:flex;border:none;outline:none;background:transparent;align-items:center;flex:1;flex-direction:column;padding:1.125rem .25rem;grid-gap:.5rem;gap:.5rem;font-size:var(--tui-font-text-s);max-width:calc(50% - 1.75rem);cursor:pointer;color:var(--tui-link);text-align:center}[tuiButtonGroup]>button:active,[tuiButtonGroup]>a:active{background:var(--tui-clear)}[tuiButtonGroup]>button:before,[tuiButtonGroup]>a:before,[tuiButtonGroup]>button:after,[tuiButtonGroup]>a:after{position:absolute;top:0;background:inherit;width:1.75rem;height:100%}[tuiButtonGroup]>button:first-child:before,[tuiButtonGroup]>a:first-child:before{content:\"\";left:-1.75rem}[tuiButtonGroup]>button:last-child:after,[tuiButtonGroup]>a:last-child:after{content:\"\";right:-1.75rem}[tuiButtonGroup]>button tui-icon,[tuiButtonGroup]>a tui-icon{font-size:1.75rem}[tuiButtonGroup]:has(button:only-child){border-radius:var(--tui-radius-l)}[tuiButtonGroup]>button:only-child,[tuiButtonGroup]>a:only-child{width:100%;flex-direction:row;font:var(--tui-font-text-l);max-width:100%;justify-content:center}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./button-group.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-button-group-styles',
                            },
                        }]
                }] });

        var TuiButtonGroupDirective = /** @class */ (function () {
            function TuiButtonGroupDirective(directiveStyles) {
                directiveStyles.addComponent(TuiButtonGroupComponent);
            }
            return TuiButtonGroupDirective;
        }());
        TuiButtonGroupDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiButtonGroupDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonGroupDirective, selector: "[tuiButtonGroup]", ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiButtonGroup]',
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            } });

        var TuiButtonGroupModule = /** @class */ (function () {
            function TuiButtonGroupModule() {
            }
            return TuiButtonGroupModule;
        }());
        TuiButtonGroupModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiButtonGroupModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupModule, declarations: [TuiButtonGroupDirective, TuiButtonGroupComponent], exports: [TuiButtonGroupDirective] });
        TuiButtonGroupModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonGroupModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiButtonGroupDirective, TuiButtonGroupComponent],
                            exports: [TuiButtonGroupDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiButtonGroupComponent = TuiButtonGroupComponent;
        exports.TuiButtonGroupDirective = TuiButtonGroupDirective;
        exports.TuiButtonGroupModule = TuiButtonGroupModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
