// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-progress-segmented.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/progress-segmented', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["progress-segmented"] = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiProgressSegmentedComponent = /** @class */ (function () {
            function TuiProgressSegmentedComponent() {
            }
            return TuiProgressSegmentedComponent;
        }());
        TuiProgressSegmentedComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiProgressSegmentedComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiProgressSegmentedComponent, selector: "ng-component", host: { classAttribute: "tui-progress-segmented" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiProgressBar]._segmented{--tui-segment-gap: .5rem;-webkit-mask-image:radial-gradient(circle closest-side at calc(var(--t-height) / 2) center,#999 0 99%,transparent calc(99% + .6px) 100%),radial-gradient(circle closest-side at calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)) center,#999 0 99%,transparent calc(99% + .6px) 100%),linear-gradient(to right,transparent 0 calc(var(--t-height) / 2),#999 calc(var(--t-height) / 2) calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)),transparent calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)));mask-image:radial-gradient(circle closest-side at calc(var(--t-height) / 2) center,#999 0 99%,transparent calc(99% + .6px) 100%),radial-gradient(circle closest-side at calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)) center,#999 0 99%,transparent calc(99% + .6px) 100%),linear-gradient(to right,transparent 0 calc(var(--t-height) / 2),#999 calc(var(--t-height) / 2) calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)),transparent calc(100% - calc(var(--t-height) / 2) - var(--tui-segment-gap)));-webkit-mask-size:calc(100% * var(--t-segment-width) + var(--tui-segment-gap) * var(--t-segment-width));mask-size:calc(100% * var(--t-segment-width) + var(--tui-segment-gap) * var(--t-segment-width))}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            host: { class: 'tui-progress-segmented' },
                            template: '',
                            styleUrls: ['./progress-segmented.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        }]
                }] });

        var TuiProgressSegmentedDirective = /** @class */ (function () {
            function TuiProgressSegmentedDirective(directiveStyles) {
                this.segments = 1;
                directiveStyles.addComponent(TuiProgressSegmentedComponent);
            }
            return TuiProgressSegmentedDirective;
        }());
        TuiProgressSegmentedDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiProgressSegmentedDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiProgressSegmentedDirective, selector: "[tuiProgressBar][segments]", inputs: { segments: "segments" }, host: { properties: { "style.--t-segment-width": "1 / segments", "attr.new": "\"\"" }, classAttribute: "_segmented" }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiProgressBar][segments]',
                            host: {
                                class: '_segmented',
                                '[style.--t-segment-width]': '1 / segments',
                                '[attr.new]': '""', // TODO: drop in v4.0
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { segments: [{
                        type: i0.Input
                    }] } });

        /**
         * TODO: Move to TuiProgressModule (from `@taiga-ui/kit`) in 4.0
         * TODO: delete the previous version of the same component
         */
        var TuiProgressSegmentedModule = /** @class */ (function () {
            function TuiProgressSegmentedModule() {
            }
            return TuiProgressSegmentedModule;
        }());
        TuiProgressSegmentedModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiProgressSegmentedModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedModule, declarations: [TuiProgressSegmentedComponent, TuiProgressSegmentedDirective], exports: [TuiProgressSegmentedDirective] });
        TuiProgressSegmentedModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProgressSegmentedModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiProgressSegmentedComponent, TuiProgressSegmentedDirective],
                            exports: [TuiProgressSegmentedDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiProgressSegmentedComponent = TuiProgressSegmentedComponent;
        exports.TuiProgressSegmentedDirective = TuiProgressSegmentedDirective;
        exports.TuiProgressSegmentedModule = TuiProgressSegmentedModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
