// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-radio.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/radio', ['exports', '@angular/core', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components.radio = {}), global.ng.core, global.ng.forms));
})(this, (function (exports, i0, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiRadioComponent = /** @class */ (function () {
        function TuiRadioComponent(control) {
            this.control = control;
            this.size = 'm';
        }
        return TuiRadioComponent;
    }());
    TuiRadioComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioComponent, deps: [{ token: i1.NgControl, optional: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiRadioComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRadioComponent, selector: "input[type=\"radio\"][tuiRadio]", inputs: { size: "size" }, host: { attributes: { "tuiAppearance": "" }, properties: { "disabled": "!control || control.disabled", "attr.data-size": "size", "class._invalid": "control?.invalid && control?.touched", "class._readonly": "!control" } }, ngImport: i0__namespace, template: '', isInline: true, styles: [":host{--t-size: 1.5rem;transition-property:background,box-shadow;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;width:var(--t-size);height:var(--t-size);cursor:pointer;margin:0;border-radius:100%;box-shadow:inset 0 0 0 .0625rem var(--tui-clear-active);color:var(--tui-primary-text)}:host:disabled._readonly{opacity:1}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;content:\"\";border-radius:100%;background:currentColor;transform:scale(0)}:host:checked{box-shadow:none;background:var(--tui-primary)}:host:checked:hover{background:var(--tui-primary-hover)}:host:checked:active{background:var(--tui-primary-active)}:host:checked:before{transform:scale(.5)}:host[data-size=s]{--t-size: 1rem}:host._invalid{box-shadow:inset 0 0 0 .0625rem var(--tui-error-bg-hover)}:host._invalid:checked{background:var(--tui-error-fill)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'input[type="radio"][tuiRadio]',
                        template: '',
                        styleUrls: ['./radio.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            tuiAppearance: '',
                            '[disabled]': '!control || control.disabled',
                            '[attr.data-size]': 'size',
                            '[class._invalid]': 'control?.invalid && control?.touched',
                            '[class._readonly]': '!control',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }] } });

    var TuiRadioModule = /** @class */ (function () {
        function TuiRadioModule() {
        }
        return TuiRadioModule;
    }());
    TuiRadioModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiRadioModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioModule, declarations: [TuiRadioComponent], exports: [TuiRadioComponent] });
    TuiRadioModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiRadioComponent],
                        exports: [TuiRadioComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiRadioComponent = TuiRadioComponent;
    exports.TuiRadioModule = TuiRadioModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
