// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-badged-content.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@angular/common')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/badged-content', ['exports', '@angular/core', '@taiga-ui/cdk', '@angular/common'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components["badged-content"] = {}), global.ng.core, global.i1, global.ng.common));
})(this, (function (exports, i0, i1, common) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiBadgedContentComponent = /** @class */ (function () {
            function TuiBadgedContentComponent() {
            }
            return TuiBadgedContentComponent;
        }());
        TuiBadgedContentComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiBadgedContentComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgedContentComponent, selector: "tui-badged-content", ngImport: i0__namespace, template: "<ng-content></ng-content>\n\n<div\n    #badgeTop\n    class=\"t-badge t-badge_top\"\n    [style.--t-badge-height.px]=\"badgeTop.offsetHeight\"\n    (tuiResize)=\"(0)\"\n>\n    <ng-content select=\"tui-badge-notification[tuiSlot='top']\"></ng-content>\n    <div class=\"t-border\">\n        <ng-content select=\"[tuiSlot='top']\"></ng-content>\n    </div>\n</div>\n\n<div\n    #badgeBottom\n    class=\"t-badge t-badge_bottom\"\n    [style.--t-badge-height.px]=\"badgeBottom.offsetHeight\"\n    (tuiResize)=\"(0)\"\n>\n    <ng-content select=\"tui-badge-notification[tuiSlot='bottom']\"></ng-content>\n    <div class=\"t-border\">\n        <ng-content select=\"[tuiSlot='bottom']\"></ng-content>\n    </div>\n</div>\n", styles: [":host{--tui-radius: var(--tui-radius-l);--t-badge-height: 0;--t-corner-offset: calc((var(--tui-radius) * 1.4142 - var(--tui-radius)) * 1 / 1.4142);position:relative;display:inline-block;height:-webkit-fit-content;height:-moz-fit-content;height:fit-content;width:-webkit-max-content;width:max-content;color:var(--tui-base-01)}.t-badge{--t-offset: calc(var(--t-badge-height) * -.5 + var(--t-corner-offset));position:absolute;display:flex}.t-badge_top{top:var(--t-offset);right:var(--t-offset)}.t-badge_bottom{bottom:var(--t-offset);right:var(--t-offset)}.t-border:not(:empty){display:flex;border-radius:6rem;box-shadow:0 0 0 1px currentColor;background:currentColor}\n"], directives: [{ type: i1__namespace.TuiResizeDirective, selector: "[tuiResize]", outputs: ["tuiResize"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            selector: 'tui-badged-content',
                            templateUrl: './badged-content.template.html',
                            styleUrls: ['./badged-content.style.less'],
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        }]
                }] });

        var TuiBadgedContentDirective = /** @class */ (function () {
            function TuiBadgedContentDirective() {
                this.tuiSlot = 'top';
            }
            return TuiBadgedContentDirective;
        }());
        TuiBadgedContentDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiBadgedContentDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgedContentDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiSlot]',
                        }]
                }], propDecorators: { tuiSlot: [{
                        type: i0.Input
                    }] } });

        var TuiBadgedContentModule = /** @class */ (function () {
            function TuiBadgedContentModule() {
            }
            return TuiBadgedContentModule;
        }());
        TuiBadgedContentModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiBadgedContentModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective], imports: [common.CommonModule, i1.TuiResizeModule], exports: [TuiBadgedContentComponent, TuiBadgedContentDirective] });
        TuiBadgedContentModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, imports: [[common.CommonModule, i1.TuiResizeModule]] });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            imports: [common.CommonModule, i1.TuiResizeModule],
                            declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                            exports: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiBadgedContentComponent = TuiBadgedContentComponent;
        exports.TuiBadgedContentDirective = TuiBadgedContentDirective;
        exports.TuiBadgedContentModule = TuiBadgedContentModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
