// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-thumbnail-card.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-commerce'), require('@taiga-ui/experimental/tokens'), require('@taiga-ui/experimental/components/icon'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/thumbnail-card', ['exports', '@angular/core', '@taiga-ui/addon-commerce', '@taiga-ui/experimental/tokens', '@taiga-ui/experimental/components/icon', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components["thumbnail-card"] = {}), global.ng.core, global.addonCommerce, global["taiga-ui"].experimental.tokens, global["taiga-ui"].experimental.components.icon, global.ng.common));
})(this, (function (exports, i0, addonCommerce, tokens, i1, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TuiThumbnailCardComponent = /** @class */ (function () {
        function TuiThumbnailCardComponent(resolver, options) {
            this.resolver = resolver;
            this.options = options;
            this.size = 'm';
            this.paymentSystem = null;
            this.iconLeft = '';
            this.iconRight = '';
        }
        Object.defineProperty(TuiThumbnailCardComponent.prototype, "isMono", {
            // TODO: Revisit this approach in 4.0 when icons are moved away from InputCard options
            get: function () {
                switch (this.paymentSystem) {
                    case 'mir':
                    case 'visa':
                    case 'electron':
                        return true;
                    default:
                        return false;
                }
            },
            enumerable: false,
            configurable: true
        });
        return TuiThumbnailCardComponent;
    }());
    TuiThumbnailCardComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardComponent, deps: [{ token: tokens.TUI_ICON_RESOLVER }, { token: addonCommerce.TUI_INPUT_CARD_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiThumbnailCardComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiThumbnailCardComponent, selector: "tui-thumbnail-card", inputs: { size: "size", paymentSystem: "paymentSystem", iconLeft: "iconLeft", iconRight: "iconRight" }, host: { properties: { "attr.data-size": "this.size" } }, ngImport: i0__namespace, template: "<div class=\"t-front\">\n    <div\n        *ngIf=\"iconLeft\"\n        class=\"t-left\"\n    >\n        <tui-icon\n            class=\"t-icon\"\n            [icon]=\"iconLeft\"\n        ></tui-icon>\n    </div>\n    <div\n        *ngIf=\"iconRight\"\n        class=\"t-right\"\n    >\n        <tui-icon\n            class=\"t-icon\"\n            [icon]=\"iconRight\"\n        ></tui-icon>\n    </div>\n    <ng-content></ng-content>\n    <ng-container *ngIf=\"paymentSystem\">\n        <tui-icon\n            *ngIf=\"isMono; else colored\"\n            class=\"t-logo\"\n            [icon]=\"options.icons[paymentSystem]\"\n        ></tui-icon>\n        <ng-template #colored>\n            <img\n                class=\"t-logo\"\n                [alt]=\"paymentSystem\"\n                [src]=\"resolver(options.icons[paymentSystem]).replace('Outline', '')\"\n            />\n        </ng-template>\n    </ng-container>\n</div>\n<div class=\"t-back\"></div>\n", styles: [":host{position:relative;display:inline-flex;flex-shrink:0;background:#000;color:#fff;transform-style:preserve-3d;overflow:hidden}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\"}:host[data-size=s]{width:2.5rem;height:1.625rem;border-radius:.25rem;font-size:0}:host[data-size=s] .t-left{transform:translate(.125rem,.125rem) scale(.625)}:host[data-size=s] .t-right{display:none}:host[data-size=s] .t-logo{bottom:-.375rem;transform:scale(.55)}:host[data-size=m]{width:3rem;height:2rem;border-radius:.3125rem;text-indent:.1875rem;font:normal .5625rem/.875rem var(--tui-font-text)}:host[data-size=m] .t-left{transform:translate(.1875rem,.25rem) scale(.75)}:host[data-size=m] .t-right{transform:translate(-.1875rem,.1875rem) scale(.875)}:host[data-size=m] .t-right .t-icon{transform:scale(.714)}:host[data-size=m] .t-logo{bottom:-.3125rem;transform:scale(.55)}:host[data-size=l]{width:3.375rem;height:2.25rem;border-radius:.375rem;text-indent:.3125rem;font:normal .625rem/1.125rem var(--tui-font-text)}:host[data-size=l] .t-left{transform:translate(.1875rem,.1875rem)}:host[data-size=l] .t-right{transform:translate(-.1875rem,.1875rem)}:host[data-size=l] .t-logo{bottom:-.1875rem;transform:scale(.75)}.t-left,.t-right{position:absolute;top:0;display:flex}.t-left{left:0;transform-origin:top left}.t-right{right:0;transform-origin:top right}.t-right:before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;border-radius:100%;background:currentColor;opacity:.35}.t-right .t-icon{color:#fff;transform:scale(.625)}.t-logo{position:absolute;right:.1875rem;width:1.5rem;height:1.5rem;transform-origin:right}.t-icon{width:1rem;height:1rem}.t-front{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;align-items:flex-end;-webkit-backface-visibility:hidden;backface-visibility:hidden;box-shadow:inset 0 0 0 1px var(--tui-clear);border-radius:inherit}.t-back{position:absolute;top:0;left:0;width:100%;height:100%;-webkit-backface-visibility:hidden;backface-visibility:hidden;transform:rotateY(180deg) translateZ(1px);background:#b0b0b0;border-radius:inherit}.t-back:after{position:absolute;top:0;left:0;bottom:0;right:0;content:\"\";top:20%;bottom:60%;background:#959595}\n"], components: [{ type: i1__namespace.TuiIconComponent, selector: "tui-icon", inputs: ["icon", "background"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-thumbnail-card',
                        templateUrl: './thumbnail-card.template.html',
                        styleUrls: ['./thumbnail-card.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ICON_RESOLVER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [addonCommerce.TUI_INPUT_CARD_OPTIONS]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], paymentSystem: [{
                    type: i0.Input
                }], iconLeft: [{
                    type: i0.Input
                }], iconRight: [{
                    type: i0.Input
                }] } });

    var TuiThumbnailCardModule = /** @class */ (function () {
        function TuiThumbnailCardModule() {
        }
        return TuiThumbnailCardModule;
    }());
    TuiThumbnailCardModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiThumbnailCardModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardModule, declarations: [TuiThumbnailCardComponent], imports: [i2.CommonModule, i1.TuiIconModule], exports: [TuiThumbnailCardComponent] });
    TuiThumbnailCardModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardModule, imports: [[i2.CommonModule, i1.TuiIconModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiThumbnailCardModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i2.CommonModule, i1.TuiIconModule],
                        declarations: [TuiThumbnailCardComponent],
                        exports: [TuiThumbnailCardComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiThumbnailCardComponent = TuiThumbnailCardComponent;
    exports.TuiThumbnailCardModule = TuiThumbnailCardModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
