// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-badge-notification.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core'), require('@taiga-ui/cdk')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/badge-notification', ['exports', '@angular/core', '@taiga-ui/core', '@taiga-ui/cdk'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components["badge-notification"] = {}), global.ng.core, global.core, global.cdk));
})(this, (function (exports, i0, core, cdk) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS = {
        size: 'm',
    };
    var TUI_BADGE_NOTIFICATION_OPTIONS = cdk.tuiCreateToken(TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS);
    function tuiBadgeNotificationOptionsProvider(options) {
        return cdk.tuiProvideOptions(TUI_BADGE_NOTIFICATION_OPTIONS, options, TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS);
    }

    var TuiBadgeNotificationComponent = /** @class */ (function () {
        function TuiBadgeNotificationComponent(animation, options) {
            this.animation = animation;
            this.options = options;
            this.size = this.options.size;
        }
        return TuiBadgeNotificationComponent;
    }());
    TuiBadgeNotificationComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationComponent, deps: [{ token: core.TUI_ANIMATION_OPTIONS }, { token: TUI_BADGE_NOTIFICATION_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiBadgeNotificationComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgeNotificationComponent, selector: "tui-badge-notification", inputs: { size: "size" }, host: { properties: { "attr.data-size": "size", "@tuiPop": "animation" } }, ngImport: i0__namespace, template: '<ng-content></ng-content>', isInline: true, styles: [":host{--t-size: 1.5rem;position:relative;display:inline-flex;align-items:center;justify-content:center;box-sizing:border-box;color:var(--tui-text-01-night);border-radius:2rem;white-space:nowrap;overflow:hidden;vertical-align:middle;font:var(--tui-font-text-s);max-width:100%;padding:0 .25rem;background:#f52222;height:var(--t-size);min-width:var(--t-size)}:host[data-size=m]{--t-size: 1.25rem}:host[data-size=s]{--t-size: 1rem;padding:0 .125rem;font:var(--tui-font-text-xs)}:host[data-size=xs]{--t-size: .375rem;padding:0;font-size:0}:host :host-context([tuiIconButton]){position:absolute;right:25%;top:25%}\n"], animations: [core.tuiPop], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-badge-notification',
                        template: '<ng-content></ng-content>',
                        styleUrls: ['./badge-notification.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        animations: [core.tuiPop],
                        host: {
                            '[attr.data-size]': 'size',
                            '[@tuiPop]': 'animation',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [core.TUI_ANIMATION_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_BADGE_NOTIFICATION_OPTIONS]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }] } });

    var TuiBadgeNotificationModule = /** @class */ (function () {
        function TuiBadgeNotificationModule() {
        }
        return TuiBadgeNotificationModule;
    }());
    TuiBadgeNotificationModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiBadgeNotificationModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationModule, declarations: [TuiBadgeNotificationComponent], exports: [TuiBadgeNotificationComponent] });
    TuiBadgeNotificationModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeNotificationModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiBadgeNotificationComponent],
                        exports: [TuiBadgeNotificationComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS = TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS;
    exports.TUI_BADGE_NOTIFICATION_OPTIONS = TUI_BADGE_NOTIFICATION_OPTIONS;
    exports.TuiBadgeNotificationComponent = TuiBadgeNotificationComponent;
    exports.TuiBadgeNotificationModule = TuiBadgeNotificationModule;
    exports.tuiBadgeNotificationOptionsProvider = tuiBadgeNotificationOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
