// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-header.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/experimental/components')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/header', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/experimental/components'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.header = {}), global.ng.core, global.i1, global["taiga-ui"].experimental.components));
})(this, (function (exports, i0, i1, components) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiHeaderStylesComponent = /** @class */ (function () {
            function TuiHeaderStylesComponent() {
            }
            return TuiHeaderStylesComponent;
        }());
        TuiHeaderStylesComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderStylesComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiHeaderStylesComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiHeaderStylesComponent, selector: "ng-component", host: { classAttribute: "tui-header" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiHeader]{position:relative;display:flex;align-items:flex-start;box-sizing:content-box;color:var(--tui-text-01)}[tuiHeader] [tuiTitle]{flex-grow:2}[tuiHeader] [tuiAccessories]{display:flex;align-items:center;grid-gap:.75rem;gap:.75rem}[tuiHeader] [tuiAccessories]:before{content:\"\\200b\";visibility:hidden}[tuiHeader] [tuiSubtitle]{font:var(--tui-font-text-m);color:var(--tui-text-02)}[tuiHeader] [tuiCaption]{font:var(--tui-font-text-s);color:var(--tui-text-02);font-weight:500;text-transform:uppercase}[tuiHeader][data-size=xxl] [tuiTitle],[tuiHeader][data-size=xxl] [tuiAccessories]{font:var(--tui-font-heading-1)}[tuiHeader][data-size=xl] [tuiTitle],[tuiHeader][data-size=xl] [tuiAccessories]{font:var(--tui-font-heading-2)}[tuiHeader][data-size=l] [tuiTitle],[tuiHeader][data-size=l] [tuiAccessories]{font:var(--tui-font-heading-3)}[tuiHeader][data-size=m] [tuiTitle],[tuiHeader][data-size=m] [tuiAccessories]{font:var(--tui-font-heading-4)}[tuiHeader][data-size=s] [tuiTitle],[tuiHeader][data-size=s] [tuiAccessories]{font:var(--tui-font-heading-5)}[tuiHeader][data-size=xs] [tuiTitle],[tuiHeader][data-size=xs] [tuiAccessories]{font:var(--tui-font-heading-6)}[tuiHeader][data-size=xxs] [tuiTitle],[tuiHeader][data-size=xxs] [tuiAccessories]{font:var(--tui-font-text-xl)}[tuiHeader][data-size=xxl] [tuiSubtitle],[tuiHeader][data-size=xl] [tuiSubtitle],[tuiHeader][data-size=l] [tuiSubtitle]{font:var(--tui-font-text-l)}tui-root._mobile [tuiHeader] [tuiAccessories]{grid-gap:1rem;gap:1rem}tui-root._mobile [tuiHeader][data-size=xxs] [tuiTitle],tui-root._mobile [tuiHeader][data-size=xxs] [tuiAccessories]{font:var(--tui-font-heading-6)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderStylesComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./header.styles.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-header',
                            },
                        }]
                }] });

        var TuiHeaderDirective = /** @class */ (function () {
            function TuiHeaderDirective(directiveStyles) {
                this.size = 's';
                directiveStyles.addComponent(TuiHeaderStylesComponent);
            }
            return TuiHeaderDirective;
        }());
        TuiHeaderDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiHeaderDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHeaderDirective, selector: "[tuiHeader]", inputs: { size: ["tuiHeader", "size"] }, host: { attributes: { "tuiHeader": "" }, properties: { "attr.data-size": "size || \"s\"" } }, providers: [
                components.tuiAvatarOptionsProvider({ size: 's' }),
                components.tuiButtonOptionsProvider({ size: 's' }),
                components.tuiBadgeOptionsProvider({ size: 'm' }),
            ], ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiHeader]',
                            providers: [
                                components.tuiAvatarOptionsProvider({ size: 's' }),
                                components.tuiButtonOptionsProvider({ size: 's' }),
                                components.tuiBadgeOptionsProvider({ size: 'm' }),
                            ],
                            host: {
                                tuiHeader: '',
                                '[attr.data-size]': 'size || "s"',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { size: [{
                        type: i0.Input,
                        args: ['tuiHeader']
                    }] } });

        var TuiHeaderModule = /** @class */ (function () {
            function TuiHeaderModule() {
            }
            return TuiHeaderModule;
        }());
        TuiHeaderModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiHeaderModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderModule, declarations: [TuiHeaderStylesComponent, TuiHeaderDirective], exports: [TuiHeaderDirective] });
        TuiHeaderModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiHeaderModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiHeaderStylesComponent, TuiHeaderDirective],
                            exports: [TuiHeaderDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiHeaderDirective = TuiHeaderDirective;
        exports.TuiHeaderModule = TuiHeaderModule;
        exports.TuiHeaderStylesComponent = TuiHeaderStylesComponent;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
