// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-button-close.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/experimental/components'), require('@taiga-ui/experimental/tokens'), require('@taiga-ui/cdk')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/button-close', ['exports', '@angular/core', '@taiga-ui/experimental/components', '@taiga-ui/experimental/tokens', '@taiga-ui/cdk'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["button-close"] = {}), global.ng.core, global["taiga-ui"].experimental.components, global["taiga-ui"].experimental.tokens, global.cdk));
})(this, (function (exports, i0, components, tokens, cdk) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TUI_BUTTON_CLOSE_ICON = cdk.tuiCreateToken('tuiIconClose');

    // TODO: 4.0 revisit after turning tuiIcon to host directive
    var TuiButtonCloseDirective = /** @class */ (function () {
        function TuiButtonCloseDirective(resolver, icon) {
            this.resolver = resolver;
            this.icon = icon;
        }
        return TuiButtonCloseDirective;
    }());
    TuiButtonCloseDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseDirective, deps: [{ token: tokens.TUI_ICON_RESOLVER }, { token: TUI_BUTTON_CLOSE_ICON }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiButtonCloseDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonCloseDirective, selector: "[tuiIconButton][tuiButtonClose]", host: { properties: { "style.--t-mask-left": "\"url(\" + resolver(icon) + \")\"", "style.--t-radius.%": "100", "style.--tui-height-s.rem": "1.875", "class._icon-left": "true" } }, providers: [components.tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' })], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiIconButton][tuiButtonClose]',
                        providers: [components.tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' })],
                        host: {
                            '[style.--t-mask-left]': '"url(" + resolver(icon) + ")"',
                            '[style.--t-radius.%]': '100',
                            '[style.--tui-height-s.rem]': '1.875',
                            '[class._icon-left]': 'true',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ICON_RESOLVER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_BUTTON_CLOSE_ICON]
                        }] }];
        } });

    var TuiButtonCloseModule = /** @class */ (function () {
        function TuiButtonCloseModule() {
        }
        return TuiButtonCloseModule;
    }());
    TuiButtonCloseModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiButtonCloseModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseModule, declarations: [TuiButtonCloseDirective], exports: [TuiButtonCloseDirective] });
    TuiButtonCloseModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonCloseModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiButtonCloseDirective],
                        exports: [TuiButtonCloseDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_BUTTON_CLOSE_ICON = TUI_BUTTON_CLOSE_ICON;
    exports.TuiButtonCloseDirective = TuiButtonCloseDirective;
    exports.TuiButtonCloseModule = TuiButtonCloseModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
