// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-appearance.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/appearance', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.appearance = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiAppearanceDirective = /** @class */ (function () {
        function TuiAppearanceDirective() {
            this.tuiAppearance = '';
            this.tuiAppearanceState = null;
            this.tuiAppearanceFocus = null;
        }
        return TuiAppearanceDirective;
    }());
    TuiAppearanceDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiAppearanceDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppearanceDirective, selector: "[tuiAppearance]", inputs: { tuiAppearance: "tuiAppearance", tuiAppearanceState: "tuiAppearanceState", tuiAppearanceFocus: "tuiAppearanceFocus" }, host: { attributes: { "tuiAppearance": "" }, properties: { "attr.data-appearance": "tuiAppearance", "attr.data-state": "tuiAppearanceState", "attr.data-focus": "tuiAppearanceFocus" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiAppearance]',
                        host: {
                            tuiAppearance: '',
                            '[attr.data-appearance]': 'tuiAppearance',
                            '[attr.data-state]': 'tuiAppearanceState',
                            '[attr.data-focus]': 'tuiAppearanceFocus',
                        },
                    }]
            }], propDecorators: { tuiAppearance: [{
                    type: i0.Input
                }], tuiAppearanceState: [{
                    type: i0.Input
                }], tuiAppearanceFocus: [{
                    type: i0.Input
                }] } });

    var TuiAppearanceModule = /** @class */ (function () {
        function TuiAppearanceModule() {
        }
        return TuiAppearanceModule;
    }());
    TuiAppearanceModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAppearanceModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceModule, declarations: [TuiAppearanceDirective], exports: [TuiAppearanceDirective] });
    TuiAppearanceModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAppearanceModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiAppearanceDirective],
                        exports: [TuiAppearanceDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiAppearanceDirective = TuiAppearanceDirective;
    exports.TuiAppearanceModule = TuiAppearanceModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
