// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-compass.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/compass', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components.compass = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiCompassComponent = /** @class */ (function () {
        function TuiCompassComponent() {
            this.degrees = NaN;
        }
        return TuiCompassComponent;
    }());
    TuiCompassComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiCompassComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCompassComponent, selector: "tui-compass", inputs: { degrees: "degrees" }, host: { properties: { "style.--t-degrees.deg": "this.degrees" } }, ngImport: i0__namespace, template: '', isInline: true, styles: [":host{position:relative;display:inline-block;color:var(--tui-primary);border-radius:50%;width:1rem;height:1rem;border:.1875rem solid var(--tui-base-01);box-shadow:0 0 .1875rem #0000004d;transform-style:preserve-3d}:host:before{content:\"\";position:absolute;top:0;left:.125rem;-webkit-mask-image:url('data:image/svg+xml,<svg xmlns=\"http://www.w3.org/2000/svg\"><path d=\"M0 8.99993C4 8 8 8 12 8.99993C9.91509 5.73239 8.5 3 6 0C3.5 3 2 5.5 0 8.99993Z\"/></svg>');mask-image:url('data:image/svg+xml,<svg xmlns=\"http://www.w3.org/2000/svg\"><path d=\"M0 8.99993C4 8 8 8 12 8.99993C9.91509 5.73239 8.5 3 6 0C3.5 3 2 5.5 0 8.99993Z\"/></svg>');transform:rotate(var(--t-degrees)) translateY(-.625rem) translateZ(-1rem);width:.75rem;height:1rem;background:currentColor}:host:after{position:absolute;top:0;left:0;bottom:0;right:0;content:\"\";border-radius:inherit;background:currentColor linear-gradient(-45deg,rgba(0,0,0,.2),transparent)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-compass',
                        template: '',
                        styleUrls: ['./compass.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { degrees: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['style.--t-degrees.deg']
                }] } });

    var TuiCompassModule = /** @class */ (function () {
        function TuiCompassModule() {
        }
        return TuiCompassModule;
    }());
    TuiCompassModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiCompassModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassModule, declarations: [TuiCompassComponent], exports: [TuiCompassComponent] });
    TuiCompassModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCompassModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiCompassComponent],
                        exports: [TuiCompassComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiCompassComponent = TuiCompassComponent;
    exports.TuiCompassModule = TuiCompassModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
