// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-button-vertical.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/button-vertical', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["button-vertical"] = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiButtonVerticalComponent = /** @class */ (function () {
            function TuiButtonVerticalComponent() {
            }
            return TuiButtonVerticalComponent;
        }());
        TuiButtonVerticalComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiButtonVerticalComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonVerticalComponent, selector: "ng-component", host: { classAttribute: "tui-button-vertical-styles" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiButton][tuiButtonVertical]{--t-line-height: 1rem;--t-margin: 0rem !important;flex-direction:column;height:auto;padding:.75rem;grid-gap:.375rem;gap:.375rem;min-width:5rem;white-space:pre-line;font:var(--tui-font-text-s-2)}[tuiButton][tuiButtonVertical]>*{max-height:calc(var(--t-line-height) * 2)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./button-vertical.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-button-vertical-styles',
                            },
                        }]
                }] });

        var TuiButtonVerticalDirective = /** @class */ (function () {
            function TuiButtonVerticalDirective(directiveStyles) {
                directiveStyles.addComponent(TuiButtonVerticalComponent);
            }
            return TuiButtonVerticalDirective;
        }());
        TuiButtonVerticalDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiButtonVerticalDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonVerticalDirective, selector: "[tuiButtonVertical]", ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiButtonVertical]',
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            } });

        var TuiButtonVerticalModule = /** @class */ (function () {
            function TuiButtonVerticalModule() {
            }
            return TuiButtonVerticalModule;
        }());
        TuiButtonVerticalModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiButtonVerticalModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalModule, declarations: [TuiButtonVerticalDirective, TuiButtonVerticalComponent], exports: [TuiButtonVerticalDirective] });
        TuiButtonVerticalModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiButtonVerticalModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiButtonVerticalDirective, TuiButtonVerticalComponent],
                            exports: [TuiButtonVerticalDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiButtonVerticalComponent = TuiButtonVerticalComponent;
        exports.TuiButtonVerticalDirective = TuiButtonVerticalDirective;
        exports.TuiButtonVerticalModule = TuiButtonVerticalModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
