// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-title.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/title', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.title = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiTitleComponent = /** @class */ (function () {
            function TuiTitleComponent() {
            }
            return TuiTitleComponent;
        }());
        TuiTitleComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiTitleComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTitleComponent, selector: "ng-component", host: { classAttribute: "tui-title" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiTitle]{position:relative;display:flex;flex-direction:column;text-align:left;grid-gap:.25rem;gap:.25rem;margin:0;font:var(--tui-font-text-m-2)}[tuiTitle][data-size=s]{grid-gap:.125rem;gap:.125rem;font:var(--tui-font-text-s)}[tuiTitle][data-size=s] [tuiSubtitle]{font:var(--tui-font-text-xs)}[tuiTitle][data-size=m]{grid-gap:.125rem;gap:.125rem;font:var(--tui-font-heading-5)}[tuiTitle][data-size=m] [tuiSubtitle]{font:var(--tui-font-text-m)}[tuiTitle][data-size=l]{grid-gap:.5rem;gap:.5rem;font:var(--tui-font-heading-3)}[tuiTitle][data-size=l] [tuiSubtitle]{font:var(--tui-font-text-m)}[tuiSubtitle]{font:var(--tui-font-text-s-2)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./title.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-title',
                            },
                        }]
                }] });

        var TuiTitleDirective = /** @class */ (function () {
            function TuiTitleDirective(directiveStyles) {
                this.size = '';
                directiveStyles.addComponent(TuiTitleComponent);
            }
            return TuiTitleDirective;
        }());
        TuiTitleDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiTitleDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTitleDirective, selector: "[tuiTitle]", inputs: { size: ["tuiTitle", "size"] }, host: { attributes: { "tuiTitle": "" }, properties: { "attr.data-size": "size || null" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiTitle]',
                            host: {
                                tuiTitle: '',
                                '[attr.data-size]': 'size || null',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { size: [{
                        type: i0.Input,
                        args: ['tuiTitle']
                    }] } });

        var TuiTitleModule = /** @class */ (function () {
            function TuiTitleModule() {
            }
            return TuiTitleModule;
        }());
        TuiTitleModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiTitleModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleModule, declarations: [TuiTitleDirective, TuiTitleComponent], exports: [TuiTitleDirective] });
        TuiTitleModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTitleModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiTitleDirective, TuiTitleComponent],
                            exports: [TuiTitleDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiTitleComponent = TuiTitleComponent;
        exports.TuiTitleDirective = TuiTitleDirective;
        exports.TuiTitleModule = TuiTitleModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
