// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-cell.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/experimental/components')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/cell', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/experimental/components'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.cell = {}), global.ng.core, global.i1, global["taiga-ui"].experimental.components));
})(this, (function (exports, i0, i1, components) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiCellComponent = /** @class */ (function () {
            function TuiCellComponent() {
            }
            return TuiCellComponent;
        }());
        TuiCellComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiCellComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCellComponent, selector: "ng-component", host: { classAttribute: "tui-cell" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiCell]{--tui-height: calc(var(--tui-height-s) - .5rem);--t-padding: .125rem 1rem;-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:none;font-size:inherit;line-height:inherit;text-decoration:none;position:relative;display:flex;align-items:center;box-sizing:content-box;isolation:isolate;color:var(--tui-text-01);padding:var(--t-padding);min-height:var(--t-height)}[tuiCell] [tuiAccessories]{position:relative;display:flex;max-height:var(--t-height);align-items:center;align-self:stretch}[tuiCell] [tuiCellActions]{position:absolute;right:0;z-index:1;padding-right:inherit}[tuiCell] [tuiCellActions]~*{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out}[tuiCell] [tuiCellActions] button,[tuiCell] [tuiCellActions] a{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;opacity:0}[tuiCell] [tuiCellActions] button:focus-visible,[tuiCell] [tuiCellActions] a:focus-visible{opacity:1}[tuiCell] [tuiSubtitle]{display:flex;align-items:center;grid-gap:.25rem;gap:.25rem;color:var(--tui-text-02)}[tuiCell] [tuiTitle]{flex-shrink:7;-webkit-margin-end:auto;margin-inline-end:auto;align-items:normal;text-align:left}[tuiCell] [tuiTitle]~[tuiTitle]{flex-shrink:3;-webkit-margin-end:0;margin-inline-end:0;text-align:right;align-items:flex-end}[tuiCell] tui-badge-notification[data-size=xs]{position:absolute;top:50%;transform:translateY(-50%);left:-.625rem}[tuiCell][data-size=s]{--t-height: calc(var(--tui-height-s) - .5rem);--t-padding: .125rem 1rem}[tuiCell][data-size=s] [tuiTitle]{max-height:100%;font:var(--tui-font-text-s-2);grid-gap:0;gap:0}[tuiCell][data-size=s] [tuiSubtitle]{font:var(--tui-font-text-xs-s)}[tuiCell][data-size=s]>*:not(:first-child),[tuiCell][data-size=s]>[tuiAccessories]>*:not(:first-child){-webkit-margin-start:.5rem;margin-inline-start:.5rem}[tuiCell][data-size=s]>tui-avatar,[tuiCell][data-size=s]>tui-avatar-stack tui-avatar,[tuiCell][data-size=s]>[tuiAccessories] tui-avatar,[tuiCell][data-size=s]>[tuiAccessories] tui-avatar-stack tui-avatar{--t-size: 1.5rem}[tuiCell][data-size=m]{--t-height: calc(var(--tui-height-m) - .75rem);--t-padding: .375rem 1rem}[tuiCell][data-size=m] [tuiTitle]{font:var(--tui-font-text-s-2);grid-gap:.125rem;gap:.125rem}[tuiCell][data-size=m] [tuiSubtitle]{font:var(--tui-font-text-xs-2)}[tuiCell][data-size=m]>*:not(:first-child),[tuiCell][data-size=m]>[tuiAccessories]>*:not(:first-child){-webkit-margin-start:.75rem;margin-inline-start:.75rem}[tuiCell][data-size=m]>tui-avatar,[tuiCell][data-size=m]>tui-avatar-stack tui-avatar,[tuiCell][data-size=m]>[tuiAccessories] tui-avatar,[tuiCell][data-size=m]>[tuiAccessories] tui-avatar-stack tui-avatar{--t-size: 2rem}[tuiCell][data-size=l]{--t-height: calc(var(--tui-height-l) - 1rem);--t-padding: .5rem 1rem}[tuiCell][data-size=l]>*:not(:first-child),[tuiCell][data-size=l]>[tuiAccessories]>*:not(:first-child){-webkit-margin-start:1rem;margin-inline-start:1rem}[tuiCell][data-size=l]>tui-avatar,[tuiCell][data-size=l]>tui-avatar-stack tui-avatar,[tuiCell][data-size=l]>[tuiAccessories] tui-avatar,[tuiCell][data-size=l]>[tuiAccessories] tui-avatar-stack tui-avatar{--t-size: 2.5rem}[tuiCell]:hover [tuiCellActions]~*{opacity:0}[tuiCell]:hover [tuiCellActions] button,[tuiCell]:hover [tuiCellActions] a{opacity:1}[tuiOption] [tuiCell],tui-primitive-textfield [tuiCell]{padding:0}@media (hover: hover) and (pointer: fine){a[tuiCell]:not([tuiSurface]):hover,button[tuiCell]:not([tuiSurface]):hover{background:var(--tui-clear);cursor:pointer}}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./cell.styles.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-cell',
                            },
                        }]
                }] });

        var TuiCellDirective = /** @class */ (function () {
            function TuiCellDirective(directiveStyles) {
                this.size = 'l';
                directiveStyles.addComponent(TuiCellComponent);
            }
            return TuiCellDirective;
        }());
        TuiCellDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiCellDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCellDirective, selector: "[tuiCell]", inputs: { size: ["tuiCell", "size"] }, host: { attributes: { "tuiCell": "" }, properties: { "attr.data-size": "size || \"l\"" } }, providers: [
                components.tuiAvatarOptionsProvider({ size: 's' }),
                components.tuiButtonOptionsProvider({ size: 's' }),
            ], ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiCell]',
                            providers: [
                                components.tuiAvatarOptionsProvider({ size: 's' }),
                                components.tuiButtonOptionsProvider({ size: 's' }),
                            ],
                            host: {
                                tuiCell: '',
                                '[attr.data-size]': 'size || "l"',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { size: [{
                        type: i0.Input,
                        args: ['tuiCell']
                    }] } });

        var TuiCellModule = /** @class */ (function () {
            function TuiCellModule() {
            }
            return TuiCellModule;
        }());
        TuiCellModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiCellModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellModule, declarations: [TuiCellComponent, TuiCellDirective], exports: [TuiCellComponent, TuiCellDirective] });
        TuiCellModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCellModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiCellComponent, TuiCellDirective],
                            exports: [TuiCellComponent, TuiCellDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiCellComponent = TuiCellComponent;
        exports.TuiCellDirective = TuiCellDirective;
        exports.TuiCellModule = TuiCellModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
