// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-badge.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/experimental/directives/icons')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/badge', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/experimental/directives/icons'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components.badge = {}), global.ng.core, global.i1, global["taiga-ui"].experimental.directives.icons));
})(this, (function (exports, i0, i1, icons) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiBadgeComponent = /** @class */ (function () {
            function TuiBadgeComponent() {
            }
            return TuiBadgeComponent;
        }());
        TuiBadgeComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiBadgeComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgeComponent, selector: "ng-component", host: { classAttribute: "tui-badge" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiBadgeNew]{--t-gap: 0rem;--t-dot: currentColor;--t-icon-size: 1rem;-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:none;font-size:inherit;line-height:inherit;text-decoration:none;position:relative;display:inline-flex;align-items:center;flex-shrink:0;box-sizing:border-box;white-space:nowrap;overflow:hidden;vertical-align:middle;max-width:100%;grid-gap:calc(var(--t-gap) - 2 * var(--t-margin));gap:calc(var(--t-gap) - 2 * var(--t-margin));border-radius:6rem;background:#959595;color:var(--tui-base-01);padding:var(--t-padding);height:var(--t-size);min-width:var(--t-size);width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;font:var(--tui-font-text-s)}[tuiBadgeNew]>img,[tuiBadgeNew]>tui-svg,[tuiBadgeNew]>tui-icon,[tuiBadgeNew]>tui-avatar,[tuiBadgeNew]>tui-badge,[tuiBadgeNew]>[tuiBadge],[tuiBadgeNew]>[tuiRadio],[tuiBadgeNew]>[tuiToggle],[tuiBadgeNew]>[tuiCheckbox],[tuiBadgeNew]._icon-left:before,[tuiBadgeNew]._icon-right:after{margin:var(--t-margin)}[tuiBadgeNew]._dot:before{content:\"\";display:block;width:.375rem;height:.375rem;background:var(--t-dot);border-radius:100%}[tuiBadgeNew]>tui-icon,[tuiBadgeNew][tuiIcons]:before,[tuiBadgeNew][tuiIcons]:after{font-size:var(--t-icon-size)!important}[tuiBadgeNew][data-appearance=error]{--t-dot: var(--tui-error-fill)}[tuiBadgeNew][data-appearance=success]{--t-dot: var(--tui-success-fill)}[tuiBadgeNew][data-appearance=warning]{--t-dot: var(--tui-warning-fill)}[tuiBadgeNew][data-appearance=info]{--t-dot: var(--tui-info-fill)}[tuiBadgeNew][data-appearance=neutral]{--t-dot: var(--tui-neutral-fill)}[tuiBadgeNew][data-size=s]{--t-padding: 0 .3125rem;--t-size: 1rem;--t-icon-size: .625rem;--t-margin: -.125rem}[tuiBadgeNew][data-size=s]._dot:before{width:.25rem;height:.25rem;-webkit-margin-end:-.125rem;margin-inline-end:-.125rem}[tuiBadgeNew][data-size=m]{--t-padding: 0 .375rem;--t-size: 1.25rem;--t-icon-size: .75rem;--t-margin: -.125rem}[tuiBadgeNew][data-size=l]{--t-padding: 0 .5rem;--t-size: var(--tui-height-xs);--t-margin: -.25rem}[tuiBadgeNew][data-size=l]._dot:before{-webkit-margin-end:-.25rem;margin-inline-end:-.25rem}[tuiBadgeNew][data-size=xl]{--t-margin: -.25rem;--t-padding: 0 .75rem;--t-size: var(--tui-height-s);font:var(--tui-font-text-m)}[tuiBadgeNew][data-size=xl]._dot:before{width:.5rem;height:.5rem;-webkit-margin-end:-.125rem;margin-inline-end:-.125rem}img[tuiBadge]{padding:0}tui-icon[tuiBadge]{--t-margin: 0 !important;-webkit-mask:none;mask:none;height:var(--t-size);width:var(--t-size)}tui-icon[tuiBadge][data-size=s]:after{-webkit-mask-size:.625rem;mask-size:.625rem}tui-icon[tuiBadge][data-size=m]:after{-webkit-mask-size:.75rem;mask-size:.75rem}tui-icon[tuiBadge][data-size=l]:after,tui-icon[tuiBadge][data-size=xl]:after{-webkit-mask-size:1rem;mask-size:1rem}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./badge.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-badge',
                            },
                        }]
                }] });

        var TUI_BADGE_DEFAULT_OPTIONS = {
            size: 'l',
            appearance: 'default',
            dot: false,
        };
        var TUI_BADGE_OPTIONS = i1.tuiCreateToken(TUI_BADGE_DEFAULT_OPTIONS);
        function tuiBadgeOptionsProvider(options) {
            return i1.tuiProvideOptions(TUI_BADGE_OPTIONS, options, TUI_BADGE_DEFAULT_OPTIONS);
        }

        var TuiBadgeDirective = /** @class */ (function () {
            function TuiBadgeDirective(options, directiveStyles) {
                this.options = options;
                this.size = this.options.size;
                this.appearance = this.options.appearance;
                this.dot = this.options.dot;
                directiveStyles.addComponent(TuiBadgeComponent);
            }
            return TuiBadgeDirective;
        }());
        TuiBadgeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeDirective, deps: [{ token: TUI_BADGE_OPTIONS }, { token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiBadgeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgeDirective, selector: "tui-badge,[tuiBadge]", inputs: { size: "size", appearance: "appearance", dot: "dot" }, host: { attributes: { "tuiBadgeNew": "", "tuiAppearance": "" }, properties: { "class._dot": "dot", "attr.data-appearance": "appearance", "attr.data-size": "size" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: 'tui-badge,[tuiBadge]',
                            host: {
                                tuiBadgeNew: '',
                                tuiAppearance: '',
                                '[class._dot]': 'dot',
                                '[attr.data-appearance]': 'appearance',
                                '[attr.data-size]': 'size',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: undefined, decorators: [{
                                type: i0.Inject,
                                args: [TUI_BADGE_OPTIONS]
                            }] }, { type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { size: [{
                        type: i0.Input
                    }], appearance: [{
                        type: i0.Input
                    }], dot: [{
                        type: i0.Input
                    }] } });

        var TuiBadgeModule = /** @class */ (function () {
            function TuiBadgeModule() {
            }
            return TuiBadgeModule;
        }());
        TuiBadgeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiBadgeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeModule, declarations: [TuiBadgeComponent, TuiBadgeDirective], imports: [icons.TuiIconsModule], exports: [TuiBadgeDirective, icons.TuiIconsDirective] });
        TuiBadgeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeModule, imports: [[icons.TuiIconsModule]] });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgeModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            imports: [icons.TuiIconsModule],
                            declarations: [TuiBadgeComponent, TuiBadgeDirective],
                            exports: [TuiBadgeDirective, icons.TuiIconsDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TUI_BADGE_DEFAULT_OPTIONS = TUI_BADGE_DEFAULT_OPTIONS;
        exports.TUI_BADGE_OPTIONS = TUI_BADGE_OPTIONS;
        exports.TuiBadgeComponent = TuiBadgeComponent;
        exports.TuiBadgeDirective = TuiBadgeDirective;
        exports.TuiBadgeModule = TuiBadgeModule;
        exports.tuiBadgeOptionsProvider = tuiBadgeOptionsProvider;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
