// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-card.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/card', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.card = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiCardComponent = /** @class */ (function () {
            function TuiCardComponent() {
            }
            return TuiCardComponent;
        }());
        TuiCardComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiCardComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardComponent, selector: "ng-component", host: { classAttribute: "tui-card" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiCardMedium],[tuiCardLarge]{-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:none;font-size:inherit;line-height:inherit;position:relative;display:flex;align-items:flex-start;flex-shrink:0;text-decoration:none;overscroll-behavior:contain}[tuiCardMedium][tuiTitle],[tuiCardLarge][tuiTitle],[tuiCardMedium] [tuiTitle],[tuiCardLarge] [tuiTitle]{font-weight:bold}[tuiCardMedium] [tuiTitle],[tuiCardLarge] [tuiTitle],[tuiCardMedium] [tuiSubtitle],[tuiCardLarge] [tuiSubtitle]{max-width:100%}[tuiCardMedium]>*,[tuiCardLarge]>*{scrollbar-width:none;-ms-overflow-style:none}[tuiCardMedium]>*::-webkit-scrollbar,[tuiCardLarge]>*::-webkit-scrollbar,[tuiCardMedium]>*::-webkit-scrollbar-thumb,[tuiCardLarge]>*::-webkit-scrollbar-thumb{display:none}[tuiCardMedium]{width:8.75rem;height:8.75rem;flex-direction:column;justify-content:space-between;padding:.75rem;border-radius:var(--tui-radius-l);box-sizing:border-box}[tuiCardMedium][tuiTitle]{padding:.625rem .75rem}[tuiCardMedium] [tuiTitle]{margin:-.125rem 0}[tuiCardMedium] [tuiSubtitle]{color:var(--tui-text-01)}[tuiCardLarge][data-space]{--t-gap: .75rem;--t-radius: var(--tui-radius-l);--t-comp: -.25rem;--t-padding: .75rem;--t-dim: calc(var(--t-padding) + var(--t-comp));font:var(--tui-font-text-m);padding:var(--t-padding);border-radius:var(--t-radius);box-sizing:border-box}[tuiCardLarge][data-space][data-space=normal]{--t-radius: 1.5rem;--t-padding: 1.5rem;--t-gap: 1.5rem}[tuiCardLarge][data-space][data-space=compact]{--t-radius: 1rem;--t-padding: 1.25rem;--t-gap: 1.25rem}[tuiCardLarge][data-space]:not([tuiCell],[tuiHeader]){flex-direction:column;grid-gap:var(--t-gap);gap:var(--t-gap);align-items:stretch}[tuiCardLarge][data-space]:not([tuiCell],[tuiHeader])>:last-child:not([tuiCell]){margin-top:auto}[tuiCardLarge][data-space][tuiHeader]{padding:var(--t-dim) var(--t-dim) var(--t-dim) var(--t-padding)}[tuiCardLarge][data-space][tuiHeader] [tuiLink]:last-child{margin-right:calc(-1 * var(--t-comp))}[tuiCardLarge][data-space]>[tuiHeader]{margin:var(--t-comp) var(--t-comp) calc(1.5 * var(--t-comp)) 0}[tuiCardLarge][data-space]>[tuiHeader] [tuiLink]:last-child{margin-right:calc(-1 * var(--t-comp))}[tuiCardLarge][data-space] [tuiCell]{width:-webkit-fill-available;width:stretch;padding:.5rem;margin:-.5rem -.5rem -.75rem;border-radius:var(--tui-radius-l)}[tuiCardLarge][data-space]>[tuiLink]:last-child{display:block;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;margin-bottom:var(--t-comp)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./card.styles.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-card',
                            },
                        }]
                }] });

        var TuiCardMediumDirective = /** @class */ (function () {
            function TuiCardMediumDirective(directiveStyles) {
                directiveStyles.addComponent(TuiCardComponent);
            }
            return TuiCardMediumDirective;
        }());
        TuiCardMediumDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardMediumDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiCardMediumDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardMediumDirective, selector: "[tuiCardMedium]", host: { attributes: { "tuiCardMedium": "" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardMediumDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiCardMedium]',
                            host: {
                                tuiCardMedium: '',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            } });
        var TuiCardLargeDirective = /** @class */ (function () {
            function TuiCardLargeDirective(directiveStyles) {
                this.space = 'normal';
                directiveStyles.addComponent(TuiCardComponent);
            }
            return TuiCardLargeDirective;
        }());
        TuiCardLargeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardLargeDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiCardLargeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardLargeDirective, selector: "[tuiCardLarge]", inputs: { space: ["tuiCardLarge", "space"] }, host: { attributes: { "tuiCardLarge": "" }, properties: { "attr.data-space": "space || \"normal\"" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardLargeDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiCardLarge]',
                            host: {
                                tuiCardLarge: '',
                                '[attr.data-space]': 'space || "normal"',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { space: [{
                        type: i0.Input,
                        args: ['tuiCardLarge']
                    }] } });

        var TuiCardModule = /** @class */ (function () {
            function TuiCardModule() {
            }
            return TuiCardModule;
        }());
        TuiCardModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiCardModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardModule, declarations: [TuiCardComponent, TuiCardLargeDirective, TuiCardMediumDirective], exports: [TuiCardLargeDirective, TuiCardMediumDirective] });
        TuiCardModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCardModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiCardComponent, TuiCardLargeDirective, TuiCardMediumDirective],
                            exports: [TuiCardLargeDirective, TuiCardMediumDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiCardComponent = TuiCardComponent;
        exports.TuiCardLargeDirective = TuiCardLargeDirective;
        exports.TuiCardMediumDirective = TuiCardMediumDirective;
        exports.TuiCardModule = TuiCardModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
