// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-block-details.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/block-details', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["block-details"] = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiBlockDetailsComponent = /** @class */ (function () {
            function TuiBlockDetailsComponent() {
            }
            return TuiBlockDetailsComponent;
        }());
        TuiBlockDetailsComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiBlockDetailsComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockDetailsComponent, selector: "ng-component", host: { classAttribute: "tui-block-details-styles" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiBlockDetails]{display:flex;flex-direction:column;align-items:center;font:var(--tui-font-heading-3);min-height:2rem}[tuiBlockDetails] [tuiTitle]{text-align:center;font:var(--tui-font-text-l);grid-gap:0;gap:0}[tuiBlockDetails] [tuiSubtitle]{font:var(--tui-font-text-m);color:var(--tui-text-02);margin-top:.25rem}:host-context(tui-root._mobile) [tuiBlockDetails] [tuiSubtitle]{font:var(--tui-font-text-s)}[tuiBlockDetails] [tuiComment]{margin-top:1.0625rem}[tuiBlockDetails] [tuiAccessory]{display:flex;flex-gap:.5rem;margin-top:1.25rem}[tuiBlockDetails] [tuiAccessories],[tuiBlockDetails] tui-badge:only-of-type{margin-top:1.25rem}[tuiBlockDetails] [tuiAccessories]{display:flex;grid-gap:.5rem;gap:.5rem}[tuiBlockDetails] tui-avatar,[tuiBlockDetails] [tuiTitle]{margin-bottom:1rem}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./block-details.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-block-details-styles',
                            },
                        }]
                }] });

        var TuiBlockDetailsDirective = /** @class */ (function () {
            function TuiBlockDetailsDirective(directiveStyles) {
                directiveStyles.addComponent(TuiBlockDetailsComponent);
            }
            return TuiBlockDetailsDirective;
        }());
        TuiBlockDetailsDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiBlockDetailsDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockDetailsDirective, selector: "[tuiBlockDetails]", ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiBlockDetails]',
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            } });

        var TuiBlockDetailsModule = /** @class */ (function () {
            function TuiBlockDetailsModule() {
            }
            return TuiBlockDetailsModule;
        }());
        TuiBlockDetailsModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiBlockDetailsModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsModule, declarations: [TuiBlockDetailsDirective, TuiBlockDetailsComponent], exports: [TuiBlockDetailsDirective] });
        TuiBlockDetailsModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockDetailsModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiBlockDetailsDirective, TuiBlockDetailsComponent],
                            exports: [TuiBlockDetailsDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiBlockDetailsComponent = TuiBlockDetailsComponent;
        exports.TuiBlockDetailsDirective = TuiBlockDetailsDirective;
        exports.TuiBlockDetailsModule = TuiBlockDetailsModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
