// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-avatar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/experimental/tokens'), require('@angular/common'), require('@taiga-ui/kit'), require('@taiga-ui/experimental/components/icon')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/avatar', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/experimental/tokens', '@angular/common', '@taiga-ui/kit', '@taiga-ui/experimental/components/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components.avatar = {}), global.ng.core, global.cdk, global["taiga-ui"].experimental.tokens, global.ng.common, global.i2, global["taiga-ui"].experimental.components.icon));
})(this, (function (exports, i0, cdk, tokens, i1, i2, icon) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TUI_AVATAR_DEFAULT_OPTIONS = {
        appearance: '',
        round: true,
        size: 'l',
    };
    var TUI_AVATAR_OPTIONS = cdk.tuiCreateToken(TUI_AVATAR_DEFAULT_OPTIONS);
    function tuiAvatarOptionsProvider(options) {
        return cdk.tuiProvideOptions(TUI_AVATAR_OPTIONS, options, TUI_AVATAR_DEFAULT_OPTIONS);
    }

    var TuiAvatarComponent = /** @class */ (function () {
        function TuiAvatarComponent(options, resolver) {
            this.options = options;
            this.resolver = resolver;
            this.size = this.options.size;
            this.round = this.options.round;
            this.src = null;
            this.appearance = this.options.appearance;
        }
        Object.defineProperty(TuiAvatarComponent.prototype, "safeSrc", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.src) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiAvatarComponent.prototype, "value", {
            get: function () {
                return this.src || '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiAvatarComponent.prototype, "type", {
            get: function () {
                if (this.value && !cdk.tuiIsString(this.value)) {
                    return 'img';
                }
                if (this.value.startsWith('tuiIcon') || this.value.endsWith('.svg')) {
                    return 'icon';
                }
                if (this.value.length > 0 && this.value.length < 3) {
                    return 'text';
                }
                return this.value.length ? 'img' : 'content';
            },
            enumerable: false,
            configurable: true
        });
        return TuiAvatarComponent;
    }());
    TuiAvatarComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarComponent, deps: [{ token: TUI_AVATAR_OPTIONS }, { token: tokens.TUI_ICON_RESOLVER }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiAvatarComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarComponent, selector: "tui-avatar", inputs: { size: "size", round: "round", src: "src", appearance: "appearance" }, host: { attributes: { "tuiAppearance": "" }, properties: { "attr.data-appearance": "appearance", "attr.data-size": "size", "attr.data-type": "type", "style.--t-mask": "\"url(\" + resolver(safeSrc) + \")\"", "class._round": "round" } }, ngImport: i0__namespace, template: "<ng-container [ngSwitch]=\"type\">\n    <img\n        *ngSwitchCase=\"'img'\"\n        alt=\"\"\n        loading=\"lazy\"\n        [src]=\"value\"\n    />\n    <ng-container *ngSwitchCase=\"'text'\">{{ value }}</ng-container>\n    <ng-container *ngSwitchDefault>\n        <ng-content></ng-content>\n    </ng-container>\n</ng-container>\n", styles: [":host{--t-size: 3.5rem;--t-radius: .75rem;position:relative;display:inline-flex;flex-shrink:0;width:var(--t-size);height:var(--t-size);align-items:center;justify-content:center;overflow:hidden;white-space:nowrap;border-radius:var(--t-radius);background:var(--tui-secondary);color:var(--tui-text-02);vertical-align:middle;box-sizing:border-box;padding:.25rem;opacity:.999}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;background:currentColor;-webkit-mask:var(--t-mask) no-repeat center / 60%;mask:var(--t-mask) no-repeat center / 60%}:host[data-size=xs]{--t-size: var(--tui-height-xs);--t-radius: .5rem;font:var(--tui-font-text-xs);font-weight:bold}:host[data-size=xs][data-type=content]{font:var(--tui-font-text-m);font-size:.5625rem}:host[data-size=s]{--t-size: var(--tui-height-s);--t-radius: .5rem;font:var(--tui-font-text-s);font-weight:bold}:host[data-size=s][data-type=content]{font:var(--tui-font-text-xs);font-weight:bold}:host[data-size=m]{--t-size: calc(var(--tui-height-m) - .25rem);--t-radius: .75rem;font:var(--tui-font-text-l);font-weight:bold}:host[data-size=m][data-type=content]{font:var(--tui-font-text-m);font-weight:bold}:host[data-size=l]{--t-size: var(--tui-height-l);--t-radius: .75rem;font:var(--tui-font-heading-5)}:host[data-size=l][data-type=content]{font:var(--tui-font-text-l);font-weight:bold}:host[data-size=xl]{--t-size: 5rem;--t-radius: .75rem;font:var(--tui-font-heading-3)}:host[data-size=xl][data-type=content]{font:var(--tui-font-heading-4)}:host[data-size=xxl]{--t-size: 6rem;--t-radius: 1rem;font:var(--tui-font-heading-3)}:host[data-size=xxl][data-type=content]{font:var(--tui-font-heading-3)}:host[data-size=xxxl]{--t-size: 8rem;--t-radius: 1.25rem;font:var(--tui-font-heading-2)}:host[data-size=xxxl][data-type=content]{font:var(--tui-font-heading-3)}:host[data-type=img]{background:transparent}:host[data-type=icon]:before{content:\"\"}:host._round{--t-radius: calc(var(--t-size) / 2)}:host ::ng-deep img,:host ::ng-deep picture,:host ::ng-deep video{position:absolute;top:0;left:0;width:100%;height:100%;object-fit:cover}\n"], directives: [{ type: i1__namespace.NgSwitch, selector: "[ngSwitch]", inputs: ["ngSwitch"] }, { type: i1__namespace.NgSwitchCase, selector: "[ngSwitchCase]", inputs: ["ngSwitchCase"] }, { type: i2__namespace.TuiLazyLoadingDirective, selector: "img[loading=\"lazy\"]", inputs: ["src"] }, { type: i1__namespace.NgSwitchDefault, selector: "[ngSwitchDefault]" }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-avatar',
                        templateUrl: './avatar.template.html',
                        styleUrls: ['./avatar.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            tuiAppearance: '',
                            '[attr.data-appearance]': 'appearance',
                            '[attr.data-size]': 'size',
                            '[attr.data-type]': 'type',
                            '[style.--t-mask]': '"url(" + resolver(safeSrc) + ")"',
                            '[class._round]': 'round',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_AVATAR_OPTIONS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ICON_RESOLVER]
                        }] }];
        }, propDecorators: { size: [{
                    type: i0.Input
                }], round: [{
                    type: i0.Input
                }], src: [{
                    type: i0.Input
                }], appearance: [{
                    type: i0.Input
                }] } });

    var TuiAvatarModule = /** @class */ (function () {
        function TuiAvatarModule() {
        }
        return TuiAvatarModule;
    }());
    TuiAvatarModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAvatarModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarModule, declarations: [TuiAvatarComponent], imports: [i1.CommonModule, icon.TuiIconModule, i2.TuiLazyLoadingModule], exports: [TuiAvatarComponent] });
    TuiAvatarModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarModule, imports: [[i1.CommonModule, icon.TuiIconModule, i2.TuiLazyLoadingModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule, icon.TuiIconModule, i2.TuiLazyLoadingModule],
                        declarations: [TuiAvatarComponent],
                        exports: [TuiAvatarComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_AVATAR_DEFAULT_OPTIONS = TUI_AVATAR_DEFAULT_OPTIONS;
    exports.TUI_AVATAR_OPTIONS = TUI_AVATAR_OPTIONS;
    exports.TuiAvatarComponent = TuiAvatarComponent;
    exports.TuiAvatarModule = TuiAvatarModule;
    exports.tuiAvatarOptionsProvider = tuiAvatarOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
