// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-comment.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/comment', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives.comment = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiCommentComponent = /** @class */ (function () {
            function TuiCommentComponent() {
            }
            return TuiCommentComponent;
        }());
        TuiCommentComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiCommentComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCommentComponent, selector: "ng-component", host: { classAttribute: "tui-comment-styles" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiComment]{position:relative;display:inline-flex;font:var(--tui-font-text-m);color:var(--tui-text-01-night);padding:.5rem .75rem;border-radius:1rem;line-height:1rem;background:var(--tui-accent);align-items:center;justify-content:center}[tuiComment]:before{content:\"\";position:absolute;bottom:100%;left:50%;width:1.5625rem;height:.5625rem;background:inherit;transform:translate(-50%);-webkit-mask:url(\"data:image/svg+xml,%3Csvg width='25' height='9' viewBox='0 0 25 9' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M5.23309 6.67813L7.56191 3.83178C9.4546 1.5185 10.4009 0.361853 11.5998 0.10488C12.0834 0.00123882 12.5834 0.00123882 13.0669 0.10488C14.2658 0.361853 15.2121 1.5185 17.1048 3.83178L19.4337 6.67813C20.636 8.14771 22.4346 9 24.3334 9H0.333374C2.23217 9 4.0307 8.14772 5.23309 6.67813Z' fill='black'/%3E%3C/svg%3E%0A\") no-repeat;mask:url(\"data:image/svg+xml,%3Csvg width='25' height='9' viewBox='0 0 25 9' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M5.23309 6.67813L7.56191 3.83178C9.4546 1.5185 10.4009 0.361853 11.5998 0.10488C12.0834 0.00123882 12.5834 0.00123882 13.0669 0.10488C14.2658 0.361853 15.2121 1.5185 17.1048 3.83178L19.4337 6.67813C20.636 8.14771 22.4346 9 24.3334 9H0.333374C2.23217 9 4.0307 8.14772 5.23309 6.67813Z' fill='black'/%3E%3C/svg%3E%0A\") no-repeat}[tuiComment][data-direction=bottom]:before{top:100%;bottom:auto;transform:translate(-50%) rotate(180deg)}[tuiComment][data-direction=left]:before{left:auto;top:50%;right:100%;transform:translate(.785rem,-50%) rotate(-90deg)}[tuiComment][data-direction=right]:before{top:50%;left:100%;transform:translate(-.785rem,-50%) rotate(90deg)}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./comment.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-comment-styles',
                            },
                        }]
                }] });

        var TuiCommentDirective = /** @class */ (function () {
            function TuiCommentDirective(directiveStyles) {
                this.tuiComment = 'top';
                directiveStyles.addComponent(TuiCommentComponent);
            }
            return TuiCommentDirective;
        }());
        TuiCommentDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiCommentDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCommentDirective, selector: "[tuiComment]", inputs: { tuiComment: "tuiComment" }, host: { properties: { "attr.data-direction": "tuiComment" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiComment]',
                            host: {
                                '[attr.data-direction]': 'tuiComment',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { tuiComment: [{
                        type: i0.Input
                    }] } });

        var TuiCommentModule = /** @class */ (function () {
            function TuiCommentModule() {
            }
            return TuiCommentModule;
        }());
        TuiCommentModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiCommentModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentModule, declarations: [TuiCommentDirective, TuiCommentComponent], exports: [TuiCommentDirective] });
        TuiCommentModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCommentModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiCommentDirective, TuiCommentComponent],
                            exports: [TuiCommentDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiCommentComponent = TuiCommentComponent;
        exports.TuiCommentDirective = TuiCommentDirective;
        exports.TuiCommentModule = TuiCommentModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
