// @taiga-ui/kit/fesm2015/taiga-ui-kit-pipes-iso-to-country-code.js
import * as i0 from '@angular/core';
import { Pipe, Inject, NgModule } from '@angular/core';
import { TUI_COUNTRIES_MASKS } from '@taiga-ui/kit/tokens';
import { tuiIsoToCountryCode } from '@taiga-ui/kit/utils';

class TuiIsoToCountryCodePipe {
    constructor(countriesMasks) {
        this.countriesMasks = countriesMasks;
    }
    transform(isoCode) {
        return tuiIsoToCountryCode(this.countriesMasks, isoCode);
    }
}
TuiIsoToCountryCodePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, deps: [{ token: TUI_COUNTRIES_MASKS }], target: i0.ɵɵFactoryTarget.Pipe });
TuiIsoToCountryCodePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, name: "tuiIsoToCountryCode" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiIsoToCountryCode',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_COUNTRIES_MASKS]
                }] }]; } });

class TuiIsoToCountryCodeModule {
}
TuiIsoToCountryCodeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiIsoToCountryCodeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodeModule, declarations: [TuiIsoToCountryCodePipe], exports: [TuiIsoToCountryCodePipe] });
TuiIsoToCountryCodeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiIsoToCountryCodePipe],
                    exports: [TuiIsoToCountryCodePipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiIsoToCountryCodeModule, TuiIsoToCountryCodePipe };
