// @taiga-ui/kit/esm2015/pipes/iso-to-country-code/iso-to-country-code.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TUI_COUNTRIES_MASKS } from '@taiga-ui/kit/tokens';
import { tuiIsoToCountryCode } from '@taiga-ui/kit/utils';
import * as i0 from "@angular/core";
export class TuiIsoToCountryCodePipe {
    constructor(countriesMasks) {
        this.countriesMasks = countriesMasks;
    }
    transform(isoCode) {
        return tuiIsoToCountryCode(this.countriesMasks, isoCode);
    }
}
TuiIsoToCountryCodePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, deps: [{ token: TUI_COUNTRIES_MASKS }], target: i0.ɵɵFactoryTarget.Pipe });
TuiIsoToCountryCodePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, name: "tuiIsoToCountryCode" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIsoToCountryCodePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiIsoToCountryCode',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_COUNTRIES_MASKS]
                }] }]; } });
