// @taiga-ui/cdk/schematics/ng-update/v4/steps/templates/toggles/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.closeStartTag = exports.removeClosingTag = exports.replaceSizeAttr = exports.replaceOpenTag = void 0;
const sizeMap = {
    l: 'm',
    m: 's',
};
function replaceOpenTag(sourceCodeLocation, recorder, templateOffset, { tag, directive, type }) {
    const { startTag } = sourceCodeLocation;
    const { startOffset, startCol } = startTag;
    const spaces = ' '.repeat(startCol + 3);
    recorder.remove(templateOffset + startOffset, `<${tag}`.length);
    recorder.insertRight(templateOffset + startOffset, `<input\n${spaces}${directive}\n${spaces}type="${type}"`);
}
exports.replaceOpenTag = replaceOpenTag;
function replaceSizeAttr(attrs, sourceCodeLocation, recorder, templateOffset) {
    var _a;
    const sizeAttr = attrs.find(attr => attr.name === 'size');
    if (sizeAttr) {
        const { startOffset, endOffset } = ((_a = sourceCodeLocation.attrs) === null || _a === void 0 ? void 0 : _a[sizeAttr.name]) || {
            startOffset: 0,
            endOffset: 0,
        };
        recorder.remove(templateOffset + startOffset, endOffset - startOffset);
        recorder.insertRight(templateOffset + startOffset, `size="${sizeMap[sizeAttr.value] || sizeAttr.value}"`);
    }
}
exports.replaceSizeAttr = replaceSizeAttr;
function removeClosingTag(sourceCodeLocation, recorder, templateOffset) {
    const { endTag } = sourceCodeLocation;
    const { startOffset, endOffset } = endTag;
    const from = templateOffset + startOffset;
    const to = endOffset - startOffset;
    recorder.remove(from, to);
}
exports.removeClosingTag = removeClosingTag;
function closeStartTag({ startTag }, recorder, templateOffset) {
    recorder.insertRight(templateOffset + startTag.endOffset - 1, '/');
}
exports.closeStartTag = closeStartTag;
