// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-click-outside.js
import { DOCUMENT } from '@angular/common';
import * as i0 from '@angular/core';
import { NgZone, ElementRef, Directive, Inject, Output, NgModule } from '@angular/core';
import { tuiZoneOptimized } from '@taiga-ui/cdk/observables';
import { tuiGetActualTarget, tuiContainsOrAfter } from '@taiga-ui/cdk/utils';
import { fromEvent } from 'rxjs';
import { map, filter } from 'rxjs/operators';

class TuiClickOutsideDirective {
    constructor(zone, doc, el) {
        this.zone = zone;
        this.doc = doc;
        this.el = el;
        this.tuiClickOutside = fromEvent(this.doc, 'mouseup').pipe(map(tuiGetActualTarget), filter(target => this.isOutside(target)), tuiZoneOptimized(this.zone));
    }
    isOutside(target) {
        return (target === this.el.nativeElement ||
            !tuiContainsOrAfter(this.el.nativeElement, target));
    }
}
TuiClickOutsideDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideDirective, deps: [{ token: NgZone }, { token: DOCUMENT }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiClickOutsideDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiClickOutsideDirective, selector: "[tuiClickOutside]", outputs: { tuiClickOutside: "tuiClickOutside" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiClickOutside]',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { tuiClickOutside: [{
                type: Output
            }] } });

class TuiClickOutsideModule {
}
TuiClickOutsideModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiClickOutsideModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideModule, declarations: [TuiClickOutsideDirective], exports: [TuiClickOutsideDirective] });
TuiClickOutsideModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiClickOutsideDirective],
                    exports: [TuiClickOutsideDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiClickOutsideDirective, TuiClickOutsideModule };
