// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-platform.js
import * as i0 from '@angular/core';
import { Directive, Input, HostBinding, NgModule } from '@angular/core';
import { TUI_PLATFORM, tuiPlatformFactory } from '@taiga-ui/cdk/tokens';

class TuiPlatformDirective {
    constructor() {
        this.tuiPlatform = '';
    }
}
TuiPlatformDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiPlatformDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPlatformDirective, selector: "[tuiPlatform]", inputs: { tuiPlatform: "tuiPlatform" }, host: { properties: { "attr.data-platform": "this.tuiPlatform" } }, providers: [
        {
            provide: TUI_PLATFORM,
            deps: [TuiPlatformDirective],
            useFactory: ({ tuiPlatform }) => tuiPlatform || tuiPlatformFactory(),
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPlatform]',
                    providers: [
                        {
                            provide: TUI_PLATFORM,
                            deps: [TuiPlatformDirective],
                            useFactory: ({ tuiPlatform }) => tuiPlatform || tuiPlatformFactory(),
                        },
                    ],
                }]
        }], propDecorators: { tuiPlatform: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-platform']
            }] } });

class TuiPlatformModule {
}
TuiPlatformModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPlatformModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformModule, declarations: [TuiPlatformDirective], exports: [TuiPlatformDirective] });
TuiPlatformModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiPlatformDirective],
                    exports: [TuiPlatformDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPlatformDirective, TuiPlatformModule };
