// @taiga-ui/addon-charts/fesm2015/taiga-ui-addon-charts-components-bar.js
import { __decorate } from 'tslib';
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, HostBinding, NgModule } from '@angular/core';
import { tuiSum, tuiPure } from '@taiga-ui/cdk';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

class TuiBarComponent {
    constructor() {
        this.value = [];
        this.size = 'm';
    }
    getHeight(value) {
        return (100 * value) / this.getSum(this.value);
    }
    getColor(index) {
        return `var(--tui-chart-${index})`;
    }
    getSum(value) {
        return tuiSum(...value);
    }
}
TuiBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBarComponent, selector: "tui-bar", inputs: { value: "value", size: "size" }, host: { properties: { "attr.data-size": "this.size" } }, ngImport: i0, template: "<div\n    *ngFor=\"let item of value; let index = index\"\n    automation-id=\"tui-bar__bar\"\n    [style.background]=\"getColor(index)\"\n    [style.height.%]=\"getHeight(item)\"\n></div>\n", styles: [":host{display:flex;min-width:.125rem;flex-direction:column-reverse;border-top-left-radius:var(--tui-radius-l);border-top-right-radius:var(--tui-radius-l);overflow:hidden}:host[data-size=l]{width:1rem;max-width:1rem;margin:0 .1875rem}:host[data-size=m]{width:.5rem;max-width:.5rem;margin:0 .1875rem}:host[data-size=s]{width:.25rem;max-width:.5rem;margin:0 .125rem}\n"], directives: [{ type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
__decorate([
    tuiPure
], TuiBarComponent.prototype, "getSum", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-bar',
                    templateUrl: './bar.template.html',
                    styleUrls: ['./bar.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { value: [{
                type: Input
            }], size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }], getSum: [] } });

class TuiBarModule {
}
TuiBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, declarations: [TuiBarComponent], imports: [CommonModule], exports: [TuiBarComponent] });
TuiBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBarComponent],
                    exports: [TuiBarComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBarComponent, TuiBarModule };
