// @taiga-ui/addon-charts/fesm2015/taiga-ui-addon-charts-components-ring-chart.js
import * as i0 from '@angular/core';
import { EventEmitter, Component, ChangeDetectionStrategy, Input, HostBinding, Output, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/addon-charts/components/pie-chart';
import { TuiPieChartModule } from '@taiga-ui/addon-charts/components/pie-chart';
import { CommonModule } from '@angular/common';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';

class TuiRingChartComponent {
    constructor() {
        this.value = [];
        this.size = 'm';
        this.activeItemIndex = NaN;
        this.activeItemIndexChange = new EventEmitter();
    }
    onActiveItemIndexChange(index) {
        this.updateActiveItemIndex(index);
    }
    updateActiveItemIndex(index) {
        if (index === this.activeItemIndex) {
            return;
        }
        this.activeItemIndex = index;
        this.activeItemIndexChange.next(index);
    }
}
TuiRingChartComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiRingChartComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRingChartComponent, selector: "tui-ring-chart", inputs: { value: "value", size: "size", activeItemIndex: "activeItemIndex" }, outputs: { activeItemIndexChange: "activeItemIndexChange" }, host: { properties: { "attr.data-size": "this.size" } }, ngImport: i0, template: "<div class=\"t-content\">\n    <div class=\"t-wrapper\">\n        <ng-content></ng-content>\n    </div>\n</div>\n\n<tui-pie-chart\n    class=\"t-chart\"\n    [activeItemIndex]=\"activeItemIndex\"\n    [masked]=\"true\"\n    [size]=\"size\"\n    [value]=\"value\"\n    (activeItemIndexChange)=\"onActiveItemIndexChange($event)\"\n></tui-pie-chart>\n\n<div class=\"t-shield\"></div>\n", styles: [":host{position:relative;display:block;width:var(--t-size);height:var(--t-size);flex-shrink:0}:host[data-size=xs]{--t-size: 3rem}:host[data-size=s]{--t-size: 4rem}:host[data-size=m]{--t-size: 9rem}:host[data-size=l]{--t-size: 11rem}:host[data-size=xl]{--t-size: 16rem}.t-content{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;font:var(--tui-font-text-m);max-height:100%;flex-direction:column;justify-content:center;text-align:center;padding:1.5rem;border-radius:100%;box-sizing:border-box;overflow:hidden;word-break:break-word;white-space:pre-wrap;color:var(--tui-text-02)}:host[data-size=m] .t-content{font:var(--tui-font-text-xs)}.t-wrapper:first-line{color:var(--tui-text-01)}:host[data-size=l] .t-wrapper:first-line{font:var(--tui-font-text-l);font-weight:bold}:host[data-size=m] .t-wrapper:first-line{font:var(--tui-font-text-m);font-weight:bold}.t-shield{position:absolute;top:25%;left:25%;right:25%;bottom:25%;border-radius:100%}.t-chart{width:100%;height:100%}\n"], components: [{ type: i1.TuiPieChartComponent, selector: "tui-pie-chart", inputs: ["value", "size", "masked", "activeItemIndex"], outputs: ["activeItemIndexChange"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-ring-chart',
                    templateUrl: './ring-chart.template.html',
                    styleUrls: ['./ring-chart.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { value: [{
                type: Input
            }], size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }], activeItemIndex: [{
                type: Input
            }], activeItemIndexChange: [{
                type: Output
            }] } });

class TuiRingChartModule {
}
TuiRingChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRingChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, declarations: [TuiRingChartComponent], imports: [CommonModule, PolymorpheusModule, TuiPieChartModule], exports: [TuiRingChartComponent] });
TuiRingChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, imports: [[CommonModule, PolymorpheusModule, TuiPieChartModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiPieChartModule],
                    declarations: [TuiRingChartComponent],
                    exports: [TuiRingChartComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiRingChartComponent, TuiRingChartModule };
