// @taiga-ui/addon-charts/fesm2015/taiga-ui-addon-charts-components-legend-item.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, HostBinding, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/core';
import { TuiButtonModule } from '@taiga-ui/core';
import { CommonModule } from '@angular/common';

class TuiLegendItemComponent {
    constructor() {
        this.active = false;
        this.color = '';
        this.text = '';
        this.size = 'm';
        this.disabled = false;
    }
    get computedColor() {
        return this.color === 'var(--tui-primary)' ? null : this.color;
    }
}
TuiLegendItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiLegendItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLegendItemComponent, selector: "tui-legend-item", inputs: { active: "active", color: "color", text: "text", size: "size", disabled: "disabled" }, host: { properties: { "attr.data-size": "this.size", "class._disabled": "this.disabled", "style.--tui-primary": "this.computedColor" } }, ngImport: i0, template: "<button\n    appearance=\"whiteblock\"\n    size=\"s\"\n    tuiButton\n    type=\"button\"\n    class=\"t-button\"\n    [pseudoHover]=\"active || null\"\n>\n    <div class=\"t-wrapper\">\n        <ng-content select=\"tui-primitive-checkbox\"></ng-content>\n        <span\n            class=\"t-dot\"\n            [style.backgroundColor]=\"color || 'var(--tui-primary)'\"\n        ></span>\n        <span class=\"t-text\">{{ text }}</span>\n        <ng-content></ng-content>\n    </div>\n</button>\n", styles: [":host{display:inline-block}.t-button{height:2rem;font-weight:bold}:host[data-size=m] .t-button{height:2.25rem;font:var(--tui-font-text-m);font-weight:bold}.t-wrapper{display:flex;align-items:center;pointer-events:none}:host._disabled .t-wrapper{color:var(--tui-text-03)}.t-dot{width:.5rem;height:.5rem;border-radius:100%;flex-shrink:0}:host._disabled .t-dot{background:var(--tui-base-04)}.t-dot:not(:first-child){display:none}.t-text{margin:0 .5rem;font-weight:normal}\n"], components: [{ type: i1.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-legend-item',
                    templateUrl: './legend-item.template.html',
                    styleUrls: ['./legend-item.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { active: [{
                type: Input
            }], color: [{
                type: Input
            }], text: [{
                type: Input
            }], size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }], disabled: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class._disabled']
            }], computedColor: [{
                type: HostBinding,
                args: ['style.--tui-primary']
            }] } });

class TuiLegendItemModule {
}
TuiLegendItemModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLegendItemModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, declarations: [TuiLegendItemComponent], imports: [CommonModule, TuiButtonModule], exports: [TuiLegendItemComponent] });
TuiLegendItemModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, imports: [[CommonModule, TuiButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiButtonModule],
                    declarations: [TuiLegendItemComponent],
                    exports: [TuiLegendItemComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiLegendItemComponent, TuiLegendItemModule };
