// @taiga-ui/addon-charts/esm2015/components/legend-item/legend-item.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule } from '@taiga-ui/core';
import { TuiLegendItemComponent } from './legend-item.component';
import * as i0 from "@angular/core";
export class TuiLegendItemModule {
}
TuiLegendItemModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLegendItemModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, declarations: [TuiLegendItemComponent], imports: [CommonModule, TuiButtonModule], exports: [TuiLegendItemComponent] });
TuiLegendItemModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, imports: [[CommonModule, TuiButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLegendItemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiButtonModule],
                    declarations: [TuiLegendItemComponent],
                    exports: [TuiLegendItemComponent],
                }]
        }] });
