// @taiga-ui/core/esm2015/directives/dropdown/dropdown-hover.directive.js
import { Directive, Inject, Input, Optional, SkipSelf } from '@angular/core';
import { TuiHoveredService } from '@taiga-ui/cdk';
import { tuiAsDriver, TuiDriver } from '@taiga-ui/core/abstract';
import { merge, of, Subject } from 'rxjs';
import { delay, share, switchMap, tap } from 'rxjs/operators';
import { TUI_DROPDOWN_HOVER_OPTIONS, } from './dropdown-hover.options';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
export class TuiDropdownHoverDirective extends TuiDriver {
    constructor(hovered$, options, parentHover) {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.hovered$ = hovered$;
        this.options = options;
        this.parentHover = parentHover;
        this.toggle$ = new Subject();
        this.stream$ = merge(this.toggle$, this.hovered$).pipe(switchMap(visible => of(visible).pipe(delay(visible ? this.showDelay : this.hideDelay))), tap(visible => {
            this.hovered = visible;
        }), share());
        this.showDelay = this.options.showDelay;
        this.hideDelay = this.options.hideDelay;
        this.hovered = false;
        this.type = 'dropdown';
    }
    toggle(visible) {
        var _a;
        (_a = this.parentHover) === null || _a === void 0 ? void 0 : _a.toggle(visible);
        this.toggle$.next(visible);
    }
}
TuiDropdownHoverDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHoverDirective, deps: [{ token: TuiHoveredService }, { token: TUI_DROPDOWN_HOVER_OPTIONS }, { token: TuiDropdownHoverDirective, optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownHoverDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownHoverDirective, selector: "[tuiDropdownHover]:not(ng-container)", inputs: { showDelay: ["tuiDropdownShowDelay", "showDelay"], hideDelay: ["tuiDropdownHideDelay", "hideDelay"] }, providers: [tuiAsDriver(TuiDropdownHoverDirective), TuiHoveredService], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHoverDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdownHover]:not(ng-container)',
                    providers: [tuiAsDriver(TuiDropdownHoverDirective), TuiHoveredService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TuiHoveredService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DROPDOWN_HOVER_OPTIONS]
                }] }, { type: TuiDropdownHoverDirective, decorators: [{
                    type: SkipSelf
                }, {
                    type: Optional
                }, {
                    type: Inject,
                    args: [TuiDropdownHoverDirective]
                }] }]; }, propDecorators: { showDelay: [{
                type: Input,
                args: ['tuiDropdownShowDelay']
            }], hideDelay: [{
                type: Input,
                args: ['tuiDropdownHideDelay']
            }] } });
