// @taiga-ui/core/esm2015/directives/dropdown/dropdown.module.js
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule, TuiHoveredModule, TuiOverscrollModule } from '@taiga-ui/cdk';
import { TuiScrollbarModule } from '@taiga-ui/core/components/scrollbar';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDropdownComponent } from './dropdown.component';
import { TuiDropdownDirective } from './dropdown.directive';
import { TuiDropdownContextDirective } from './dropdown-context.directive';
import { TuiDropdownDriverDirective } from './dropdown-driver.directive';
import { TuiDropdownHostDirective } from './dropdown-host.directive';
import { TuiDropdownHoverDirective } from './dropdown-hover.directive';
import { TuiDropdownManualDirective } from './dropdown-manual.directive';
import { TuiDropdownOpenDirective } from './dropdown-open.directive';
import { TuiDropdownOptionsDirective } from './dropdown-options.directive';
import { TuiDropdownPositionDirective } from './dropdown-position.directive';
import { TuiDropdownPositionSidedDirective } from './dropdown-position-sided.directive';
import { TuiDropdownSelectionDirective } from './dropdown-selection.directive';
import * as i0 from "@angular/core";
export class TuiDropdownModule {
}
TuiDropdownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDropdownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownModule, declarations: [TuiDropdownDirective,
        TuiDropdownComponent,
        TuiDropdownOpenDirective,
        TuiDropdownOptionsDirective,
        TuiDropdownHostDirective,
        TuiDropdownDriverDirective,
        TuiDropdownManualDirective,
        TuiDropdownHoverDirective,
        TuiDropdownContextDirective,
        TuiDropdownPositionDirective,
        TuiDropdownPositionSidedDirective,
        TuiDropdownSelectionDirective], imports: [PolymorpheusModule,
        TuiActiveZoneModule,
        TuiOverscrollModule,
        TuiScrollbarModule,
        TuiHoveredModule], exports: [TuiDropdownDirective,
        TuiDropdownComponent,
        TuiDropdownOpenDirective,
        TuiDropdownOptionsDirective,
        TuiDropdownHostDirective,
        TuiDropdownDriverDirective,
        TuiDropdownManualDirective,
        TuiDropdownHoverDirective,
        TuiDropdownContextDirective,
        TuiDropdownPositionDirective,
        TuiDropdownPositionSidedDirective,
        TuiDropdownSelectionDirective] });
TuiDropdownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownModule, imports: [[
            PolymorpheusModule,
            TuiActiveZoneModule,
            TuiOverscrollModule,
            TuiScrollbarModule,
            TuiHoveredModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        PolymorpheusModule,
                        TuiActiveZoneModule,
                        TuiOverscrollModule,
                        TuiScrollbarModule,
                        TuiHoveredModule,
                    ],
                    declarations: [
                        TuiDropdownDirective,
                        TuiDropdownComponent,
                        TuiDropdownOpenDirective,
                        TuiDropdownOptionsDirective,
                        TuiDropdownHostDirective,
                        TuiDropdownDriverDirective,
                        TuiDropdownManualDirective,
                        TuiDropdownHoverDirective,
                        TuiDropdownContextDirective,
                        TuiDropdownPositionDirective,
                        TuiDropdownPositionSidedDirective,
                        TuiDropdownSelectionDirective,
                    ],
                    exports: [
                        TuiDropdownDirective,
                        TuiDropdownComponent,
                        TuiDropdownOpenDirective,
                        TuiDropdownOptionsDirective,
                        TuiDropdownHostDirective,
                        TuiDropdownDriverDirective,
                        TuiDropdownManualDirective,
                        TuiDropdownHoverDirective,
                        TuiDropdownContextDirective,
                        TuiDropdownPositionDirective,
                        TuiDropdownPositionSidedDirective,
                        TuiDropdownSelectionDirective,
                    ],
                }]
        }] });
