// @taiga-ui/addon-doc/esm2015/components/example/example.module.js
import { ClipboardModule } from '@angular/cdk/clipboard';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiLoaderModule } from '@taiga-ui/core';
import { TuiTabsModule } from '@taiga-ui/kit';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDocCodeModule } from '../code/code.module';
import { TuiDocCopyModule } from '../copy/copy.module';
import { TuiDocExampleComponent } from './example.component';
import { TuiDocExampleGetTabsPipe } from './example-get-tabs.pipe';
import * as i0 from "@angular/core";
export class TuiDocExampleModule {
}
TuiDocExampleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocExampleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocExampleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocExampleModule, declarations: [TuiDocExampleComponent, TuiDocExampleGetTabsPipe], imports: [CommonModule,
        ClipboardModule,
        TuiTabsModule,
        TuiButtonModule,
        TuiDocCopyModule,
        TuiDocCodeModule,
        PolymorpheusModule,
        TuiMapperPipeModule,
        TuiLoaderModule,
        RouterModule], exports: [TuiDocExampleComponent, TuiDocExampleGetTabsPipe] });
TuiDocExampleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocExampleModule, imports: [[
            CommonModule,
            ClipboardModule,
            TuiTabsModule,
            TuiButtonModule,
            TuiDocCopyModule,
            TuiDocCodeModule,
            PolymorpheusModule,
            TuiMapperPipeModule,
            TuiLoaderModule,
            RouterModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocExampleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ClipboardModule,
                        TuiTabsModule,
                        TuiButtonModule,
                        TuiDocCopyModule,
                        TuiDocCodeModule,
                        PolymorpheusModule,
                        TuiMapperPipeModule,
                        TuiLoaderModule,
                        RouterModule,
                    ],
                    declarations: [TuiDocExampleComponent, TuiDocExampleGetTabsPipe],
                    exports: [TuiDocExampleComponent, TuiDocExampleGetTabsPipe],
                }]
        }] });
