// @taiga-ui/addon-mobile/esm2015/components/sheet/directives/sheet-top/sheet-top.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { tuiClamp } from '@taiga-ui/cdk';
import { map } from 'rxjs/operators';
import { TUI_SHEET, TUI_SHEET_SCROLL } from '../../sheet-tokens';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
// So that borders get rounded when image is visible for at least 10px
const OFFSET = 10;
export class TuiSheetTopDirective {
    constructor(scroll$, component, win) {
        this.scroll$ = scroll$;
        this.component = component;
        this.win = win;
        this.stop = 0;
        this.transform$ = this.scroll$.pipe(map(y => `translateY(${this.getY(y)}%) scaleX(-1)`));
        this.rounded$ = this.scroll$.pipe(map(y => y < this.stop + OFFSET));
        this.clickthrough$ = this.scroll$.pipe(map(y => !!Math.round(this.getY(y))));
    }
    getY(scrollTop) {
        const value = scrollTop - this.stop;
        const total = this.win.innerHeight - this.component.item.offset - this.stop;
        return this.stop && tuiClamp(100 - (value / total) * 100, 0, 100);
    }
}
TuiSheetTopDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetTopDirective, deps: [{ token: TUI_SHEET_SCROLL }, { token: TUI_SHEET }, { token: WINDOW }], target: i0.ɵɵFactoryTarget.Directive });
TuiSheetTopDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetTopDirective, selector: "[tuiSheetTop]", inputs: { stop: ["tuiSheetTop", "stop"] }, host: { listeners: { "$.style.transform": "transform$", "$.class._rounded": "rounded$", "$.class._clickthrough": "clickthrough$" }, properties: { "$.style.transform": "transform$", "$.class._rounded": "rounded$", "$.class._clickthrough": "clickthrough$" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetTopDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSheetTop]',
                    host: {
                        '[$.style.transform]': 'transform$',
                        '($.style.transform)': 'transform$',
                        '[$.class._rounded]': 'rounded$',
                        '($.class._rounded)': 'rounded$',
                        '[$.class._clickthrough]': 'clickthrough$',
                        '($.class._clickthrough)': 'clickthrough$',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_SHEET_SCROLL]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_SHEET]
                }] }, { type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }]; }, propDecorators: { stop: [{
                type: Input,
                args: ['tuiSheetTop']
            }] } });
