// @taiga-ui/experimental/esm2015/directives/card/card.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiCardComponent } from './card.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiCardMediumDirective {
    constructor(directiveStyles) {
        directiveStyles.addComponent(TuiCardComponent);
    }
}
TuiCardMediumDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardMediumDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCardMediumDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardMediumDirective, selector: "[tuiCardMedium]", host: { attributes: { "tuiCardMedium": "" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardMediumDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCardMedium]',
                    host: {
                        tuiCardMedium: '',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });
export class TuiCardLargeDirective {
    constructor(directiveStyles) {
        this.space = 'normal';
        directiveStyles.addComponent(TuiCardComponent);
    }
}
TuiCardLargeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardLargeDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCardLargeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardLargeDirective, selector: "[tuiCardLarge]", inputs: { space: ["tuiCardLarge", "space"] }, host: { attributes: { "tuiCardLarge": "" }, properties: { "attr.data-space": "space || \"normal\"" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardLargeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCardLarge]',
                    host: {
                        tuiCardLarge: '',
                        '[attr.data-space]': 'space || "normal"',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { space: [{
                type: Input,
                args: ['tuiCardLarge']
            }] } });
