// @taiga-ui/addon-preview/esm2015/components/preview/title/preview-title.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiPreviewTitleComponent {
}
TuiPreviewTitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewTitleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiPreviewTitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewTitleComponent, selector: "tui-preview-title", ngImport: i0, template: `
        <ng-content></ng-content>
    `, isInline: true, styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:inline-block;padding:.375rem .75rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewTitleComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-preview-title',
                    template: `
        <ng-content></ng-content>
    `,
                    styleUrls: ['./preview-title.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
