// @taiga-ui/addon-mobile/esm2015/components/app-bar/app-bar.component.js
import { ChangeDetectionStrategy, Component, Inject, ViewChildren, } from '@angular/core';
import { MutationObserverService } from '@ng-web-apis/mutation-observer';
import { EMPTY_QUERY, TuiResizeService } from '@taiga-ui/cdk';
import { merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { TUI_APP_BAR_PROVIDERS } from './app-bar.providers';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "rxjs";
export class TuiAppBarComponent {
    constructor(resize$, mutation$) {
        this.resize$ = resize$;
        this.mutation$ = mutation$;
        this.side = EMPTY_QUERY;
        this.width$ = merge(this.resize$, this.mutation$).pipe(map(() => {
            var _a, _b;
            return 2 *
                Math.max((_a = this.side.first) === null || _a === void 0 ? void 0 : _a.nativeElement.clientWidth, (_b = this.side.last) === null || _b === void 0 ? void 0 : _b.nativeElement.clientWidth);
        }));
    }
}
TuiAppBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, deps: [{ token: TuiResizeService }, { token: MutationObserverService }], target: i0.ɵɵFactoryTarget.Component });
TuiAppBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarComponent, selector: "tui-app-bar", providers: TUI_APP_BAR_PROVIDERS, viewQueries: [{ propertyName: "side", predicate: ["side"], descendants: true }], ngImport: i0, template: "<div\n    #side\n    class=\"t-left\"\n>\n    <ng-content select=\"[tuiSlot='left']\"></ng-content>\n</div>\n<div\n    class=\"t-content\"\n    [style.--sides.px]=\"width$ | async\"\n>\n    <ng-content></ng-content>\n</div>\n<div\n    #side\n    class=\"t-right\"\n>\n    <ng-content select=\"[tuiSlot='right']\"></ng-content>\n</div>\n", styles: [":host{--tui-padding: .625rem;position:relative;display:flex;height:3rem;align-items:center;justify-content:space-between;border:solid transparent;border-width:0 .375rem;box-sizing:border-box;font:var(--tui-font-text-m);font-weight:bold;color:var(--tui-text-01);background:var(--tui-base-01)}:host ::ng-deep [tuiButton][data-size=m]{font-weight:normal}.t-content{top:50%;left:50%;transform:translate(-50%,-50%);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;position:absolute;max-width:calc(100% - var(--sides, 0px));flex:1;padding:0 .375rem;box-sizing:border-box;text-align:center;color:var(--tui-text-01)}\n"], pipes: { "async": i1.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-app-bar',
                    templateUrl: './app-bar.template.html',
                    styleUrls: ['./app-bar.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: TUI_APP_BAR_PROVIDERS,
                }]
        }], ctorParameters: function () { return [{ type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }, { type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [MutationObserverService]
                }] }]; }, propDecorators: { side: [{
                type: ViewChildren,
                args: ['side']
            }] } });
