// @taiga-ui/addon-mobile/esm2015/directives/sidebar/sidebar.module.js
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiSidebarComponent } from './sidebar.component';
import { TuiSidebarDirective } from './sidebar.directive';
import * as i0 from "@angular/core";
export class TuiSidebarModule {
}
TuiSidebarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSidebarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarModule, declarations: [TuiSidebarDirective, TuiSidebarComponent], imports: [TuiActiveZoneModule, PolymorpheusModule], exports: [TuiSidebarDirective] });
TuiSidebarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarModule, imports: [[TuiActiveZoneModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiActiveZoneModule, PolymorpheusModule],
                    declarations: [TuiSidebarDirective, TuiSidebarComponent],
                    exports: [TuiSidebarDirective],
                }]
        }] });
