// @taiga-ui/addon-mobile/esm2015/directives/sidebar/sidebar.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Inject, } from '@angular/core';
import { TUI_ANIMATION_OPTIONS, tuiSlideIn, } from '@taiga-ui/core';
import { TuiSidebarDirective } from './sidebar.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
import * as i2 from "@tinkoff/ng-polymorpheus";
import * as i3 from "./sidebar.directive";
export class TuiSidebarComponent {
    constructor(options, directive) {
        this.options = options;
        this.directive = directive;
        this.left = Object.assign(Object.assign({}, this.options), { value: 'left' });
        this.right = Object.assign(Object.assign({}, this.options), { value: 'right' });
    }
    get animation() {
        return this.direction === 'left' ? this.left : this.right;
    }
    get directionHostClass() {
        return `t-${this.directive.direction}`;
    }
    get direction() {
        return this.directive.direction;
    }
    get content() {
        return this.directive.content;
    }
    get autoWidth() {
        return this.directive.autoWidth;
    }
    ngDoCheck() {
        this.directive.check();
    }
}
TuiSidebarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarComponent, deps: [{ token: TUI_ANIMATION_OPTIONS }, { token: TuiSidebarDirective }], target: i0.ɵɵFactoryTarget.Component });
TuiSidebarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSidebarComponent, selector: "aside[tuiSidebar]", host: { properties: { "@tuiSlideIn": "this.animation", "class": "this.directionHostClass" } }, ngImport: i0, template: "<div\n    tuiActiveZone\n    class=\"t-wrapper\"\n    [class.t-wrapper_auto-width]=\"autoWidth\"\n>\n    <ng-container *polymorpheusOutlet=\"content as text\">\n        {{ text }}\n    </ng-container>\n</div>\n", styles: [":host{position:fixed;top:0;left:0;width:100%;height:100%;display:flex}:host:before,:host:after{position:absolute;top:0;left:0;width:100%;content:\"\";height:100%;background:rgba(0,0,0,.38);animation:tuiFadeIn var(--tui-duration)}:host:before{left:-100%;width:200%}:host:after{left:100%}:host.ng-animating:before,:host.ng-animating:after{opacity:0;transition:opacity var(--tui-duration)}:host.t-right{justify-content:flex-end}:host.t-left{justify-content:flex-start}.t-wrapper{position:relative;display:flex;flex-direction:column;background:var(--tui-base-01);box-shadow:0 10rem var(--tui-base-01);width:17.25rem}.t-wrapper_auto-width{width:auto}\n"], directives: [{ type: i1.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i2.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [tuiSlideIn], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSidebarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'aside[tuiSidebar]',
                    templateUrl: './sidebar.template.html',
                    styleUrls: ['./sidebar.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiSlideIn],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATION_OPTIONS]
                }] }, { type: i3.TuiSidebarDirective, decorators: [{
                    type: Inject,
                    args: [TuiSidebarDirective]
                }] }]; }, propDecorators: { animation: [{
                type: HostBinding,
                args: ['@tuiSlideIn']
            }], directionHostClass: [{
                type: HostBinding,
                args: ['class']
            }] } });
