// @taiga-ui/experimental/esm2015/directives/button-vertical/button-vertical.component.js
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiButtonVerticalComponent {
}
TuiButtonVerticalComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiButtonVerticalComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonVerticalComponent, selector: "ng-component", host: { classAttribute: "tui-button-vertical-styles" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiButton][tuiButtonVertical]{--t-line-height: 1rem;--t-margin: 0rem !important;flex-direction:column;height:auto;padding:.75rem;grid-gap:.375rem;gap:.375rem;min-width:5rem;white-space:pre-line;font:var(--tui-font-text-s-2)}[tuiButton][tuiButtonVertical]>*{max-height:calc(var(--t-line-height) * 2)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./button-vertical.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-button-vertical-styles',
                    },
                }]
        }] });
