// @taiga-ui/addon-mobile/esm2015/components/mobile-calendar-dialog/mobile-calendar-dialog.component.js
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { ALWAYS_FALSE_HANDLER, TUI_FIRST_DAY, TUI_LAST_DAY, } from '@taiga-ui/cdk';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/addon-mobile/components/mobile-calendar";
export class TuiMobileCalendarDialogComponent {
    constructor(context) {
        this.context = context;
    }
    get single() {
        var _a;
        return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.single) === true;
    }
    get multi() {
        var _a;
        return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.multi) === true;
    }
    get min() {
        var _a;
        return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.min) || TUI_FIRST_DAY;
    }
    get max() {
        var _a;
        return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.max) || TUI_LAST_DAY;
    }
    get disabledItemHandler() {
        var _a;
        return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.disabledItemHandler) || ALWAYS_FALSE_HANDLER;
    }
}
TuiMobileCalendarDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogComponent, deps: [{ token: POLYMORPHEUS_CONTEXT }], target: i0.ɵɵFactoryTarget.Component });
TuiMobileCalendarDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileCalendarDialogComponent, selector: "tui-mobile-calendar-dialog", ngImport: i0, template: "<tui-mobile-calendar\n    [disabledItemHandler]=\"disabledItemHandler\"\n    [max]=\"max\"\n    [min]=\"min\"\n    [multi]=\"multi\"\n    [single]=\"single\"\n    (cancel)=\"context.$implicit.complete()\"\n    (confirm)=\"context.completeWith($event)\"\n></tui-mobile-calendar>\n", styles: [":host{display:block;height:100vh;margin:-1.5rem -1.5rem -2rem}\n"], components: [{ type: i1.TuiMobileCalendarComponent, selector: "tui-mobile-calendar", inputs: ["single", "multi", "min", "max", "disabledItemHandler"], outputs: ["cancel", "confirm"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-mobile-calendar-dialog',
                    templateUrl: './mobile-calendar-dialog.template.html',
                    styleUrls: ['./mobile-calendar-dialog.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [POLYMORPHEUS_CONTEXT]
                }] }]; } });
