// @taiga-ui/addon-charts/esm2015/components/arc-chart/arc-chart.module.js
import { NgModule } from '@angular/core';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiArcChartComponent } from './arc-chart.component';
import * as i0 from "@angular/core";
export class TuiArcChartModule {
}
TuiArcChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArcChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiArcChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArcChartModule, declarations: [TuiArcChartComponent], imports: [TuiRepeatTimesModule], exports: [TuiArcChartComponent] });
TuiArcChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArcChartModule, imports: [[TuiRepeatTimesModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiArcChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiRepeatTimesModule],
                    declarations: [TuiArcChartComponent],
                    exports: [TuiArcChartComponent],
                }]
        }] });
