// @taiga-ui/kit/esm2015/components/input-time/input-time.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiDataListModule, TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiSelectOptionModule } from '@taiga-ui/kit/components/select-option';
import { TuiValueAccessorModule } from '@taiga-ui/kit/directives';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputTimeComponent } from './input-time.component';
import { TuiInputTimeDirective } from './input-time.directive';
import { TuiNativeTimeComponent } from './native-time/native-time.component';
import * as i0 from "@angular/core";
export class TuiInputTimeModule {
}
TuiInputTimeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTimeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputTimeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTimeModule, declarations: [TuiInputTimeComponent, TuiNativeTimeComponent, TuiInputTimeDirective], imports: [CommonModule,
        MaskitoModule,
        TuiSelectOptionModule,
        TuiDataListModule,
        TuiWrapperModule,
        TuiHostedDropdownModule,
        TuiPrimitiveTextfieldModule,
        TuiValueAccessorModule,
        TuiTextfieldControllerModule,
        TuiSvgModule,
        PolymorpheusModule], exports: [TuiInputTimeComponent, TuiInputTimeDirective] });
TuiInputTimeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTimeModule, imports: [[
            CommonModule,
            MaskitoModule,
            TuiSelectOptionModule,
            TuiDataListModule,
            TuiWrapperModule,
            TuiHostedDropdownModule,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiTextfieldControllerModule,
            TuiSvgModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputTimeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoModule,
                        TuiSelectOptionModule,
                        TuiDataListModule,
                        TuiWrapperModule,
                        TuiHostedDropdownModule,
                        TuiPrimitiveTextfieldModule,
                        TuiValueAccessorModule,
                        TuiTextfieldControllerModule,
                        TuiSvgModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiInputTimeComponent, TuiNativeTimeComponent, TuiInputTimeDirective],
                    exports: [TuiInputTimeComponent, TuiInputTimeDirective],
                }]
        }] });
