// @taiga-ui/addon-doc/esm2015/components/main/main.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TuiAlertModule, TuiButtonModule, TuiDialogModule, TuiModeModule, TuiRootModule, TuiThemeNightModule, } from '@taiga-ui/core';
import { TuiDocHeaderModule } from '../internal/header/header.module';
import { TuiDocNavigationModule } from '../navigation/navigation.module';
import { TuiDocMainComponent } from './main.component';
import * as i0 from "@angular/core";
export class TuiDocMainModule {
}
TuiDocMainModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocMainModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocMainModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocMainModule, declarations: [TuiDocMainComponent], imports: [CommonModule,
        RouterModule,
        TuiThemeNightModule,
        TuiDocHeaderModule,
        TuiRootModule,
        TuiDialogModule,
        TuiAlertModule,
        TuiButtonModule,
        TuiModeModule,
        TuiDocNavigationModule], exports: [TuiDocMainComponent] });
TuiDocMainModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocMainModule, imports: [[
            CommonModule,
            RouterModule,
            TuiThemeNightModule,
            TuiDocHeaderModule,
            TuiRootModule,
            TuiDialogModule,
            TuiAlertModule,
            TuiButtonModule,
            TuiModeModule,
            TuiDocNavigationModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocMainModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        TuiThemeNightModule,
                        TuiDocHeaderModule,
                        TuiRootModule,
                        TuiDialogModule,
                        TuiAlertModule,
                        TuiButtonModule,
                        TuiModeModule,
                        TuiDocNavigationModule,
                    ],
                    declarations: [TuiDocMainComponent],
                    exports: [TuiDocMainComponent],
                }]
        }] });
