// @taiga-ui/kit/esm2015/pipes/filter-by-input/filter-by-input.base.js
import { __decorate } from "tslib";
import { tuiIsPresent, tuiPure, } from '@taiga-ui/cdk';
import { tuiIsFlat } from '@taiga-ui/kit/utils';
export class AbstractTuiFilterByInput {
    get query() {
        return this.accessor.nativeFocusableElement
            ? this.accessor.nativeFocusableElement.value || ''
            : '';
    }
    filter(items, matcher, stringify, query) {
        if (!items) {
            return null;
        }
        return tuiIsFlat(items)
            ? this.filterFlat(items, matcher, stringify, query)
            : this.filter2d(items, matcher, stringify, query);
    }
    filterFlat(items, matcher, stringify, query) {
        const match = this.getMatch(items, stringify, query);
        return tuiIsPresent(match) && !this.multiSelect
            ? items
            : items.filter(item => matcher(item, query, stringify));
    }
    filter2d(items, matcher, stringify, query) {
        const match = items.find(item => tuiIsPresent(this.getMatch(item, stringify, query)));
        return tuiIsPresent(match) && !this.multiSelect
            ? items
            : items.map(inner => this.filterFlat(inner, matcher, stringify, query));
    }
    getMatch(items, stringify, query) {
        return items.find(item => stringify(item).toLocaleLowerCase() === query.toLocaleLowerCase());
    }
}
__decorate([
    tuiPure
], AbstractTuiFilterByInput.prototype, "filter", null);
