// @taiga-ui/kit/esm2015/pipes/filter-by-input/filter-by-input-with.pipe.js
import { Inject, Optional, Pipe } from '@angular/core';
import { TUI_DEFAULT_MATCHER, TUI_FOCUSABLE_ITEM_ACCESSOR, } from '@taiga-ui/cdk';
import { TuiMultiSelectDirective } from '@taiga-ui/kit/components/multi-select';
import { AbstractTuiFilterByInput } from './filter-by-input.base';
import * as i0 from "@angular/core";
/**
 * @deprecated Use {@link TuiFilterByInputPipe} instead
 */
export class TuiFilterByInputWithPipe extends AbstractTuiFilterByInput {
    constructor(accessor, multiSelect) {
        super();
        this.accessor = accessor;
        this.multiSelect = multiSelect;
    }
    transform(items, stringify) {
        return this.filter(items, TUI_DEFAULT_MATCHER, stringify, this.query);
    }
}
TuiFilterByInputWithPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterByInputWithPipe, deps: [{ token: TUI_FOCUSABLE_ITEM_ACCESSOR }, { token: TuiMultiSelectDirective, optional: true }], target: i0.ɵɵFactoryTarget.Pipe });
TuiFilterByInputWithPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterByInputWithPipe, name: "tuiFilterByInputWith", pure: false });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilterByInputWithPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiFilterByInputWith',
                    pure: false,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_FOCUSABLE_ITEM_ACCESSOR]
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TuiMultiSelectDirective]
                }] }]; } });
