// @taiga-ui/layout/bundles/taiga-ui-layout-components-block-status.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/layout/components/block-status', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].layout = global["taiga-ui"].layout || {}, global["taiga-ui"].layout.components = global["taiga-ui"].layout.components || {}, global["taiga-ui"].layout.components["block-status"] = {}), global.ng.core, global.ng.common));
})(this, (function (exports, i0, common) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiBlockStatusComponent = /** @class */ (function () {
        function TuiBlockStatusComponent() {
            this.card = false;
        }
        return TuiBlockStatusComponent;
    }());
    TuiBlockStatusComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiBlockStatusComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockStatusComponent, selector: "tui-block-status", inputs: { card: "card" }, host: { properties: { "class._card": "this.card" } }, ngImport: i0__namespace, template: "<div class=\"t-block-image\">\n    <ng-content select=\"[tuiSlot='top']\"></ng-content>\n</div>\n\n<ng-content select=\"h1,h2,h3,h4,h5,h6\"></ng-content>\n\n<div class=\"t-block-text\">\n    <ng-content></ng-content>\n</div>\n\n<div class=\"t-block-actions\">\n    <ng-content select=\"a,button,[tuiSlot='action']\"></ng-content>\n</div>\n", styles: ["tui-block-status{display:flex;flex-direction:column;align-items:center;justify-content:center;text-align:center;box-sizing:border-box;overflow:hidden;width:100%;height:100%}tui-root._mobile tui-block-status{padding:2rem}tui-block-status._card{border-radius:var(--tui-radius-xl);padding:2rem}tui-root._mobile tui-block-status._card{padding:1.25rem}tui-root._mobile tui-block-status._card .t-block-image{margin-top:.75rem}tui-block-status .t-block-image{margin-bottom:2rem}tui-root._mobile tui-block-status .t-block-image{margin-bottom:1.5rem}tui-block-status .t-block-text{font:var(--tui-font-text-m);color:var(--tui-text-02);margin-bottom:2rem;white-space:pre-line}tui-root._mobile tui-block-status .t-block-text{margin-bottom:1.5rem}tui-block-status .t-block-actions{display:flex;flex-direction:column;align-items:center;width:100%}tui-block-status h1,tui-block-status h2,tui-block-status h3,tui-block-status h4,tui-block-status h5,tui-block-status h6{font:var(--tui-font-heading-4);margin-top:0;margin-bottom:1rem}tui-root._mobile tui-block-status h1,tui-root._mobile tui-block-status h2,tui-root._mobile tui-block-status h3,tui-root._mobile tui-block-status h4,tui-root._mobile tui-block-status h5,tui-root._mobile tui-block-status h6{font:var(--tui-font-heading-5);margin-bottom:.5rem}tui-block-status img{width:9rem;height:9rem}tui-root._mobile tui-block-status img{width:8rem;height:8rem}tui-block-status a:not(:last-child),tui-block-status button:not(:last-child){margin-bottom:.75rem}tui-root._mobile tui-block-status a:not(:last-child),tui-root._mobile tui-block-status button:not(:last-child){margin-bottom:.5rem}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-block-status',
                        templateUrl: './block-status.template.html',
                        styleUrls: ['./block-status.style.less'],
                        encapsulation: i0.ViewEncapsulation.None,
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { card: [{
                    type: i0.HostBinding,
                    args: ['class._card']
                }, {
                    type: i0.Input
                }] } });

    var TuiBlockStatusDirective = /** @class */ (function () {
        function TuiBlockStatusDirective() {
            this.tuiSlot = 'top';
        }
        return TuiBlockStatusDirective;
    }());
    TuiBlockStatusDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiBlockStatusDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockStatusDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiSlot]',
                    }]
            }], propDecorators: { tuiSlot: [{
                    type: i0.Input
                }] } });

    var TuiBlockStatusModule = /** @class */ (function () {
        function TuiBlockStatusModule() {
        }
        return TuiBlockStatusModule;
    }());
    TuiBlockStatusModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiBlockStatusModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusModule, declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective], imports: [common.CommonModule], exports: [TuiBlockStatusComponent, TuiBlockStatusDirective] });
    TuiBlockStatusModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusModule, imports: [[common.CommonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBlockStatusModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [common.CommonModule],
                        declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                        exports: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiBlockStatusComponent = TuiBlockStatusComponent;
    exports.TuiBlockStatusDirective = TuiBlockStatusDirective;
    exports.TuiBlockStatusModule = TuiBlockStatusModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
