// @taiga-ui/core/esm2015/directives/hint/hint.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiHintComponent } from './hint.component';
import { TuiHintDirective } from './hint.directive';
import { TuiHintDescribeDirective } from './hint-describe.directive';
import { TuiHintDriverDirective } from './hint-driver.directive';
import { TuiHintHostDirective } from './hint-host.directive';
import { TuiHintHoverDirective } from './hint-hover.directive';
import { TuiHintManualDirective } from './hint-manual.directive';
import { TuiHintOptionsDirective } from './hint-options.directive';
import { TuiHintPointerDirective } from './hint-pointer.directive';
import { TuiHintPositionDirective } from './hint-position.directive';
import { TuiHintUnstyledComponent, TuiHintUnstyledDirective, } from './hint-unstyled.component';
import * as i0 from "@angular/core";
export class TuiHintModule {
}
TuiHintModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiHintModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintModule, declarations: [TuiHintComponent,
        TuiHintDirective,
        TuiHintDriverDirective,
        TuiHintHostDirective,
        TuiHintHoverDirective,
        TuiHintManualDirective,
        TuiHintPointerDirective,
        TuiHintDescribeDirective,
        TuiHintPositionDirective,
        TuiHintOptionsDirective,
        TuiHintUnstyledComponent,
        TuiHintUnstyledDirective], imports: [CommonModule, PolymorpheusModule], exports: [TuiHintComponent,
        TuiHintDirective,
        TuiHintDriverDirective,
        TuiHintHostDirective,
        TuiHintHoverDirective,
        TuiHintManualDirective,
        TuiHintPointerDirective,
        TuiHintDescribeDirective,
        TuiHintPositionDirective,
        TuiHintOptionsDirective,
        TuiHintUnstyledDirective] });
TuiHintModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintModule, imports: [[CommonModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule],
                    declarations: [
                        TuiHintComponent,
                        TuiHintDirective,
                        TuiHintDriverDirective,
                        TuiHintHostDirective,
                        TuiHintHoverDirective,
                        TuiHintManualDirective,
                        TuiHintPointerDirective,
                        TuiHintDescribeDirective,
                        TuiHintPositionDirective,
                        TuiHintOptionsDirective,
                        TuiHintUnstyledComponent,
                        TuiHintUnstyledDirective,
                    ],
                    exports: [
                        TuiHintComponent,
                        TuiHintDirective,
                        TuiHintDriverDirective,
                        TuiHintHostDirective,
                        TuiHintHoverDirective,
                        TuiHintManualDirective,
                        TuiHintPointerDirective,
                        TuiHintDescribeDirective,
                        TuiHintPositionDirective,
                        TuiHintOptionsDirective,
                        TuiHintUnstyledDirective,
                    ],
                }]
        }] });
