// @taiga-ui/core/esm2015/directives/hint/hint-unstyled.component.js
import { ChangeDetectionStrategy, Component, Directive, Inject, TemplateRef, } from '@angular/core';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk';
import { POLYMORPHEUS_CONTEXT, PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiHintDirective } from './hint.directive';
import * as i0 from "@angular/core";
import * as i1 from "@tinkoff/ng-polymorpheus";
import * as i2 from "./hint.directive";
export class TuiHintUnstyledComponent {
    constructor(context) {
        this.context = context;
    }
}
TuiHintUnstyledComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintUnstyledComponent, deps: [{ token: POLYMORPHEUS_CONTEXT }], target: i0.ɵɵFactoryTarget.Component });
TuiHintUnstyledComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintUnstyledComponent, selector: "ng-component", host: { properties: { "@tuiParentAnimation": "" } }, ngImport: i0, template: `
        <ng-container
            *polymorpheusOutlet="context.$implicit.content; context: context"
        ></ng-container>
    `, isInline: true, directives: [{ type: i1.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [TUI_PARENT_ANIMATION], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintUnstyledComponent, decorators: [{
            type: Component,
            args: [{
                    template: `
        <ng-container
            *polymorpheusOutlet="context.$implicit.content; context: context"
        ></ng-container>
    `,
                    host: { '[@tuiParentAnimation]': '' },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [TUI_PARENT_ANIMATION],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [POLYMORPHEUS_CONTEXT]
                }] }]; } });
export class TuiHintUnstyledDirective {
    constructor(templateRef, hint) {
        hint.component = new PolymorpheusComponent(TuiHintUnstyledComponent);
        hint.content = templateRef;
    }
}
TuiHintUnstyledDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintUnstyledDirective, deps: [{ token: TemplateRef }, { token: TuiHintDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiHintUnstyledDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintUnstyledDirective, selector: "ng-template[tuiHint]", providers: [
        {
            provide: POLYMORPHEUS_CONTEXT,
            useValue: { $implicit: {} },
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintUnstyledDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiHint]',
                    providers: [
                        {
                            provide: POLYMORPHEUS_CONTEXT,
                            useValue: { $implicit: {} },
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }, { type: i2.TuiHintDirective, decorators: [{
                    type: Inject,
                    args: [TuiHintDirective]
                }] }]; } });
