// @taiga-ui/addon-doc/esm2015/components/doc-tab/doc-tab.module.js
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiDocTabComponent } from './doc-tab.component';
import * as i0 from "@angular/core";
export class TuiDocTabModule {
}
TuiDocTabModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocTabModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTabModule, declarations: [TuiDocTabComponent], imports: [TuiSvgModule], exports: [TuiDocTabComponent] });
TuiDocTabModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTabModule, imports: [[TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTabModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiSvgModule],
                    declarations: [TuiDocTabComponent],
                    exports: [TuiDocTabComponent],
                }]
        }] });
