// @taiga-ui/addon-mobile/esm2015/directives/elastic-sticky/elastic-sticky.directive.js
import { Directive, Inject, Output } from '@angular/core';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { TuiElasticStickyService } from './elastic-sticky.service';
import * as i0 from "@angular/core";
import * as i1 from "./elastic-sticky.service";
export class TuiElasticStickyDirective {
    constructor(elasticStickyService) {
        this.tuiElasticSticky = elasticStickyService;
    }
}
TuiElasticStickyDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElasticStickyDirective, deps: [{ token: TuiElasticStickyService }], target: i0.ɵɵFactoryTarget.Directive });
TuiElasticStickyDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiElasticStickyDirective, selector: "[tuiElasticSticky]", outputs: { tuiElasticSticky: "tuiElasticSticky" }, providers: [TuiElasticStickyService, TuiDestroyService], exportAs: ["tuiElasticSticky"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiElasticStickyDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiElasticSticky]',
                    providers: [TuiElasticStickyService, TuiDestroyService],
                    exportAs: 'tuiElasticSticky',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiElasticStickyService, decorators: [{
                    type: Inject,
                    args: [TuiElasticStickyService]
                }] }]; }, propDecorators: { tuiElasticSticky: [{
                type: Output
            }] } });
