// @taiga-ui/addon-table/esm2015/components/reorder/reorder.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule, TuiSvgModule } from '@taiga-ui/core';
import { TuiTilesModule } from '@taiga-ui/kit';
import { TuiReorderComponent } from './reorder.component';
import * as i0 from "@angular/core";
export class TuiReorderModule {
}
TuiReorderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReorderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiReorderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReorderModule, declarations: [TuiReorderComponent], imports: [CommonModule, TuiSvgModule, TuiButtonModule, TuiTilesModule], exports: [TuiReorderComponent] });
TuiReorderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReorderModule, imports: [[CommonModule, TuiSvgModule, TuiButtonModule, TuiTilesModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiReorderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiButtonModule, TuiTilesModule],
                    declarations: [TuiReorderComponent],
                    exports: [TuiReorderComponent],
                }]
        }] });
