// @taiga-ui/addon-preview/esm2015/components/preview/preview.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MutationObserverModule } from '@ng-web-apis/mutation-observer';
import { TuiPreviewDialogService } from '@taiga-ui/addon-preview/components/preview-dialog';
import { tuiAsDialog, TuiDragModule, TuiPanModule, TuiResizeModule, TuiZoomModule, } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiHintModule, TuiLoaderModule, TuiSvgModule, } from '@taiga-ui/core';
import { TuiSliderModule } from '@taiga-ui/kit';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPreviewPaginationComponent } from './pagination/preview-pagination.component';
import { TuiPreviewComponent } from './preview.component';
import { TuiPreviewActionDirective } from './preview-action/preview-action.directive';
import { TuiPreviewActionModule } from './preview-action/preview-action.module';
import { TuiPreviewTitleComponent } from './title/preview-title.component';
import { TuiPreviewZoomComponent } from './zoom/preview-zoom.component';
import * as i0 from "@angular/core";
export class TuiPreviewModule {
}
TuiPreviewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPreviewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewModule, declarations: [TuiPreviewComponent,
        TuiPreviewTitleComponent,
        TuiPreviewPaginationComponent,
        TuiPreviewZoomComponent], imports: [CommonModule,
        PolymorpheusModule,
        MutationObserverModule,
        TuiButtonModule,
        FormsModule,
        TuiSliderModule,
        TuiHintModule,
        TuiDragModule,
        TuiResizeModule,
        TuiSvgModule,
        TuiLoaderModule,
        TuiPanModule,
        TuiZoomModule,
        TuiPreviewActionModule], exports: [TuiPreviewComponent,
        TuiPreviewActionDirective,
        TuiPreviewTitleComponent,
        TuiPreviewPaginationComponent,
        TuiPreviewZoomComponent] });
TuiPreviewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewModule, providers: [tuiAsDialog(TuiPreviewDialogService)], imports: [[
            CommonModule,
            PolymorpheusModule,
            MutationObserverModule,
            TuiButtonModule,
            FormsModule,
            TuiSliderModule,
            TuiHintModule,
            TuiDragModule,
            TuiResizeModule,
            TuiSvgModule,
            TuiLoaderModule,
            TuiPanModule,
            TuiZoomModule,
            TuiPreviewActionModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        MutationObserverModule,
                        TuiButtonModule,
                        FormsModule,
                        TuiSliderModule,
                        TuiHintModule,
                        TuiDragModule,
                        TuiResizeModule,
                        TuiSvgModule,
                        TuiLoaderModule,
                        TuiPanModule,
                        TuiZoomModule,
                        TuiPreviewActionModule,
                    ],
                    declarations: [
                        TuiPreviewComponent,
                        TuiPreviewTitleComponent,
                        TuiPreviewPaginationComponent,
                        TuiPreviewZoomComponent,
                    ],
                    providers: [tuiAsDialog(TuiPreviewDialogService)],
                    exports: [
                        TuiPreviewComponent,
                        TuiPreviewActionDirective,
                        TuiPreviewTitleComponent,
                        TuiPreviewPaginationComponent,
                        TuiPreviewZoomComponent,
                    ],
                }]
        }] });
