// @taiga-ui/addon-commerce/esm2015/components/input-expire/input-expire.module.js
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiPrimitiveTextfieldModule, TuiTextfieldControllerModule } from '@taiga-ui/core';
import { TuiValueAccessorModule } from '@taiga-ui/kit';
import { TuiInputExpireComponent } from './input-expire.component';
import * as i0 from "@angular/core";
export class TuiInputExpireModule {
}
TuiInputExpireModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputExpireModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputExpireModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputExpireModule, declarations: [TuiInputExpireComponent], imports: [MaskitoModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiValueAccessorModule], exports: [TuiInputExpireComponent] });
TuiInputExpireModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputExpireModule, imports: [[
            MaskitoModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiValueAccessorModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputExpireModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiValueAccessorModule,
                    ],
                    declarations: [TuiInputExpireComponent],
                    exports: [TuiInputExpireComponent],
                }]
        }] });
