// @taiga-ui/experimental/esm2015/components/segmented/segmented.directive.js
import { ContentChildren, Directive, ElementRef, HostListener, Inject, Self, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { RouterLinkActive } from '@angular/router';
import { EMPTY_QUERY, TuiDestroyService, tuiQueryListChanges } from '@taiga-ui/cdk';
import { EMPTY } from 'rxjs';
import { switchMap, takeUntil } from 'rxjs/operators';
import { TuiSegmentedComponent } from './segmented.component';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
import * as i2 from "./segmented.component";
export class TuiSegmentedDirective {
    constructor(destroy$, component, el) {
        this.destroy$ = destroy$;
        this.component = component;
        this.el = el;
        this.controls = EMPTY_QUERY;
        this.links = EMPTY_QUERY;
        this.elements = EMPTY_QUERY;
    }
    update(target) {
        const index = this.getIndex(target);
        if (index >= 0) {
            this.component.update(index);
        }
    }
    ngAfterContentInit() {
        tuiQueryListChanges(this.controls)
            .pipe(switchMap(() => { var _a; return ((_a = this.controls.last) === null || _a === void 0 ? void 0 : _a.valueChanges) || EMPTY; }), takeUntil(this.destroy$))
            .subscribe(() => {
            this.update(this.el.nativeElement.querySelector(':checked'));
        });
    }
    ngAfterContentChecked() {
        var _a;
        if (this.links.length) {
            this.update(((_a = this.elements.get(this.linkIndex)) === null || _a === void 0 ? void 0 : _a.nativeElement) || null);
        }
    }
    get linkIndex() {
        return this.links.toArray().findIndex(link => link.isActive);
    }
    getIndex(element) {
        return Array.from(this.component.tabs).findIndex(tab => tab.contains(element));
    }
}
TuiSegmentedDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSegmentedDirective, deps: [{ token: TuiDestroyService, self: true }, { token: TuiSegmentedComponent }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiSegmentedDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSegmentedDirective, selector: "tui-segmented", host: { listeners: { "click": "update($event.target)" } }, queries: [{ propertyName: "controls", predicate: NgControl, descendants: true }, { propertyName: "links", predicate: RouterLinkActive }, { propertyName: "elements", predicate: RouterLinkActive, read: ElementRef }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSegmentedDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-segmented',
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: i2.TuiSegmentedComponent, decorators: [{
                    type: Inject,
                    args: [TuiSegmentedComponent]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { controls: [{
                type: ContentChildren,
                args: [NgControl, { descendants: true }]
            }], links: [{
                type: ContentChildren,
                args: [RouterLinkActive]
            }], elements: [{
                type: ContentChildren,
                args: [RouterLinkActive, { read: ElementRef }]
            }], update: [{
                type: HostListener,
                args: ['click', ['$event.target']]
            }] } });
