// @taiga-ui/cdk/esm2015/components/dialog-host/dialog-host.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiScrollControlsModule } from '@taiga-ui/cdk/components/scroll-controls';
import { TuiFocusTrapModule, TuiLetModule, TuiOverscrollModule, } from '@taiga-ui/cdk/directives';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDialogHostComponent } from './dialog-host.component';
import * as i0 from "@angular/core";
export class TuiDialogHostModule {
}
TuiDialogHostModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogHostModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDialogHostModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogHostModule, declarations: [TuiDialogHostComponent], imports: [CommonModule,
        PolymorpheusModule,
        TuiOverscrollModule,
        TuiFocusTrapModule,
        TuiLetModule,
        TuiScrollControlsModule], exports: [TuiDialogHostComponent] });
TuiDialogHostModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogHostModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiOverscrollModule,
            TuiFocusTrapModule,
            TuiLetModule,
            TuiScrollControlsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDialogHostModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiOverscrollModule,
                        TuiFocusTrapModule,
                        TuiLetModule,
                        TuiScrollControlsModule,
                    ],
                    declarations: [TuiDialogHostComponent],
                    exports: [TuiDialogHostComponent],
                }]
        }] });
