// @taiga-ui/addon-doc/esm2015/components/code/code.module.js
import { ClipboardModule } from '@angular/cdk/clipboard';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule, TuiSvgModule } from '@taiga-ui/core';
import { HighlightModule } from 'ngx-highlightjs';
import { TuiDocCodeComponent } from './code.component';
import * as i0 from "@angular/core";
export class TuiDocCodeModule {
}
TuiDocCodeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocCodeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocCodeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocCodeModule, declarations: [TuiDocCodeComponent], imports: [CommonModule,
        TuiSvgModule,
        HighlightModule,
        TuiButtonModule,
        ClipboardModule], exports: [TuiDocCodeComponent] });
TuiDocCodeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocCodeModule, imports: [[
            CommonModule,
            TuiSvgModule,
            HighlightModule,
            TuiButtonModule,
            ClipboardModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocCodeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiSvgModule,
                        HighlightModule,
                        TuiButtonModule,
                        ClipboardModule,
                    ],
                    declarations: [TuiDocCodeComponent],
                    exports: [TuiDocCodeComponent],
                }]
        }] });
