// @taiga-ui/addon-charts/esm2015/components/axes/axes.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiAxesComponent } from './axes.component';
import * as i0 from "@angular/core";
export class TuiAxesModule {
}
TuiAxesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAxesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAxesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAxesModule, declarations: [TuiAxesComponent], imports: [CommonModule, TuiRepeatTimesModule], exports: [TuiAxesComponent] });
TuiAxesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAxesModule, imports: [[CommonModule, TuiRepeatTimesModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAxesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiRepeatTimesModule],
                    declarations: [TuiAxesComponent],
                    exports: [TuiAxesComponent],
                }]
        }] });
