// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-platform.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/tokens')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/platform', ['exports', '@angular/core', '@taiga-ui/cdk/tokens'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.platform = {}), global.ng.core, global["taiga-ui"].cdk.tokens));
})(this, (function (exports, i0, tokens) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiPlatformDirective = /** @class */ (function () {
        function TuiPlatformDirective() {
            this.tuiPlatform = '';
        }
        return TuiPlatformDirective;
    }());
    TuiPlatformDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPlatformDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPlatformDirective, selector: "[tuiPlatform]", inputs: { tuiPlatform: "tuiPlatform" }, host: { properties: { "attr.data-platform": "this.tuiPlatform" } }, providers: [
            {
                provide: tokens.TUI_PLATFORM,
                deps: [TuiPlatformDirective],
                useFactory: function (_a) {
                    var tuiPlatform = _a.tuiPlatform;
                    return tuiPlatform || tokens.tuiPlatformFactory();
                },
            },
        ], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPlatform]',
                        providers: [
                            {
                                provide: tokens.TUI_PLATFORM,
                                deps: [TuiPlatformDirective],
                                useFactory: function (_a) {
                                    var tuiPlatform = _a.tuiPlatform;
                                    return tuiPlatform || tokens.tuiPlatformFactory();
                                },
                            },
                        ],
                    }]
            }], propDecorators: { tuiPlatform: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-platform']
                }] } });

    var TuiPlatformModule = /** @class */ (function () {
        function TuiPlatformModule() {
        }
        return TuiPlatformModule;
    }());
    TuiPlatformModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPlatformModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformModule, declarations: [TuiPlatformDirective], exports: [TuiPlatformDirective] });
    TuiPlatformModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPlatformModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPlatformDirective],
                        exports: [TuiPlatformDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPlatformDirective = TuiPlatformDirective;
    exports.TuiPlatformModule = TuiPlatformModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
