// @taiga-ui/cdk/bundles/taiga-ui-cdk-components-alert-host.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/constants'), require('@taiga-ui/cdk/services'), require('@taiga-ui/cdk/tokens'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/cdk/pipes')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/components/alert-host', ['exports', '@angular/core', '@taiga-ui/cdk/constants', '@taiga-ui/cdk/services', '@taiga-ui/cdk/tokens', '@tinkoff/ng-polymorpheus', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/cdk/pipes'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.components = global["taiga-ui"].cdk.components || {}, global["taiga-ui"].cdk.components["alert-host"] = {}), global.ng.core, global["taiga-ui"].cdk.constants, global["taiga-ui"].cdk.services, global["taiga-ui"].cdk.tokens, global.ngPolymorpheus, global.rxjs, global.rxjs.operators, global.ng.common, global["taiga-ui"].cdk.pipes));
})(this, (function (exports, i0, constants, services, tokens, ngPolymorpheus, i3, operators, i1, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TuiAlertHostComponent = /** @class */ (function () {
        function TuiAlertHostComponent(allAlerts, injector, destroy$, cdr) {
            var _this = this;
            this.allAlerts = allAlerts;
            this.injector = injector;
            this.destroy$ = destroy$;
            this.cdr = cdr;
            this.alerts = [];
            this.trackBy = function (index) { return index; };
            this.mapper = function (useValue) { return i0.Injector.create({
                providers: [
                    {
                        provide: ngPolymorpheus.POLYMORPHEUS_CONTEXT,
                        useValue: useValue,
                    },
                ],
                parent: _this.injector,
            }); };
        }
        TuiAlertHostComponent.prototype.ngOnInit = function () {
            var _this = this;
            // Due to this view being parallel to app content, `markForCheck` from `async` pipe
            // can happen after view was checked, so calling `detectChanges` instead
            i3.combineLatest(this.allAlerts)
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (alerts) {
                _this.alerts = alerts;
                _this.cdr.detectChanges();
            });
        };
        return TuiAlertHostComponent;
    }());
    TuiAlertHostComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostComponent, deps: [{ token: tokens.TUI_ALERTS }, { token: i0.INJECTOR }, { token: services.TuiDestroyService, self: true }, { token: i0.ChangeDetectorRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiAlertHostComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAlertHostComponent, selector: "tui-alert-host", providers: [services.TuiDestroyService], ngImport: i0__namespace, template: "<div\n    *ngFor=\"let alert of alerts; trackBy: trackBy\"\n    class=\"t-wrapper\"\n    @tuiParentAnimation\n>\n    <ng-container\n        *ngFor=\"let item of alert\"\n        [ngComponentOutlet]=\"item.component.component\"\n        [ngComponentOutletInjector]=\"item | tuiMapper: mapper\"\n    ></ng-container>\n</div>\n", styles: ["tui-alert-host>.t-wrapper{position:fixed;top:0;left:0;display:flex;width:100%;height:100%;flex-direction:column;pointer-events:none}tui-alert-host>.t-wrapper>*{pointer-events:auto}\n"], directives: [{ type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i1__namespace.NgComponentOutlet, selector: "[ngComponentOutlet]", inputs: ["ngComponentOutlet", "ngComponentOutletInjector", "ngComponentOutletContent", "ngComponentOutletNgModuleFactory"] }], pipes: { "tuiMapper": i2__namespace.TuiMapperPipe }, animations: [constants.TUI_PARENT_ANIMATION], changeDetection: i0__namespace.ChangeDetectionStrategy.Default, encapsulation: i0__namespace.ViewEncapsulation.None });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-alert-host',
                        templateUrl: './alert-host.template.html',
                        styleUrls: ['./alert-host.style.less'],
                        encapsulation: i0.ViewEncapsulation.None,
                        // So that we do not force OnPush on custom alerts
                        // eslint-disable-next-line @angular-eslint/prefer-on-push-component-change-detection
                        changeDetection: i0.ChangeDetectionStrategy.Default,
                        providers: [services.TuiDestroyService],
                        animations: [constants.TUI_PARENT_ANIMATION],
                    }]
            }], ctorParameters: function () {
            return [{ type: Array, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_ALERTS]
                        }] }, { type: i0__namespace.Injector, decorators: [{
                            type: i0.Inject,
                            args: [i0.INJECTOR]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [services.TuiDestroyService]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }];
        } });

    var TuiAlertHostModule = /** @class */ (function () {
        function TuiAlertHostModule() {
        }
        return TuiAlertHostModule;
    }());
    TuiAlertHostModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiAlertHostModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostModule, declarations: [TuiAlertHostComponent], imports: [i1.CommonModule, i2.TuiMapperPipeModule], exports: [TuiAlertHostComponent] });
    TuiAlertHostModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostModule, imports: [[i1.CommonModule, i2.TuiMapperPipeModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAlertHostModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule, i2.TuiMapperPipeModule],
                        declarations: [TuiAlertHostComponent],
                        exports: [TuiAlertHostComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiAlertHostComponent = TuiAlertHostComponent;
    exports.TuiAlertHostModule = TuiAlertHostModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
