// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-click-outside.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@taiga-ui/cdk/observables'), require('@taiga-ui/cdk/utils'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/click-outside', ['exports', '@angular/common', '@angular/core', '@taiga-ui/cdk/observables', '@taiga-ui/cdk/utils', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives["click-outside"] = {}), global.ng.common, global.ng.core, global["taiga-ui"].cdk.observables, global["taiga-ui"].cdk.utils, global.rxjs, global.rxjs.operators));
})(this, (function (exports, common, i0, observables, utils, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiClickOutsideDirective = /** @class */ (function () {
        function TuiClickOutsideDirective(zone, doc, el) {
            var _this = this;
            this.zone = zone;
            this.doc = doc;
            this.el = el;
            this.tuiClickOutside = rxjs.fromEvent(this.doc, 'mouseup').pipe(operators.map(utils.tuiGetActualTarget), operators.filter(function (target) { return _this.isOutside(target); }), observables.tuiZoneOptimized(this.zone));
        }
        TuiClickOutsideDirective.prototype.isOutside = function (target) {
            return (target === this.el.nativeElement ||
                !utils.tuiContainsOrAfter(this.el.nativeElement, target));
        };
        return TuiClickOutsideDirective;
    }());
    TuiClickOutsideDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideDirective, deps: [{ token: i0.NgZone }, { token: common.DOCUMENT }, { token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiClickOutsideDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiClickOutsideDirective, selector: "[tuiClickOutside]", outputs: { tuiClickOutside: "tuiClickOutside" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiClickOutside]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: Document, decorators: [{
                            type: i0.Inject,
                            args: [common.DOCUMENT]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { tuiClickOutside: [{
                    type: i0.Output
                }] } });

    var TuiClickOutsideModule = /** @class */ (function () {
        function TuiClickOutsideModule() {
        }
        return TuiClickOutsideModule;
    }());
    TuiClickOutsideModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiClickOutsideModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideModule, declarations: [TuiClickOutsideDirective], exports: [TuiClickOutsideDirective] });
    TuiClickOutsideModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiClickOutsideModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiClickOutsideDirective],
                        exports: [TuiClickOutsideDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiClickOutsideDirective = TuiClickOutsideDirective;
    exports.TuiClickOutsideModule = TuiClickOutsideModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
