// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-repeat-times.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk/utils/math')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/repeat-times', ['exports', '@angular/core', '@taiga-ui/cdk/utils/math'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives["repeat-times"] = {}), global.ng.core, global["taiga-ui"].cdk.utils.math));
})(this, (function (exports, i0, math) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var MAX_VALUE = 0x10000;
    var TuiRepeatTimesContext = /** @class */ (function () {
        function TuiRepeatTimesContext($implicit) {
            this.$implicit = $implicit;
        }
        return TuiRepeatTimesContext;
    }());
    /**
     * Directive similar to ngFor but using a number of repetitions rather than an array
     *
     * {@link TuiRepeatTimesDirective.tuiRepeatTimesOf requested number of times}.
     * {@link TuiRepeatTimesContext context} for every instance of the template inherits outer context and stores
     * {@link TuiRepeatTimesContext.$implicit index} of a template instance.
     */
    var TuiRepeatTimesDirective = /** @class */ (function () {
        function TuiRepeatTimesDirective(viewContainer, templateRef) {
            this.viewContainer = viewContainer;
            this.templateRef = templateRef;
        }
        Object.defineProperty(TuiRepeatTimesDirective.prototype, "tuiRepeatTimesOf", {
            set: function (count) {
                var safeCount = Math.floor(math.tuiClamp(count, 0, MAX_VALUE));
                var length = this.viewContainer.length;
                if (count < length) {
                    this.removeContainers(length - count);
                }
                else {
                    this.addContainers(safeCount);
                }
            },
            enumerable: false,
            configurable: true
        });
        TuiRepeatTimesDirective.prototype.addContainers = function (count) {
            for (var index = this.viewContainer.length; index < count; index++) {
                this.viewContainer.createEmbeddedView(this.templateRef, new TuiRepeatTimesContext(index));
            }
        };
        TuiRepeatTimesDirective.prototype.removeContainers = function (amount) {
            for (var index = 0; index < amount; index++) {
                this.viewContainer.remove();
            }
        };
        return TuiRepeatTimesDirective;
    }());
    TuiRepeatTimesDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiRepeatTimesDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: { tuiRepeatTimesOf: "tuiRepeatTimesOf" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiRepeatTimes][tuiRepeatTimesOf]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ViewContainerRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ViewContainerRef]
                        }] }, { type: i0__namespace.TemplateRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.TemplateRef]
                        }] }];
        }, propDecorators: { tuiRepeatTimesOf: [{
                    type: i0.Input
                }] } });

    var TuiRepeatTimesModule = /** @class */ (function () {
        function TuiRepeatTimesModule() {
        }
        return TuiRepeatTimesModule;
    }());
    TuiRepeatTimesModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiRepeatTimesModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesModule, declarations: [TuiRepeatTimesDirective], exports: [TuiRepeatTimesDirective] });
    TuiRepeatTimesModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRepeatTimesModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiRepeatTimesDirective],
                        exports: [TuiRepeatTimesDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiRepeatTimesContext = TuiRepeatTimesContext;
    exports.TuiRepeatTimesDirective = TuiRepeatTimesDirective;
    exports.TuiRepeatTimesModule = TuiRepeatTimesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
