// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-validator.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk/constants')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/validator', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk/constants'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.directives = global["taiga-ui"].cdk.directives || {}, global["taiga-ui"].cdk.directives.validator = {}), global.ng.core, global.ng.forms, global["taiga-ui"].cdk.constants));
})(this, (function (exports, i0, forms, constants) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiValidatorDirective = /** @class */ (function () {
        function TuiValidatorDirective() {
            this.onChange = constants.EMPTY_FUNCTION;
            this.tuiValidator = forms.Validators.nullValidator;
        }
        TuiValidatorDirective.prototype.validate = function (control) {
            return this.tuiValidator(control);
        };
        TuiValidatorDirective.prototype.registerOnValidatorChange = function (onChange) {
            this.onChange = onChange;
        };
        TuiValidatorDirective.prototype.ngOnChanges = function () {
            this.onChange();
        };
        TuiValidatorDirective.prototype.ngOnDestroy = function () {
            this.tuiValidator = forms.Validators.nullValidator;
            this.onChange();
        };
        return TuiValidatorDirective;
    }());
    TuiValidatorDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiValidatorDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiValidatorDirective, selector: "[tuiValidator]", inputs: { tuiValidator: "tuiValidator" }, providers: [
            {
                provide: forms.NG_VALIDATORS,
                useExisting: i0.forwardRef(function () { return TuiValidatorDirective; }),
                multi: true,
            },
        ], usesOnChanges: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiValidator]',
                        providers: [
                            {
                                provide: forms.NG_VALIDATORS,
                                useExisting: i0.forwardRef(function () { return TuiValidatorDirective; }),
                                multi: true,
                            },
                        ],
                    }]
            }], propDecorators: { tuiValidator: [{
                    type: i0.Input
                }] } });

    var TuiValidatorModule = /** @class */ (function () {
        function TuiValidatorModule() {
        }
        return TuiValidatorModule;
    }());
    TuiValidatorModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiValidatorModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorModule, declarations: [TuiValidatorDirective], exports: [TuiValidatorDirective] });
    TuiValidatorModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiValidatorModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiValidatorDirective],
                        exports: [TuiValidatorDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiValidatorDirective = TuiValidatorDirective;
    exports.TuiValidatorModule = TuiValidatorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
