// @taiga-ui/addon-mobile/esm2015/components/sheet-dialog/sheet-dialog.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { tuiAsDialog, TuiClickOutsideModule } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiSheetDialogComponent } from './sheet-dialog.component';
import { TuiSheetDialogDirective } from './sheet-dialog.directive';
import { TuiSheetDialogService } from './sheet-dialog.service';
import * as i0 from "@angular/core";
export class TuiSheetDialogModule {
}
TuiSheetDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSheetDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogModule, declarations: [TuiSheetDialogComponent, TuiSheetDialogDirective], imports: [CommonModule, PolymorpheusModule, TuiClickOutsideModule, TuiButtonModule], exports: [TuiSheetDialogComponent, TuiSheetDialogDirective] });
TuiSheetDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogModule, providers: [tuiAsDialog(TuiSheetDialogService)], imports: [[CommonModule, PolymorpheusModule, TuiClickOutsideModule, TuiButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiClickOutsideModule, TuiButtonModule],
                    declarations: [TuiSheetDialogComponent, TuiSheetDialogDirective],
                    providers: [tuiAsDialog(TuiSheetDialogService)],
                    exports: [TuiSheetDialogComponent, TuiSheetDialogDirective],
                }]
        }] });
