// @taiga-ui/addon-mobile/esm2015/components/sheet-dialog/sheet-dialog.service.js
import { inject, Injectable } from '@angular/core';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiSheetDialogComponent } from './sheet-dialog.component';
import { TUI_SHEET_DIALOG_OPTIONS } from './sheet-dialog.options';
import * as i0 from "@angular/core";
const DIALOG = new PolymorpheusComponent(TuiSheetDialogComponent);
export class TuiSheetDialogService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.component = DIALOG;
        this.defaultOptions = Object.assign(Object.assign({}, inject(TUI_SHEET_DIALOG_OPTIONS)), { data: undefined });
    }
}
TuiSheetDialogService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogService, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
TuiSheetDialogService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
