// @taiga-ui/addon-mobile/esm2015/components/sheet-dialog/sheet-dialog.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import { TuiSheetDialogService } from './sheet-dialog.service';
import * as i0 from "@angular/core";
export class TuiSheetDialogDirective extends AbstractTuiDialogDirective {
}
TuiSheetDialogDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiSheetDialogDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDialogDirective, selector: "ng-template[tuiSheetDialog]", inputs: { options: ["tuiSheetDialogOptions", "options"], open: ["tuiSheetDialog", "open"] }, outputs: { openChange: "tuiSheetDialogChange" }, providers: [
        {
            provide: AbstractTuiDialogService,
            useExisting: TuiSheetDialogService,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiSheetDialog]',
                    inputs: ['options: tuiSheetDialogOptions', 'open: tuiSheetDialog'],
                    outputs: ['openChange: tuiSheetDialogChange'],
                    providers: [
                        {
                            provide: AbstractTuiDialogService,
                            useExisting: TuiSheetDialogService,
                        },
                    ],
                }]
        }] });
