// @taiga-ui/cdk/esm2015/directives/platform/platform.directive.js
import { Directive, HostBinding, Input } from '@angular/core';
import { TUI_PLATFORM, tuiPlatformFactory } from '@taiga-ui/cdk/tokens';
import * as i0 from "@angular/core";
export class TuiPlatformDirective {
    constructor() {
        this.tuiPlatform = '';
    }
}
TuiPlatformDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiPlatformDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPlatformDirective, selector: "[tuiPlatform]", inputs: { tuiPlatform: "tuiPlatform" }, host: { properties: { "attr.data-platform": "this.tuiPlatform" } }, providers: [
        {
            provide: TUI_PLATFORM,
            deps: [TuiPlatformDirective],
            useFactory: ({ tuiPlatform }) => tuiPlatform || tuiPlatformFactory(),
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPlatformDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiPlatform]',
                    providers: [
                        {
                            provide: TUI_PLATFORM,
                            deps: [TuiPlatformDirective],
                            useFactory: ({ tuiPlatform }) => tuiPlatform || tuiPlatformFactory(),
                        },
                    ],
                }]
        }], propDecorators: { tuiPlatform: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-platform']
            }] } });
