// @taiga-ui/experimental/esm2015/components/badge/badge.module.js
import { NgModule } from '@angular/core';
import { TuiIconsDirective, TuiIconsModule } from '@taiga-ui/experimental/directives/icons';
import { TuiBadgeComponent } from './badge.component';
import { TuiBadgeDirective } from './badge.directive';
import * as i0 from "@angular/core";
export class TuiBadgeModule {
}
TuiBadgeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, declarations: [TuiBadgeComponent, TuiBadgeDirective], imports: [TuiIconsModule], exports: [TuiBadgeDirective, TuiIconsDirective] });
TuiBadgeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, imports: [[TuiIconsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiIconsModule],
                    declarations: [TuiBadgeComponent, TuiBadgeDirective],
                    exports: [TuiBadgeDirective, TuiIconsDirective],
                }]
        }] });
