// @taiga-ui/experimental/esm2015/components/badge/badge.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiBadgeComponent } from './badge.component';
import { TUI_BADGE_OPTIONS } from './badge.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiBadgeDirective {
    constructor(options, directiveStyles) {
        this.options = options;
        this.size = this.options.size;
        this.appearance = this.options.appearance;
        this.dot = this.options.dot;
        directiveStyles.addComponent(TuiBadgeComponent);
    }
}
TuiBadgeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeDirective, deps: [{ token: TUI_BADGE_OPTIONS }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiBadgeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgeDirective, selector: "tui-badge,[tuiBadge]", inputs: { size: "size", appearance: "appearance", dot: "dot" }, host: { attributes: { "tuiBadgeNew": "", "tuiAppearance": "" }, properties: { "class._dot": "dot", "attr.data-appearance": "appearance", "attr.data-size": "size" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-badge,[tuiBadge]',
                    host: {
                        tuiBadgeNew: '',
                        tuiAppearance: '',
                        '[class._dot]': 'dot',
                        '[attr.data-appearance]': 'appearance',
                        '[attr.data-size]': 'size',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_BADGE_OPTIONS]
                }] }, { type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { size: [{
                type: Input
            }], appearance: [{
                type: Input
            }], dot: [{
                type: Input
            }] } });
