// @taiga-ui/addon-doc/esm2015/components/internal/source-code/source-code.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { TUI_DOC_ICONS, TUI_DOC_SOURCE_CODE, TUI_DOC_SOURCE_CODE_TEXT, } from '@taiga-ui/addon-doc/tokens';
import { tuiPure } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
import * as i2 from "@tinkoff/ng-polymorpheus";
export class TuiDocSourceCodeComponent {
    constructor(icons, sourceCode, text) {
        this.icons = icons;
        this.sourceCode = sourceCode;
        this.text = text;
        this.header = '';
        this.package = '';
        this.type = '';
        this.path = '';
    }
    get pathOptions() {
        return this.getPathOptions(this.header, this.package, this.type, this.path);
    }
    pathIsUrl(path) {
        return path.startsWith('http');
    }
    getPathOptions(header, packageName, type, path) {
        return {
            header,
            package: packageName,
            type,
            path,
        };
    }
}
TuiDocSourceCodeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocSourceCodeComponent, deps: [{ token: TUI_DOC_ICONS }, { token: TUI_DOC_SOURCE_CODE }, { token: TUI_DOC_SOURCE_CODE_TEXT }], target: i0.ɵɵFactoryTarget.Component });
TuiDocSourceCodeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiDocSourceCodeComponent, selector: "tui-doc-source-code", inputs: { header: "header", package: "package", type: "type", path: "path" }, ngImport: i0, template: "<a\n    *polymorpheusOutlet=\"pathIsUrl(path) ? path : sourceCode as link; context: pathOptions\"\n    appearance=\"icon\"\n    size=\"s\"\n    target=\"_blank\"\n    tuiIconButton\n    type=\"button\"\n    [href]=\"link\"\n    [icon]=\"icons.code\"\n    [title]=\"text\"\n></a>\n", components: [{ type: i1.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
__decorate([
    tuiPure
], TuiDocSourceCodeComponent.prototype, "pathIsUrl", null);
__decorate([
    tuiPure
], TuiDocSourceCodeComponent.prototype, "getPathOptions", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocSourceCodeComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-doc-source-code',
                    templateUrl: './source-code.template.html',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_ICONS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_SOURCE_CODE]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_SOURCE_CODE_TEXT]
                }] }]; }, propDecorators: { header: [{
                type: Input
            }], package: [{
                type: Input
            }], type: [{
                type: Input
            }], path: [{
                type: Input
            }], pathIsUrl: [], getPathOptions: [] } });
