// @taiga-ui/addon-charts/esm2015/components/bar-set/bar-set.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBarModule } from '@taiga-ui/addon-charts/components/bar';
import { TuiBarSetComponent } from './bar-set.component';
import * as i0 from "@angular/core";
export class TuiBarSetModule {
}
TuiBarSetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarSetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBarSetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarSetModule, declarations: [TuiBarSetComponent], imports: [CommonModule, TuiBarModule], exports: [TuiBarSetComponent] });
TuiBarSetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarSetModule, imports: [[CommonModule, TuiBarModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarSetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiBarModule],
                    declarations: [TuiBarSetComponent],
                    exports: [TuiBarSetComponent],
                }]
        }] });
