// @taiga-ui/addon-mobile/esm2015/components/sheet/directives/sheet-close/sheet-close.directive.js
import { Directive, ElementRef, Inject, NgZone, Output } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { tuiIfMap, tuiIsFalsy, tuiTypedFromEvent, tuiZonefull } from '@taiga-ui/cdk';
import { merge } from 'rxjs';
import { distinctUntilChanged, filter, startWith } from 'rxjs/operators';
import { TuiSheetComponent } from '../../components/sheet/sheet.component';
import { TUI_SHEET_CLOSE } from '../../components/sheet-heading/sheet-heading.component';
import { TUI_SHEET_DRAGGED, TUI_SHEET_SCROLL } from '../../sheet-tokens';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
import * as i2 from "../../components/sheet/sheet.component";
export class TuiSheetCloseDirective {
    constructor(zone, dragged$, scroll$, win, el, sheet) {
        this.zone = zone;
        this.dragged$ = dragged$;
        this.scroll$ = scroll$;
        this.win = win;
        this.el = el;
        this.sheet = sheet;
        this.close = merge(tuiTypedFromEvent(this.el.nativeElement, TUI_SHEET_CLOSE), this.dragged$.pipe(tuiIfMap(() => this.scroll$.pipe(startWith(this.el.nativeElement.scrollTop)), tuiIsFalsy), filter(y => { var _a; return ((_a = this.sheet.item) === null || _a === void 0 ? void 0 : _a.closeable) && this.shouldClose(y); }), distinctUntilChanged(), tuiZonefull(this.zone)));
    }
    shouldClose(scrollTop) {
        const height = Math.min(this.win.innerHeight, this.el.nativeElement.scrollHeight - this.win.innerHeight);
        const min = Math.min(height, this.sheet.stops[0] || Infinity);
        return scrollTop < min / 2;
    }
}
TuiSheetCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetCloseDirective, deps: [{ token: NgZone }, { token: TUI_SHEET_DRAGGED }, { token: TUI_SHEET_SCROLL }, { token: WINDOW }, { token: ElementRef }, { token: TuiSheetComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiSheetCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetCloseDirective, selector: "tui-sheet[close]", outputs: { close: "close" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-sheet[close]',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_SHEET_DRAGGED]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_SHEET_SCROLL]
                }] }, { type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i2.TuiSheetComponent, decorators: [{
                    type: Inject,
                    args: [TuiSheetComponent]
                }] }]; }, propDecorators: { close: [{
                type: Output
            }] } });
