// @taiga-ui/core/esm2015/components/alert/alert.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import { TuiAlertService } from './alert.service';
import * as i0 from "@angular/core";
export class TuiAlertDirective extends AbstractTuiDialogDirective {
}
TuiAlertDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiAlertDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAlertDirective, selector: "ng-template[tuiAlert]", inputs: { options: ["tuiAlertOptions", "options"], open: ["tuiAlert", "open"] }, outputs: { openChange: "tuiAlertChange" }, providers: [
        {
            provide: AbstractTuiDialogService,
            useExisting: TuiAlertService,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAlertDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiAlert]',
                    inputs: ['options: tuiAlertOptions', 'open: tuiAlert'],
                    outputs: ['openChange: tuiAlertChange'],
                    providers: [
                        {
                            provide: AbstractTuiDialogService,
                            useExisting: TuiAlertService,
                        },
                    ],
                }]
        }] });
