// @taiga-ui/addon-commerce/bundles/taiga-ui-addon-commerce-components-input-card.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/addon-commerce/constants'), require('@taiga-ui/core'), require('@taiga-ui/addon-commerce/utils'), require('@taiga-ui/kit'), require('@maskito/angular')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-commerce/components/input-card', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/addon-commerce/constants', '@taiga-ui/core', '@taiga-ui/addon-commerce/utils', '@taiga-ui/kit', '@maskito/angular'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-commerce"] = global["taiga-ui"]["addon-commerce"] || {}, global["taiga-ui"]["addon-commerce"].components = global["taiga-ui"]["addon-commerce"].components || {}, global["taiga-ui"]["addon-commerce"].components["input-card"] = {}), global.ng.core, global.ng.forms, global.cdk, global["taiga-ui"]["addon-commerce"].constants, global.i1$1, global["taiga-ui"]["addon-commerce"].utils, global.i2, global.i3));
})(this, (function (exports, i0, i1, cdk, constants, i1$1, utils, i2, i3) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var AbstractTuiInputCard = /** @class */ (function (_super) {
        __extends(AbstractTuiInputCard, _super);
        function AbstractTuiInputCard(control, cdr, options) {
            var _this = _super.call(this, control, cdr) || this;
            _this.options = options;
            _this.autocompleteEnabled = _this.options.autocompleteEnabled;
            _this.binChange = new i0.EventEmitter();
            return _this;
        }
        Object.defineProperty(AbstractTuiInputCard.prototype, "defaultIcon", {
            get: function () {
                var paymentSystem = this.getPaymentSystem(this.card);
                return paymentSystem && this.options.icons[paymentSystem];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(AbstractTuiInputCard.prototype, "paymentSystem", {
            get: function () {
                return this.getPaymentSystem(this.card);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(AbstractTuiInputCard.prototype, "icon", {
            /** @deprecated remove in 4.0 */
            get: function () {
                return this.cardSrc || this.defaultIcon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(AbstractTuiInputCard.prototype, "autocomplete", {
            get: function () {
                return this.autocompleteEnabled ? 'cc-number' : 'off';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(AbstractTuiInputCard.prototype, "autocompleteCard", {
            /**
             * @deprecated: drop in v4.0
             * use {@link autocomplete}
             */
            get: function () {
                return this.autocomplete;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(AbstractTuiInputCard.prototype, "bin", {
            get: function () {
                return this.card.length < 6 ? null : this.card.slice(0, 6);
            },
            enumerable: false,
            configurable: true
        });
        AbstractTuiInputCard.prototype.getPaymentSystem = function (cardNumber) {
            return this.options.paymentSystemHandler(cardNumber);
        };
        return AbstractTuiInputCard;
    }(cdk.AbstractTuiNullableControl));
    AbstractTuiInputCard.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: AbstractTuiInputCard, deps: "invalid", target: i0__namespace.ɵɵFactoryTarget.Directive });
    AbstractTuiInputCard.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiInputCard, inputs: { cardSrc: "cardSrc", autocompleteEnabled: "autocompleteEnabled" }, outputs: { binChange: "binChange" }, usesInheritance: true, ngImport: i0__namespace });
    __decorate([
        cdk.tuiPure
    ], AbstractTuiInputCard.prototype, "getPaymentSystem", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: AbstractTuiInputCard, decorators: [{
                type: i0.Directive
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: undefined }];
        }, propDecorators: { cardSrc: [{
                    type: i0.Input
                }], autocompleteEnabled: [{
                    type: i0.Input
                }], binChange: [{
                    type: i0.Output
                }], getPaymentSystem: [] } });

    // TODO: Move payment system icons into its own token in 4.0
    var TUI_INPUT_CARD_DEFAULT_OPTIONS = {
        icons: utils.TUI_PAYMENT_SYSTEM_ICONS,
        cardSrc: '',
        paymentSystemHandler: utils.tuiGetPaymentSystem,
        autocompleteEnabled: false,
    };
    var TUI_INPUT_CARD_OPTIONS = cdk.tuiCreateToken(TUI_INPUT_CARD_DEFAULT_OPTIONS);
    function tuiInputCardOptionsProvider(options) {
        return cdk.tuiProvideOptions(TUI_INPUT_CARD_OPTIONS, options, TUI_INPUT_CARD_DEFAULT_OPTIONS);
    }

    var TuiInputCardComponent = /** @class */ (function (_super) {
        __extends(TuiInputCardComponent, _super);
        function TuiInputCardComponent(control, cdr, textfieldSize, options) {
            var _this = _super.call(this, control, cdr, options) || this;
            _this.textfieldSize = textfieldSize;
            _this.maskOptions = {
                mask: constants.TUI_CARD_MASK,
            };
            return _this;
        }
        Object.defineProperty(TuiInputCardComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputCardComponent.prototype, "card", {
            get: function () {
                var _a;
                return (_a = this.value) !== null && _a !== void 0 ? _a : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputCardComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.input) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) !== null && _b !== void 0 ? _b : null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputCardComponent.prototype, "focused", {
            get: function () {
                return !!this.input && this.input.focused;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputCardComponent.prototype.onValueChange = function (value) {
            var parsed = value.split(' ').join('');
            var currentBin = this.bin;
            this.value = parsed;
            var newBin = this.bin;
            if (currentBin !== newBin) {
                this.binChange.emit(newBin);
            }
        };
        TuiInputCardComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiInputCardComponent.prototype.writeValue = function (value) {
            var currentBin = this.bin;
            _super.prototype.writeValue.call(this, value);
            var newBin = this.bin;
            if (currentBin !== newBin) {
                this.binChange.emit(newBin);
            }
        };
        TuiInputCardComponent.prototype.getFallbackValue = function () {
            return '';
        };
        return TuiInputCardComponent;
    }(AbstractTuiInputCard));
    TuiInputCardComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardComponent, deps: [{ token: i1.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i1$1.TUI_TEXTFIELD_SIZE }, { token: TUI_INPUT_CARD_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputCardComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputCardComponent, selector: "tui-input-card", host: { properties: { "attr.data-size": "this.size" } }, providers: [
            cdk.tuiAsFocusableItemAccessor(TuiInputCardComponent),
            cdk.tuiAsControl(TuiInputCardComponent),
        ], viewQueries: [{ propertyName: "input", first: true, predicate: i1$1.TuiPrimitiveTextfieldComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-primitive-textfield\n    tuiValueAccessor\n    class=\"t-input\"\n    [disabled]=\"disabled\"\n    [focusable]=\"focusable\"\n    [invalid]=\"computedInvalid\"\n    [maskito]=\"maskOptions\"\n    [nativeId]=\"nativeId\"\n    [pseudoActive]=\"pseudoActive\"\n    [pseudoFocus]=\"pseudoFocus\"\n    [pseudoHover]=\"pseudoHover\"\n    [readOnly]=\"readOnly\"\n    [tuiTextfieldCustomContent]=\"cardSrc || (defaultIcon && logo)\"\n    [value]=\"card | maskito: maskOptions\"\n    (focusedChange)=\"onFocused($event)\"\n    (valueChange)=\"onValueChange($event)\"\n>\n    <ng-content></ng-content>\n    <ng-content\n        ngProjectAs=\"input\"\n        select=\"input\"\n    ></ng-content>\n</tui-primitive-textfield>\n<ng-template #logo>\n    <tui-svg\n        class=\"t-payment-system\"\n        [src]=\"defaultIcon\"\n    ></tui-svg>\n</ng-template>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-input{border-radius:inherit;text-align:inherit}.t-payment-system{width:2rem;height:2rem;color:#1434cb}\n"], components: [{ type: i1__namespace$1.TuiPrimitiveTextfieldComponent, selector: "tui-primitive-textfield", inputs: ["editable", "filler", "iconCleaner", "readOnly", "invalid", "disabled", "prefix", "postfix", "value"], outputs: ["valueChange"] }, { type: i1__namespace$1.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i1__namespace$1.TuiPrimitiveTextfieldDirective, selector: "tui-primitive-textfield" }, { type: i2__namespace.TuiValueAccessorDirective, selector: "[tuiValueAccessor]" }, { type: i3__namespace.MaskitoDirective, selector: "[maskito]", inputs: ["maskito", "maskitoElement"] }, { type: i1__namespace$1.TuiTextfieldCustomContentDirective, selector: "[tuiTextfieldCustomContent]", inputs: ["tuiTextfieldCustomContent"] }], pipes: { "maskito": i3__namespace.MaskitoPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-card',
                        templateUrl: './input-card.template.html',
                        styleUrls: ['./input-card.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            cdk.tuiAsFocusableItemAccessor(TuiInputCardComponent),
                            cdk.tuiAsControl(TuiInputCardComponent),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i1__namespace$1.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_INPUT_CARD_OPTIONS]
                        }] }];
        }, propDecorators: { input: [{
                    type: i0.ViewChild,
                    args: [i1$1.TuiPrimitiveTextfieldComponent]
                }], size: [{
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }] } });

    var TuiInputCardDirective = /** @class */ (function (_super) {
        __extends(TuiInputCardDirective, _super);
        function TuiInputCardDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TuiInputCardDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        TuiInputCardDirective.prototype.ngAfterViewInit = function () {
            var nativeFocusableElement = this.host.nativeFocusableElement;
            if (!nativeFocusableElement) {
                return;
            }
            nativeFocusableElement.inputMode = 'numeric';
            nativeFocusableElement.placeholder =
                nativeFocusableElement.placeholder || '0000 0000 0000 0000';
        };
        return TuiInputCardDirective;
    }(i1$1.AbstractTuiTextfieldHost));
    TuiInputCardDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiInputCardDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputCardDirective, selector: "tui-input-card", providers: [i1$1.tuiAsTextfieldHost(TuiInputCardDirective)], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-input-card',
                        providers: [i1$1.tuiAsTextfieldHost(TuiInputCardDirective)],
                    }]
            }] });

    var TuiInputCardModule = /** @class */ (function () {
        function TuiInputCardModule() {
        }
        return TuiInputCardModule;
    }());
    TuiInputCardModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputCardModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardModule, declarations: [TuiInputCardComponent, TuiInputCardDirective], imports: [i3.MaskitoModule,
            i1$1.TuiPrimitiveTextfieldModule,
            i1$1.TuiTextfieldControllerModule,
            i2.TuiValueAccessorModule,
            i1$1.TuiSvgModule], exports: [TuiInputCardComponent, TuiInputCardDirective, i1$1.TuiTextfieldComponent] });
    TuiInputCardModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardModule, imports: [[
                i3.MaskitoModule,
                i1$1.TuiPrimitiveTextfieldModule,
                i1$1.TuiTextfieldControllerModule,
                i2.TuiValueAccessorModule,
                i1$1.TuiSvgModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputCardModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.MaskitoModule,
                            i1$1.TuiPrimitiveTextfieldModule,
                            i1$1.TuiTextfieldControllerModule,
                            i2.TuiValueAccessorModule,
                            i1$1.TuiSvgModule,
                        ],
                        declarations: [TuiInputCardComponent, TuiInputCardDirective],
                        exports: [TuiInputCardComponent, TuiInputCardDirective, i1$1.TuiTextfieldComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AbstractTuiInputCard = AbstractTuiInputCard;
    exports.TUI_INPUT_CARD_DEFAULT_OPTIONS = TUI_INPUT_CARD_DEFAULT_OPTIONS;
    exports.TUI_INPUT_CARD_OPTIONS = TUI_INPUT_CARD_OPTIONS;
    exports.TuiInputCardComponent = TuiInputCardComponent;
    exports.TuiInputCardDirective = TuiInputCardDirective;
    exports.TuiInputCardModule = TuiInputCardModule;
    exports.tuiInputCardOptionsProvider = tuiInputCardOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
