// @taiga-ui/addon-commerce/bundles/taiga-ui-addon-commerce-pipes.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-commerce/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-commerce/pipes', ['exports', '@angular/core', '@taiga-ui/addon-commerce/utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-commerce"] = global["taiga-ui"]["addon-commerce"] || {}, global["taiga-ui"]["addon-commerce"].pipes = {}), global.ng.core, global["taiga-ui"]["addon-commerce"].utils));
})(this, (function (exports, i0, utils) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiCurrencyPipe = /** @class */ (function () {
        function TuiCurrencyPipe() {
        }
        TuiCurrencyPipe.prototype.transform = function (currency) {
            return utils.tuiFormatCurrency(currency);
        };
        return TuiCurrencyPipe;
    }());
    TuiCurrencyPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiCurrencyPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipe, name: "tuiCurrency" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiCurrency',
                    }]
            }] });

    var TuiCurrencyPipeModule = /** @class */ (function () {
        function TuiCurrencyPipeModule() {
        }
        return TuiCurrencyPipeModule;
    }());
    TuiCurrencyPipeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiCurrencyPipeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipeModule, declarations: [TuiCurrencyPipe], exports: [TuiCurrencyPipe] });
    TuiCurrencyPipeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiCurrencyPipeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiCurrencyPipe],
                        exports: [TuiCurrencyPipe],
                    }]
            }] });

    var TuiFormatCardPipe = /** @class */ (function () {
        function TuiFormatCardPipe() {
        }
        TuiFormatCardPipe.prototype.transform = function (value, cardPrefilled) {
            if (value === void 0) { value = ''; }
            if (cardPrefilled === void 0) { cardPrefilled = false; }
            return value && !cardPrefilled
                ? value
                    .split('')
                    .map(function (char, index) { return (index && index % 4 === 0 ? " " + char : char); })
                    .join('')
                : '';
        };
        return TuiFormatCardPipe;
    }());
    TuiFormatCardPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiFormatCardPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardPipe, name: "tuiFormatCard" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: 'tuiFormatCard' }]
            }] });

    var TuiFormatCardModule = /** @class */ (function () {
        function TuiFormatCardModule() {
        }
        return TuiFormatCardModule;
    }());
    TuiFormatCardModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiFormatCardModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardModule, declarations: [TuiFormatCardPipe], exports: [TuiFormatCardPipe] });
    TuiFormatCardModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiFormatCardModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiFormatCardPipe],
                        exports: [TuiFormatCardPipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiCurrencyPipe = TuiCurrencyPipe;
    exports.TuiCurrencyPipeModule = TuiCurrencyPipeModule;
    exports.TuiFormatCardModule = TuiFormatCardModule;
    exports.TuiFormatCardPipe = TuiFormatCardPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
