// @taiga-ui/addon-doc/esm2015/services/theme.service.js
import { Inject, Injectable } from '@angular/core';
import { LOCAL_STORAGE } from '@ng-web-apis/common';
import { BehaviorSubject } from 'rxjs';
import { TUI_THEME_DEFAULT_NAME, TUI_THEME_NAME, TUI_THEME_STORAGE_KEY, } from './theme.options';
import * as i0 from "@angular/core";
export class TuiThemeService extends BehaviorSubject {
    constructor(initialTheme, key, storage) {
        super(storage.getItem(key) || initialTheme);
        this.initialTheme = initialTheme;
        this.key = key;
        this.storage = storage;
    }
    next(theme) {
        this.storage.setItem(this.key, theme);
        super.next(theme);
    }
    get isDefaultTheme() {
        return this.value === TUI_THEME_DEFAULT_NAME;
    }
}
TuiThemeService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeService, deps: [{ token: TUI_THEME_NAME }, { token: TUI_THEME_STORAGE_KEY }, { token: LOCAL_STORAGE }], target: i0.ɵɵFactoryTarget.Injectable });
TuiThemeService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_THEME_NAME]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_THEME_STORAGE_KEY]
                }] }, { type: Storage, decorators: [{
                    type: Inject,
                    args: [LOCAL_STORAGE]
                }] }]; } });
