// @taiga-ui/addon-doc/esm2015/services/theme-night.service.js
import { Inject, Injectable } from '@angular/core';
import { LOCAL_STORAGE, WINDOW } from '@ng-web-apis/common';
import { BehaviorSubject } from 'rxjs';
import { TUI_THEME_NIGHT_STORAGE_KEY, TUI_USE_DEFAULT_NIGHT_THEME, } from './theme-night.options';
import * as i0 from "@angular/core";
export class TuiThemeNightService extends BehaviorSubject {
    constructor(win, storage, key, useDefaultNightTheme) {
        super(storage.getItem(key) === 'true' ||
            (storage.getItem(key) === null &&
                win.matchMedia('(prefers-color-scheme: dark)').matches));
        this.win = win;
        this.storage = storage;
        this.key = key;
        this.useDefaultNightTheme = useDefaultNightTheme;
    }
    next(night) {
        this.storage.setItem(this.key, String(night));
        super.next(night);
    }
    toggle() {
        this.next(!this.value);
    }
}
TuiThemeNightService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeNightService, deps: [{ token: WINDOW }, { token: LOCAL_STORAGE }, { token: TUI_THEME_NIGHT_STORAGE_KEY }, { token: TUI_USE_DEFAULT_NIGHT_THEME }], target: i0.ɵɵFactoryTarget.Injectable });
TuiThemeNightService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeNightService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeNightService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: Storage, decorators: [{
                    type: Inject,
                    args: [LOCAL_STORAGE]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_THEME_NIGHT_STORAGE_KEY]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_USE_DEFAULT_NIGHT_THEME]
                }] }]; } });
