// @taiga-ui/addon-charts/esm2015/components/ring-chart/ring-chart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiPieChartModule } from '@taiga-ui/addon-charts/components/pie-chart';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiRingChartComponent } from './ring-chart.component';
import * as i0 from "@angular/core";
export class TuiRingChartModule {
}
TuiRingChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRingChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, declarations: [TuiRingChartComponent], imports: [CommonModule, PolymorpheusModule, TuiPieChartModule], exports: [TuiRingChartComponent] });
TuiRingChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, imports: [[CommonModule, PolymorpheusModule, TuiPieChartModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRingChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiPieChartModule],
                    declarations: [TuiRingChartComponent],
                    exports: [TuiRingChartComponent],
                }]
        }] });
