// @taiga-ui/cdk/esm2015/directives/click-outside/click-outside.directive.js
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, Inject, NgZone, Output } from '@angular/core';
import { tuiZoneOptimized } from '@taiga-ui/cdk/observables';
import { tuiContainsOrAfter, tuiGetActualTarget } from '@taiga-ui/cdk/utils';
import { fromEvent } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiClickOutsideDirective {
    constructor(zone, doc, el) {
        this.zone = zone;
        this.doc = doc;
        this.el = el;
        this.tuiClickOutside = fromEvent(this.doc, 'mouseup').pipe(map(tuiGetActualTarget), filter(target => this.isOutside(target)), tuiZoneOptimized(this.zone));
    }
    isOutside(target) {
        return (target === this.el.nativeElement ||
            !tuiContainsOrAfter(this.el.nativeElement, target));
    }
}
TuiClickOutsideDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideDirective, deps: [{ token: NgZone }, { token: DOCUMENT }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiClickOutsideDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiClickOutsideDirective, selector: "[tuiClickOutside]", outputs: { tuiClickOutside: "tuiClickOutside" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiClickOutsideDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiClickOutside]',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { tuiClickOutside: [{
                type: Output
            }] } });
