// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-block-details.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiBlockDetailsComponent {
}
TuiBlockDetailsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiBlockDetailsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockDetailsComponent, selector: "ng-component", host: { classAttribute: "tui-block-details-styles" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiBlockDetails]{display:flex;flex-direction:column;align-items:center;font:var(--tui-font-heading-3);min-height:2rem}[tuiBlockDetails] [tuiTitle]{text-align:center;font:var(--tui-font-text-l);grid-gap:0;gap:0}[tuiBlockDetails] [tuiSubtitle]{font:var(--tui-font-text-m);color:var(--tui-text-02);margin-top:.25rem}:host-context(tui-root._mobile) [tuiBlockDetails] [tuiSubtitle]{font:var(--tui-font-text-s)}[tuiBlockDetails] [tuiComment]{margin-top:1.0625rem}[tuiBlockDetails] [tuiAccessory]{display:flex;flex-gap:.5rem;margin-top:1.25rem}[tuiBlockDetails] [tuiAccessories],[tuiBlockDetails] tui-badge:only-of-type{margin-top:1.25rem}[tuiBlockDetails] [tuiAccessories]{display:flex;grid-gap:.5rem;gap:.5rem}[tuiBlockDetails] tui-avatar,[tuiBlockDetails] [tuiTitle]{margin-bottom:1rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./block-details.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-block-details-styles',
                    },
                }]
        }] });

class TuiBlockDetailsDirective {
    constructor(directiveStyles) {
        directiveStyles.addComponent(TuiBlockDetailsComponent);
    }
}
TuiBlockDetailsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiBlockDetailsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockDetailsDirective, selector: "[tuiBlockDetails]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiBlockDetails]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });

class TuiBlockDetailsModule {
}
TuiBlockDetailsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBlockDetailsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsModule, declarations: [TuiBlockDetailsDirective, TuiBlockDetailsComponent], exports: [TuiBlockDetailsDirective] });
TuiBlockDetailsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockDetailsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiBlockDetailsDirective, TuiBlockDetailsComponent],
                    exports: [TuiBlockDetailsDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBlockDetailsComponent, TuiBlockDetailsDirective, TuiBlockDetailsModule };
