// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-button-vertical.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiButtonVerticalComponent {
}
TuiButtonVerticalComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiButtonVerticalComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonVerticalComponent, selector: "ng-component", host: { classAttribute: "tui-button-vertical-styles" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiButton][tuiButtonVertical]{--t-line-height: 1rem;--t-margin: 0rem !important;flex-direction:column;height:auto;padding:.75rem;grid-gap:.375rem;gap:.375rem;min-width:5rem;white-space:pre-line;font:var(--tui-font-text-s-2)}[tuiButton][tuiButtonVertical]>*{max-height:calc(var(--t-line-height) * 2)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./button-vertical.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-button-vertical-styles',
                    },
                }]
        }] });

class TuiButtonVerticalDirective {
    constructor(directiveStyles) {
        directiveStyles.addComponent(TuiButtonVerticalComponent);
    }
}
TuiButtonVerticalDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiButtonVerticalDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonVerticalDirective, selector: "[tuiButtonVertical]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiButtonVertical]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });

class TuiButtonVerticalModule {
}
TuiButtonVerticalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiButtonVerticalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalModule, declarations: [TuiButtonVerticalDirective, TuiButtonVerticalComponent], exports: [TuiButtonVerticalDirective] });
TuiButtonVerticalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonVerticalModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiButtonVerticalDirective, TuiButtonVerticalComponent],
                    exports: [TuiButtonVerticalDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiButtonVerticalComponent, TuiButtonVerticalDirective, TuiButtonVerticalModule };
