// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-pipes-initials.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';

class TuiInitialsPipe {
    transform(text) {
        return text
            .toUpperCase()
            .split(' ')
            .map(([char]) => char)
            .join('')
            .slice(0, 2);
    }
}
TuiInitialsPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiInitialsPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsPipe, name: "tuiInitials" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiInitials',
                }]
        }] });

class TuiInitialsModule {
}
TuiInitialsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInitialsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsModule, declarations: [TuiInitialsPipe], exports: [TuiInitialsPipe] });
TuiInitialsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInitialsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiInitialsPipe],
                    exports: [TuiInitialsPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiInitialsModule, TuiInitialsPipe };
