// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-app-bar.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Inject, ViewChildren, Input, HostBinding, Directive, NgModule } from '@angular/core';
import { MutationObserverService, MUTATION_OBSERVER_INIT } from '@ng-web-apis/mutation-observer';
import { TuiDestroyService, TuiResizeService, EMPTY_QUERY } from '@taiga-ui/cdk';
import * as i3 from 'rxjs';
import { merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { tuiButtonOptionsProvider } from '@taiga-ui/experimental/components/button';
import * as i1 from '@taiga-ui/experimental/directives/fade';
import { TuiFadeModule } from '@taiga-ui/experimental/directives/fade';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import { TuiSvgModule } from '@taiga-ui/core';

const TUI_APP_BAR_PROVIDERS = [
    TuiDestroyService,
    TuiResizeService,
    MutationObserverService,
    {
        provide: MUTATION_OBSERVER_INIT,
        useValue: {
            characterData: true,
            childList: true,
            subtree: true,
        },
    },
    tuiButtonOptionsProvider({
        size: 'm',
        appearance: 'link',
    }),
];

class TuiAppBarComponent {
    constructor(resize$, mutation$) {
        this.resize$ = resize$;
        this.mutation$ = mutation$;
        this.side = EMPTY_QUERY;
        this.size = 'm';
        this.width$ = merge(this.resize$, this.mutation$).pipe(map(() => {
            var _a, _b;
            return 2 *
                Math.max((_a = this.side.first) === null || _a === void 0 ? void 0 : _a.nativeElement.clientWidth, (_b = this.side.last) === null || _b === void 0 ? void 0 : _b.nativeElement.clientWidth);
        }));
    }
}
TuiAppBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, deps: [{ token: TuiResizeService }, { token: MutationObserverService }], target: i0.ɵɵFactoryTarget.Component });
TuiAppBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarComponent, selector: "tui-app-bar", inputs: { size: "size" }, host: { properties: { "attr.data-size": "this.size" } }, providers: TUI_APP_BAR_PROVIDERS, viewQueries: [{ propertyName: "side", predicate: ["side"], descendants: true }], ngImport: i0, template: "<div #side>\n    <ng-content select=\"[tuiSlot='left']\"></ng-content>\n</div>\n<div\n    tuiFade\n    class=\"t-content\"\n    [style.--sides.px]=\"width$ | async\"\n>\n    <ng-content></ng-content>\n</div>\n<div #side>\n    <ng-content select=\"[tuiSlot='right']\"></ng-content>\n</div>\n", styles: ["tui-app-bar{position:relative;display:flex;height:3rem;align-items:center;justify-content:space-between;border:solid transparent;border-width:0 .375rem;box-sizing:border-box;font:var(--tui-font-text-m);font-weight:bold;text-align:center;color:var(--tui-text-01);background:var(--tui-base-01);-webkit-backdrop-filter:blur(2rem);backdrop-filter:blur(2rem)}tui-app-bar>.t-content{left:50%;transform:translate(-50%);position:absolute;display:flex;height:100%;max-width:calc(100% - var(--sides, 0px));flex-direction:column;justify-content:center;flex:1;padding:0 .375rem;box-sizing:border-box;text-align:inherit;white-space:nowrap}tui-app-bar [tuiButton][data-size=m]{font-weight:normal}tui-app-bar [tuiTitle]{text-align:inherit;font:inherit}tui-app-bar [tuiSubtitle]{color:var(--tui-text-02)}tui-app-bar[data-size=m] [tuiTitle]{line-height:1.2em;grid-gap:0;gap:0}tui-app-bar[data-size=l]{border:0;height:4rem}tui-app-bar[data-size=l]>:first-child,tui-app-bar[data-size=l]>:last-child{margin:0 -.75rem}\n"], directives: [{ type: i1.TuiFadeDirective, selector: "[tuiFade]", inputs: ["tuiFadeHeight", "tuiFadeSize", "tuiFadeOffset", "tuiFade"] }], pipes: { "async": i2.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-app-bar',
                    templateUrl: './app-bar.template.html',
                    styleUrls: ['./app-bar.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: TUI_APP_BAR_PROVIDERS,
                }]
        }], ctorParameters: function () { return [{ type: i3.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }, { type: i3.Observable, decorators: [{
                    type: Inject,
                    args: [MutationObserverService]
                }] }]; }, propDecorators: { side: [{
                type: ViewChildren,
                args: ['side']
            }], size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }] } });

class TuiAppBarDirective {
    constructor() {
        this.tuiSlot = 'left';
    }
}
TuiAppBarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiAppBarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSlot]',
                }]
        }], propDecorators: { tuiSlot: [{
                type: Input
            }] } });

class TuiAppBarBackComponent {
}
TuiAppBarBackComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarBackComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAppBarBackComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarBackComponent, selector: "button[tuiAppBarBack], a[tuiAppBarBack]", ngImport: i0, template: "<svg\n    height=\"24\"\n    viewBox=\"0 0 24 24\"\n    width=\"24\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    class=\"t-ios\"\n>\n    <path\n        d=\"M7.44025 12L14.9826 4.43872C15.5391 3.88083 15.5391 2.97631 14.9826 2.41842C14.4261 1.86053 13.5239 1.86053 12.9674 2.41842L4.41737 10.9898C3.86088 11.5477 3.86088 12.4523 4.41737 13.0102L12.9674 21.5816C13.5239 22.1395 14.4261 22.1395 14.9826 21.5816C15.5391 21.0237 15.5391 20.1192 14.9826 19.5613L7.44025 12Z\"\n        fill=\"currentColor\"\n    />\n</svg>\n<svg\n    height=\"24\"\n    viewBox=\"0 0 24 24\"\n    width=\"24\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n    class=\"t-android\"\n>\n    <path\n        d=\"M19.6999 11.5899C19.6999 11.0377 19.2522 10.5899 18.6999 10.5899H7.49992L12.3999 5.68995C12.7865 5.30335 12.7865 4.67655 12.3999 4.28995V4.28995C12.0133 3.90335 11.3865 3.90335 10.9999 4.28995L4.40703 10.8828C4.0165 11.2734 4.01651 11.9065 4.40703 12.2971L10.9999 18.89C11.3865 19.2765 12.0133 19.2765 12.3999 18.8899V18.8899C12.7865 18.5034 12.7865 17.8765 12.3999 17.49L7.49992 12.5899H18.6999C19.2522 12.5899 19.6999 12.1422 19.6999 11.5899V11.5899Z\"\n        fill=\"currentColor\"\n    />\n</svg>\n<ng-content></ng-content>\n", styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;border:0;background:none;font-size:inherit;line-height:inherit;text-decoration:none;display:flex;align-items:center;padding:0 .625rem 0 .125rem;color:var(--tui-link)}:host :host-context([data-platform=\"android\"]){font-size:0;padding:0 .625rem}:host :host-context([data-platform=\"android\"]) .t-ios{display:none}:host :host-context([data-platform=\"android\"]) .t-android{display:block}.t-android{display:none}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarBackComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'button[tuiAppBarBack], a[tuiAppBarBack]',
                    templateUrl: './app-bar-back.template.html',
                    styleUrls: ['./app-bar-back.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiAppBarModule {
}
TuiAppBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAppBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent], imports: [CommonModule, TuiSvgModule, TuiFadeModule], exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent] });
TuiAppBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, imports: [[CommonModule, TuiSvgModule, TuiFadeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiFadeModule],
                    declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                    exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_APP_BAR_PROVIDERS, TuiAppBarBackComponent, TuiAppBarComponent, TuiAppBarDirective, TuiAppBarModule };
