// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-comment.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiCommentComponent {
}
TuiCommentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiCommentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCommentComponent, selector: "ng-component", host: { classAttribute: "tui-comment-styles" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiComment]{position:relative;display:inline-flex;font:var(--tui-font-text-m);color:var(--tui-text-01-night);padding:.5rem .75rem;border-radius:1rem;line-height:1rem;background:var(--tui-accent);align-items:center;justify-content:center}[tuiComment]:before{content:\"\";position:absolute;bottom:100%;left:50%;width:1.5625rem;height:.5625rem;background:inherit;transform:translate(-50%);-webkit-mask:url(\"data:image/svg+xml,%3Csvg width='25' height='9' viewBox='0 0 25 9' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M5.23309 6.67813L7.56191 3.83178C9.4546 1.5185 10.4009 0.361853 11.5998 0.10488C12.0834 0.00123882 12.5834 0.00123882 13.0669 0.10488C14.2658 0.361853 15.2121 1.5185 17.1048 3.83178L19.4337 6.67813C20.636 8.14771 22.4346 9 24.3334 9H0.333374C2.23217 9 4.0307 8.14772 5.23309 6.67813Z' fill='black'/%3E%3C/svg%3E%0A\") no-repeat;mask:url(\"data:image/svg+xml,%3Csvg width='25' height='9' viewBox='0 0 25 9' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M5.23309 6.67813L7.56191 3.83178C9.4546 1.5185 10.4009 0.361853 11.5998 0.10488C12.0834 0.00123882 12.5834 0.00123882 13.0669 0.10488C14.2658 0.361853 15.2121 1.5185 17.1048 3.83178L19.4337 6.67813C20.636 8.14771 22.4346 9 24.3334 9H0.333374C2.23217 9 4.0307 8.14772 5.23309 6.67813Z' fill='black'/%3E%3C/svg%3E%0A\") no-repeat}[tuiComment][data-direction=bottom]:before{top:100%;bottom:auto;transform:translate(-50%) rotate(180deg)}[tuiComment][data-direction=left]:before{left:auto;top:50%;right:100%;transform:translate(.785rem,-50%) rotate(-90deg)}[tuiComment][data-direction=right]:before{top:50%;left:100%;transform:translate(-.785rem,-50%) rotate(90deg)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./comment.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-comment-styles',
                    },
                }]
        }] });

class TuiCommentDirective {
    constructor(directiveStyles) {
        this.tuiComment = 'top';
        directiveStyles.addComponent(TuiCommentComponent);
    }
}
TuiCommentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCommentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCommentDirective, selector: "[tuiComment]", inputs: { tuiComment: "tuiComment" }, host: { properties: { "attr.data-direction": "tuiComment" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiComment]',
                    host: {
                        '[attr.data-direction]': 'tuiComment',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiComment: [{
                type: Input
            }] } });

class TuiCommentModule {
}
TuiCommentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCommentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentModule, declarations: [TuiCommentDirective, TuiCommentComponent], exports: [TuiCommentDirective] });
TuiCommentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiCommentDirective, TuiCommentComponent],
                    exports: [TuiCommentDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiCommentComponent, TuiCommentDirective, TuiCommentModule };
