// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-badge-notification.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Inject, Input, NgModule } from '@angular/core';
import { TUI_ANIMATION_OPTIONS, tuiPop } from '@taiga-ui/core';
import { tuiCreateToken, tuiProvideOptions } from '@taiga-ui/cdk';

const TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS = {
    size: 'm',
};
const TUI_BADGE_NOTIFICATION_OPTIONS = tuiCreateToken(TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS);
function tuiBadgeNotificationOptionsProvider(options) {
    return tuiProvideOptions(TUI_BADGE_NOTIFICATION_OPTIONS, options, TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS);
}

class TuiBadgeNotificationComponent {
    constructor(animation, options) {
        this.animation = animation;
        this.options = options;
        this.size = this.options.size;
    }
}
TuiBadgeNotificationComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationComponent, deps: [{ token: TUI_ANIMATION_OPTIONS }, { token: TUI_BADGE_NOTIFICATION_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiBadgeNotificationComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgeNotificationComponent, selector: "tui-badge-notification", inputs: { size: "size" }, host: { properties: { "attr.data-size": "size", "@tuiPop": "animation" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{--t-size: 1.5rem;position:relative;display:inline-flex;align-items:center;justify-content:center;box-sizing:border-box;color:var(--tui-text-01-night);border-radius:2rem;white-space:nowrap;overflow:hidden;vertical-align:middle;font:var(--tui-font-text-s);max-width:100%;padding:0 .25rem;background:#f52222;height:var(--t-size);min-width:var(--t-size)}:host[data-size=m]{--t-size: 1.25rem}:host[data-size=s]{--t-size: 1rem;padding:0 .125rem;font:var(--tui-font-text-xs)}:host[data-size=xs]{--t-size: .375rem;padding:0;font-size:0}:host :host-context([tuiIconButton]){position:absolute;right:25%;top:25%}\n"], animations: [tuiPop], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-badge-notification',
                    template: '<ng-content></ng-content>',
                    styleUrls: ['./badge-notification.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiPop],
                    host: {
                        '[attr.data-size]': 'size',
                        '[@tuiPop]': 'animation',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATION_OPTIONS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_BADGE_NOTIFICATION_OPTIONS]
                }] }]; }, propDecorators: { size: [{
                type: Input
            }] } });

class TuiBadgeNotificationModule {
}
TuiBadgeNotificationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgeNotificationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationModule, declarations: [TuiBadgeNotificationComponent], exports: [TuiBadgeNotificationComponent] });
TuiBadgeNotificationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgeNotificationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiBadgeNotificationComponent],
                    exports: [TuiBadgeNotificationComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_BADGE_NOTIFICATION_DEFAULT_OPTIONS, TUI_BADGE_NOTIFICATION_OPTIONS, TuiBadgeNotificationComponent, TuiBadgeNotificationModule, tuiBadgeNotificationOptionsProvider };
