// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-button-group.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiButtonGroupComponent {
}
TuiButtonGroupComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiButtonGroupComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonGroupComponent, selector: "ng-component", host: { classAttribute: "tui-button-group-styles" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiButtonGroup]{display:flex;border-radius:var(--tui-radius-xl);justify-content:center;overflow:hidden}[tuiButtonGroup]>button,[tuiButtonGroup]>a{transition-property:background;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:relative;display:flex;border:none;outline:none;background:transparent;align-items:center;flex:1;flex-direction:column;padding:1.125rem .25rem;grid-gap:.5rem;gap:.5rem;font-size:var(--tui-font-text-s);max-width:calc(50% - 1.75rem);cursor:pointer;color:var(--tui-link);text-align:center}[tuiButtonGroup]>button:active,[tuiButtonGroup]>a:active{background:var(--tui-clear)}[tuiButtonGroup]>button:before,[tuiButtonGroup]>a:before,[tuiButtonGroup]>button:after,[tuiButtonGroup]>a:after{position:absolute;top:0;background:inherit;width:1.75rem;height:100%}[tuiButtonGroup]>button:first-child:before,[tuiButtonGroup]>a:first-child:before{content:\"\";left:-1.75rem}[tuiButtonGroup]>button:last-child:after,[tuiButtonGroup]>a:last-child:after{content:\"\";right:-1.75rem}[tuiButtonGroup]>button tui-icon,[tuiButtonGroup]>a tui-icon{font-size:1.75rem}[tuiButtonGroup]:has(button:only-child){border-radius:var(--tui-radius-l)}[tuiButtonGroup]>button:only-child,[tuiButtonGroup]>a:only-child{width:100%;flex-direction:row;font:var(--tui-font-text-l);max-width:100%;justify-content:center}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./button-group.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-button-group-styles',
                    },
                }]
        }] });

class TuiButtonGroupDirective {
    constructor(directiveStyles) {
        directiveStyles.addComponent(TuiButtonGroupComponent);
    }
}
TuiButtonGroupDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiButtonGroupDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonGroupDirective, selector: "[tuiButtonGroup]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiButtonGroup]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });

class TuiButtonGroupModule {
}
TuiButtonGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiButtonGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupModule, declarations: [TuiButtonGroupDirective, TuiButtonGroupComponent], exports: [TuiButtonGroupDirective] });
TuiButtonGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiButtonGroupDirective, TuiButtonGroupComponent],
                    exports: [TuiButtonGroupDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiButtonGroupComponent, TuiButtonGroupDirective, TuiButtonGroupModule };
