// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-compass.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, HostBinding, NgModule } from '@angular/core';

class TuiCompassComponent {
    constructor() {
        this.degrees = NaN;
    }
}
TuiCompassComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiCompassComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCompassComponent, selector: "tui-compass", inputs: { degrees: "degrees" }, host: { properties: { "style.--t-degrees.deg": "this.degrees" } }, ngImport: i0, template: '', isInline: true, styles: [":host{position:relative;display:inline-block;color:var(--tui-primary);border-radius:50%;width:1rem;height:1rem;border:.1875rem solid var(--tui-base-01);box-shadow:0 0 .1875rem #0000004d;transform-style:preserve-3d}:host:before{content:\"\";position:absolute;top:0;left:.125rem;-webkit-mask-image:url('data:image/svg+xml,<svg xmlns=\"http://www.w3.org/2000/svg\"><path d=\"M0 8.99993C4 8 8 8 12 8.99993C9.91509 5.73239 8.5 3 6 0C3.5 3 2 5.5 0 8.99993Z\"/></svg>');mask-image:url('data:image/svg+xml,<svg xmlns=\"http://www.w3.org/2000/svg\"><path d=\"M0 8.99993C4 8 8 8 12 8.99993C9.91509 5.73239 8.5 3 6 0C3.5 3 2 5.5 0 8.99993Z\"/></svg>');transform:rotate(var(--t-degrees)) translateY(-.625rem) translateZ(-1rem);width:.75rem;height:1rem;background:currentColor}:host:after{position:absolute;top:0;left:0;bottom:0;right:0;content:\"\";border-radius:inherit;background:currentColor linear-gradient(-45deg,rgba(0,0,0,.2),transparent)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-compass',
                    template: '',
                    styleUrls: ['./compass.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { degrees: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['style.--t-degrees.deg']
            }] } });

class TuiCompassModule {
}
TuiCompassModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCompassModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassModule, declarations: [TuiCompassComponent], exports: [TuiCompassComponent] });
TuiCompassModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCompassModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiCompassComponent],
                    exports: [TuiCompassComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiCompassComponent, TuiCompassModule };
