// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-card.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiCardComponent {
}
TuiCardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiCardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardComponent, selector: "ng-component", host: { classAttribute: "tui-card" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiCardMedium],[tuiCardLarge]{-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:none;font-size:inherit;line-height:inherit;position:relative;display:flex;align-items:flex-start;flex-shrink:0;text-decoration:none;overscroll-behavior:contain}[tuiCardMedium][tuiTitle],[tuiCardLarge][tuiTitle],[tuiCardMedium] [tuiTitle],[tuiCardLarge] [tuiTitle]{font-weight:bold}[tuiCardMedium] [tuiTitle],[tuiCardLarge] [tuiTitle],[tuiCardMedium] [tuiSubtitle],[tuiCardLarge] [tuiSubtitle]{max-width:100%}[tuiCardMedium]>*,[tuiCardLarge]>*{scrollbar-width:none;-ms-overflow-style:none}[tuiCardMedium]>*::-webkit-scrollbar,[tuiCardLarge]>*::-webkit-scrollbar,[tuiCardMedium]>*::-webkit-scrollbar-thumb,[tuiCardLarge]>*::-webkit-scrollbar-thumb{display:none}[tuiCardMedium]{width:8.75rem;height:8.75rem;flex-direction:column;justify-content:space-between;padding:.75rem;border-radius:var(--tui-radius-l);box-sizing:border-box}[tuiCardMedium][tuiTitle]{padding:.625rem .75rem}[tuiCardMedium] [tuiTitle]{margin:-.125rem 0}[tuiCardMedium] [tuiSubtitle]{color:var(--tui-text-01)}[tuiCardLarge][data-space]{--t-gap: .75rem;--t-radius: var(--tui-radius-l);--t-comp: -.25rem;--t-padding: .75rem;--t-dim: calc(var(--t-padding) + var(--t-comp));font:var(--tui-font-text-m);padding:var(--t-padding);border-radius:var(--t-radius);box-sizing:border-box}[tuiCardLarge][data-space][data-space=normal]{--t-radius: 1.5rem;--t-padding: 1.5rem;--t-gap: 1.5rem}[tuiCardLarge][data-space][data-space=compact]{--t-radius: 1rem;--t-padding: 1.25rem;--t-gap: 1.25rem}[tuiCardLarge][data-space]:not([tuiCell],[tuiHeader]){flex-direction:column;grid-gap:var(--t-gap);gap:var(--t-gap);align-items:stretch}[tuiCardLarge][data-space]:not([tuiCell],[tuiHeader])>:last-child:not([tuiCell]){margin-top:auto}[tuiCardLarge][data-space][tuiHeader]{padding:var(--t-dim) var(--t-dim) var(--t-dim) var(--t-padding)}[tuiCardLarge][data-space][tuiHeader] [tuiLink]:last-child{margin-right:calc(-1 * var(--t-comp))}[tuiCardLarge][data-space]>[tuiHeader]{margin:var(--t-comp) var(--t-comp) calc(1.5 * var(--t-comp)) 0}[tuiCardLarge][data-space]>[tuiHeader] [tuiLink]:last-child{margin-right:calc(-1 * var(--t-comp))}[tuiCardLarge][data-space] [tuiCell]{width:-webkit-fill-available;width:stretch;padding:.5rem;margin:-.5rem -.5rem -.75rem;border-radius:var(--tui-radius-l)}[tuiCardLarge][data-space]>[tuiLink]:last-child{display:block;width:-webkit-fit-content;width:-moz-fit-content;width:fit-content;margin-bottom:var(--t-comp)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./card.styles.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-card',
                    },
                }]
        }] });

class TuiCardMediumDirective {
    constructor(directiveStyles) {
        directiveStyles.addComponent(TuiCardComponent);
    }
}
TuiCardMediumDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardMediumDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCardMediumDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardMediumDirective, selector: "[tuiCardMedium]", host: { attributes: { "tuiCardMedium": "" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardMediumDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCardMedium]',
                    host: {
                        tuiCardMedium: '',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });
class TuiCardLargeDirective {
    constructor(directiveStyles) {
        this.space = 'normal';
        directiveStyles.addComponent(TuiCardComponent);
    }
}
TuiCardLargeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardLargeDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCardLargeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardLargeDirective, selector: "[tuiCardLarge]", inputs: { space: ["tuiCardLarge", "space"] }, host: { attributes: { "tuiCardLarge": "" }, properties: { "attr.data-space": "space || \"normal\"" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardLargeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCardLarge]',
                    host: {
                        tuiCardLarge: '',
                        '[attr.data-space]': 'space || "normal"',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { space: [{
                type: Input,
                args: ['tuiCardLarge']
            }] } });

class TuiCardModule {
}
TuiCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, declarations: [TuiCardComponent, TuiCardLargeDirective, TuiCardMediumDirective], exports: [TuiCardLargeDirective, TuiCardMediumDirective] });
TuiCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiCardComponent, TuiCardLargeDirective, TuiCardMediumDirective],
                    exports: [TuiCardLargeDirective, TuiCardMediumDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiCardComponent, TuiCardLargeDirective, TuiCardMediumDirective, TuiCardModule };
