// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-title.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiTitleComponent {
}
TuiTitleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiTitleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTitleComponent, selector: "ng-component", host: { classAttribute: "tui-title" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiTitle]{position:relative;display:flex;flex-direction:column;text-align:left;grid-gap:.25rem;gap:.25rem;margin:0;font:var(--tui-font-text-m-2)}[tuiTitle][data-size=s]{grid-gap:.125rem;gap:.125rem;font:var(--tui-font-text-s)}[tuiTitle][data-size=s] [tuiSubtitle]{font:var(--tui-font-text-xs)}[tuiTitle][data-size=m]{grid-gap:.125rem;gap:.125rem;font:var(--tui-font-heading-5)}[tuiTitle][data-size=m] [tuiSubtitle]{font:var(--tui-font-text-m)}[tuiTitle][data-size=l]{grid-gap:.5rem;gap:.5rem;font:var(--tui-font-heading-3)}[tuiTitle][data-size=l] [tuiSubtitle]{font:var(--tui-font-text-m)}[tuiSubtitle]{font:var(--tui-font-text-s-2)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./title.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-title',
                    },
                }]
        }] });

class TuiTitleDirective {
    constructor(directiveStyles) {
        this.size = '';
        directiveStyles.addComponent(TuiTitleComponent);
    }
}
TuiTitleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiTitleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTitleDirective, selector: "[tuiTitle]", inputs: { size: ["tuiTitle", "size"] }, host: { attributes: { "tuiTitle": "" }, properties: { "attr.data-size": "size || null" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTitle]',
                    host: {
                        tuiTitle: '',
                        '[attr.data-size]': 'size || null',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { size: [{
                type: Input,
                args: ['tuiTitle']
            }] } });

class TuiTitleModule {
}
TuiTitleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTitleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleModule, declarations: [TuiTitleDirective, TuiTitleComponent], exports: [TuiTitleDirective] });
TuiTitleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTitleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiTitleDirective, TuiTitleComponent],
                    exports: [TuiTitleDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiTitleComponent, TuiTitleDirective, TuiTitleModule };
