// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-badged-content.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Directive, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiResizeModule } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

class TuiBadgedContentComponent {
}
TuiBadgedContentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiBadgedContentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgedContentComponent, selector: "tui-badged-content", ngImport: i0, template: "<ng-content></ng-content>\n\n<div\n    #badgeTop\n    class=\"t-badge t-badge_top\"\n    [style.--t-badge-height.px]=\"badgeTop.offsetHeight\"\n    (tuiResize)=\"(0)\"\n>\n    <ng-content select=\"tui-badge-notification[tuiSlot='top']\"></ng-content>\n    <div class=\"t-border\">\n        <ng-content select=\"[tuiSlot='top']\"></ng-content>\n    </div>\n</div>\n\n<div\n    #badgeBottom\n    class=\"t-badge t-badge_bottom\"\n    [style.--t-badge-height.px]=\"badgeBottom.offsetHeight\"\n    (tuiResize)=\"(0)\"\n>\n    <ng-content select=\"tui-badge-notification[tuiSlot='bottom']\"></ng-content>\n    <div class=\"t-border\">\n        <ng-content select=\"[tuiSlot='bottom']\"></ng-content>\n    </div>\n</div>\n", styles: [":host{--tui-radius: var(--tui-radius-l);--t-badge-height: 0;--t-corner-offset: calc((var(--tui-radius) * 1.4142 - var(--tui-radius)) * 1 / 1.4142);position:relative;display:inline-block;height:-webkit-fit-content;height:-moz-fit-content;height:fit-content;width:-webkit-max-content;width:max-content;color:var(--tui-base-01)}.t-badge{--t-offset: calc(var(--t-badge-height) * -.5 + var(--t-corner-offset));position:absolute;display:flex}.t-badge_top{top:var(--t-offset);right:var(--t-offset)}.t-badge_bottom{bottom:var(--t-offset);right:var(--t-offset)}.t-border:not(:empty){display:flex;border-radius:6rem;box-shadow:0 0 0 1px currentColor;background:currentColor}\n"], directives: [{ type: i1.TuiResizeDirective, selector: "[tuiResize]", outputs: ["tuiResize"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-badged-content',
                    templateUrl: './badged-content.template.html',
                    styleUrls: ['./badged-content.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiBadgedContentDirective {
    constructor() {
        this.tuiSlot = 'top';
    }
}
TuiBadgedContentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiBadgedContentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgedContentDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSlot]',
                }]
        }], propDecorators: { tuiSlot: [{
                type: Input
            }] } });

class TuiBadgedContentModule {
}
TuiBadgedContentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgedContentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective], imports: [CommonModule, TuiResizeModule], exports: [TuiBadgedContentComponent, TuiBadgedContentDirective] });
TuiBadgedContentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, imports: [[CommonModule, TuiResizeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiResizeModule],
                    declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                    exports: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBadgedContentComponent, TuiBadgedContentDirective, TuiBadgedContentModule };
