// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-pipes-fallback-src.js
import * as i0 from '@angular/core';
import { ElementRef, Pipe, Inject, NgModule } from '@angular/core';
import { TUI_ICON_ERROR } from '@taiga-ui/core';
import { merge, fromEvent } from 'rxjs';
import { map, startWith } from 'rxjs/operators';

class TuiFallbackSrcPipe {
    constructor(el) {
        this.el = el;
    }
    transform(src, fallback) {
        return merge(fromEvent(this.el.nativeElement, TUI_ICON_ERROR), fromEvent(this.el.nativeElement, 'error', { capture: true })).pipe(map(() => fallback), startWith(src));
    }
}
TuiFallbackSrcPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Pipe });
TuiFallbackSrcPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, name: "tuiFallbackSrc" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiFallbackSrc',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; } });

class TuiFallbackSrcModule {
}
TuiFallbackSrcModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiFallbackSrcModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcModule, declarations: [TuiFallbackSrcPipe], exports: [TuiFallbackSrcPipe] });
TuiFallbackSrcModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiFallbackSrcPipe],
                    exports: [TuiFallbackSrcPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFallbackSrcModule, TuiFallbackSrcPipe };
