// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-pipes-auto-color.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import { tuiStringHashToHsl } from '@taiga-ui/kit';

class TuiAutoColorPipe {
    transform(text) {
        return tuiStringHashToHsl(text);
    }
}
TuiAutoColorPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
TuiAutoColorPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorPipe, name: "tuiAutoColor" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiAutoColor',
                }]
        }] });

class TuiAutoColorModule {
}
TuiAutoColorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAutoColorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorModule, declarations: [TuiAutoColorPipe], exports: [TuiAutoColorPipe] });
TuiAutoColorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutoColorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiAutoColorPipe],
                    exports: [TuiAutoColorPipe],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAutoColorModule, TuiAutoColorPipe };
