// @taiga-ui/addon-mobile/esm2015/components/mobile-dialog/mobile-dialog.service.js
import { inject, Injectable } from '@angular/core';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiMobileDialogComponent } from './mobile-dialog.component';
import { TUI_MOBILE_DIALOG_OPTIONS } from './mobile-dialog.options';
import * as i0 from "@angular/core";
const DIALOG = new PolymorpheusComponent(TuiMobileDialogComponent);
export class TuiMobileDialogService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.component = DIALOG;
        this.defaultOptions = Object.assign(Object.assign({}, inject(TUI_MOBILE_DIALOG_OPTIONS)), { data: undefined });
    }
    open(content, options = {}) {
        return super.open(content, options);
    }
}
TuiMobileDialogService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogService, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
TuiMobileDialogService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
