// @taiga-ui/addon-mobile/esm2015/components/mobile-dialog/mobile-dialog.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRippleModule } from '@taiga-ui/addon-mobile/directives/ripple';
import { tuiAsDialog } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiMobileDialogComponent } from './mobile-dialog.component';
import { TuiMobileDialogService } from './mobile-dialog.service';
import * as i0 from "@angular/core";
export class TuiMobileDialogModule {
}
TuiMobileDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMobileDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogModule, declarations: [TuiMobileDialogComponent], imports: [CommonModule, TuiButtonModule, TuiRippleModule, PolymorpheusModule], exports: [TuiMobileDialogComponent] });
TuiMobileDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogModule, providers: [tuiAsDialog(TuiMobileDialogService)], imports: [[CommonModule, TuiButtonModule, TuiRippleModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiButtonModule, TuiRippleModule, PolymorpheusModule],
                    declarations: [TuiMobileDialogComponent],
                    providers: [tuiAsDialog(TuiMobileDialogService)],
                    exports: [TuiMobileDialogComponent],
                }]
        }] });
