// @taiga-ui/experimental/esm2015/components/avatar-stack/avatar-stack.module.js
import { NgModule } from '@angular/core';
import { TuiAvatarComponent, TuiAvatarModule, } from '@taiga-ui/experimental/components/avatar';
import { TuiAvatarStackComponent } from './avatar-stack.component';
import * as i0 from "@angular/core";
export class TuiAvatarStackModule {
}
TuiAvatarStackModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarStackModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAvatarStackModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarStackModule, declarations: [TuiAvatarStackComponent], imports: [TuiAvatarModule], exports: [TuiAvatarStackComponent, TuiAvatarComponent] });
TuiAvatarStackModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarStackModule, imports: [[TuiAvatarModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarStackModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiAvatarModule],
                    declarations: [TuiAvatarStackComponent],
                    exports: [TuiAvatarStackComponent, TuiAvatarComponent],
                }]
        }] });
