// @tinkoff/ng-dompurify/esm2015/lib/ng-dompurify.pipe.js
import { Pipe, SecurityContext } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ng-dompurify.service";
import * as i2 from "@angular/platform-browser";
/**
 * Pipe that transforms dirty content to clean via {@link NgDompurifySanitizer}
 */
export class NgDompurifyPipe {
    constructor(sanitizer, domSanitizer) {
        this.sanitizer = sanitizer;
        this.domSanitizer = domSanitizer;
    }
    transform(value, context = SecurityContext.HTML, config) {
        return this.bypassSecurityTrust(context, this.sanitizer.sanitize(context, value, config));
    }
    bypassSecurityTrust(context, purifiedValue) {
        switch (context) {
            case SecurityContext.HTML:
                return this.domSanitizer.bypassSecurityTrustHtml(purifiedValue);
            case SecurityContext.STYLE:
                return this.domSanitizer.bypassSecurityTrustStyle(purifiedValue);
            case SecurityContext.URL:
                return this.domSanitizer.bypassSecurityTrustUrl(purifiedValue);
            case SecurityContext.RESOURCE_URL:
                return this.domSanitizer.bypassSecurityTrustResourceUrl(purifiedValue);
            default:
                return null;
        }
    }
}
NgDompurifyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgDompurifyPipe, deps: [{ token: i1.NgDompurifySanitizer }, { token: i2.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
NgDompurifyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgDompurifyPipe, name: "dompurify" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: NgDompurifyPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'dompurify' }]
        }], ctorParameters: function () { return [{ type: i1.NgDompurifySanitizer }, { type: i2.DomSanitizer }]; } });
