// @taiga-ui/core/bundles/taiga-ui-core-components-label.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/core/providers'), require('@taiga-ui/core/tokens'), require('@tinkoff/ng-polymorpheus'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/label', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/core/providers', '@taiga-ui/core/tokens', '@tinkoff/ng-polymorpheus', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components.label = {}), global.ng.core, global.ng.forms, global["taiga-ui"].core.providers, global["taiga-ui"].core.tokens, global.i1, global.rxjs));
})(this, (function (exports, i0, forms, providers, tokens, i1, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TuiLabelComponent = /** @class */ (function () {
        function TuiLabelComponent(mode$) {
            this.mode$ = mode$;
            this.context = {
                $implicit: null,
            };
        }
        return TuiLabelComponent;
    }());
    TuiLabelComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelComponent, deps: [{ token: tokens.TUI_MODE }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiLabelComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLabelComponent, selector: "label[tuiLabel]", inputs: { tuiLabel: "tuiLabel", context: "context" }, host: { listeners: { "$.data-mode.attr": "mode$" }, properties: { "class._control": "this.control" } }, providers: [providers.MODE_PROVIDER], queries: [{ propertyName: "control", first: true, predicate: forms.NgControl, descendants: true }], ngImport: i0__namespace, template: "<span class=\"t-label\">\n    <span\n        *polymorpheusOutlet=\"tuiLabel as text; context: context\"\n        class=\"t-text\"\n        [textContent]=\"text\"\n    ></span>\n</span>\n<span class=\"t-content\">\n    <ng-content></ng-content>\n</span>\n", styles: [":host{display:flex;min-height:2.75rem;justify-content:space-between;flex-direction:column;pointer-events:none}.t-label{font:var(--tui-font-text-s);max-width:100%;align-self:inherit;overflow:inherit;pointer-events:auto}:host._control .t-label{margin-bottom:.25rem}.t-text{display:block;overflow:inherit;text-overflow:ellipsis}:host:not(._control) .t-text{color:var(--tui-text-02)}:host([data-mode=\"onDark\"]) .t-text{color:var(--tui-text-02-night)}.t-content{line-height:1.5rem;overflow:inherit;text-overflow:ellipsis;pointer-events:auto}\n"], directives: [{ type: i1__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'label[tuiLabel]',
                        templateUrl: './label.template.html',
                        styleUrls: ['./label.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [providers.MODE_PROVIDER],
                        host: {
                            '($.data-mode.attr)': 'mode$',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_MODE]
                        }] }];
        }, propDecorators: { tuiLabel: [{
                    type: i0.Input
                }], context: [{
                    type: i0.Input
                }], control: [{
                    type: i0.ContentChild,
                    args: [forms.NgControl]
                }, {
                    type: i0.HostBinding,
                    args: ['class._control']
                }] } });

    var TuiLabelModule = /** @class */ (function () {
        function TuiLabelModule() {
        }
        return TuiLabelModule;
    }());
    TuiLabelModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiLabelModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelModule, declarations: [TuiLabelComponent], imports: [i1.PolymorpheusModule], exports: [TuiLabelComponent] });
    TuiLabelModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelModule, imports: [[i1.PolymorpheusModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLabelModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.PolymorpheusModule],
                        declarations: [TuiLabelComponent],
                        exports: [TuiLabelComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiLabelComponent = TuiLabelComponent;
    exports.TuiLabelModule = TuiLabelModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
