// @taiga-ui/core/bundles/taiga-ui-core-internal-primitive-year-month-pagination.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/components/primitive-spin-button'), require('@taiga-ui/core/components/link'), require('@angular/common'), require('@taiga-ui/core/pipes')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/internal/primitive-year-month-pagination', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/components/primitive-spin-button', '@taiga-ui/core/components/link', '@angular/common', '@taiga-ui/core/pipes'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.internal = global["taiga-ui"].core.internal || {}, global["taiga-ui"].core.internal["primitive-year-month-pagination"] = {}), global.ng.core, global.i4, global["taiga-ui"].core.components["primitive-spin-button"], global["taiga-ui"].core.components.link, global.ng.common, global["taiga-ui"].core.pipes));
})(this, (function (exports, i0, i4, i1, i2, i3, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    var TuiPrimitiveYearMonthPaginationComponent = /** @class */ (function () {
        function TuiPrimitiveYearMonthPaginationComponent() {
            this.value = i4.TuiMonth.currentLocal();
            this.min = i4.TUI_FIRST_DAY;
            this.max = i4.TUI_LAST_DAY;
            this.valueChange = new i0.EventEmitter();
            this.yearClick = new i0.EventEmitter();
        }
        Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "computedMin", {
            get: function () {
                var _a;
                return (_a = this.min) !== null && _a !== void 0 ? _a : i4.TUI_FIRST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "computedMax", {
            get: function () {
                var _a;
                return (_a = this.max) !== null && _a !== void 0 ? _a : i4.TUI_LAST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "prevMonthDisabled", {
            get: function () {
                return this.value.monthSameOrBefore(this.computedMin);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "nextMonthDisabled", {
            get: function () {
                return this.value.monthSameOrAfter(this.computedMax);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "oneYear", {
            get: function () {
                var _b = this, computedMin = _b.computedMin, computedMax = _b.computedMax;
                return computedMin.year === computedMax.year;
            },
            enumerable: false,
            configurable: true
        });
        TuiPrimitiveYearMonthPaginationComponent.prototype.onYearClick = function () {
            this.yearClick.next(this.value);
        };
        TuiPrimitiveYearMonthPaginationComponent.prototype.onPrevMonthClick = function () {
            this.appendValueWithLimit({ month: -1 });
        };
        TuiPrimitiveYearMonthPaginationComponent.prototype.onNextMonthClick = function () {
            this.appendValueWithLimit({ month: 1 });
        };
        TuiPrimitiveYearMonthPaginationComponent.prototype.appendValueWithLimit = function (date) {
            var newMonth = this.value.append(date);
            var _b = this, computedMin = _b.computedMin, computedMax = _b.computedMax;
            if (computedMin.monthSameOrAfter(newMonth)) {
                this.updateValue(computedMin);
                return;
            }
            this.updateValue(computedMax.monthSameOrBefore(newMonth) ? computedMax : newMonth);
        };
        TuiPrimitiveYearMonthPaginationComponent.prototype.updateValue = function (value) {
            if (this.value.monthSame(value)) {
                return;
            }
            this.value = value;
            this.valueChange.emit(value);
        };
        return TuiPrimitiveYearMonthPaginationComponent;
    }());
    TuiPrimitiveYearMonthPaginationComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPrimitiveYearMonthPaginationComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPrimitiveYearMonthPaginationComponent, selector: "tui-primitive-year-month-pagination", inputs: { value: "value", min: "min", max: "max" }, outputs: { valueChange: "valueChange", yearClick: "yearClick" }, ngImport: i0__namespace, template: "<tui-primitive-spin-button\n    [focusable]=\"false\"\n    [leftDisabled]=\"prevMonthDisabled\"\n    [rightDisabled]=\"nextMonthDisabled\"\n    (leftClick)=\"onPrevMonthClick()\"\n    (rightClick)=\"onNextMonthClick()\"\n>\n    {{ value | tuiMonth | async }}\n    <ng-container *ngIf=\"oneYear; else button\">\n        {{ value.formattedYear }}\n    </ng-container>\n    <ng-template #button>\n        <button\n            id=\"year-btn\"\n            automation-id=\"tui-primitive-year-month-pagination__year-button\"\n            tuiLink\n            type=\"button\"\n            [tuiFocusable]=\"false\"\n            (click)=\"onYearClick()\"\n        >\n            {{ value.formattedYear }}\n        </button>\n    </ng-template>\n</tui-primitive-spin-button>\n", styles: [":host{display:block}\n"], components: [{ type: i1__namespace.TuiPrimitiveSpinButtonComponent, selector: "tui-primitive-spin-button", inputs: ["disabled", "leftDisabled", "rightDisabled"], outputs: ["leftClick", "rightClick"] }, { type: i2__namespace.TuiLinkComponent, selector: "a[tuiLink], button[tuiLink]", inputs: ["pseudo", "icon", "iconAlign", "iconRotated", "mode"], exportAs: ["tuiLink"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.TuiFocusableDirective, selector: "[tuiFocusable]", inputs: ["tuiFocusable"] }], pipes: { "async": i3__namespace.AsyncPipe, "tuiMonth": i5__namespace.TuiMonthPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-primitive-year-month-pagination',
                        templateUrl: './primitive-year-month-pagination.template.html',
                        styleUrls: ['./primitive-year-month-pagination.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { value: [{
                    type: i0.Input
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], valueChange: [{
                    type: i0.Output
                }], yearClick: [{
                    type: i0.Output
                }] } });

    var TuiPrimitiveYearMonthPaginationModule = /** @class */ (function () {
        function TuiPrimitiveYearMonthPaginationModule() {
        }
        return TuiPrimitiveYearMonthPaginationModule;
    }());
    TuiPrimitiveYearMonthPaginationModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPrimitiveYearMonthPaginationModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationModule, declarations: [TuiPrimitiveYearMonthPaginationComponent], imports: [i3.CommonModule,
            i4.TuiFocusableModule,
            i1.TuiPrimitiveSpinButtonModule,
            i2.TuiLinkModule,
            i5.TuiMonthPipeModule], exports: [TuiPrimitiveYearMonthPaginationComponent] });
    TuiPrimitiveYearMonthPaginationModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationModule, imports: [[
                i3.CommonModule,
                i4.TuiFocusableModule,
                i1.TuiPrimitiveSpinButtonModule,
                i2.TuiLinkModule,
                i5.TuiMonthPipeModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearMonthPaginationModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            i4.TuiFocusableModule,
                            i1.TuiPrimitiveSpinButtonModule,
                            i2.TuiLinkModule,
                            i5.TuiMonthPipeModule,
                        ],
                        declarations: [TuiPrimitiveYearMonthPaginationComponent],
                        exports: [TuiPrimitiveYearMonthPaginationComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPrimitiveYearMonthPaginationComponent = TuiPrimitiveYearMonthPaginationComponent;
    exports.TuiPrimitiveYearMonthPaginationModule = TuiPrimitiveYearMonthPaginationModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
