// @taiga-ui/core/bundles/taiga-ui-core-components-primitive-year-picker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/enums'), require('@taiga-ui/core/directives/scroll-into-view')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/primitive-year-picker', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/enums', '@taiga-ui/core/directives/scroll-into-view'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components["primitive-year-picker"] = {}), global.ng.core, global.i1, global["taiga-ui"].core.enums, global["taiga-ui"].core.directives["scroll-into-view"]));
})(this, (function (exports, i0, i1, enums, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var LIMIT = 100;
    var ITEMS_IN_ROW = 4;
    var TuiPrimitiveYearPickerComponent = /** @class */ (function () {
        function TuiPrimitiveYearPickerComponent() {
            this.hoveredItem = null;
            this.pressedItem = null;
            this.currentYear = i1.TuiMonth.currentLocal().year;
            this.value = null;
            this.initialItem = i1.TuiMonth.currentLocal();
            this.min = i1.TUI_FIRST_DAY;
            this.max = i1.TUI_LAST_DAY;
            this.disabledItemHandler = i1.ALWAYS_FALSE_HANDLER;
            this.yearClick = new i0.EventEmitter();
        }
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "computedMin", {
            get: function () {
                var _a;
                return (_a = this.min) !== null && _a !== void 0 ? _a : i1.TUI_FIRST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "computedMax", {
            get: function () {
                var _a;
                return (_a = this.max) !== null && _a !== void 0 ? _a : i1.TUI_LAST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "isSingle", {
            get: function () {
                return this.isRange(this.value) && this.value.from.yearSame(this.value.to);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "rows", {
            get: function () {
                return Math.ceil((this.calculatedMax - this.calculatedMin) / ITEMS_IN_ROW);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "calculatedMin", {
            get: function () {
                var initial = this.initialItem.year - LIMIT;
                var min = this.computedMin;
                return min.year > initial ? min.year : initial;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveYearPickerComponent.prototype, "calculatedMax", {
            get: function () {
                var initial = this.initialItem.year + LIMIT;
                var max = this.computedMax;
                return max.year < initial ? max.year + 1 : initial;
            },
            enumerable: false,
            configurable: true
        });
        TuiPrimitiveYearPickerComponent.prototype.isRange = function (item) {
            return item instanceof i1.TuiMonthRange;
        };
        TuiPrimitiveYearPickerComponent.prototype.scrollItemIntoView = function (item) {
            return this.initialItem.year === item;
        };
        TuiPrimitiveYearPickerComponent.prototype.getItem = function (rowIndex, colIndex) {
            return rowIndex * ITEMS_IN_ROW + colIndex + this.calculatedMin;
        };
        TuiPrimitiveYearPickerComponent.prototype.getItemState = function (item) {
            var _b = this, disabledItemHandler = _b.disabledItemHandler, pressedItem = _b.pressedItem, hoveredItem = _b.hoveredItem;
            var max = this.computedMax;
            if (max.year < item ||
                (disabledItemHandler !== i1.ALWAYS_FALSE_HANDLER && disabledItemHandler(item))) {
                return enums.TuiInteractiveState.Disabled;
            }
            if (pressedItem === item) {
                return enums.TuiInteractiveState.Active;
            }
            if (hoveredItem === item) {
                return enums.TuiInteractiveState.Hover;
            }
            return null;
        };
        TuiPrimitiveYearPickerComponent.prototype.getItemRange = function (item) {
            var _b = this, value = _b.value, hoveredItem = _b.hoveredItem;
            if (value === null) {
                return null;
            }
            if (value instanceof i1.TuiYear) {
                return value.year === item ? enums.TuiRangeState.Single : null;
            }
            if (!(value instanceof i1.TuiMonthRange)) {
                return value.find(function (day) { return day.year === item; }) ? enums.TuiRangeState.Single : null;
            }
            if ((value.from.year === item && !value.from.yearSame(value.to)) ||
                (hoveredItem !== null &&
                    hoveredItem > value.from.year &&
                    value.from.year === item &&
                    value.from.yearSame(value.to)) ||
                (hoveredItem !== null &&
                    hoveredItem === item &&
                    hoveredItem < value.from.year &&
                    value.from.yearSame(value.to))) {
                return enums.TuiRangeState.Start;
            }
            if ((value.to.year === item && !value.from.yearSame(value.to)) ||
                (hoveredItem !== null &&
                    hoveredItem < value.from.year &&
                    value.from.year === item &&
                    value.from.yearSame(value.to)) ||
                (hoveredItem !== null &&
                    hoveredItem === item &&
                    hoveredItem > value.from.year &&
                    value.from.yearSame(value.to))) {
                return enums.TuiRangeState.End;
            }
            return value.from.yearSame(value.to) && value.from.year === item
                ? enums.TuiRangeState.Single
                : null;
        };
        TuiPrimitiveYearPickerComponent.prototype.itemIsToday = function (item) {
            return this.currentYear === item;
        };
        TuiPrimitiveYearPickerComponent.prototype.itemIsInterval = function (item) {
            var _b = this, value = _b.value, hoveredItem = _b.hoveredItem;
            if (value === null || !this.isRange(value)) {
                return false;
            }
            if (!value.from.yearSame(value.to)) {
                return value.from.year <= item && value.to.year > item;
            }
            if (hoveredItem === null || value.from.year === hoveredItem) {
                return false;
            }
            return i1.tuiInRange(item, Math.min(value.from.year, hoveredItem), Math.max(value.from.year, hoveredItem));
        };
        TuiPrimitiveYearPickerComponent.prototype.onItemHovered = function (hovered, item) {
            this.updateHoveredItem(hovered, item);
        };
        TuiPrimitiveYearPickerComponent.prototype.onItemPressed = function (pressed, item) {
            this.updatePressedItem(pressed, item);
        };
        TuiPrimitiveYearPickerComponent.prototype.onItemClick = function (item) {
            this.yearClick.emit(new i1.TuiYear(item));
        };
        TuiPrimitiveYearPickerComponent.prototype.updateHoveredItem = function (hovered, item) {
            this.hoveredItem = hovered ? item : null;
        };
        TuiPrimitiveYearPickerComponent.prototype.updatePressedItem = function (pressed, item) {
            this.pressedItem = pressed ? item : null;
        };
        return TuiPrimitiveYearPickerComponent;
    }());
    TuiPrimitiveYearPickerComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPrimitiveYearPickerComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPrimitiveYearPickerComponent, selector: "tui-primitive-year-picker", inputs: { value: "value", initialItem: "initialItem", min: "min", max: "max", disabledItemHandler: "disabledItemHandler" }, outputs: { yearClick: "yearClick" }, host: { properties: { "class._single": "this.isSingle" } }, ngImport: i0__namespace, template: "<div\n    *tuiRepeatTimes=\"let rowIndex of rows\"\n    automation-id=\"tui-primitive-year-picker__row\"\n    class=\"t-row\"\n>\n    <ng-container *tuiRepeatTimes=\"let colIndex of 4\">\n        <div\n            *tuiLet=\"getItem(rowIndex, colIndex) as item\"\n            automation-id=\"tui-primitive-year-picker__cell\"\n            class=\"t-cell\"\n            [attr.data-range]=\"getItemRange(item)\"\n            [attr.data-state]=\"getItemState(item)\"\n            [class.t-cell_interval]=\"itemIsInterval(item)\"\n            [class.t-cell_today]=\"itemIsToday(item)\"\n            [tuiScrollIntoView]=\"scrollItemIntoView(item)\"\n            (click)=\"onItemClick(item)\"\n            (tuiHoveredChange)=\"onItemHovered($event, item)\"\n            (tuiPressedChange)=\"onItemPressed($event, item)\"\n        >\n            <div class=\"t-item\">{{ item }}</div>\n        </div>\n    </ng-container>\n</div>\n", styles: [":host{display:block;font:var(--tui-font-text-m)}.t-row{position:relative;display:flex;justify-content:space-between;height:2.25rem;isolation:isolate}.t-item{position:relative;flex:1;line-height:2rem;border-radius:var(--tui-radius-m)}.t-item:before,.t-item:after{content:\"\";position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.t-item:after{border-radius:.5rem}.t-cell{position:relative;display:flex;align-items:center;justify-content:center;width:3.9375rem;text-align:center;outline:none;cursor:pointer;background-clip:content-box;box-sizing:border-box;border:.125rem solid transparent}.t-cell:before{content:\"\";position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.t-cell_today:after{position:absolute;left:50%;transform:translate(-50%);content:\"\";bottom:.3125rem;height:.125rem;width:.75rem;border-radius:.375rem;background-color:var(--tui-text-01)}.t-cell_interval:before{background:var(--tui-base-02)}:host._single .t-cell_interval:before{background:var(--tui-secondary-hover)}.t-cell_interval:not(:last-child):before{right:-.25rem;border-top-right-radius:0;border-bottom-right-radius:0}.t-cell_interval:not([data-range=\"start\"]):not(:first-child):before{border-top-left-radius:0;border-bottom-left-radius:0}.t-cell_interval:last-child:first-child:before{right:0}.t-cell_interval:first-child>.t-item{border-top-left-radius:var(--tui-radius-m);border-bottom-left-radius:var(--tui-radius-m)}.t-cell_interval:last-child>.t-item{border-top-right-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.t-cell_interval>.t-item{border-radius:0}.t-cell[data-range]:after{background-color:var(--tui-primary-text)}.t-cell[data-range]>.t-item{color:var(--tui-primary-text)}.t-cell[data-range]>.t-item:before,.t-cell[data-range]>.t-item:after{background-color:var(--tui-primary)}.t-cell[data-range][data-state=hover]>.t-item:before,.t-cell[data-range][data-state=hover]>.t-item:after{background-color:var(--tui-primary-hover)}.t-cell[data-range][data-state=active]>.t-item:before,.t-cell[data-range][data-state=active]>.t-item:after{background-color:var(--tui-primary-active)}.t-cell[data-range=end]:before{background:var(--tui-base-02)}:host._single .t-cell[data-range=end]:before{background:var(--tui-secondary-hover)}.t-cell[data-range=end]:not(:first-child):before{border-top-left-radius:0;border-bottom-left-radius:0}.t-cell[data-range=end]>.t-item:before{left:.625rem;border-top-left-radius:0;border-bottom-left-radius:0}.t-cell[data-range=end]>.t-item:after{left:-2rem;right:100%;transform:translate(1.6rem) scaleY(.83) scaleX(.5) rotate(45deg)}.t-cell[data-range=start]>.t-item:before{right:.625rem;border-top-right-radius:0;border-bottom-right-radius:0}.t-cell[data-range=start]>.t-item:after{left:100%;right:-2rem;transform:translate(-1.6rem) scaleY(.83) scaleX(.5) rotate(45deg)}.t-cell[data-range=single]>.t-item:after{display:none}.t-cell[data-state=disabled]{pointer-events:none}.t-cell[data-state=disabled]>.t-item{opacity:.36}.t-cell[data-state=hover]:hover:not([data-range])>.t-item{background-color:var(--tui-secondary-hover)}.t-cell[data-state=active]:hover:not([data-range])>.t-item{background-color:var(--tui-secondary-active)}:host{width:15.75rem}.t-row{margin:.875rem 0}.t-row:first-child{margin-top:0}.t-row:last-child{margin-bottom:0}\n"], directives: [{ type: i1__namespace.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }, { type: i1__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2__namespace.TuiScrollIntoViewDirective, selector: "[tuiScrollIntoView]", inputs: ["tuiScrollIntoView"] }, { type: i1__namespace.TuiHoveredDirective, selector: "[tuiHoveredChange]", outputs: ["tuiHoveredChange"] }, { type: i1__namespace.TuiPressedDirective, selector: "[tuiPressedChange]", outputs: ["tuiPressedChange"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-primitive-year-picker',
                        templateUrl: './primitive-year-picker.template.html',
                        styleUrls: ['./primitive-year-picker.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { value: [{
                    type: i0.Input
                }], initialItem: [{
                    type: i0.Input
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], disabledItemHandler: [{
                    type: i0.Input
                }], yearClick: [{
                    type: i0.Output
                }], isSingle: [{
                    type: i0.HostBinding,
                    args: ['class._single']
                }] } });

    var TuiPrimitiveYearPickerModule = /** @class */ (function () {
        function TuiPrimitiveYearPickerModule() {
        }
        return TuiPrimitiveYearPickerModule;
    }());
    TuiPrimitiveYearPickerModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPrimitiveYearPickerModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerModule, declarations: [TuiPrimitiveYearPickerComponent], imports: [i1.TuiHoveredModule,
            i1.TuiPressedModule,
            i1.TuiRepeatTimesModule,
            i1.TuiLetModule,
            i2.TuiScrollIntoViewModule], exports: [TuiPrimitiveYearPickerComponent] });
    TuiPrimitiveYearPickerModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerModule, imports: [[
                i1.TuiHoveredModule,
                i1.TuiPressedModule,
                i1.TuiRepeatTimesModule,
                i1.TuiLetModule,
                i2.TuiScrollIntoViewModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveYearPickerModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i1.TuiHoveredModule,
                            i1.TuiPressedModule,
                            i1.TuiRepeatTimesModule,
                            i1.TuiLetModule,
                            i2.TuiScrollIntoViewModule,
                        ],
                        declarations: [TuiPrimitiveYearPickerComponent],
                        exports: [TuiPrimitiveYearPickerComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPrimitiveYearPickerComponent = TuiPrimitiveYearPickerComponent;
    exports.TuiPrimitiveYearPickerModule = TuiPrimitiveYearPickerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
