// @taiga-ui/core/bundles/taiga-ui-core-components-link.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/providers'), require('@taiga-ui/core/tokens'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/components/svg'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/link', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/providers', '@taiga-ui/core/tokens', 'rxjs', 'rxjs/operators', '@taiga-ui/core/components/svg', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components.link = {}), global.ng.core, global.i4, global["taiga-ui"].core.providers, global["taiga-ui"].core.tokens, global.rxjs, global.rxjs.operators, global["taiga-ui"].core.components.svg, global.ng.common));
})(this, (function (exports, i0, i4, providers, tokens, i3, operators, i1, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    // @bad TODO: Think about extending Interactive
    var TuiLinkComponent = /** @class */ (function () {
        function TuiLinkComponent(el, mode$, focusVisible$) {
            var _this = this;
            this.el = el;
            this.mode$ = mode$;
            this.pseudo = false;
            this.icon = '';
            this.iconAlign = 'right';
            this.iconRotated = false;
            this.mode = null;
            this.focusVisible = false;
            this.focusedChange = i3.merge(i4.tuiTypedFromEvent(this.el.nativeElement, 'focusin').pipe(operators.map(i4.ALWAYS_TRUE_HANDLER)), i4.tuiTypedFromEvent(this.el.nativeElement, 'focusout').pipe(operators.map(i4.ALWAYS_FALSE_HANDLER)));
            focusVisible$.subscribe(function (visible) {
                _this.focusVisible = visible;
            });
        }
        Object.defineProperty(TuiLinkComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLinkComponent.prototype, "focused", {
            get: function () {
                return i4.tuiIsNativeFocused(this.nativeFocusableElement);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLinkComponent.prototype, "hasIcon", {
            get: function () {
                return !!this.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLinkComponent.prototype, "iconAlignLeft", {
            get: function () {
                return this.hasIcon && this.iconAlign === 'left';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiLinkComponent.prototype, "iconAlignRight", {
            get: function () {
                return this.hasIcon && this.iconAlign === 'right';
            },
            enumerable: false,
            configurable: true
        });
        return TuiLinkComponent;
    }());
    TuiLinkComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkComponent, deps: [{ token: i0.ElementRef }, { token: tokens.TUI_MODE }, { token: i4.TuiFocusVisibleService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiLinkComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLinkComponent, selector: "a[tuiLink], button[tuiLink]", inputs: { pseudo: "pseudo", icon: "icon", iconAlign: "iconAlign", iconRotated: "iconRotated", mode: "mode" }, host: { listeners: { "$.data-mode.attr": "mode$" }, properties: { "class._pseudo": "this.pseudo", "class._icon-rotated": "this.iconRotated", "attr.data-host-mode": "this.mode", "class._focus-visible": "this.focusVisible" } }, providers: [
            i4.tuiAsFocusableItemAccessor(TuiLinkComponent),
            i4.TuiFocusVisibleService,
            i4.TuiDestroyService,
            providers.MODE_PROVIDER,
        ], exportAs: ["tuiLink"], ngImport: i0__namespace, template: "<tui-svg\n    *ngIf=\"iconAlignLeft\"\n    class=\"t-icon t-icon_left\"\n    [src]=\"icon || ''\"\n></tui-svg>\n<span class=\"t-content\">\n    <ng-content></ng-content>\n</span>\n<tui-svg\n    *ngIf=\"iconAlignRight\"\n    class=\"t-icon t-icon_right\"\n    [src]=\"icon || ''\"\n></tui-svg>\n", styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:none;font-size:inherit;line-height:inherit;text-decoration:none;transition-property:color;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;color:var(--tui-link);text-align:left;text-transform:inherit;font-weight:inherit;cursor:pointer;outline:none}:host:hover{color:var(--tui-link-hover)}:host:disabled{opacity:var(--tui-disabled-opacity);cursor:default}:host[data-mode=onLight]{color:var(--tui-text-02)}:host[data-mode=onLight]:hover,:host[data-mode=onLight]._active{color:var(--tui-text-01)}:host[data-mode=onLight]._pseudo{color:var(--tui-text-01)}:host[data-mode=onLight]._pseudo .t-content{-webkit-text-decoration-color:rgba(0,0,0,.48);text-decoration-color:#0000007a}:host[data-mode=onLight]._pseudo:hover,:host[data-mode=onLight]._pseudo._active{color:var(--tui-text-03)}:host[data-mode=onLight][data-host-mode=negative]{color:var(--tui-negative-night)}:host[data-mode=onLight][data-host-mode=negative]:hover,:host[data-mode=onLight][data-host-mode=negative]:active,:host[data-mode=onLight][data-host-mode=negative]._active{color:var(--tui-negative-night-hover)}:host[data-mode=onDark]{color:var(--tui-text-03-night)}:host[data-mode=onDark]:hover,:host[data-mode=onDark]:active,:host[data-mode=onDark]._active{color:var(--tui-text-01-night)}:host[data-mode=onDark]._pseudo{color:var(--tui-text-01-night)}:host[data-mode=onDark]._pseudo .t-content{-webkit-text-decoration-color:rgba(255,255,255,.48);text-decoration-color:#ffffff7a}:host[data-mode=onDark]._pseudo:hover,:host[data-mode=onDark]._pseudo:active,:host[data-mode=onDark]._pseudo._active{color:var(--tui-text-03-night)}:host[data-mode=onDark][data-host-mode=positive]{color:var(--tui-positive-night)}:host[data-mode=onDark][data-host-mode=positive]:hover,:host[data-mode=onDark][data-host-mode=positive]:active,:host[data-mode=onDark][data-host-mode=positive]._active{color:var(--tui-positive-night-hover)}:host[data-host-mode=negative]{color:var(--tui-negative)}:host[data-host-mode=negative]:hover{color:var(--tui-negative-hover)}:host[data-host-mode=negative]._pseudo .t-content{-webkit-text-decoration-color:rgba(222,76,30,.48);text-decoration-color:#de4c1e7a}:host[data-host-mode=positive]{color:var(--tui-positive)}:host[data-host-mode=positive]:hover{color:var(--tui-positive-hover)}:host[data-host-mode=positive]._pseudo .t-content{-webkit-text-decoration-color:rgba(58,169,129,.48);text-decoration-color:#3aa9817a}:host._focus-visible .t-content{background:var(--tui-selection)}:host[data-mode=onLight]._focus-visible .t-content{background:var(--tui-clear);color:var(--tui-text-01)}:host[data-mode=onDark]._focus-visible .t-content{background:var(--tui-clear-inverse);color:var(--tui-text-01-night)}:host[data-host-mode=positive]._focus-visible .t-content{background:var(--tui-success-bg)}:host[data-host-mode=negative]._focus-visible .t-content{background:var(--tui-error-bg)}:host._pseudo .t-content{padding-bottom:.15em;-webkit-text-decoration:underline dashed rgba(51,111,238,.48);text-decoration:underline dashed rgba(51,111,238,.48);text-underline-offset:.2em;text-decoration-thickness:.7px}.t-icon{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;width:var(--tui-link-icon-size, 1.5rem);height:var(--tui-link-icon-size, 1.5rem);margin-top:-.125rem}.t-icon_left{margin-right:.25rem}.t-icon_right{margin-left:.25rem}:host._icon-rotated .t-icon{transform:rotate(180deg)}\n"], components: [{ type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'a[tuiLink], button[tuiLink]',
                        templateUrl: './link.template.html',
                        styleUrls: ['./link.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i4.tuiAsFocusableItemAccessor(TuiLinkComponent),
                            i4.TuiFocusVisibleService,
                            i4.TuiDestroyService,
                            providers.MODE_PROVIDER,
                        ],
                        host: {
                            '($.data-mode.attr)': 'mode$',
                        },
                        exportAs: 'tuiLink',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_MODE]
                        }] }, { type: i4__namespace.TuiFocusVisibleService, decorators: [{
                            type: i0.Inject,
                            args: [i4.TuiFocusVisibleService]
                        }] }];
        }, propDecorators: { pseudo: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._pseudo']
                }], icon: [{
                    type: i0.Input
                }], iconAlign: [{
                    type: i0.Input
                }], iconRotated: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._icon-rotated']
                }], mode: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-host-mode']
                }], focusVisible: [{
                    type: i0.HostBinding,
                    args: ['class._focus-visible']
                }] } });

    var TuiLinkModule = /** @class */ (function () {
        function TuiLinkModule() {
        }
        return TuiLinkModule;
    }());
    TuiLinkModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiLinkModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkModule, declarations: [TuiLinkComponent], imports: [i2.CommonModule, i1.TuiSvgModule], exports: [TuiLinkComponent] });
    TuiLinkModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkModule, imports: [[i2.CommonModule, i1.TuiSvgModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiLinkModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i2.CommonModule, i1.TuiSvgModule],
                        declarations: [TuiLinkComponent],
                        exports: [TuiLinkComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiLinkComponent = TuiLinkComponent;
    exports.TuiLinkModule = TuiLinkModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
