// @taiga-ui/core/bundles/taiga-ui-core-enums.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/enums', ['exports'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.enums = {})));
})(this, (function (exports) { 'use strict';

    // TODO: change type in v4.0
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiAppearance = void 0;
    (function (TuiAppearance) {
        TuiAppearance["Accent"] = "accent";
        TuiAppearance["Flat"] = "flat";
        TuiAppearance["Icon"] = "icon";
        TuiAppearance["Outline"] = "outline";
        TuiAppearance["Primary"] = "primary";
        TuiAppearance["Secondary"] = "secondary";
        TuiAppearance["SecondaryDestructive"] = "secondary-destructive";
        TuiAppearance["Table"] = "table";
        TuiAppearance["Textfield"] = "textfield";
        TuiAppearance["Whiteblock"] = "whiteblock";
        TuiAppearance["WhiteblockActive"] = "whiteblock-active";
    })(exports.TuiAppearance || (exports.TuiAppearance = {}));

    /** @deprecated not used anymore TODO: remove in 4.0 **/
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiDropdownAnimation = void 0;
    (function (TuiDropdownAnimation) {
        TuiDropdownAnimation["FadeInBottom"] = "fadeInBottom";
        TuiDropdownAnimation["FadeInTop"] = "fadeInTop";
    })(exports.TuiDropdownAnimation || (exports.TuiDropdownAnimation = {}));

    /**
     * @internal used in calendar, year picker and wrapper
     */
    // TODO: change type in v4.0
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiInteractiveState = void 0;
    (function (TuiInteractiveState) {
        TuiInteractiveState["Active"] = "active";
        TuiInteractiveState["Disabled"] = "disabled";
        TuiInteractiveState["Hover"] = "hover";
        TuiInteractiveState["Readonly"] = "readonly";
    })(exports.TuiInteractiveState || (exports.TuiInteractiveState = {}));

    /**
     * @deprecated use union type TuiNotificationT instead
     */
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiNotification = void 0;
    (function (TuiNotification) {
        TuiNotification["Error"] = "error";
        TuiNotification["Info"] = "info";
        TuiNotification["Success"] = "success";
        TuiNotification["Warning"] = "warning";
    })(exports.TuiNotification || (exports.TuiNotification = {}));

    /**
     * @internal used in calendar and year picker
     */
    // TODO: change type in v4.0
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiRangeState = void 0;
    (function (TuiRangeState) {
        TuiRangeState["End"] = "end";
        TuiRangeState["Single"] = "single";
        TuiRangeState["Start"] = "start";
    })(exports.TuiRangeState || (exports.TuiRangeState = {}));

    /**
     * Generated bundle index. Do not edit.
     */

    Object.defineProperty(exports, '__esModule', { value: true });

}));
