// @taiga-ui/core/bundles/taiga-ui-core-components-primitive-calendar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/constants'), require('@taiga-ui/core/enums'), require('@taiga-ui/core/tokens'), require('@angular/common'), require('@taiga-ui/core/pipes')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/primitive-calendar', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/constants', '@taiga-ui/core/enums', '@taiga-ui/core/tokens', '@angular/common', '@taiga-ui/core/pipes'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].core = global["taiga-ui"].core || {}, global["taiga-ui"].core.components = global["taiga-ui"].core.components || {}, global["taiga-ui"].core.components["primitive-calendar"] = {}), global.ng.core, global.i2, global["taiga-ui"].core.constants, global["taiga-ui"].core.enums, global["taiga-ui"].core.tokens, global.ng.common, global["taiga-ui"].core.pipes));
})(this, (function (exports, i0, i2, constants, enums, tokens, i1, i3) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);

    var TuiPrimitiveCalendarComponent = /** @class */ (function () {
        function TuiPrimitiveCalendarComponent(unorderedWeekDays$, dayTypeHandler) {
            this.unorderedWeekDays$ = unorderedWeekDays$;
            this.dayTypeHandler = dayTypeHandler;
            this.pressedItem = null;
            this.today = i2.TuiDay.currentLocal();
            this.month = i2.TuiMonth.currentLocal();
            this.disabledItemHandler = i2.ALWAYS_FALSE_HANDLER;
            this.markerHandler = constants.TUI_DEFAULT_MARKER_HANDLER;
            this.value = null;
            this.hoveredItem = null;
            this.showAdjacent = true;
            this.hoveredItemChange = new i0.EventEmitter();
            this.dayClick = new i0.EventEmitter();
            this.toMarkers = function (day, today, inRange, markerHandler) {
                if (today || inRange) {
                    return null;
                }
                var markers = markerHandler(day);
                return markers.length === 0 ? null : markers;
            };
        }
        Object.defineProperty(TuiPrimitiveCalendarComponent.prototype, "isSingleDayRange", {
            get: function () {
                return this.value instanceof i2.TuiDayRange && this.value.isSingleDay;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveCalendarComponent.prototype, "isSingle", {
            /**
             * @deprecated: use {@link this.isSingleDayRange}
             */
            get: function () {
                return this.isSingleDayRange;
            },
            enumerable: false,
            configurable: true
        });
        TuiPrimitiveCalendarComponent.prototype.getItemState = function (item) {
            var _a = this, disabledItemHandler = _a.disabledItemHandler, pressedItem = _a.pressedItem, hoveredItem = _a.hoveredItem;
            if (disabledItemHandler(item)) {
                return enums.TuiInteractiveState.Disabled;
            }
            if (pressedItem === null || pressedItem === void 0 ? void 0 : pressedItem.daySame(item)) {
                return enums.TuiInteractiveState.Active;
            }
            if (hoveredItem === null || hoveredItem === void 0 ? void 0 : hoveredItem.daySame(item)) {
                return enums.TuiInteractiveState.Hover;
            }
            return null;
        };
        TuiPrimitiveCalendarComponent.prototype.getItemRange = function (item) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (!value) {
                return null;
            }
            if (value instanceof i2.TuiDay) {
                return value.daySame(item) ? enums.TuiRangeState.Single : null;
            }
            if (!(value instanceof i2.TuiDayRange)) {
                return value.find(function (day) { return day.daySame(item); }) ? enums.TuiRangeState.Single : null;
            }
            if ((value.from.daySame(item) && !value.isSingleDay) ||
                ((hoveredItem === null || hoveredItem === void 0 ? void 0 : hoveredItem.dayAfter(value.from)) &&
                    value.from.daySame(item) &&
                    value.isSingleDay) ||
                ((hoveredItem === null || hoveredItem === void 0 ? void 0 : hoveredItem.daySame(item)) &&
                    hoveredItem.dayBefore(value.from) &&
                    value.isSingleDay)) {
                return enums.TuiRangeState.Start;
            }
            if ((value.to.daySame(item) && !value.isSingleDay) ||
                ((hoveredItem === null || hoveredItem === void 0 ? void 0 : hoveredItem.dayBefore(value.from)) &&
                    value.from.daySame(item) &&
                    value.isSingleDay) ||
                ((hoveredItem === null || hoveredItem === void 0 ? void 0 : hoveredItem.daySame(item)) &&
                    hoveredItem.dayAfter(value.from) &&
                    value.isSingleDay)) {
                return enums.TuiRangeState.End;
            }
            return value.isSingleDay && value.from.daySame(item)
                ? enums.TuiRangeState.Single
                : null;
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsToday = function (item) {
            return this.today.daySame(item);
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsUnavailable = function (item) {
            return !this.month.monthSame(item);
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsInterval = function (day) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (!(value instanceof i2.TuiDayRange)) {
                return false;
            }
            if (!value.isSingleDay) {
                return value.from.daySameOrBefore(day) && value.to.dayAfter(day);
            }
            if (hoveredItem === null) {
                return false;
            }
            var range = i2.TuiDayRange.sort(value.from, hoveredItem);
            return range.from.daySameOrBefore(day) && range.to.dayAfter(day);
        };
        TuiPrimitiveCalendarComponent.prototype.onItemHovered = function (item) {
            this.updateHoveredItem(item || null);
        };
        TuiPrimitiveCalendarComponent.prototype.onItemPressed = function (item) {
            this.pressedItem = item || null;
        };
        TuiPrimitiveCalendarComponent.prototype.onItemClick = function (item) {
            this.dayClick.emit(item);
        };
        TuiPrimitiveCalendarComponent.prototype.updateHoveredItem = function (day) {
            if (i2.tuiNullableSame(this.hoveredItem, day, function (a, b) { return a.daySame(b); })) {
                return;
            }
            this.hoveredItem = day;
            this.hoveredItemChange.emit(day);
        };
        return TuiPrimitiveCalendarComponent;
    }());
    TuiPrimitiveCalendarComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarComponent, deps: [{ token: tokens.TUI_SHORT_WEEK_DAYS }, { token: tokens.TUI_DAY_TYPE_HANDLER }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPrimitiveCalendarComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPrimitiveCalendarComponent, selector: "tui-primitive-calendar", inputs: { month: "month", disabledItemHandler: "disabledItemHandler", markerHandler: "markerHandler", value: "value", hoveredItem: "hoveredItem", showAdjacent: "showAdjacent" }, outputs: { hoveredItemChange: "hoveredItemChange", dayClick: "dayClick" }, host: { properties: { "class._single": "this.isSingleDayRange" } }, ngImport: i0__namespace, template: "<div class=\"t-row t-row_weekday\">\n    <div\n        *ngFor=\"let day of unorderedWeekDays$ | tuiOrderWeekDays | async\"\n        class=\"t-cell\"\n        [textContent]=\"day\"\n    ></div>\n</div>\n<div\n    *tuiLet=\"month | tuiCalendarSheet: true as sheet\"\n    id=\"date-rows\"\n>\n    <div\n        *tuiRepeatTimes=\"let rowIndex of sheet.length\"\n        automation-id=\"tui-primitive-calendar__row\"\n        class=\"t-row\"\n    >\n        <ng-container *tuiRepeatTimes=\"let colIndex of sheet[rowIndex].length\">\n            <ng-container *tuiLet=\"sheet[rowIndex][colIndex] as item\">\n                <div\n                    *ngIf=\"!itemIsUnavailable(item) || showAdjacent\"\n                    automation-id=\"tui-primitive-calendar__cell\"\n                    class=\"t-cell\"\n                    [attr.data-range]=\"getItemRange(item)\"\n                    [attr.data-state]=\"getItemState(item)\"\n                    [class.t-cell_interval]=\"itemIsInterval(item)\"\n                    [class.t-cell_today]=\"itemIsToday(item)\"\n                    [class.t-cell_unavailable]=\"itemIsUnavailable(item)\"\n                    (click)=\"onItemClick(item)\"\n                    (tuiHoveredChange)=\"onItemHovered($event && item)\"\n                    (tuiPressedChange)=\"onItemPressed($event && item)\"\n                >\n                    <div\n                        automation-id=\"tui-primitive-calendar__item\"\n                        class=\"t-item\"\n                        [attr.data-type]=\"item | tuiMapper: dayTypeHandler\"\n                        [class.t-item_unavailable]=\"itemIsUnavailable(item)\"\n                    >\n                        {{ item.day }}\n                        <div\n                            *ngIf=\"\n                                item\n                                    | tuiMapper\n                                        : toMarkers\n                                        : itemIsToday(item)\n                                        : !!getItemRange(item)\n                                        : markerHandler as markers\n                            \"\n                            class=\"t-dots\"\n                        >\n                            <div\n                                class=\"t-dot\"\n                                [style.backgroundColor]=\"markers?.[0]\"\n                            ></div>\n                            <div\n                                *ngIf=\"markers.length > 1\"\n                                class=\"t-dot\"\n                                [style.backgroundColor]=\"markers?.[1] || ''\"\n                            ></div>\n                        </div>\n                    </div>\n                </div>\n            </ng-container>\n        </ng-container>\n    </div>\n</div>\n", styles: [":host{display:block;font:var(--tui-font-text-m)}.t-row{position:relative;display:flex;justify-content:space-between;height:2.25rem;isolation:isolate}.t-item{position:relative;flex:1;line-height:2rem;border-radius:var(--tui-radius-m)}.t-item:before,.t-item:after{content:\"\";position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.t-item:after{border-radius:.5rem}.t-cell{position:relative;display:flex;align-items:center;justify-content:center;width:2.25rem;text-align:center;outline:none;cursor:pointer;background-clip:content-box;box-sizing:border-box;border:.125rem solid transparent}.t-cell:before{content:\"\";position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.t-cell_today:after{position:absolute;left:50%;transform:translate(-50%);content:\"\";bottom:.3125rem;height:.125rem;width:.75rem;border-radius:.375rem;background-color:var(--tui-text-01)}.t-cell_interval:before{background:var(--tui-base-02)}:host._single .t-cell_interval:before{background:var(--tui-secondary-hover)}.t-cell_interval:not(:last-child):before{right:-.25rem;border-top-right-radius:0;border-bottom-right-radius:0}.t-cell_interval:not([data-range=\"start\"]):not(:first-child):before{border-top-left-radius:0;border-bottom-left-radius:0}.t-cell_interval:last-child:first-child:before{right:0}.t-cell_interval:first-child>.t-item{border-top-left-radius:var(--tui-radius-m);border-bottom-left-radius:var(--tui-radius-m)}.t-cell_interval:last-child>.t-item{border-top-right-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.t-cell_interval>.t-item{border-radius:0}.t-cell[data-range]:after{background-color:var(--tui-primary-text)}.t-cell[data-range]>.t-item{color:var(--tui-primary-text)}.t-cell[data-range]>.t-item:before,.t-cell[data-range]>.t-item:after{background-color:var(--tui-primary)}.t-cell[data-range][data-state=hover]>.t-item:before,.t-cell[data-range][data-state=hover]>.t-item:after{background-color:var(--tui-primary-hover)}.t-cell[data-range][data-state=active]>.t-item:before,.t-cell[data-range][data-state=active]>.t-item:after{background-color:var(--tui-primary-active)}.t-cell[data-range=end]:before{background:var(--tui-base-02)}:host._single .t-cell[data-range=end]:before{background:var(--tui-secondary-hover)}.t-cell[data-range=end]:not(:first-child):before{border-top-left-radius:0;border-bottom-left-radius:0}.t-cell[data-range=end]>.t-item:before{left:.625rem;border-top-left-radius:0;border-bottom-left-radius:0}.t-cell[data-range=end]>.t-item:after{left:-2rem;right:100%;transform:translate(1.6rem) scaleY(.83) scaleX(.5) rotate(45deg)}.t-cell[data-range=start]>.t-item:before{right:.625rem;border-top-right-radius:0;border-bottom-right-radius:0}.t-cell[data-range=start]>.t-item:after{left:100%;right:-2rem;transform:translate(-1.6rem) scaleY(.83) scaleX(.5) rotate(45deg)}.t-cell[data-range=single]>.t-item:after{display:none}.t-cell[data-state=disabled]{pointer-events:none}.t-cell[data-state=disabled]>.t-item{opacity:.36}.t-cell[data-state=hover]:hover:not([data-range])>.t-item{background-color:var(--tui-secondary-hover)}.t-cell[data-state=active]:hover:not([data-range])>.t-item{background-color:var(--tui-secondary-active)}:host{width:15.75rem}.t-row{justify-content:flex-start}.t-row:first-child{justify-content:flex-end}.t-row_weekday{font:var(--tui-font-text-s);color:var(--tui-text-02);pointer-events:none}.t-item{display:flex;flex-direction:column}.t-item_unavailable{opacity:var(--tui-disabled-opacity)}.t-dots{display:flex;justify-content:center;margin-top:-.5rem;padding-bottom:.25rem}.t-dot{display:inline-block;width:.25rem;height:.25rem;border-radius:100%;margin:0 .0625rem}\n"], directives: [{ type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2__namespace.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }, { type: i1__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2__namespace.TuiHoveredDirective, selector: "[tuiHoveredChange]", outputs: ["tuiHoveredChange"] }, { type: i2__namespace.TuiPressedDirective, selector: "[tuiPressedChange]", outputs: ["tuiPressedChange"] }], pipes: { "async": i1__namespace.AsyncPipe, "tuiOrderWeekDays": i3__namespace.TuiOrderWeekDaysPipe, "tuiCalendarSheet": i3__namespace.TuiCalendarSheetPipe, "tuiMapper": i2__namespace.TuiMapperPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-primitive-calendar',
                        templateUrl: './primitive-calendar.template.html',
                        styleUrls: ['./primitive-calendar.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_SHORT_WEEK_DAYS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DAY_TYPE_HANDLER]
                        }] }];
        }, propDecorators: { month: [{
                    type: i0.Input
                }], disabledItemHandler: [{
                    type: i0.Input
                }], markerHandler: [{
                    type: i0.Input
                }], value: [{
                    type: i0.Input
                }], hoveredItem: [{
                    type: i0.Input
                }], showAdjacent: [{
                    type: i0.Input
                }], hoveredItemChange: [{
                    type: i0.Output
                }], dayClick: [{
                    type: i0.Output
                }], isSingleDayRange: [{
                    type: i0.HostBinding,
                    args: ['class._single']
                }] } });

    var TuiPrimitiveCalendarModule = /** @class */ (function () {
        function TuiPrimitiveCalendarModule() {
        }
        return TuiPrimitiveCalendarModule;
    }());
    TuiPrimitiveCalendarModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPrimitiveCalendarModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarModule, declarations: [TuiPrimitiveCalendarComponent], imports: [i1.CommonModule,
            i2.TuiLetModule,
            i2.TuiMapperPipeModule,
            i2.TuiRepeatTimesModule,
            i2.TuiHoveredModule,
            i2.TuiPressedModule,
            i3.TuiCalendarSheetPipeModule,
            i3.TuiOrderWeekDaysPipeModule], exports: [TuiPrimitiveCalendarComponent] });
    TuiPrimitiveCalendarModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarModule, imports: [[
                i1.CommonModule,
                i2.TuiLetModule,
                i2.TuiMapperPipeModule,
                i2.TuiRepeatTimesModule,
                i2.TuiHoveredModule,
                i2.TuiPressedModule,
                i3.TuiCalendarSheetPipeModule,
                i3.TuiOrderWeekDaysPipeModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPrimitiveCalendarModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i1.CommonModule,
                            i2.TuiLetModule,
                            i2.TuiMapperPipeModule,
                            i2.TuiRepeatTimesModule,
                            i2.TuiHoveredModule,
                            i2.TuiPressedModule,
                            i3.TuiCalendarSheetPipeModule,
                            i3.TuiOrderWeekDaysPipeModule,
                        ],
                        declarations: [TuiPrimitiveCalendarComponent],
                        exports: [TuiPrimitiveCalendarComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPrimitiveCalendarComponent = TuiPrimitiveCalendarComponent;
    exports.TuiPrimitiveCalendarModule = TuiPrimitiveCalendarModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
