// @taiga-ui/core/esm2015/components/notification/notification.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule } from '@taiga-ui/core/components/button';
import { TuiSvgModule } from '@taiga-ui/core/components/svg';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiNotificationComponent } from './notification.component';
import * as i0 from "@angular/core";
export class TuiNotificationModule {
}
TuiNotificationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNotificationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiNotificationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNotificationModule, declarations: [TuiNotificationComponent], imports: [CommonModule, TuiSvgModule, TuiButtonModule, PolymorpheusModule], exports: [TuiNotificationComponent] });
TuiNotificationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNotificationModule, imports: [[CommonModule, TuiSvgModule, TuiButtonModule, PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNotificationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiButtonModule, PolymorpheusModule],
                    declarations: [TuiNotificationComponent],
                    exports: [TuiNotificationComponent],
                }]
        }] });
