// @taiga-ui/layout/fesm2015/taiga-ui-layout-components-block-status.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, HostBinding, Input, Directive, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

class TuiBlockStatusComponent {
    constructor() {
        this.card = false;
    }
}
TuiBlockStatusComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiBlockStatusComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockStatusComponent, selector: "tui-block-status", inputs: { card: "card" }, host: { properties: { "class._card": "this.card" } }, ngImport: i0, template: "<div class=\"t-block-image\">\n    <ng-content select=\"[tuiSlot='top']\"></ng-content>\n</div>\n\n<ng-content select=\"h1,h2,h3,h4,h5,h6\"></ng-content>\n\n<div class=\"t-block-text\">\n    <ng-content></ng-content>\n</div>\n\n<div class=\"t-block-actions\">\n    <ng-content select=\"a,button,[tuiSlot='action']\"></ng-content>\n</div>\n", styles: ["tui-block-status{display:flex;flex-direction:column;align-items:center;justify-content:center;text-align:center;box-sizing:border-box;overflow:hidden;width:100%;height:100%}tui-root._mobile tui-block-status{padding:2rem}tui-block-status._card{border-radius:var(--tui-radius-xl);padding:2rem}tui-root._mobile tui-block-status._card{padding:1.25rem}tui-root._mobile tui-block-status._card .t-block-image{margin-top:.75rem}tui-block-status .t-block-image{margin-bottom:2rem}tui-root._mobile tui-block-status .t-block-image{margin-bottom:1.5rem}tui-block-status .t-block-text{font:var(--tui-font-text-m);color:var(--tui-text-02);margin-bottom:2rem;white-space:pre-line}tui-root._mobile tui-block-status .t-block-text{margin-bottom:1.5rem}tui-block-status .t-block-actions{display:flex;flex-direction:column;align-items:center;width:100%}tui-block-status h1,tui-block-status h2,tui-block-status h3,tui-block-status h4,tui-block-status h5,tui-block-status h6{font:var(--tui-font-heading-4);margin-top:0;margin-bottom:1rem}tui-root._mobile tui-block-status h1,tui-root._mobile tui-block-status h2,tui-root._mobile tui-block-status h3,tui-root._mobile tui-block-status h4,tui-root._mobile tui-block-status h5,tui-root._mobile tui-block-status h6{font:var(--tui-font-heading-5);margin-bottom:.5rem}tui-block-status img{width:9rem;height:9rem}tui-root._mobile tui-block-status img{width:8rem;height:8rem}tui-block-status a:not(:last-child),tui-block-status button:not(:last-child){margin-bottom:.75rem}tui-root._mobile tui-block-status a:not(:last-child),tui-root._mobile tui-block-status button:not(:last-child){margin-bottom:.5rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-block-status',
                    templateUrl: './block-status.template.html',
                    styleUrls: ['./block-status.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { card: [{
                type: HostBinding,
                args: ['class._card']
            }, {
                type: Input
            }] } });

class TuiBlockStatusDirective {
    constructor() {
        this.tuiSlot = 'top';
    }
}
TuiBlockStatusDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiBlockStatusDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiBlockStatusDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSlot]',
                }]
        }], propDecorators: { tuiSlot: [{
                type: Input
            }] } });

class TuiBlockStatusModule {
}
TuiBlockStatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBlockStatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective], imports: [CommonModule], exports: [TuiBlockStatusComponent, TuiBlockStatusDirective] });
TuiBlockStatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                    exports: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBlockStatusComponent, TuiBlockStatusDirective, TuiBlockStatusModule };
