// @taiga-ui/addon-doc/esm2015/components/internal/input-opacity/input-opacity.directive.js
import { Directive, Inject, Input, Renderer2 } from '@angular/core';
import { TUI_FOCUSABLE_ITEM_ACCESSOR } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiInputOpacityDirective {
    constructor(renderer, focusable) {
        this.renderer = renderer;
        this.focusable = focusable;
    }
    set tuiInputOpacity(opacity) {
        const { nativeFocusableElement } = this.focusable;
        if (nativeFocusableElement) {
            this.renderer.setStyle(nativeFocusableElement, 'opacity', opacity / 100);
        }
    }
}
TuiInputOpacityDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputOpacityDirective, deps: [{ token: Renderer2 }, { token: TUI_FOCUSABLE_ITEM_ACCESSOR }], target: i0.ɵɵFactoryTarget.Directive });
TuiInputOpacityDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputOpacityDirective, selector: "[tuiInputOpacity]", inputs: { tuiInputOpacity: "tuiInputOpacity" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputOpacityDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiInputOpacity]',
                }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2, decorators: [{
                    type: Inject,
                    args: [Renderer2]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_FOCUSABLE_ITEM_ACCESSOR]
                }] }]; }, propDecorators: { tuiInputOpacity: [{
                type: Input
            }] } });
