// @taiga-ui/addon-mobile/esm2015/components/pull-to-refresh/pull-to-refresh.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiLoaderModule, TuiSvgModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiMobileLoaderAndroidComponent } from './loader-android/loader-android.component';
import { TuiMobileLoaderIOSComponent } from './loader-ios/loader-ios.component';
import { TuiPullToRefreshComponent } from './pull-to-refresh.component';
import * as i0 from "@angular/core";
export class TuiPullToRefreshModule {
}
TuiPullToRefreshModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPullToRefreshModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshModule, declarations: [TuiPullToRefreshComponent,
        TuiMobileLoaderAndroidComponent,
        TuiMobileLoaderIOSComponent], imports: [CommonModule,
        TuiLoaderModule,
        TuiSvgModule,
        TuiRepeatTimesModule,
        PolymorpheusModule], exports: [TuiPullToRefreshComponent] });
TuiPullToRefreshModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshModule, imports: [[
            CommonModule,
            TuiLoaderModule,
            TuiSvgModule,
            TuiRepeatTimesModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiLoaderModule,
                        TuiSvgModule,
                        TuiRepeatTimesModule,
                        PolymorpheusModule,
                    ],
                    declarations: [
                        TuiPullToRefreshComponent,
                        TuiMobileLoaderAndroidComponent,
                        TuiMobileLoaderIOSComponent,
                    ],
                    exports: [TuiPullToRefreshComponent],
                }]
        }] });
