// @taiga-ui/addon-mobile/esm2015/components/pull-to-refresh/pull-to-refresh.component.js
import { ChangeDetectionStrategy, Component, Inject, Input, NgZone, Output, Self, } from '@angular/core';
import { TUI_IS_IOS, TUI_SCROLL_REF, TuiDestroyService, tuiPx, tuiScrollFrom, tuiZonefree, } from '@taiga-ui/cdk';
import { distinctUntilChanged, filter, map, startWith, takeUntil } from 'rxjs/operators';
import { TUI_PULL_TO_REFRESH_COMPONENT, TUI_PULL_TO_REFRESH_THRESHOLD, } from './pull-to-refresh.providers';
import { MICRO_OFFSET, TuiPullToRefreshService } from './pull-to-refresh.service';
import * as i0 from "@angular/core";
import * as i1 from "@tinkoff/ng-polymorpheus";
import * as i2 from "@angular/common";
import * as i3 from "rxjs";
export class TuiPullToRefreshComponent {
    constructor(zone, destroy$, { nativeElement }, isIOS, threshold, component, pulling$) {
        this.isIOS = isIOS;
        this.threshold = threshold;
        this.component = component;
        this.pulling$ = pulling$;
        this.styleHandler = this.isIOS
            ? distance => ({ top: tuiPx(distance / 2) })
            : () => null;
        this.pulled = this.pulling$.pipe(filter(distance => distance === this.threshold));
        this.dropped$ = this.pulling$.pipe(map(distance => distance <= MICRO_OFFSET || distance === this.threshold), distinctUntilChanged());
        // Ensure scrolling down is impossible while pulling
        if (this.component) {
            tuiScrollFrom(nativeElement)
                .pipe(startWith(null), tuiZonefree(zone), takeUntil(destroy$))
                .subscribe(() => {
                if (nativeElement.scrollTop) {
                    nativeElement.style.touchAction = '';
                }
                else {
                    nativeElement.style.touchAction = 'pan-down';
                }
            });
        }
    }
}
TuiPullToRefreshComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshComponent, deps: [{ token: NgZone }, { token: TuiDestroyService, self: true }, { token: TUI_SCROLL_REF }, { token: TUI_IS_IOS }, { token: TUI_PULL_TO_REFRESH_THRESHOLD }, { token: TUI_PULL_TO_REFRESH_COMPONENT }, { token: TuiPullToRefreshService }], target: i0.ɵɵFactoryTarget.Component });
TuiPullToRefreshComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPullToRefreshComponent, selector: "tui-pull-to-refresh", inputs: { styleHandler: "styleHandler" }, outputs: { pulled: "pulled" }, providers: [TuiPullToRefreshService, TuiDestroyService], ngImport: i0, template: "<ng-container *polymorpheusOutlet=\"component; context: {$implicit: (pulling$ | async) || 0}\"></ng-container>\n\n<div\n    [class.t-drop]=\"dropped$ | async\"\n    [ngStyle]=\"styleHandler((pulling$ | async) || 0)\"\n    [style.position]=\"'relative'\"\n>\n    <ng-content></ng-content>\n</div>\n", styles: [".t-drop{transition:all var(--tui-duration) cubic-bezier(.4,0,.2,1)}\n"], directives: [{ type: i1.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], pipes: { "async": i2.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPullToRefreshComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-pull-to-refresh',
                    templateUrl: './pull-to-refresh.template.html',
                    styleUrls: ['./pull-to-refresh.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [TuiPullToRefreshService, TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: i3.Observable, decorators: [{
                    type: Inject,
                    args: [TuiDestroyService]
                }, {
                    type: Self
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [TUI_SCROLL_REF]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_IOS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_PULL_TO_REFRESH_THRESHOLD]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_PULL_TO_REFRESH_COMPONENT]
                }] }, { type: i3.Observable, decorators: [{
                    type: Inject,
                    args: [TuiPullToRefreshService]
                }] }]; }, propDecorators: { styleHandler: [{
                type: Input
            }], pulled: [{
                type: Output
            }] } });
