// @taiga-ui/experimental/esm2015/components/navigation/aside.component.js
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, } from '@angular/core';
import { tuiDropdownOptionsProvider } from '@taiga-ui/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/experimental/components/button';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
import * as i2 from "@taiga-ui/experimental/directives/fade";
export class TuiAsideComponent {
    constructor() {
        this.tuiNavigationAside = false;
    }
}
TuiAsideComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAsideComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAsideComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAsideComponent, selector: "aside[tuiNavigationAside]", inputs: { tuiNavigationAside: "tuiNavigationAside" }, host: { attributes: { "tuiNavigationAside": "", "tuiTheme": "night" }, properties: { "class._expanded": "tuiNavigationAside" } }, providers: [
        tuiButtonOptionsProvider({ size: 's', appearance: 'flat' }),
        tuiDropdownOptionsProvider({ appearance: 'dropdown-navigation', align: 'right' }),
    ], ngImport: i0, template: `
        <ng-content select="header"></ng-content>
        <tui-scrollbar class="t-nav-scrollbar">
            <div
                tuiFade="vertical"
                tuiScrollable
                class="t-nav-content"
            >
                <ng-content></ng-content>
            </div>
        </tui-scrollbar>
        <ng-content select="footer"></ng-content>
    `, isInline: true, styles: ["aside[tuiNavigationAside]{--tui-link: var(--tui-text-01-night);--tui-link-hover: var(--tui-text-01-night);position:-webkit-sticky;position:sticky;top:3rem;display:flex;width:3rem;flex-direction:column;padding:.5rem .5rem .25rem;box-sizing:border-box;background:var(--tui-theme-color, #000);color:var(--tui-text-01-night);transition-property:width;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out}aside[tuiNavigationAside]._expanded{width:13.75rem}aside[tuiNavigationAside]._expanded+main[tuiNavigationMain]{max-width:calc(100% - 13.75rem)}aside[tuiNavigationAside] a[tuiButton].active{background:var(--tui-clear-hover)}aside[tuiNavigationAside] .t-nav-scrollbar{margin:0 -.25rem}aside[tuiNavigationAside] .t-nav-content{scrollbar-width:none;-ms-overflow-style:none;height:100%;padding:0 .25rem;overflow:auto}aside[tuiNavigationAside] .t-nav-content::-webkit-scrollbar,aside[tuiNavigationAside] .t-nav-content::-webkit-scrollbar-thumb{display:none}aside[tuiNavigationAside] tui-hosted-dropdown{width:100%}aside[tuiNavigationAside] [tuiButton]{width:100%;justify-content:flex-start;grid-gap:.625rem;gap:.625rem;margin-bottom:.25rem}aside[tuiNavigationAside] [tuiButton]:after{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;margin-left:auto;color:var(--tui-text-03-night)}aside[tuiNavigationAside] tui-expand [tuiButton]{-webkit-padding-start:2rem;padding-inline-start:2rem}aside[tuiNavigationAside] hr{background:var(--tui-clear-inverse);height:1px;margin:0 0 .25rem;border:none}aside[tuiNavigationAside] header{padding-bottom:.25rem;box-shadow:inset 0 -1px var(--tui-clear);margin-bottom:1rem}aside[tuiNavigationAside] footer{margin-top:auto;border-top:1.75rem solid transparent}aside[tuiNavigationAside]+main[tuiNavigationMain]{max-width:calc(100% - 3rem)}tui-dropdown[data-appearance=dropdown-navigation]{--tui-text-01: var(--tui-text-01-night);--tui-clear: var(--tui-clear-inverse);border:none;background:#323232;background:color-mix(in srgb,var(--tui-theme-color, #000) 80%,#fff)}\n"], components: [{ type: i1.TuiScrollbarComponent, selector: "tui-scrollbar", inputs: ["hidden"] }], directives: [{ type: i2.TuiFadeDirective, selector: "[tuiFade]", inputs: ["tuiFadeHeight", "tuiFadeSize", "tuiFadeOffset", "tuiFade"] }, { type: i1.TuiScrollableDirective, selector: "[tuiScrollable]" }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAsideComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'aside[tuiNavigationAside]',
                    template: `
        <ng-content select="header"></ng-content>
        <tui-scrollbar class="t-nav-scrollbar">
            <div
                tuiFade="vertical"
                tuiScrollable
                class="t-nav-content"
            >
                <ng-content></ng-content>
            </div>
        </tui-scrollbar>
        <ng-content select="footer"></ng-content>
    `,
                    styleUrls: ['./aside.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        tuiButtonOptionsProvider({ size: 's', appearance: 'flat' }),
                        tuiDropdownOptionsProvider({ appearance: 'dropdown-navigation', align: 'right' }),
                    ],
                    host: {
                        tuiNavigationAside: '',
                        tuiTheme: 'night',
                        '[class._expanded]': 'tuiNavigationAside',
                    },
                }]
        }], propDecorators: { tuiNavigationAside: [{
                type: Input
            }] } });
