// @taiga-ui/experimental/esm2015/components/navigation/logo.component.js
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiLogoComponent {
}
TuiLogoComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLogoComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiLogoComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiLogoComponent, selector: "[tuiNavigationLogo]", ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: ["[tuiNavigationLogo]{display:flex;align-items:center;grid-gap:.5rem;gap:.5rem;overflow:hidden;white-space:nowrap;font-weight:bold}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLogoComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[tuiNavigationLogo]',
                    template: '<ng-content></ng-content>',
                    styleUrls: ['./logo.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
