// @taiga-ui/addon-doc/esm2015/components/documentation/documentation.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TuiFilterPipeModule } from '@taiga-ui/cdk';
import { TuiDataListModule, TuiDropdownModule, TuiGroupModule, TuiLinkModule, TuiModeModule, TuiNotificationModule, TuiPrimitiveTextfieldModule, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiBadgeModule, TuiDataListWrapperModule, TuiInputNumberModule, TuiSelectModule, TuiToggleModule, } from '@taiga-ui/kit';
import { TuiInputOpacityModule } from '../internal/input-opacity/input-opacity.module';
import { TuiDocDocumentationComponent } from './documentation.component';
import { TuiDocDocumentationPropertyConnectorDirective } from './documentation-property-connector.directive';
import { TuiShowCleanerPipe } from './pipes/cleaner.pipe';
import { TuiGetColorPipe } from './pipes/color.pipe';
import { TuiInspectPipe } from './pipes/inspect.pipe';
import { TuiGetOpacityPipe } from './pipes/opacity.pipe';
import { TuiIsOptionalPipe } from './pipes/optional.pipe';
import { TuiIsPrimitivePolymorpheusContentPipe } from './pipes/primitive-polymorpheus-content.pipe';
import { TuiStripOptionalPipe } from './pipes/strip-optional.pipe';
import { TuiDocTypeReferencePipe } from './pipes/type-reference.pipe';
import * as i0 from "@angular/core";
export class TuiDocDocumentationModule {
}
TuiDocDocumentationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDocumentationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocDocumentationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDocumentationModule, declarations: [TuiInspectPipe,
        TuiGetColorPipe,
        TuiGetOpacityPipe,
        TuiIsOptionalPipe,
        TuiShowCleanerPipe,
        TuiStripOptionalPipe,
        TuiDocTypeReferencePipe,
        TuiDocDocumentationComponent,
        TuiIsPrimitivePolymorpheusContentPipe,
        TuiDocDocumentationPropertyConnectorDirective], imports: [CommonModule,
        FormsModule,
        RouterModule,
        TuiBadgeModule,
        TuiSelectModule,
        TuiToggleModule,
        TuiLinkModule,
        TuiInputNumberModule,
        TuiModeModule,
        TuiGroupModule,
        TuiInputOpacityModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiDropdownModule,
        TuiDataListModule,
        TuiDataListWrapperModule,
        TuiNotificationModule,
        TuiFilterPipeModule], exports: [TuiDocTypeReferencePipe,
        TuiDocDocumentationComponent,
        TuiDocDocumentationPropertyConnectorDirective] });
TuiDocDocumentationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDocumentationModule, imports: [[
            CommonModule,
            FormsModule,
            RouterModule,
            TuiBadgeModule,
            TuiSelectModule,
            TuiToggleModule,
            TuiLinkModule,
            TuiInputNumberModule,
            TuiModeModule,
            TuiGroupModule,
            TuiInputOpacityModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiDropdownModule,
            TuiDataListModule,
            TuiDataListWrapperModule,
            TuiNotificationModule,
            TuiFilterPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDocumentationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        TuiBadgeModule,
                        TuiSelectModule,
                        TuiToggleModule,
                        TuiLinkModule,
                        TuiInputNumberModule,
                        TuiModeModule,
                        TuiGroupModule,
                        TuiInputOpacityModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiDropdownModule,
                        TuiDataListModule,
                        TuiDataListWrapperModule,
                        TuiNotificationModule,
                        TuiFilterPipeModule,
                    ],
                    declarations: [
                        TuiInspectPipe,
                        TuiGetColorPipe,
                        TuiGetOpacityPipe,
                        TuiIsOptionalPipe,
                        TuiShowCleanerPipe,
                        TuiStripOptionalPipe,
                        TuiDocTypeReferencePipe,
                        TuiDocDocumentationComponent,
                        TuiIsPrimitivePolymorpheusContentPipe,
                        TuiDocDocumentationPropertyConnectorDirective,
                    ],
                    exports: [
                        TuiDocTypeReferencePipe,
                        TuiDocDocumentationComponent,
                        TuiDocDocumentationPropertyConnectorDirective,
                    ],
                }]
        }] });
