// @taiga-ui/cdk/esm2015/abstract/portal-service.js
import { Injectable } from '@angular/core';
import { TuiNoHostException } from '@taiga-ui/cdk/exceptions';
import * as i0 from "@angular/core";
/**
 * Abstract service for displaying portals
 */
export class AbstractTuiPortalService {
    get safeHost() {
        if (!this.host) {
            throw new TuiNoHostException();
        }
        return this.host;
    }
    attach(host) {
        this.host = host;
    }
    add(component) {
        return this.safeHost.addComponentChild(component);
    }
    remove({ hostView }) {
        if (!hostView.destroyed) {
            hostView.destroy();
        }
    }
    addTemplate(templateRef, context) {
        return this.safeHost.addTemplateChild(templateRef, context);
    }
    removeTemplate(viewRef) {
        if (!viewRef.destroyed) {
            viewRef.destroy();
        }
    }
}
AbstractTuiPortalService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiPortalService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
AbstractTuiPortalService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiPortalService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiPortalService, decorators: [{
            type: Injectable
        }] });
