// @taiga-ui/kit/bundles/taiga-ui-kit-pipes-iso-to-country-code.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/pipes/iso-to-country-code', ['exports', '@angular/core', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.pipes = global["taiga-ui"].kit.pipes || {}, global["taiga-ui"].kit.pipes["iso-to-country-code"] = {}), global.ng.core, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils));
})(this, (function (exports, i0, tokens, utils) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiIsoToCountryCodePipe = /** @class */ (function () {
        function TuiIsoToCountryCodePipe(countriesMasks) {
            this.countriesMasks = countriesMasks;
        }
        TuiIsoToCountryCodePipe.prototype.transform = function (isoCode) {
            return utils.tuiIsoToCountryCode(this.countriesMasks, isoCode);
        };
        return TuiIsoToCountryCodePipe;
    }());
    TuiIsoToCountryCodePipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodePipe, deps: [{ token: tokens.TUI_COUNTRIES_MASKS }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiIsoToCountryCodePipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodePipe, name: "tuiIsoToCountryCode" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodePipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiIsoToCountryCode',
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_COUNTRIES_MASKS]
                        }] }];
        } });

    var TuiIsoToCountryCodeModule = /** @class */ (function () {
        function TuiIsoToCountryCodeModule() {
        }
        return TuiIsoToCountryCodeModule;
    }());
    TuiIsoToCountryCodeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiIsoToCountryCodeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodeModule, declarations: [TuiIsoToCountryCodePipe], exports: [TuiIsoToCountryCodePipe] });
    TuiIsoToCountryCodeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIsoToCountryCodeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiIsoToCountryCodePipe],
                        exports: [TuiIsoToCountryCodePipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiIsoToCountryCodeModule = TuiIsoToCountryCodeModule;
    exports.TuiIsoToCountryCodePipe = TuiIsoToCountryCodePipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
