// @taiga-ui/kit/bundles/taiga-ui-kit-directives-project-class.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/directives/project-class', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.directives = global["taiga-ui"].kit.directives || {}, global["taiga-ui"].kit.directives["project-class"] = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    /**
     * A directive for projecting classes from nested children to host
     */
    var TuiProjectClassDirective = /** @class */ (function () {
        function TuiProjectClassDirective(el) {
            this.el = el;
            this.classNames = [];
        }
        TuiProjectClassDirective.prototype.ngAfterViewChecked = function () {
            var _this = this;
            this.classNames.forEach(function (className) {
                var hostElement = _this.el.nativeElement;
                hostElement.classList.toggle(className, !!hostElement.querySelector("." + className));
            });
        };
        return TuiProjectClassDirective;
    }());
    TuiProjectClassDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassDirective, deps: [{ token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiProjectClassDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiProjectClassDirective, selector: "[tuiProjectClass]", inputs: { classNames: ["tuiProjectClass", "classNames"] }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiProjectClass]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { classNames: [{
                    type: i0.Input,
                    args: ['tuiProjectClass']
                }] } });

    var TuiProjectClassModule = /** @class */ (function () {
        function TuiProjectClassModule() {
        }
        return TuiProjectClassModule;
    }());
    TuiProjectClassModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiProjectClassModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassModule, declarations: [TuiProjectClassDirective], exports: [TuiProjectClassDirective] });
    TuiProjectClassModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiProjectClassModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiProjectClassDirective],
                        exports: [TuiProjectClassDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiProjectClassDirective = TuiProjectClassDirective;
    exports.TuiProjectClassModule = TuiProjectClassModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
