// @taiga-ui/kit/bundles/taiga-ui-kit-pipes-to-country-code.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/i18n'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/pipes/to-country-code', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/i18n', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.pipes = global["taiga-ui"].kit.pipes || {}, global["taiga-ui"].kit.pipes["to-country-code"] = {}), global.ng.core, global.cdk, global.i18n, global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils));
})(this, (function (exports, i0, cdk, i18n, tokens, utils) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiToCountryCodePipe = /** @class */ (function () {
        function TuiToCountryCodePipe(countriesMasks) {
            this.countriesMasks = countriesMasks;
        }
        TuiToCountryCodePipe.prototype.transform = function (value, countries) {
            var _this = this;
            return countries.find(function (countryIsoCode) {
                var ruCodeTest = countryIsoCode === i18n.TuiCountryIsoCode.RU &&
                    /^[78]/.test(value) &&
                    /^(?!880[1-9 ])/.test(value) &&
                    value.length + 1 ===
                        utils.tuiGetMaxAllowedPhoneLength(_this.countriesMasks, i18n.TuiCountryIsoCode.RU);
                var matched = ruCodeTest ||
                    (value.startsWith(utils.tuiIsoToCountryCode(_this.countriesMasks, countryIsoCode).replace(cdk.CHAR_PLUS, '')) &&
                        value.length + 1 ===
                            utils.tuiGetMaxAllowedPhoneLength(_this.countriesMasks, countryIsoCode));
                if (matched) {
                    switch (countryIsoCode) {
                        case i18n.TuiCountryIsoCode.RU:
                            return utils.tuiNotKzRegion(value);
                        case i18n.TuiCountryIsoCode.KZ:
                            return !utils.tuiNotKzRegion(value);
                        default:
                            return true;
                    }
                }
                return false;
            });
        };
        return TuiToCountryCodePipe;
    }());
    TuiToCountryCodePipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToCountryCodePipe, deps: [{ token: tokens.TUI_COUNTRIES_MASKS }], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    TuiToCountryCodePipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToCountryCodePipe, name: "tuiToCountryCode" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiToCountryCodePipe, decorators: [{
                type: i0.Pipe,
                args: [{
                        name: 'tuiToCountryCode',
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_COUNTRIES_MASKS]
                        }] }];
        } });

    var TuiExtractCountryCodeModule = /** @class */ (function () {
        function TuiExtractCountryCodeModule() {
        }
        return TuiExtractCountryCodeModule;
    }());
    TuiExtractCountryCodeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiExtractCountryCodeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiExtractCountryCodeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiExtractCountryCodeModule, declarations: [TuiToCountryCodePipe], exports: [TuiToCountryCodePipe] });
    TuiExtractCountryCodeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiExtractCountryCodeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiExtractCountryCodeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiToCountryCodePipe],
                        exports: [TuiToCountryCodePipe],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiExtractCountryCodeModule = TuiExtractCountryCodeModule;
    exports.TuiToCountryCodePipe = TuiToCountryCodePipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
