// @taiga-ui/kit/bundles/taiga-ui-kit-components-island.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/island', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.island = {}), global.ng.core, global.ng.common));
})(this, (function (exports, i0, common) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiIslandComponent = /** @class */ (function () {
        function TuiIslandComponent() {
            this.size = 'm';
            this.textAlign = 'left';
            this.hoverable = false;
            this.transparent = false;
        }
        Object.defineProperty(TuiIslandComponent.prototype, "sizeS", {
            get: function () {
                return this.size === 's';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiIslandComponent.prototype, "sizeM", {
            get: function () {
                return this.size === 'm';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiIslandComponent.prototype, "sizeL", {
            get: function () {
                return this.size === 'l';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiIslandComponent.prototype, "textAlignLeft", {
            get: function () {
                return this.textAlign === 'left';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiIslandComponent.prototype, "textAlignCenter", {
            get: function () {
                return this.textAlign === 'center';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiIslandComponent.prototype, "textAlignRight", {
            get: function () {
                return this.textAlign === 'right';
            },
            enumerable: false,
            configurable: true
        });
        return TuiIslandComponent;
    }());
    TuiIslandComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiIslandComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiIslandComponent, selector: "tui-island, a[tuiIsland]", inputs: { size: "size", textAlign: "textAlign", hoverable: "hoverable", transparent: "transparent" }, host: { properties: { "class.tui-island_hoverable": "this.hoverable", "class.tui-island_transparent": "this.transparent", "class.tui-island_size_s": "this.sizeS", "class.tui-island_size_m": "this.sizeM", "class.tui-island_size_l": "this.sizeL", "class.tui-island_text-align_left": "this.textAlignLeft", "class.tui-island_text-align_center": "this.textAlignCenter", "class.tui-island_text-align_right": "this.textAlignRight" }, classAttribute: "tui-island" }, ngImport: i0__namespace, template: "<ng-content></ng-content>\n", styles: [":host{display:block}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-island, a[tuiIsland]',
                        templateUrl: './island.template.html',
                        styleUrls: ['./island.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            class: 'tui-island',
                        },
                    }]
            }], propDecorators: { size: [{
                    type: i0.Input
                }], textAlign: [{
                    type: i0.Input
                }], hoverable: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class.tui-island_hoverable']
                }], transparent: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class.tui-island_transparent']
                }], sizeS: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_size_s']
                }], sizeM: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_size_m']
                }], sizeL: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_size_l']
                }], textAlignLeft: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_text-align_left']
                }], textAlignCenter: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_text-align_center']
                }], textAlignRight: [{
                    type: i0.HostBinding,
                    args: ['class.tui-island_text-align_right']
                }] } });

    var TuiIslandModule = /** @class */ (function () {
        function TuiIslandModule() {
        }
        return TuiIslandModule;
    }());
    TuiIslandModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiIslandModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandModule, declarations: [TuiIslandComponent], imports: [common.CommonModule], exports: [TuiIslandComponent] });
    TuiIslandModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandModule, imports: [[common.CommonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiIslandModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [common.CommonModule],
                        declarations: [TuiIslandComponent],
                        exports: [TuiIslandComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiIslandComponent = TuiIslandComponent;
    exports.TuiIslandModule = TuiIslandModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
