// @taiga-ui/kit/bundles/taiga-ui-kit-components-tiles.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/mutation-observer'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators'), require('@tinkoff/ng-event-plugins')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/tiles', ['exports', '@angular/core', '@ng-web-apis/mutation-observer', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators', '@tinkoff/ng-event-plugins'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components.tiles = {}), global.ng.core, global.mutationObserver, global.cdk, global.rxjs, global.rxjs.operators, global.ngEventPlugins));
})(this, (function (exports, i0, mutationObserver, cdk, i2, operators, ngEventPlugins) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiTilesComponent = /** @class */ (function () {
        function TuiTilesComponent(el) {
            var _this = this;
            this.el = el;
            this.el$ = new i2.Subject();
            this.debounce = 0;
            this.orderChange = this.el$.pipe(operators.debounce(function () { return i2.timer(_this.debounce); }), operators.filter(this.filter.bind(this)), operators.map(function (element) { return _this.reorder(element); }));
            this.element = null;
            this.order$ = new i2.BehaviorSubject(new Map());
        }
        Object.defineProperty(TuiTilesComponent.prototype, "order", {
            get: function () {
                return this.order$.value;
            },
            set: function (map) {
                this.order$.next(map);
            },
            enumerable: false,
            configurable: true
        });
        TuiTilesComponent.prototype.rearrange = function (element) {
            this.el$.next(element);
        };
        TuiTilesComponent.prototype.filter = function (element) {
            return !!this.element && !!element && this.element !== element;
        };
        TuiTilesComponent.prototype.reorder = function (element) {
            var _a, _b;
            var elements = Array.from(this.el.nativeElement.children);
            var currentIndex = elements.indexOf(this.element || element);
            var newIndex = elements.indexOf(element);
            var order = this.order.size
                ? new Map(this.order)
                : new Map(elements.map(function (_, index) { return [index, index]; }));
            var dragged = (_a = order.get(currentIndex)) !== null && _a !== void 0 ? _a : currentIndex;
            var placement = (_b = order.get(newIndex)) !== null && _b !== void 0 ? _b : newIndex;
            order.set(currentIndex, placement);
            order.set(newIndex, dragged);
            this.order$.next(order);
            return order;
        };
        return TuiTilesComponent;
    }());
    TuiTilesComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesComponent, deps: [{ token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiTilesComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTilesComponent, selector: "tui-tiles", inputs: { debounce: "debounce", order: "order" }, outputs: { orderChange: "orderChange" }, host: { listeners: { "pointerleave.silent": "rearrange()" }, properties: { "class._dragged": "this.element" } }, providers: [
            cdk.TuiDestroyService,
            cdk.TuiResizeService,
            mutationObserver.MutationObserverService,
            {
                provide: mutationObserver.MUTATION_OBSERVER_INIT,
                useValue: { childList: true },
            },
        ], ngImport: i0__namespace, template: '<ng-content></ng-content>', isInline: true, styles: ["tui-tiles{position:relative;z-index:0;display:grid;grid-auto-flow:dense;justify-items:stretch}tui-tiles._dragged tui-tile>.t-wrapper{pointer-events:none}tui-tiles._dragged tui-tile:not(._dragged)>.t-wrapper{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;transition-delay:1ms}tui-tiles:not(._dragged) tui-tile._dragged>.t-wrapper{transition-property:all;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;transition-delay:1ms}tui-tile>.t-wrapper{position:absolute;z-index:0;border-radius:inherit}tui-tile._dragged>.t-wrapper{z-index:1;transition:none}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-tiles',
                        template: '<ng-content></ng-content>',
                        styleUrls: ['./tiles.style.less'],
                        encapsulation: i0.ViewEncapsulation.None,
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            cdk.TuiDestroyService,
                            cdk.TuiResizeService,
                            mutationObserver.MutationObserverService,
                            {
                                provide: mutationObserver.MUTATION_OBSERVER_INIT,
                                useValue: { childList: true },
                            },
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { debounce: [{
                    type: i0.Input
                }], order: [{
                    type: i0.Input
                }], orderChange: [{
                    type: i0.Output
                }], element: [{
                    type: i0.HostBinding,
                    args: ['class._dragged']
                }], rearrange: [{
                    type: i0.HostListener,
                    args: ['pointerleave.silent']
                }] } });

    var TuiTileService = /** @class */ (function () {
        function TuiTileService(el, tiles, resize$, mutation$) {
            this.el = el;
            this.tiles = tiles;
            this.resize$ = resize$;
            this.mutation$ = mutation$;
            this.sub = new i2.Subscription();
            this.offset$ = new i2.BehaviorSubject([NaN, NaN]);
            this.position$ = i2.combineLatest([
                this.offset$.pipe(operators.distinctUntilChanged(cdk.tuiArrayShallowEquals)),
                this.resize$.pipe(operators.startWith(null)),
                this.mutation$.pipe(operators.startWith(null)),
                this.tiles.order$.pipe(operators.debounceTime(0)),
            ]).pipe(operators.map(function (_a) {
                var _b = __read(_a, 1), offset = _b[0];
                return offset;
            }));
        }
        TuiTileService.prototype.init = function (element) {
            var _this = this;
            this.sub.add(this.position$.subscribe(function (offset) {
                _this.setPosition(element, offset);
                _this.setRect(element, offset);
            }));
        };
        TuiTileService.prototype.setOffset = function (offset) {
            this.offset$.next(offset);
        };
        TuiTileService.prototype.ngOnDestroy = function () {
            this.sub.unsubscribe();
        };
        TuiTileService.prototype.getRect = function (_a) {
            var _b = __read(_a, 2), left = _b[0], top = _b[1];
            return {
                top: Number.isNaN(top) ? this.el.nativeElement.offsetTop : top,
                left: Number.isNaN(left) ? this.el.nativeElement.offsetLeft : left,
                width: this.el.nativeElement.clientWidth,
                height: this.el.nativeElement.clientHeight,
                right: NaN,
                bottom: NaN,
            };
        };
        TuiTileService.prototype.setRect = function (_a, offset) {
            var style = _a.style;
            var _b = this.getRect(offset), top = _b.top, left = _b.left, width = _b.width, height = _b.height;
            style.top = cdk.tuiPx(top);
            style.left = cdk.tuiPx(left);
            style.width = cdk.tuiPx(width);
            style.height = cdk.tuiPx(height);
        };
        TuiTileService.prototype.setPosition = function (element, _a) {
            var _b = __read(_a, 1), left = _b[0];
            if (!Number.isNaN(left)) {
                element.style.setProperty('position', 'fixed');
                return;
            }
            var style = element.style;
            var rect = element.getBoundingClientRect();
            var host = this.el.nativeElement.getBoundingClientRect();
            style.removeProperty('position');
            style.top = cdk.tuiPx(rect.top - host.top + this.el.nativeElement.offsetTop);
            style.left = cdk.tuiPx(rect.left - host.left + this.el.nativeElement.offsetLeft);
        };
        return TuiTileService;
    }());
    TuiTileService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileService, deps: [{ token: i0.ElementRef }, { token: TuiTilesComponent }, { token: cdk.TuiResizeService }, { token: mutationObserver.MutationObserverService }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiTileService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileService });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileService, decorators: [{
                type: i0.Injectable
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: TuiTilesComponent, decorators: [{
                            type: i0.Inject,
                            args: [TuiTilesComponent]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [cdk.TuiResizeService]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [mutationObserver.MutationObserverService]
                        }] }];
        } });

    var TuiTileComponent = /** @class */ (function () {
        function TuiTileComponent(service, el, tiles) {
            this.service = service;
            this.el = el;
            this.tiles = tiles;
            this.width = 1;
            this.height = 1;
            this.dragged = false;
        }
        Object.defineProperty(TuiTileComponent.prototype, "column", {
            get: function () {
                return "span var(--tui-width, " + this.width + ")";
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiTileComponent.prototype, "row", {
            get: function () {
                return "span var(--tui-height, " + this.height + ")";
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiTileComponent.prototype, "element", {
            get: function () {
                return this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        TuiTileComponent.prototype.onEnter = function () {
            this.tiles.rearrange(this.element);
        };
        TuiTileComponent.prototype.onDrag = function (offset) {
            var dragged = !Number.isNaN(offset[0]);
            this.dragged = this.dragged || dragged;
            this.tiles.element = dragged ? this.element : null;
            this.service.setOffset(offset);
        };
        TuiTileComponent.prototype.onTransitionEnd = function () {
            this.dragged = false;
        };
        TuiTileComponent.prototype.ngAfterViewInit = function () {
            if (this.wrapper) {
                this.service.init(this.wrapper.nativeElement);
            }
        };
        TuiTileComponent.prototype.ngOnDestroy = function () {
            if (this.tiles.element === this.element) {
                this.tiles.element = null;
            }
        };
        return TuiTileComponent;
    }());
    TuiTileComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileComponent, deps: [{ token: TuiTileService }, { token: i0.ElementRef }, { token: TuiTilesComponent }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiTileComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTileComponent, selector: "tui-tile", inputs: { width: "width", height: "height" }, host: { listeners: { "pointerenter": "onEnter()" }, properties: { "class._dragged": "this.dragged", "style.gridColumn": "this.column", "style.gridRow": "this.row" } }, providers: [TuiTileService], viewQueries: [{ propertyName: "wrapper", first: true, predicate: ["wrapper"], descendants: true }], ngImport: i0__namespace, template: "<div\n    #wrapper\n    class=\"t-wrapper\"\n    (transitionend.self)=\"onTransitionEnd()\"\n>\n    <ng-content></ng-content>\n</div>\n", changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-tile',
                        templateUrl: './tile.template.html',
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [TuiTileService],
                    }]
            }], ctorParameters: function () {
            return [{ type: TuiTileService, decorators: [{
                            type: i0.Inject,
                            args: [TuiTileService]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: TuiTilesComponent, decorators: [{
                            type: i0.Inject,
                            args: [TuiTilesComponent]
                        }] }];
        }, propDecorators: { wrapper: [{
                    type: i0.ViewChild,
                    args: ['wrapper']
                }], width: [{
                    type: i0.Input
                }], height: [{
                    type: i0.Input
                }], dragged: [{
                    type: i0.HostBinding,
                    args: ['class._dragged']
                }], column: [{
                    type: i0.HostBinding,
                    args: ['style.gridColumn']
                }], row: [{
                    type: i0.HostBinding,
                    args: ['style.gridRow']
                }], onEnter: [{
                    type: i0.HostListener,
                    args: ['pointerenter']
                }] } });

    function isInteracting(x) {
        if (x === void 0) { x = NaN; }
        return !Number.isNaN(x) || !Number.isNaN(this['x']);
    }
    function isDragging() {
        return !Number.isNaN(this['x']);
    }
    var TuiTileHandleDirective = /** @class */ (function () {
        function TuiTileHandleDirective(tile) {
            this.tile = tile;
            this.x = NaN;
            this.y = NaN;
        }
        TuiTileHandleDirective.prototype.onStart = function (event) {
            var target = cdk.tuiGetActualTarget(event);
            var x = event.x, y = event.y, pointerId = event.pointerId;
            if (cdk.tuiIsElement(target)) {
                target.releasePointerCapture(pointerId);
            }
            this.onPointer(x, y);
        };
        TuiTileHandleDirective.prototype.onPointer = function (x, y) {
            if (x === void 0) { x = NaN; }
            if (y === void 0) { y = NaN; }
            var _a = this.tile.element.getBoundingClientRect(), left = _a.left, top = _a.top;
            this.x = x - left;
            this.y = y - top;
            this.tile.onDrag([NaN, NaN]);
        };
        TuiTileHandleDirective.prototype.onMove = function (x, y) {
            this.tile.onDrag([x - this.x, y - this.y]);
        };
        return TuiTileHandleDirective;
    }());
    TuiTileHandleDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileHandleDirective, deps: [{ token: TuiTileComponent }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiTileHandleDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTileHandleDirective, selector: "[tuiTileHandle]", host: { listeners: { "pointerdown.silent": "onStart($event)", "document:pointerup.silent": "onPointer()", "document:pointermove.silent": "onMove($event.x,$event.y)" }, properties: { "style.touchAction": "\"none\"", "style.userSelect": "\"none\"" } }, ngImport: i0__namespace });
    __decorate([
        ngEventPlugins.shouldCall(isInteracting)
    ], TuiTileHandleDirective.prototype, "onPointer", null);
    __decorate([
        ngEventPlugins.shouldCall(isDragging)
    ], TuiTileHandleDirective.prototype, "onMove", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTileHandleDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiTileHandle]',
                        host: {
                            '[style.touchAction]': '"none"',
                            '[style.userSelect]': '"none"',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: TuiTileComponent, decorators: [{
                            type: i0.Inject,
                            args: [TuiTileComponent]
                        }] }];
        }, propDecorators: { onStart: [{
                    type: i0.HostListener,
                    args: ['pointerdown.silent', ['$event']]
                }], onPointer: [{
                    type: i0.HostListener,
                    args: ['document:pointerup.silent']
                }], onMove: [{
                    type: i0.HostListener,
                    args: ['document:pointermove.silent', ['$event.x', '$event.y']]
                }] } });

    var TuiTilesModule = /** @class */ (function () {
        function TuiTilesModule() {
        }
        return TuiTilesModule;
    }());
    TuiTilesModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiTilesModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesModule, declarations: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective], exports: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective] });
    TuiTilesModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTilesModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective],
                        exports: [TuiTilesComponent, TuiTileComponent, TuiTileHandleDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiTileComponent = TuiTileComponent;
    exports.TuiTileHandleDirective = TuiTileHandleDirective;
    exports.TuiTileService = TuiTileService;
    exports.TuiTilesComponent = TuiTilesComponent;
    exports.TuiTilesModule = TuiTilesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
