// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-date-multi.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@maskito/kit'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/classes'), require('@taiga-ui/kit/components/input-tag'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils'), require('@tinkoff/ng-polymorpheus'), require('rxjs/operators'), require('@maskito/angular'), require('@angular/common'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-date-multi', ['exports', '@angular/core', '@angular/forms', '@maskito/kit', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/classes', '@taiga-ui/kit/components/input-tag', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils', '@tinkoff/ng-polymorpheus', 'rxjs/operators', '@maskito/angular', '@angular/common', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["input-date-multi"] = {}), global.ng.core, global.ng.forms, global.kit, global.i6, global.i1, global["taiga-ui"].kit.classes, global["taiga-ui"].kit.components["input-tag"], global["taiga-ui"].kit.tokens, global["taiga-ui"].kit.utils, global.i5, global.rxjs.operators, global.i3, global.ng.common, global.rxjs));
})(this, (function (exports, i0, i4, kit, i6, i1, classes, i2, tokens, utils, i5, operators, i3, i7, i8) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);
    var i8__namespace = /*#__PURE__*/_interopNamespace(i8);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var TuiInputDateMultiComponent = /** @class */ (function (_super) {
        __extends(TuiInputDateMultiComponent, _super);
        function TuiInputDateMultiComponent(control, cdr, injector, isMobile, dialogs, mobileCalendar, dateFormat, dateSeparator, dateTexts$, valueTransformer, options, textfieldSize, doneWord$) {
            var _this = this;
            var _a, _b;
            _this = _super.call(this, control, cdr, valueTransformer) || this;
            _this.injector = injector;
            _this.isMobile = isMobile;
            _this.dialogs = dialogs;
            _this.mobileCalendar = mobileCalendar;
            _this.dateFormat = dateFormat;
            _this.dateSeparator = dateSeparator;
            _this.dateTexts$ = dateTexts$;
            _this.valueTransformer = valueTransformer;
            _this.options = options;
            _this.textfieldSize = textfieldSize;
            _this.doneWord$ = doneWord$;
            _this.month = null;
            _this.min = _this.options.min;
            _this.max = _this.options.max;
            _this.disabledItemHandler = i6.ALWAYS_FALSE_HANDLER;
            _this.markerHandler = i1.TUI_DEFAULT_MARKER_HANDLER;
            _this.defaultActiveYearMonth = i6.TuiMonth.currentLocal();
            _this.expandable = false;
            _this.inputHidden = false;
            _this.editable = true;
            _this.search = '';
            _this.placeholder = '';
            _this.rows = Infinity;
            _this.maskitoOptions = kit.maskitoDateOptionsGenerator({
                mode: 'dd/mm/yyyy',
                separator: '.',
                min: (_a = _this.min) === null || _a === void 0 ? void 0 : _a.toLocalNativeDate(),
                max: (_b = _this.max) === null || _b === void 0 ? void 0 : _b.toLocalNativeDate(),
            });
            _this.open = false;
            _this.filler$ = _this.dateTexts$.pipe(operators.map(function (dateTexts) { return i6.changeDateSeparator(dateTexts[_this.dateFormat], _this.dateSeparator); }));
            _this.tagValidator = function (tag) {
                var _a, _b, _c;
                var _d = i6.tuiIsString(tag)
                    ? i6.TuiDay.parseRawDateString(tag)
                    : tag, year = _d.year, month = _d.month, day = _d.day;
                var date = new i6.TuiDay(year, month, day);
                return ((_c = (i6.TuiDay.isValidDay(year, month, day) &&
                    ((_a = _this.min) === null || _a === void 0 ? void 0 : _a.dayBefore(date)) &&
                    ((_b = _this.max) === null || _b === void 0 ? void 0 : _b.dayAfter(date)))) !== null && _c !== void 0 ? _c : false);
            };
            _this.disabledItemHandlerWrapper = function (handler) { return function (stringifiable) { return i6.tuiIsString(stringifiable) || handler(stringifiable.item); }; };
            _this.valueMapper = function (value) { return value.map(function (item) { return new classes.TuiStringifiableItem(item, function (item) { return String(item); }); }); };
            return _this;
        }
        TuiInputDateMultiComponent.prototype.onClick = function () {
            if (!this.isMobile) {
                this.open = !this.open;
            }
        };
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "nativeDropdownMode", {
            get: function () {
                return this.isMobile && !this.editable;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "computedMin", {
            get: function () {
                var _a;
                return (_a = this.min) !== null && _a !== void 0 ? _a : this.options.min;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "computedMax", {
            get: function () {
                var _a;
                return (_a = this.max) !== null && _a !== void 0 ? _a : this.options.max;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.textfield) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || null;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "focused", {
            get: function () {
                var _a;
                return !!((_a = this.textfield) === null || _a === void 0 ? void 0 : _a.focused);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "computedMobile", {
            get: function () {
                return this.isMobile && !!this.mobileCalendar;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "calendarIcon", {
            get: function () {
                return this.options.icon;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "computedActiveYearMonth", {
            get: function () {
                return (this.month ||
                    this.value[this.value.length - 1] ||
                    i6.tuiDateClamp(this.defaultActiveYearMonth, this.computedMin, this.computedMax));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiInputDateMultiComponent.prototype, "canOpen", {
            get: function () {
                return this.interactive && !this.computedMobile;
            },
            enumerable: false,
            configurable: true
        });
        TuiInputDateMultiComponent.prototype.onIconClick = function () {
            var _this = this;
            if (!this.computedMobile || !this.mobileCalendar) {
                return;
            }
            this.dialogs
                .open(new i5.PolymorpheusComponent(this.mobileCalendar, this.injector), {
                size: 'fullscreen',
                closeable: false,
                data: {
                    multi: true,
                    min: this.min,
                    max: this.max,
                    disabledItemHandler: this.disabledItemHandler,
                },
            })
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (value) {
                _this.value = value;
            });
        };
        TuiInputDateMultiComponent.prototype.onEnter = function (search) {
            if (!this.tagValidator(search)) {
                return;
            }
            this.value = utils.tuiImmutableUpdateInputDateMulti(this.value, i6.TuiDay.normalizeParse(search));
            if (this.inputTag) {
                this.inputTag.search = '';
            }
            this.done();
        };
        TuiInputDateMultiComponent.prototype.onValueChange = function (value) {
            var _a;
            (_a = this.control) === null || _a === void 0 ? void 0 : _a.updateValueAndValidity({ emitEvent: false });
            if (!value.length) {
                this.onOpenChange(true);
            }
            this.value = value.map(function (_d) {
                var item = _d.item;
                return item;
            });
        };
        TuiInputDateMultiComponent.prototype.onDayClick = function (value) {
            this.value = utils.tuiImmutableUpdateInputDateMulti(this.value, value);
        };
        TuiInputDateMultiComponent.prototype.done = function () {
            this.open = false;
        };
        TuiInputDateMultiComponent.prototype.onMonthChange = function (month) {
            this.month = month;
        };
        TuiInputDateMultiComponent.prototype.onOpenChange = function (open) {
            this.open = open;
        };
        TuiInputDateMultiComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiInputDateMultiComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        return TuiInputDateMultiComponent;
    }(i6.AbstractTuiMultipleControl));
    TuiInputDateMultiComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiComponent, deps: [{ token: i4.NgControl, optional: true, self: true }, { token: i0.ChangeDetectorRef }, { token: i0.Injector }, { token: i6.TUI_IS_MOBILE }, { token: i1.TuiDialogService }, { token: tokens.TUI_MOBILE_CALENDAR, optional: true }, { token: i6.TUI_DATE_FORMAT }, { token: i6.TUI_DATE_SEPARATOR }, { token: tokens.TUI_DATE_TEXTS }, { token: tokens.TUI_DATE_VALUE_TRANSFORMER, optional: true }, { token: tokens.TUI_INPUT_DATE_OPTIONS }, { token: i1.TUI_TEXTFIELD_SIZE }, { token: tokens.TUI_DONE_WORD }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiInputDateMultiComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputDateMultiComponent, selector: "tui-input-date[multiple]", inputs: { min: "min", max: "max", disabledItemHandler: "disabledItemHandler", markerHandler: "markerHandler", defaultActiveYearMonth: "defaultActiveYearMonth", expandable: "expandable", inputHidden: "inputHidden", editable: "editable", search: "search", placeholder: "placeholder", rows: "rows", tagValidator: "tagValidator" }, host: { listeners: { "click": "onClick()" }, properties: { "class._editable": "this.editable", "attr.data-size": "this.size" } }, providers: [
            i6.tuiAsFocusableItemAccessor(TuiInputDateMultiComponent),
            i6.tuiAsControl(TuiInputDateMultiComponent),
            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_VALUE_TRANSFORMER),
        ], viewQueries: [{ propertyName: "textfield", first: true, predicate: i1.TuiPrimitiveTextfieldComponent, descendants: true }, { propertyName: "inputTag", first: true, predicate: i2.TuiInputTagComponent, descendants: true }], usesInheritance: true, ngImport: i0__namespace, template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"canOpen\"\n    [content]=\"dropdown\"\n    [open]=\"open && canOpen\"\n    (openChange)=\"onOpenChange($event)\"\n>\n    <tui-input-tag\n        #inputTag\n        class=\"t-input\"\n        [disabled]=\"disabled\"\n        [disabledItemHandler]=\"disabledItemHandler | tuiMapper: disabledItemHandlerWrapper\"\n        [editable]=\"editable\"\n        [expandable]=\"expandable\"\n        [inputHidden]=\"inputHidden\"\n        [maskito]=\"maskitoOptions\"\n        [nativeId]=\"nativeId\"\n        [ngModel]=\"value | tuiMapper: valueMapper\"\n        [placeholder]=\"placeholder\"\n        [pseudoFocus]=\"pseudoFocus\"\n        [pseudoHover]=\"pseudoHover\"\n        [pseudoInvalid]=\"computedInvalid\"\n        [readOnly]=\"readOnly\"\n        [removable]=\"!nativeDropdownMode\"\n        [rows]=\"rows\"\n        [tagValidator]=\"tagValidator | tuiMapper: disabledItemHandlerWrapper\"\n        [tuiTextfieldFiller]=\"(filler$ | async) || ''\"\n        [tuiTextfieldIcon]=\"calendarIcon && iconContent\"\n        (keydown.enter)=\"onEnter(inputTag.search || '')\"\n        (ngModelChange)=\"onValueChange($event)\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            ngProjectAs=\"input\"\n            select=\"input\"\n        ></ng-content>\n    </tui-input-tag>\n\n    <ng-template #iconContent>\n        <tui-svg\n            *polymorpheusOutlet=\"calendarIcon as src; context: {$implicit: size}\"\n            appearance=\"icon\"\n            tuiWrapper\n            [class.t-icon]=\"!computedDisabled\"\n            [src]=\"src\"\n            (click)=\"onIconClick()\"\n        ></tui-svg>\n    </ng-template>\n\n    <ng-template #dropdown>\n        <tui-calendar\n            [disabledItemHandler]=\"disabledItemHandler\"\n            [markerHandler]=\"markerHandler\"\n            [max]=\"computedMax\"\n            [min]=\"computedMin\"\n            [month]=\"computedActiveYearMonth\"\n            [value]=\"value\"\n            (dayClick)=\"onDayClick($event)\"\n            (monthChange)=\"onMonthChange($event)\"\n            (mousedown.silent.prevent)=\"(0)\"\n        ></tui-calendar>\n        <div\n            class=\"t-button\"\n            (mousedown.silent.prevent)=\"(0)\"\n        >\n            <button\n                tuiLink\n                type=\"button\"\n                (click)=\"done()\"\n            >\n                {{ doneWord$ | async }}\n            </button>\n        </div>\n    </ng-template>\n</tui-hosted-dropdown>\n", styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-icon{pointer-events:auto}.t-button{display:flex;height:2.75rem;justify-content:center;box-shadow:inset 0 1px var(--tui-base-03)}.t-button button{flex:1;text-align:center}.t-native-input{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0;pointer-events:auto;font-size:2rem}\n"], components: [{ type: i1__namespace.TuiHostedDropdownComponent, selector: "tui-hosted-dropdown", inputs: ["content", "sided", "canOpen", "open"], outputs: ["openChange", "focusedChange"] }, { type: i2__namespace.TuiInputTagComponent, selector: "tui-input-tag", inputs: ["separator", "search", "editable", "tagValidator", "expandable", "rows", "inputHidden", "uniqueTags", "maxLength", "placeholder", "removable", "disabledItemHandler", "pseudoFocused"], outputs: ["searchChange"] }, { type: i1__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }, { type: i1__namespace.TuiCalendarComponent, selector: "tui-calendar", inputs: ["initialView", "month", "disabledItemHandler", "min", "max", "minViewedMonth", "maxViewedMonth", "hoveredItem", "showAdjacent", "markerHandler", "value"], outputs: ["dayClick", "monthChange", "hoveredItemChange"] }, { type: i1__namespace.TuiLinkComponent, selector: "a[tuiLink], button[tuiLink]", inputs: ["pseudo", "icon", "iconAlign", "iconRotated", "mode"], exportAs: ["tuiLink"] }], directives: [{ type: i3__namespace.MaskitoDirective, selector: "[maskito]", inputs: ["maskito", "maskitoElement"] }, { type: i4__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i4__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i1__namespace.TuiTextfieldFillerDirective, selector: "[tuiTextfieldFiller]", inputs: ["tuiTextfieldFiller"] }, { type: i1__namespace.TuiTextfieldIconDirective, selector: "[tuiTextfieldIcon]", inputs: ["tuiTextfieldIcon"] }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1__namespace.TuiWrapperDirective, selector: "[tuiWrapper]", inputs: ["disabled", "readOnly", "hover", "active", "focus", "invalid", "appearance"] }], pipes: { "tuiMapper": i6__namespace.TuiMapperPipe, "async": i7__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-input-date[multiple]',
                        templateUrl: './input-date-multi.template.html',
                        styleUrls: ['../input-date/input-date.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [
                            i6.tuiAsFocusableItemAccessor(TuiInputDateMultiComponent),
                            i6.tuiAsControl(TuiInputDateMultiComponent),
                            tokens.tuiDateStreamWithTransformer(tokens.TUI_DATE_VALUE_TRANSFORMER),
                        ],
                    }]
            }], ctorParameters: function () {
            return [{ type: i4__namespace.NgControl, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i4.NgControl]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i0__namespace.Injector, decorators: [{
                            type: i0.Inject,
                            args: [i0.Injector]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_IS_MOBILE]
                        }] }, { type: i1__namespace.TuiDialogService, decorators: [{
                            type: i0.Inject,
                            args: [i1.TuiDialogService]
                        }] }, { type: i0__namespace.Type, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_MOBILE_CALENDAR]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_DATE_FORMAT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i6.TUI_DATE_SEPARATOR]
                        }] }, { type: i8__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_TEXTS]
                        }] }, { type: i6__namespace.AbstractTuiValueTransformer, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [tokens.TUI_DATE_VALUE_TRANSFORMER]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_INPUT_DATE_OPTIONS]
                        }] }, { type: i1__namespace.TuiTextfieldSizeDirective, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_TEXTFIELD_SIZE]
                        }] }, { type: i8__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DONE_WORD]
                        }] }];
        }, propDecorators: { textfield: [{
                    type: i0.ViewChild,
                    args: [i1.TuiPrimitiveTextfieldComponent]
                }], inputTag: [{
                    type: i0.ViewChild,
                    args: [i2.TuiInputTagComponent]
                }], min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], disabledItemHandler: [{
                    type: i0.Input
                }], markerHandler: [{
                    type: i0.Input
                }], defaultActiveYearMonth: [{
                    type: i0.Input
                }], expandable: [{
                    type: i0.Input
                }], inputHidden: [{
                    type: i0.Input
                }], editable: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._editable']
                }], search: [{
                    type: i0.Input
                }], placeholder: [{
                    type: i0.Input
                }], rows: [{
                    type: i0.Input
                }], tagValidator: [{
                    type: i0.Input
                }], onClick: [{
                    type: i0.HostListener,
                    args: ['click']
                }], size: [{
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }] } });

    var TuiInputDateMultiModule = /** @class */ (function () {
        function TuiInputDateMultiModule() {
        }
        return TuiInputDateMultiModule;
    }());
    TuiInputDateMultiModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiInputDateMultiModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiModule, declarations: [TuiInputDateMultiComponent], imports: [i7.CommonModule,
            i3.MaskitoModule,
            i5.PolymorpheusModule,
            i1.TuiWrapperModule,
            i1.TuiCalendarModule,
            i1.TuiSvgModule,
            i1.TuiLinkModule,
            i2.TuiInputTagModule,
            i4.FormsModule,
            i6.TuiMapperPipeModule,
            i1.TuiHostedDropdownModule,
            i1.TuiTextfieldControllerModule,
            i1.TuiPrimitiveTextfieldModule], exports: [TuiInputDateMultiComponent] });
    TuiInputDateMultiModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiModule, imports: [[
                i7.CommonModule,
                i3.MaskitoModule,
                i5.PolymorpheusModule,
                i1.TuiWrapperModule,
                i1.TuiCalendarModule,
                i1.TuiSvgModule,
                i1.TuiLinkModule,
                i2.TuiInputTagModule,
                i4.FormsModule,
                i6.TuiMapperPipeModule,
                i1.TuiHostedDropdownModule,
                i1.TuiTextfieldControllerModule,
                i1.TuiPrimitiveTextfieldModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiInputDateMultiModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i7.CommonModule,
                            i3.MaskitoModule,
                            i5.PolymorpheusModule,
                            i1.TuiWrapperModule,
                            i1.TuiCalendarModule,
                            i1.TuiSvgModule,
                            i1.TuiLinkModule,
                            i2.TuiInputTagModule,
                            i4.FormsModule,
                            i6.TuiMapperPipeModule,
                            i1.TuiHostedDropdownModule,
                            i1.TuiTextfieldControllerModule,
                            i1.TuiPrimitiveTextfieldModule,
                        ],
                        declarations: [TuiInputDateMultiComponent],
                        exports: [TuiInputDateMultiComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiInputDateMultiComponent = TuiInputDateMultiComponent;
    exports.TuiInputDateMultiModule = TuiInputDateMultiModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
