// @taiga-ui/kit/bundles/taiga-ui-kit-tokens.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/kit/utils'), require('@taiga-ui/cdk'), require('@taiga-ui/i18n')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/tokens', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/kit/utils', '@taiga-ui/cdk', '@taiga-ui/i18n'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.tokens = {}), global.ng.core, global.ng.forms, global["taiga-ui"].kit.utils, global.cdk, global.i18n));
})(this, (function (exports, core, forms, utils, cdk, i18n) { 'use strict';

    /**
     * Stream that emits calendar data change
     */
    var TUI_CALENDAR_DATE_STREAM = new core.InjectionToken('[TUI_CALENDAR_DATE_STREAM]');
    function tuiDateStreamWithTransformer(transformer) {
        return {
            provide: TUI_CALENDAR_DATE_STREAM,
            deps: [
                [new core.Optional(), new core.Self(), forms.NgControl],
                [new core.Optional(), transformer],
            ],
            useFactory: utils.tuiControlValueFactory,
        };
    }

    var _a;
    var TUI_COUNTRIES_MASKS = cdk.tuiCreateToken((_a = {},
        _a[i18n.TuiCountryIsoCode.AD] = '+376###-###',
        _a[i18n.TuiCountryIsoCode.AE] = '+971-##-###-####',
        _a[i18n.TuiCountryIsoCode.AF] = '+93##-###-####',
        _a[i18n.TuiCountryIsoCode.AG] = '+1(268) ###-####',
        _a[i18n.TuiCountryIsoCode.AI] = '+1(264) ###-####',
        _a[i18n.TuiCountryIsoCode.AL] = '+355(###) ###-###',
        _a[i18n.TuiCountryIsoCode.AM] = '+374##-###-###',
        _a[i18n.TuiCountryIsoCode.AO] = '+244(###) ###-###',
        _a[i18n.TuiCountryIsoCode.AR] = '+54(###) ####-####',
        _a[i18n.TuiCountryIsoCode.AT] = '+43(###) ###-####',
        _a[i18n.TuiCountryIsoCode.AU] = '+61#-####-####',
        _a[i18n.TuiCountryIsoCode.AW] = '+297###-####',
        _a[i18n.TuiCountryIsoCode.AZ] = '+994##-###-##-##',
        _a[i18n.TuiCountryIsoCode.BA] = '+387 ###-####-##',
        _a[i18n.TuiCountryIsoCode.BB] = '+1(246) ###-####',
        _a[i18n.TuiCountryIsoCode.BD] = '+880##-###-###',
        _a[i18n.TuiCountryIsoCode.BE] = '+32(###) ###-###',
        _a[i18n.TuiCountryIsoCode.BF] = '+226##-##-####',
        _a[i18n.TuiCountryIsoCode.BG] = '+359(###) ###-###',
        _a[i18n.TuiCountryIsoCode.BH] = '+973####-####',
        _a[i18n.TuiCountryIsoCode.BI] = '+257##-##-####',
        _a[i18n.TuiCountryIsoCode.BJ] = '+229##-##-####',
        _a[i18n.TuiCountryIsoCode.BL] = '+590 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.BM] = '+1(441) ###-####',
        _a[i18n.TuiCountryIsoCode.BN] = '+673###-####',
        _a[i18n.TuiCountryIsoCode.BO] = '+591#-###-####',
        _a[i18n.TuiCountryIsoCode.BQ] = '+599-###-####',
        _a[i18n.TuiCountryIsoCode.BR] = '+55(##) ####-#####',
        _a[i18n.TuiCountryIsoCode.BS] = '+1(242) ###-####',
        _a[i18n.TuiCountryIsoCode.BT] = '+975#-###-###',
        _a[i18n.TuiCountryIsoCode.BW] = '+267##-###-###',
        _a[i18n.TuiCountryIsoCode.BY] = '+375(##) ###-##-##',
        _a[i18n.TuiCountryIsoCode.BZ] = '+501###-####',
        _a[i18n.TuiCountryIsoCode.CA] = '+1(###) ###-####',
        _a[i18n.TuiCountryIsoCode.CD] = '+243(###) ###-###',
        _a[i18n.TuiCountryIsoCode.CF] = '+236-##-##-####',
        _a[i18n.TuiCountryIsoCode.CG] = '+242##-###-####',
        _a[i18n.TuiCountryIsoCode.CH] = '+41##-###-####',
        _a[i18n.TuiCountryIsoCode.CI] = '+225##-##-###-###',
        _a[i18n.TuiCountryIsoCode.CL] = '+56#-####-####',
        _a[i18n.TuiCountryIsoCode.CM] = '+237#####-####',
        _a[i18n.TuiCountryIsoCode.CN] = '+86(###) ####-####',
        _a[i18n.TuiCountryIsoCode.CO] = '+57(###) ###-####',
        _a[i18n.TuiCountryIsoCode.CR] = '+506####-####',
        _a[i18n.TuiCountryIsoCode.CU] = '+53#-###-####',
        _a[i18n.TuiCountryIsoCode.CV] = '+238(###) ##-##',
        _a[i18n.TuiCountryIsoCode.CW] = '+5999-###-####',
        _a[i18n.TuiCountryIsoCode.CY] = '+357##-###-###',
        _a[i18n.TuiCountryIsoCode.CZ] = '+420(###) ###-###',
        _a[i18n.TuiCountryIsoCode.DE] = '+49(###) ###-###-##',
        _a[i18n.TuiCountryIsoCode.DJ] = '+253##-##-##-##',
        _a[i18n.TuiCountryIsoCode.DK] = '+45##-##-##-##',
        _a[i18n.TuiCountryIsoCode.DM] = '+1(767) ###-####',
        _a[i18n.TuiCountryIsoCode.DO] = '+1(###) ###-####',
        _a[i18n.TuiCountryIsoCode.DZ] = '+213##-###-####',
        _a[i18n.TuiCountryIsoCode.EC] = '+593##-###-####',
        _a[i18n.TuiCountryIsoCode.EE] = '+372####-####',
        _a[i18n.TuiCountryIsoCode.EG] = '+20(###) ###-####',
        _a[i18n.TuiCountryIsoCode.ER] = '+291#-###-###',
        _a[i18n.TuiCountryIsoCode.ES] = '+34(###) ###-###',
        _a[i18n.TuiCountryIsoCode.ET] = '+251##-###-####',
        _a[i18n.TuiCountryIsoCode.FI] = '+358(###) ###-##-##',
        _a[i18n.TuiCountryIsoCode.FJ] = '+679##-#####',
        _a[i18n.TuiCountryIsoCode.FK] = '+500#####',
        _a[i18n.TuiCountryIsoCode.FM] = '+691###-####',
        _a[i18n.TuiCountryIsoCode.FR] = '+33 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.GA] = '+241##-##-##-##',
        _a[i18n.TuiCountryIsoCode.GB] = '+44##-####-####',
        _a[i18n.TuiCountryIsoCode.GD] = '+1(473) ###-####',
        _a[i18n.TuiCountryIsoCode.GE] = '+995(###) ###-###',
        _a[i18n.TuiCountryIsoCode.GF] = '+594 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.GH] = '+233(###) ###-###',
        _a[i18n.TuiCountryIsoCode.GI] = '+350###-#####',
        _a[i18n.TuiCountryIsoCode.GL] = '+299##-##-##',
        _a[i18n.TuiCountryIsoCode.GM] = '+220(###) ##-##',
        _a[i18n.TuiCountryIsoCode.GN] = '+224##-###-###',
        _a[i18n.TuiCountryIsoCode.GP] = '+590 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.GQ] = '+240##-###-####',
        _a[i18n.TuiCountryIsoCode.GR] = '+30(###) ###-####',
        _a[i18n.TuiCountryIsoCode.GT] = '+502#-###-####',
        _a[i18n.TuiCountryIsoCode.GW] = '+245#-######',
        _a[i18n.TuiCountryIsoCode.GY] = '+592###-####',
        _a[i18n.TuiCountryIsoCode.HK] = '+852####-####',
        _a[i18n.TuiCountryIsoCode.HN] = '+504####-####',
        _a[i18n.TuiCountryIsoCode.HR] = '+385##-###-####',
        _a[i18n.TuiCountryIsoCode.HT] = '+509##-##-####',
        _a[i18n.TuiCountryIsoCode.HU] = '+36(###) ###-###',
        _a[i18n.TuiCountryIsoCode.ID] = '+62(###) ###-##-###',
        _a[i18n.TuiCountryIsoCode.IE] = '+353(###) ###-###',
        _a[i18n.TuiCountryIsoCode.IL] = '+972##-###-####',
        _a[i18n.TuiCountryIsoCode.IN] = '+91(####) ###-###',
        _a[i18n.TuiCountryIsoCode.IQ] = '+964(###) ###-####',
        _a[i18n.TuiCountryIsoCode.IR] = '+98(###) ###-####',
        _a[i18n.TuiCountryIsoCode.IS] = '+354###-####',
        _a[i18n.TuiCountryIsoCode.IT] = '+39(###) ####-###',
        _a[i18n.TuiCountryIsoCode.JM] = '+1(876) ###-####',
        _a[i18n.TuiCountryIsoCode.JO] = '+962#-####-####',
        _a[i18n.TuiCountryIsoCode.JP] = '+81-##-####-####',
        _a[i18n.TuiCountryIsoCode.KE] = '+254###-######',
        _a[i18n.TuiCountryIsoCode.KG] = '+996(###) ###-###',
        _a[i18n.TuiCountryIsoCode.KH] = '+855##-###-###',
        _a[i18n.TuiCountryIsoCode.KM] = '+269##-#####',
        _a[i18n.TuiCountryIsoCode.KN] = '+1(869) ###-####',
        _a[i18n.TuiCountryIsoCode.KP] = '+850####-#############',
        _a[i18n.TuiCountryIsoCode.KR] = '+82##-###-####',
        _a[i18n.TuiCountryIsoCode.KW] = '+965####-####',
        _a[i18n.TuiCountryIsoCode.KY] = '+1(345) ###-####',
        _a[i18n.TuiCountryIsoCode.KZ] = '+7(###) ###-##-##',
        _a[i18n.TuiCountryIsoCode.LA] = '+856##-##-###-###',
        _a[i18n.TuiCountryIsoCode.LB] = '+961##-###-###',
        _a[i18n.TuiCountryIsoCode.LC] = '+1(758) ###-####',
        _a[i18n.TuiCountryIsoCode.LI] = '+423(###) ###-####',
        _a[i18n.TuiCountryIsoCode.LK] = '+94##-###-####',
        _a[i18n.TuiCountryIsoCode.LR] = '+231##-###-###',
        _a[i18n.TuiCountryIsoCode.LS] = '+266#-###-####',
        _a[i18n.TuiCountryIsoCode.LT] = '+370(###) ##-###',
        _a[i18n.TuiCountryIsoCode.LU] = '+352(###) ###-###',
        _a[i18n.TuiCountryIsoCode.LV] = '+371##-###-###',
        _a[i18n.TuiCountryIsoCode.LY] = '+218##-###-####',
        _a[i18n.TuiCountryIsoCode.MA] = '+212##-####-###',
        _a[i18n.TuiCountryIsoCode.MC] = '+377###-###-###',
        _a[i18n.TuiCountryIsoCode.MD] = '+373####-####',
        _a[i18n.TuiCountryIsoCode.ME] = '+382##-###-###',
        _a[i18n.TuiCountryIsoCode.MF] = '+590 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.MG] = '+261##-##-#####',
        _a[i18n.TuiCountryIsoCode.MK] = '+389##-###-###',
        _a[i18n.TuiCountryIsoCode.ML] = '+223##-##-####',
        _a[i18n.TuiCountryIsoCode.MM] = '+95##-###-###',
        _a[i18n.TuiCountryIsoCode.MN] = '+976##-##-####',
        _a[i18n.TuiCountryIsoCode.MO] = '+853####-####',
        _a[i18n.TuiCountryIsoCode.MQ] = '+596 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.MR] = '+222##-##-####',
        _a[i18n.TuiCountryIsoCode.MS] = '+1(664) ###-####',
        _a[i18n.TuiCountryIsoCode.MT] = '+356####-####',
        _a[i18n.TuiCountryIsoCode.MU] = '+230####-####',
        _a[i18n.TuiCountryIsoCode.MV] = '+960###-####',
        _a[i18n.TuiCountryIsoCode.MW] = '+265#-####-####',
        _a[i18n.TuiCountryIsoCode.MX] = '+52(###) ###-####',
        _a[i18n.TuiCountryIsoCode.MY] = '+60(###) ###-###',
        _a[i18n.TuiCountryIsoCode.MZ] = '+258##-###-###',
        _a[i18n.TuiCountryIsoCode.NA] = '+264##-###-####',
        _a[i18n.TuiCountryIsoCode.NC] = '+687 ### ###',
        _a[i18n.TuiCountryIsoCode.NE] = '+227##-##-####',
        _a[i18n.TuiCountryIsoCode.NG] = '+234(###) ###-####',
        _a[i18n.TuiCountryIsoCode.NI] = '+505####-####',
        _a[i18n.TuiCountryIsoCode.NL] = '+31##-###-####',
        _a[i18n.TuiCountryIsoCode.NO] = '+47(###) ##-###',
        _a[i18n.TuiCountryIsoCode.NP] = '+977##-###-###',
        _a[i18n.TuiCountryIsoCode.NZ] = '+64(###) ###-####',
        _a[i18n.TuiCountryIsoCode.OM] = '+968##-###-###',
        _a[i18n.TuiCountryIsoCode.PA] = '+507###-####',
        _a[i18n.TuiCountryIsoCode.PE] = '+51(###) ###-###',
        _a[i18n.TuiCountryIsoCode.PF] = '+689 ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.PG] = '+675(###) ##-###',
        _a[i18n.TuiCountryIsoCode.PH] = '+63(###) ###-####',
        _a[i18n.TuiCountryIsoCode.PK] = '+92(###) ###-####',
        _a[i18n.TuiCountryIsoCode.PL] = '+48(###) ###-###',
        _a[i18n.TuiCountryIsoCode.PT] = '+351##-###-####',
        _a[i18n.TuiCountryIsoCode.PW] = '+680###-####',
        _a[i18n.TuiCountryIsoCode.PY] = '+595(###) ###-###',
        _a[i18n.TuiCountryIsoCode.QA] = '+974####-####',
        _a[i18n.TuiCountryIsoCode.RE] = '+262 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.RO] = '+40##-###-####',
        _a[i18n.TuiCountryIsoCode.RS] = '+381##-###-####',
        _a[i18n.TuiCountryIsoCode.RU] = '+7### ###-##-##',
        _a[i18n.TuiCountryIsoCode.RW] = '+250(###) ###-###',
        _a[i18n.TuiCountryIsoCode.SA] = '+966#-####-####',
        _a[i18n.TuiCountryIsoCode.SB] = '+677###-####',
        _a[i18n.TuiCountryIsoCode.SC] = '+248#-###-###',
        _a[i18n.TuiCountryIsoCode.SD] = '+249##-###-####',
        _a[i18n.TuiCountryIsoCode.SE] = '+46##-###-####',
        _a[i18n.TuiCountryIsoCode.SG] = '+65####-####',
        _a[i18n.TuiCountryIsoCode.SH] = '+290####',
        _a[i18n.TuiCountryIsoCode.SI] = '+386##-###-###',
        _a[i18n.TuiCountryIsoCode.SK] = '+421(###) ###-###',
        _a[i18n.TuiCountryIsoCode.SL] = '+232##-######',
        _a[i18n.TuiCountryIsoCode.SM] = '+378####-######',
        _a[i18n.TuiCountryIsoCode.SN] = '+221##-###-####',
        _a[i18n.TuiCountryIsoCode.SO] = '+252##-###-###',
        _a[i18n.TuiCountryIsoCode.SR] = '+597###-####',
        _a[i18n.TuiCountryIsoCode.ST] = '+239##-#####',
        _a[i18n.TuiCountryIsoCode.SV] = '+503##-##-####',
        _a[i18n.TuiCountryIsoCode.SX] = '+1(721) ###-####',
        _a[i18n.TuiCountryIsoCode.SY] = '+963##-####-###',
        _a[i18n.TuiCountryIsoCode.SZ] = '+268##-##-####',
        _a[i18n.TuiCountryIsoCode.TC] = '+1(649) ###-####',
        _a[i18n.TuiCountryIsoCode.TD] = '+235##-##-##-##',
        _a[i18n.TuiCountryIsoCode.TG] = '+228##-###-###',
        _a[i18n.TuiCountryIsoCode.TH] = '+66##-###-####',
        _a[i18n.TuiCountryIsoCode.TJ] = '+992##-###-####',
        _a[i18n.TuiCountryIsoCode.TL] = '+670###-#####',
        _a[i18n.TuiCountryIsoCode.TM] = '+993#-###-####',
        _a[i18n.TuiCountryIsoCode.TN] = '+216##-###-###',
        _a[i18n.TuiCountryIsoCode.TO] = '+676#####',
        _a[i18n.TuiCountryIsoCode.TR] = '+90(###) ###-####',
        _a[i18n.TuiCountryIsoCode.TT] = '+1(868) ###-####',
        _a[i18n.TuiCountryIsoCode.TW] = '+886#-####-####',
        _a[i18n.TuiCountryIsoCode.TZ] = '+255##-###-####',
        _a[i18n.TuiCountryIsoCode.UA] = '+380(##) ###-##-##',
        _a[i18n.TuiCountryIsoCode.UG] = '+256(###) ###-###',
        _a[i18n.TuiCountryIsoCode.US] = '+1(###) ###-####',
        _a[i18n.TuiCountryIsoCode.UY] = '+598#-###-##-##',
        _a[i18n.TuiCountryIsoCode.UZ] = '+998##-###-####',
        _a[i18n.TuiCountryIsoCode.VC] = '+1(784) ###-####',
        _a[i18n.TuiCountryIsoCode.VE] = '+58(###) ###-####',
        _a[i18n.TuiCountryIsoCode.VG] = '+1(284)###-####',
        _a[i18n.TuiCountryIsoCode.VN] = '+84(###) ####-###',
        _a[i18n.TuiCountryIsoCode.VU] = '+678##-#####',
        _a[i18n.TuiCountryIsoCode.WS] = '+685##-####',
        _a[i18n.TuiCountryIsoCode.XK] = '+383##-###-###',
        _a[i18n.TuiCountryIsoCode.YE] = '+967###-###-###',
        _a[i18n.TuiCountryIsoCode.YT] = '+262 ## ## ## ## ##',
        _a[i18n.TuiCountryIsoCode.ZA] = '+27##-###-####',
        _a[i18n.TuiCountryIsoCode.ZM] = '+260##-###-####',
        _a[i18n.TuiCountryIsoCode.ZW] = '+263#-######',
        _a));

    // TODO: Refactor to use `AbstractTuiValueTransformer` and add ability to provide it for all controls
    /**
     * Control value transformer of TuiDay to custom value format for InputDate* components
     */
    var TUI_DATE_VALUE_TRANSFORMER = new core.InjectionToken('[TUI_DATE_VALUE_TRANSFORMER]');
    /**
     * Control value transformer for InputDateRange component
     */
    var TUI_DATE_RANGE_VALUE_TRANSFORMER = new core.InjectionToken('[TUI_DATE_RANGE_VALUE_TRANSFORMER]');
    /**
     * Control value transformer for InputDateTime component
     */
    var TUI_DATE_TIME_VALUE_TRANSFORMER = new core.InjectionToken('[TUI_DATE_TIME_VALUE_TRANSFORMER]');

    var TUI_PROMPT_WORDS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('prompt'));
    var TUI_CANCEL_WORD = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('cancel'));
    var TUI_DONE_WORD = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('done'));
    var TUI_MORE_WORD = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('more'));
    var TUI_HIDE_TEXT = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('hide'));
    var TUI_SHOW_ALL_TEXT = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('showAll'));
    var TUI_OTHER_DATE_TEXT = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('otherDate'));
    var TUI_CHOOSE_DAY_OR_RANGE_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('mobileCalendarTexts'));
    var TUI_FROM_TO_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('range'));
    var TUI_PLUS_MINUS_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('countTexts'));
    var TUI_TIME_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('time'));
    var TUI_DATE_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('dateTexts'));
    var TUI_DIGITAL_INFORMATION_UNITS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('digitalInformationUnits'));
    var TUI_COPY_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('copyTexts'));
    var TUI_PASSWORD_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('passwordTexts'));
    var TUI_CALENDAR_MONTHS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('shortCalendarMonths'));
    var TUI_FILE_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('fileTexts'));
    var TUI_PAGINATION_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('pagination'));
    var TUI_INPUT_FILE_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('inputFileTexts'));
    var TUI_MULTI_SELECT_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('multiSelectTexts'));
    var TUI_COUNTRIES = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('countries'));

    var TUI_INPUT_DATE_DEFAULT_OPTIONS = {
        icon: function (_a) {
            var $implicit = _a.$implicit;
            return $implicit === 's' ? 'tuiIconCalendar' : 'tuiIconCalendarLarge';
        },
        min: cdk.TUI_FIRST_DAY,
        max: cdk.TUI_LAST_DAY,
        nativePicker: false,
    };
    /**
     * Default parameters for InputDate component
     */
    var TUI_INPUT_DATE_OPTIONS = cdk.tuiCreateToken(TUI_INPUT_DATE_DEFAULT_OPTIONS);
    function tuiInputDateOptionsProvider(options) {
        return cdk.tuiProvideOptions(TUI_INPUT_DATE_OPTIONS, options, TUI_INPUT_DATE_DEFAULT_OPTIONS);
    }

    var TUI_DEFAULT_ITEMS_HANDLERS = {
        stringify: cdk.TUI_DEFAULT_STRINGIFY,
        identityMatcher: cdk.TUI_DEFAULT_IDENTITY_MATCHER,
        disabledItemHandler: cdk.ALWAYS_FALSE_HANDLER,
    };
    /**
     * Default items handlers for components
     */
    var TUI_ITEMS_HANDLERS = cdk.tuiCreateToken(TUI_DEFAULT_ITEMS_HANDLERS);
    function tuiItemsHandlersProvider(options) {
        return cdk.tuiProvideOptions(TUI_ITEMS_HANDLERS, options, TUI_DEFAULT_ITEMS_HANDLERS);
    }

    /**
     * A component for mobile data picker
     */
    var TUI_MOBILE_CALENDAR = new core.InjectionToken('[TUI_MOBILE_CALENDAR]');

    /**
     * A function to get localized formatted month
     */
    var TUI_MONTH_FORMATTER = new core.InjectionToken('[TUI_MONTH_FORMATTER]');

    /**
     * Margin between tabs
     */
    var TUI_TAB_MARGIN = cdk.tuiCreateToken(24);

    /**
     * Validation errors
     */
    var TUI_VALIDATION_ERRORS = cdk.tuiCreateToken({});

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_CALENDAR_DATE_STREAM = TUI_CALENDAR_DATE_STREAM;
    exports.TUI_CALENDAR_MONTHS = TUI_CALENDAR_MONTHS;
    exports.TUI_CANCEL_WORD = TUI_CANCEL_WORD;
    exports.TUI_CHOOSE_DAY_OR_RANGE_TEXTS = TUI_CHOOSE_DAY_OR_RANGE_TEXTS;
    exports.TUI_COPY_TEXTS = TUI_COPY_TEXTS;
    exports.TUI_COUNTRIES = TUI_COUNTRIES;
    exports.TUI_COUNTRIES_MASKS = TUI_COUNTRIES_MASKS;
    exports.TUI_DATE_RANGE_VALUE_TRANSFORMER = TUI_DATE_RANGE_VALUE_TRANSFORMER;
    exports.TUI_DATE_TEXTS = TUI_DATE_TEXTS;
    exports.TUI_DATE_TIME_VALUE_TRANSFORMER = TUI_DATE_TIME_VALUE_TRANSFORMER;
    exports.TUI_DATE_VALUE_TRANSFORMER = TUI_DATE_VALUE_TRANSFORMER;
    exports.TUI_DEFAULT_ITEMS_HANDLERS = TUI_DEFAULT_ITEMS_HANDLERS;
    exports.TUI_DIGITAL_INFORMATION_UNITS = TUI_DIGITAL_INFORMATION_UNITS;
    exports.TUI_DONE_WORD = TUI_DONE_WORD;
    exports.TUI_FILE_TEXTS = TUI_FILE_TEXTS;
    exports.TUI_FROM_TO_TEXTS = TUI_FROM_TO_TEXTS;
    exports.TUI_HIDE_TEXT = TUI_HIDE_TEXT;
    exports.TUI_INPUT_DATE_DEFAULT_OPTIONS = TUI_INPUT_DATE_DEFAULT_OPTIONS;
    exports.TUI_INPUT_DATE_OPTIONS = TUI_INPUT_DATE_OPTIONS;
    exports.TUI_INPUT_FILE_TEXTS = TUI_INPUT_FILE_TEXTS;
    exports.TUI_ITEMS_HANDLERS = TUI_ITEMS_HANDLERS;
    exports.TUI_MOBILE_CALENDAR = TUI_MOBILE_CALENDAR;
    exports.TUI_MONTH_FORMATTER = TUI_MONTH_FORMATTER;
    exports.TUI_MORE_WORD = TUI_MORE_WORD;
    exports.TUI_MULTI_SELECT_TEXTS = TUI_MULTI_SELECT_TEXTS;
    exports.TUI_OTHER_DATE_TEXT = TUI_OTHER_DATE_TEXT;
    exports.TUI_PAGINATION_TEXTS = TUI_PAGINATION_TEXTS;
    exports.TUI_PASSWORD_TEXTS = TUI_PASSWORD_TEXTS;
    exports.TUI_PLUS_MINUS_TEXTS = TUI_PLUS_MINUS_TEXTS;
    exports.TUI_PROMPT_WORDS = TUI_PROMPT_WORDS;
    exports.TUI_SHOW_ALL_TEXT = TUI_SHOW_ALL_TEXT;
    exports.TUI_TAB_MARGIN = TUI_TAB_MARGIN;
    exports.TUI_TIME_TEXTS = TUI_TIME_TEXTS;
    exports.TUI_VALIDATION_ERRORS = TUI_VALIDATION_ERRORS;
    exports.tuiDateStreamWithTransformer = tuiDateStreamWithTransformer;
    exports.tuiInputDateOptionsProvider = tuiInputDateOptionsProvider;
    exports.tuiItemsHandlersProvider = tuiItemsHandlersProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
