// @taiga-ui/kit/bundles/taiga-ui-kit-components-badged-content.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/components/badge'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/badged-content', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/components/badge', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["badged-content"] = {}), global.ng.core, global.cdk, global.i2, global["taiga-ui"].kit.components.badge, global.ng.common, global.i4));
})(this, (function (exports, i0, cdk, i2, i1, i3, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    var BADGE_SIZE = {
        xs: 'xs',
        s: 's',
        m: 's',
        l: 'm',
        xl: 'm',
        xxl: 'l',
    };
    var TuiBadgedContentComponent = /** @class */ (function () {
        function TuiBadgedContentComponent() {
            this.size = 'm';
            this.colorTop = '';
            this.colorBottom = '';
            this.rounded = false;
        }
        Object.defineProperty(TuiBadgedContentComponent.prototype, "topNotification", {
            get: function () {
                return !this.contentTop && this.colorTop ? this.colorTop : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiBadgedContentComponent.prototype, "bottomNotification", {
            get: function () {
                return !this.contentBottom && this.colorBottom ? this.colorBottom : '';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiBadgedContentComponent.prototype, "badgeSize", {
            get: function () {
                return BADGE_SIZE[this.size];
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiBadgedContentComponent.prototype, "badgeHidden", {
            get: function () {
                return this.size === 'xs';
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiBadgedContentComponent.prototype, "sizeBig", {
            get: function () {
                return i2.tuiSizeBigger(this.size);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiBadgedContentComponent.prototype, "boxShadow", {
            get: function () {
                var borderWidth = this.sizeBig ? 3 : 2;
                return "0 0 0 " + cdk.tuiPx(borderWidth);
            },
            enumerable: false,
            configurable: true
        });
        TuiBadgedContentComponent.prototype.contentIsNumber = function (content) {
            return cdk.tuiIsNumber(content === null || content === void 0 ? void 0 : content.valueOf());
        };
        TuiBadgedContentComponent.prototype.contentIsString = function (content) {
            return cdk.tuiIsString(content === null || content === void 0 ? void 0 : content.valueOf());
        };
        TuiBadgedContentComponent.prototype.getStatus = function (color) {
            return color ? 'custom' : 'primary';
        };
        return TuiBadgedContentComponent;
    }());
    TuiBadgedContentComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiBadgedContentComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBadgedContentComponent, selector: "tui-badged-content", inputs: { contentTop: "contentTop", contentBottom: "contentBottom", size: "size", colorTop: "colorTop", colorBottom: "colorBottom", rounded: "rounded" }, host: { properties: { "class._with-top": "this.contentTop", "class._with-bottom": "this.contentBottom", "attr.data-size": "this.size", "class._rounded": "this.rounded" } }, ngImport: i0__namespace, template: "<ng-content></ng-content>\n\n<div\n    *ngIf=\"topNotification\"\n    class=\"t-notification t-notification_top\"\n    [style.background]=\"topNotification\"\n    [style.boxShadow]=\"boxShadow\"\n></div>\n\n<div\n    *ngIf=\"bottomNotification\"\n    class=\"t-notification t-notification_bottom\"\n    [style.background]=\"bottomNotification\"\n    [style.boxShadow]=\"boxShadow\"\n></div>\n\n<div\n    *ngIf=\"contentTop\"\n    class=\"t-content t-content_top\"\n    [style.boxShadow]=\"boxShadow\"\n>\n    <ng-container *polymorpheusOutlet=\"contentTop as content\">\n        <tui-badge\n            *ngIf=\"contentIsNumber(content)\"\n            [size]=\"badgeSize\"\n            [status]=\"getStatus(colorTop)\"\n            [style.backgroundColor]=\"colorTop\"\n            [value]=\"content\"\n        ></tui-badge>\n        <tui-svg\n            *ngIf=\"contentIsString(content)\"\n            class=\"t-icon\"\n            [src]=\"content\"\n            [style.color]=\"colorTop\"\n        ></tui-svg>\n    </ng-container>\n</div>\n\n<div\n    *ngIf=\"contentBottom && sizeBig\"\n    class=\"t-content t-content_bottom\"\n    [style.boxShadow]=\"boxShadow\"\n>\n    <ng-container *polymorpheusOutlet=\"contentBottom as content\">\n        <tui-badge\n            *ngIf=\"contentIsNumber(content)\"\n            [size]=\"badgeSize\"\n            [status]=\"getStatus(colorBottom)\"\n            [style.backgroundColor]=\"colorBottom\"\n            [value]=\"content\"\n        ></tui-badge>\n        <tui-svg\n            *ngIf=\"contentIsString(content)\"\n            class=\"t-icon\"\n            [src]=\"content\"\n            [style.color]=\"colorBottom\"\n        ></tui-svg>\n    </ng-container>\n</div>\n", styles: [":host{position:relative;display:inline-block;color:var(--tui-base-01);height:-webkit-min-content;height:min-content}.t-notification{position:absolute;border-radius:100%;box-shadow:0 0 0 2px currentColor;background-color:currentColor}:host[data-size=xs]:not(._rounded) .t-notification_top,:host[data-size=s]:not(._rounded) .t-notification_top,:host[data-size=m]:not(._rounded) .t-notification_top{top:-.125rem;right:-.125rem}:host[data-size=l]:not(._rounded) .t-notification_top,:host[data-size=xl]:not(._rounded) .t-notification_top{top:-.25rem;right:-.25rem}:host[data-size=xxl]:not(._rounded) .t-notification_top{top:-.5rem;right:-.5rem}:host[data-size=xs]._rounded .t-notification_top{top:0;right:0}:host[data-size=s]._rounded .t-notification_top{top:.0625rem;right:.0625rem}:host[data-size=m]._rounded .t-notification_top{top:.1875rem;right:.1875rem}:host[data-size=l]._rounded .t-notification_top{top:.25rem;right:.25rem}:host[data-size=xl]._rounded .t-notification_top{top:.375rem;right:.375rem}:host[data-size=xxl]._rounded .t-notification_top{top:.5rem;right:.5rem}:host[data-size=m]:not(._rounded) .t-notification_bottom{bottom:-.125rem;right:-.125rem}:host[data-size=l]:not(._rounded) .t-notification_bottom,:host[data-size=xl]:not(._rounded) .t-notification_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl]:not(._rounded) .t-notification_bottom{bottom:-.5rem;right:-.5rem}:host[data-size=m]._rounded .t-notification_bottom{bottom:.1875rem;right:.1875rem}:host[data-size=l]._rounded .t-notification_bottom{bottom:.25rem;right:.25rem}:host[data-size=xl]._rounded .t-notification_bottom{bottom:.375rem;right:.375rem}:host[data-size=xxl]._rounded .t-notification_bottom{bottom:.5rem;right:.5rem}:host[data-size=xs] .t-notification,:host[data-size=s] .t-notification,:host[data-size=m] .t-notification{width:.5rem;height:.5rem}:host[data-size=l] .t-notification{width:.75rem;height:.75rem}:host[data-size=xl] .t-notification{width:1rem;height:1rem}:host[data-size=xxl] .t-notification{width:1.5rem;height:1.5rem}.t-icon{color:var(--tui-success-fill)}:host[data-size=xs] .t-icon{transform:scale(.66666667);margin:-.25rem}:host[data-size=s] .t-icon,:host[data-size=m] .t-icon{width:1.5rem;height:1.5rem;margin:-.125rem}:host[data-size=l] .t-icon,:host[data-size=xl] .t-icon{transform:scale(1.2)}:host[data-size=xxl] .t-icon{transform:scale(1.6)}.t-content{position:absolute;display:flex;align-items:center;justify-content:center;border-radius:6.25rem;overflow:hidden;background-color:currentColor;box-shadow:0 0 0 2px currentColor}:host[data-size=xxl]._with-top:not(._rounded) .t-content_top,:host[data-size=l]._with-top:not(._rounded) .t-content_top{right:-.5rem;top:-.5rem}:host[data-size=xl]._with-top:not(._rounded) .t-content_top{right:-.375rem;top:-.375rem}:host[data-size=m]._with-top:not(._rounded) .t-content_top,:host[data-size=s]._with-top:not(._rounded) .t-content_top,:host[data-size=xs]._with-top:not(._rounded) .t-content_top{right:-.25rem;top:-.25rem}:host[data-size=xxl]._with-top._rounded .t-content_top{right:.25rem;top:.25rem}:host[data-size=xl]._with-top._rounded .t-content_top{right:.125rem;top:.125rem}:host[data-size=l]._with-top._rounded .t-content_top,:host[data-size=m]._with-top._rounded .t-content_top,:host[data-size=s]._with-top._rounded .t-content_top,:host[data-size=xs]._with-top._rounded .t-content_top{right:-.25rem;top:-.25rem}:host[data-size=xxl]._with-bottom:not(._rounded) .t-content_bottom,:host[data-size=l]._with-bottom:not(._rounded) .t-content_bottom{bottom:-.5rem;right:-.5rem}:host[data-size=xl]._with-bottom:not(._rounded) .t-content_bottom{bottom:-.375rem;right:-.375rem}:host[data-size=m]._with-bottom:not(._rounded) .t-content_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl]._with-bottom._rounded .t-content_bottom{bottom:.25rem;right:.25rem}:host[data-size=xl]._with-bottom._rounded .t-content_bottom{bottom:.125rem;right:.125rem}:host[data-size=l]._with-bottom._rounded .t-content_bottom,:host[data-size=m]._with-bottom._rounded .t-content_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl] .t-content{height:2rem;min-width:2rem}:host[data-size=xl] .t-content,:host[data-size=l] .t-content{height:1.5rem;min-width:1.5rem}:host[data-size=m] .t-content,:host[data-size=s] .t-content{height:1.25rem;min-width:1.25rem}:host[data-size=xs] .t-content{height:1rem;min-width:1rem}\n"], components: [{ type: i1__namespace.TuiBadgeComponent, selector: "tui-badge", inputs: ["value", "size", "status", "hoverable"] }, { type: i2__namespace.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-badged-content',
                        templateUrl: './badged-content.template.html',
                        styleUrls: ['./badged-content.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { contentTop: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._with-top']
                }], contentBottom: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._with-bottom']
                }], size: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['attr.data-size']
                }], colorTop: [{
                    type: i0.Input
                }], colorBottom: [{
                    type: i0.Input
                }], rounded: [{
                    type: i0.Input
                }, {
                    type: i0.HostBinding,
                    args: ['class._rounded']
                }] } });

    var TuiBadgedContentModule = /** @class */ (function () {
        function TuiBadgedContentModule() {
        }
        return TuiBadgedContentModule;
    }());
    TuiBadgedContentModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiBadgedContentModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, declarations: [TuiBadgedContentComponent], imports: [i3.CommonModule, i4.PolymorpheusModule, i2.TuiSvgModule, i1.TuiBadgeModule], exports: [TuiBadgedContentComponent] });
    TuiBadgedContentModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, imports: [[i3.CommonModule, i4.PolymorpheusModule, i2.TuiSvgModule, i1.TuiBadgeModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBadgedContentModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i3.CommonModule, i4.PolymorpheusModule, i2.TuiSvgModule, i1.TuiBadgeModule],
                        declarations: [TuiBadgedContentComponent],
                        exports: [TuiBadgedContentComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiBadgedContentComponent = TuiBadgedContentComponent;
    exports.TuiBadgedContentModule = TuiBadgedContentModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
