// @taiga-ui/kit/bundles/taiga-ui-kit-components-radio-group.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/radio-group', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].kit = global["taiga-ui"].kit || {}, global["taiga-ui"].kit.components = global["taiga-ui"].kit.components || {}, global["taiga-ui"].kit.components["radio-group"] = {}), global.ng.core));
})(this, (function (exports, i0) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiRadioGroupComponent = /** @class */ (function () {
        function TuiRadioGroupComponent() {
            this.name = "tui-radio-group-" + TuiRadioGroupComponent.index++;
        }
        return TuiRadioGroupComponent;
    }());
    TuiRadioGroupComponent.index = 0;
    TuiRadioGroupComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiRadioGroupComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRadioGroupComponent, selector: "tui-radio-group", inputs: { name: "name" }, ngImport: i0__namespace, template: "<ng-content></ng-content>\n", changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-radio-group',
                        templateUrl: './radio-group.template.html',
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], propDecorators: { name: [{
                    type: i0.Input
                }] } });

    var TuiRadioGroupModule = /** @class */ (function () {
        function TuiRadioGroupModule() {
        }
        return TuiRadioGroupModule;
    }());
    TuiRadioGroupModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiRadioGroupModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupModule, declarations: [TuiRadioGroupComponent], exports: [TuiRadioGroupComponent] });
    TuiRadioGroupModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiRadioGroupModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiRadioGroupComponent],
                        exports: [TuiRadioGroupComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiRadioGroupComponent = TuiRadioGroupComponent;
    exports.TuiRadioGroupModule = TuiRadioGroupModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
