// @taiga-ui/addon-mobile/esm2015/directives/ripple/ripple.providers.js
import { ElementRef, InjectionToken, Renderer2 } from '@angular/core';
import { TUI_IS_IOS, TUI_TOUCH_SUPPORTED, TuiDestroyService, tuiIsPresent, tuiPx, tuiTypedFromEvent, } from '@taiga-ui/cdk';
import { EMPTY } from 'rxjs';
import { filter, map, takeUntil } from 'rxjs/operators';
export const RIPPLE_ON = 'tuiRippleOn';
export const RIPPLE_OFF = 'tuiRippleOff';
/**
 * Stream of ripple elements to add
 */
export const TUI_RIPPLE_START = new InjectionToken('[TUI_RIPPLE_START]');
/**
 * Stream of ripple elements to remove
 */
export const TUI_RIPPLE_END = new InjectionToken('[TUI_RIPPLE_END]');
export const TUI_RIPPLE_PROVIDERS = [
    TuiDestroyService,
    {
        provide: TUI_RIPPLE_START,
        deps: [TUI_TOUCH_SUPPORTED, TUI_IS_IOS, ElementRef, Renderer2],
        useFactory: (touchSupported, isIOS, { nativeElement }, renderer) => isIOS || !touchSupported
            ? EMPTY
            : tuiTypedFromEvent(nativeElement, 'touchstart').pipe(map(({ touches }) => {
                const { clientX, clientY } = touches[touches.length - 1];
                const { width, height, top, left } = nativeElement.getBoundingClientRect();
                const radius = Math.sqrt(width * width + height * height);
                const dimension = radius * 2;
                const x = clientX - left - radius;
                const y = clientY - top - radius;
                const ripple = renderer.createElement('div');
                renderer.addClass(ripple, 'tui-ripple');
                renderer.setAttribute(ripple, 'style', `width: ${tuiPx(dimension)}; height: ${tuiPx(dimension)}; left: ${tuiPx(x)}; top: ${tuiPx(y)}; animation-name: ${RIPPLE_ON};`);
                return ripple;
            })),
    },
    {
        provide: TUI_RIPPLE_END,
        deps: [TUI_TOUCH_SUPPORTED, TUI_IS_IOS, ElementRef, TuiDestroyService],
        useFactory: (touchSupported, isIOS, { nativeElement }, destroy$) => isIOS || !touchSupported
            ? EMPTY
            : tuiTypedFromEvent(nativeElement, 'animationend').pipe(filter(({ animationName }) => animationName === RIPPLE_OFF), map(({ target }) => target), filter(tuiIsPresent), takeUntil(destroy$)),
    },
];
