// @maskito/angular/esm2015/lib/maskito.cva.js
import { Directive, Input } from '@angular/core';
import { DefaultValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { MASKITO_DEFAULT_OPTIONS, maskitoTransform } from '@maskito/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
export class MaskitoCva {
    constructor(accessor) {
        this.accessor = accessor;
        this.maskito = MASKITO_DEFAULT_OPTIONS;
        const original = accessor.writeValue.bind(accessor);
        accessor.writeValue = (value) => {
            var _a;
            original(maskitoTransform(String(value !== null && value !== void 0 ? value : ''), (_a = this.maskito) !== null && _a !== void 0 ? _a : MASKITO_DEFAULT_OPTIONS));
        };
    }
}
MaskitoCva.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: MaskitoCva, deps: [{ token: i1.DefaultValueAccessor }], target: i0.ɵɵFactoryTarget.Directive });
MaskitoCva.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: MaskitoCva, selector: "input[maskito], textarea[maskito]", inputs: { maskito: "maskito" }, host: { listeners: { "input": "$any(this.accessor)._handleInput($event.target.value)", "blur": "accessor.onTouched()", "compositionstart": "$any(this.accessor)._compositionStart()", "compositionend": "$any(this.accessor)._compositionEnd($event.target.value)" } }, providers: [
        DefaultValueAccessor,
        {
            provide: NG_VALUE_ACCESSOR,
            multi: true,
            useExisting: DefaultValueAccessor,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: MaskitoCva, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[maskito], textarea[maskito]',
                    providers: [
                        DefaultValueAccessor,
                        {
                            provide: NG_VALUE_ACCESSOR,
                            multi: true,
                            useExisting: DefaultValueAccessor,
                        },
                    ],
                    host: {
                        '(input)': '$any(this.accessor)._handleInput($event.target.value)',
                        '(blur)': 'accessor.onTouched()',
                        '(compositionstart)': '$any(this.accessor)._compositionStart()',
                        '(compositionend)': '$any(this.accessor)._compositionEnd($event.target.value)',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.DefaultValueAccessor }]; }, propDecorators: { maskito: [{
                type: Input
            }] } });
