// @taiga-ui/experimental/esm2015/components/avatar-labeled/avatar-labeled.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
// eslint-disable-next-line @taiga-ui/experience/no-deep-imports
import { TuiFadeModule } from '@taiga-ui/experimental/directives/fade';
import { TuiAvatarLabeledComponent } from './avatar-labeled.component';
import * as i0 from "@angular/core";
export class TuiAvatarLabeledModule {
}
TuiAvatarLabeledModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAvatarLabeledModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledModule, declarations: [TuiAvatarLabeledComponent], imports: [CommonModule, TuiFadeModule], exports: [TuiAvatarLabeledComponent] });
TuiAvatarLabeledModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledModule, imports: [[CommonModule, TuiFadeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiFadeModule],
                    declarations: [TuiAvatarLabeledComponent],
                    exports: [TuiAvatarLabeledComponent],
                }]
        }] });
