// @taiga-ui/experimental/esm2015/components/avatar-labeled/avatar-labeled.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation, } from '@angular/core';
import { tuiPure } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@taiga-ui/experimental/directives/fade";
export class TuiAvatarLabeledComponent {
    constructor() {
        this.label = '';
    }
    split(label) {
        return label.split(' ');
    }
}
TuiAvatarLabeledComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAvatarLabeledComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarLabeledComponent, selector: "tui-avatar-labeled", inputs: { label: "label" }, host: { properties: { "attr.title": "this.label" } }, ngImport: i0, template: `
        <ng-content></ng-content>
        <ng-container *ngIf="label.length">
            <span
                *ngFor="let item of split(label)"
                tuiFade
            >
                {{ item }}
            </span>
        </ng-container>
    `, isInline: true, styles: ["tui-avatar-labeled{display:flex;width:3.5rem;box-sizing:content-box;flex-direction:column;text-align:center;align-items:center;line-height:.895rem;font-size:.75rem;padding:0 .5rem;white-space:nowrap}tui-avatar-labeled tui-avatar{margin-bottom:.375rem}tui-avatar-labeled [tuiFade]{width:calc(100% + 1rem)}\n"], directives: [{ type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.TuiFadeDirective, selector: "[tuiFade]", inputs: ["tuiFadeHeight", "tuiFadeSize", "tuiFadeOffset", "tuiFade"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
__decorate([
    tuiPure
], TuiAvatarLabeledComponent.prototype, "split", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarLabeledComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-avatar-labeled',
                    template: `
        <ng-content></ng-content>
        <ng-container *ngIf="label.length">
            <span
                *ngFor="let item of split(label)"
                tuiFade
            >
                {{ item }}
            </span>
        </ng-container>
    `,
                    styleUrls: ['./avatar-labeled.styles.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { label: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.title']
            }], split: [] } });
