// @taiga-ui/core/esm2015/components/scroll-controls/scroll-controls.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLetModule } from '@taiga-ui/cdk';
import { TuiScrollControlsComponent } from './scroll-controls.component';
import { TuiScrollbarDirective } from './scrollbar.directive';
import { TuiScrollbarWrapperDirective } from './scrollbar-wrapper.directive';
import * as i0 from "@angular/core";
/** @deprecated import from `@taiga-ui/cdk` instead */
export class TuiScrollControlsModule {
}
TuiScrollControlsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiScrollControlsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, declarations: [TuiScrollbarDirective,
        TuiScrollbarWrapperDirective,
        TuiScrollControlsComponent], imports: [CommonModule, TuiLetModule], exports: [TuiScrollControlsComponent] });
TuiScrollControlsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, imports: [[CommonModule, TuiLetModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiLetModule],
                    declarations: [
                        TuiScrollbarDirective,
                        TuiScrollbarWrapperDirective,
                        TuiScrollControlsComponent,
                    ],
                    exports: [TuiScrollControlsComponent],
                }]
        }] });
