// @taiga-ui/addon-table/esm2015/components/table/directives/direction-order.directive.js
import { Directive, Inject, Input, Output } from '@angular/core';
import { map } from 'rxjs/operators';
import { TuiTableDirective } from './table.directive';
import * as i0 from "@angular/core";
import * as i1 from "./table.directive";
export class TuiDirectionOrderDirective {
    constructor(table) {
        this.table = table;
        this.directionOrderChange = this.table.directionChange.pipe(map(dir => (dir === 1 ? 'asc' : 'desc')));
    }
    set directionOrder(order) {
        this.table.direction = order === 'asc' ? 1 : -1;
    }
}
TuiDirectionOrderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDirectionOrderDirective, deps: [{ token: TuiTableDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiDirectionOrderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDirectionOrderDirective, selector: "table[tuiTable][tuiDirectionOrder]", inputs: { directionOrder: "directionOrder" }, outputs: { directionOrderChange: "directionOrderChange" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDirectionOrderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'table[tuiTable][tuiDirectionOrder]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiTableDirective, decorators: [{
                    type: Inject,
                    args: [TuiTableDirective]
                }] }]; }, propDecorators: { directionOrder: [{
                type: Input
            }], directionOrderChange: [{
                type: Output
            }] } });
