// @taiga-ui/addon-table/esm2015/components/table/directives/head.directive.js
import { Directive, Inject, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiHeadDirective {
    constructor(template) {
        this.template = template;
        this.tuiHead = '';
    }
}
TuiHeadDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHeadDirective, deps: [{ token: TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiHeadDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHeadDirective, selector: "[tuiHead]", inputs: { tuiHead: "tuiHead" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHeadDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHead]',
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }]; }, propDecorators: { tuiHead: [{
                type: Input
            }] } });
