// @taiga-ui/addon-table/esm2015/components/table/directives/sortable.directive.js
import { Directive, forwardRef, Inject } from '@angular/core';
import { TuiThComponent } from '../th/th.component';
import { TuiSortByDirective } from './sort-by.directive';
import { TuiTableDirective } from './table.directive';
import * as i0 from "@angular/core";
import * as i1 from "./sort-by.directive";
import * as i2 from "./table.directive";
import * as i3 from "../th/th.component";
export class TuiSortableDirective {
    constructor(sortBy, table, th) {
        this.sortBy = sortBy;
        this.table = table;
        this.th = th;
        this.sorter = () => 0;
    }
    get key() {
        return this.th.key;
    }
    ngOnInit() {
        this.sorter = this.match ? this.table.sorter : this.sorter;
        this.th.sorter = this.sorter;
    }
    check() {
        if (this.match && this.table.sorter !== this.sorter) {
            this.table.updateSorter(this.sorter);
        }
    }
    get match() {
        return this.sortBy.tuiSortBy === this.key;
    }
}
TuiSortableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortableDirective, deps: [{ token: forwardRef(() => TuiSortByDirective) }, { token: TuiTableDirective }, { token: TuiThComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiSortableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSortableDirective, selector: "th[tuiTh][tuiSortable]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'th[tuiTh][tuiSortable]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiSortByDirective, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => TuiSortByDirective)]
                }] }, { type: i2.TuiTableDirective, decorators: [{
                    type: Inject,
                    args: [TuiTableDirective]
                }] }, { type: i3.TuiThComponent, decorators: [{
                    type: Inject,
                    args: [TuiThComponent]
                }] }]; } });
