// @taiga-ui/addon-table/esm2015/components/table/directives/sort-by.directive.js
import { ContentChildren, Directive, Inject, Input, Output, } from '@angular/core';
import { EMPTY_QUERY } from '@taiga-ui/cdk';
import { delay, filter, map } from 'rxjs/operators';
import { TuiSortableDirective } from './sortable.directive';
import { TuiTableDirective } from './table.directive';
import * as i0 from "@angular/core";
import * as i1 from "./table.directive";
export class TuiSortByDirective {
    constructor(table) {
        this.table = table;
        this.sortables = EMPTY_QUERY;
        this.tuiSortByChange = this.table.sorterChange.pipe(
        // delay is for getting actual ContentChildren (sortables) https://github.com/angular/angular/issues/38976
        delay(0), filter(() => !!this.sortables.length), map(sorter => this.getKey(sorter)));
        this.tuiSortBy = null;
    }
    set sortBy(sortBy) {
        this.tuiSortBy = sortBy;
        this.checkSortables();
    }
    checkSortables() {
        this.sortables.forEach(s => s.check());
    }
    getKey(sorter) {
        var _a;
        return ((_a = this.sortables.find(s => s.sorter === sorter)) === null || _a === void 0 ? void 0 : _a.key) || null;
    }
}
TuiSortByDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortByDirective, deps: [{ token: TuiTableDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiSortByDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSortByDirective, selector: "table[tuiTable][tuiSortBy]", inputs: { sortBy: ["tuiSortBy", "sortBy"] }, outputs: { tuiSortByChange: "tuiSortByChange" }, queries: [{ propertyName: "sortables", predicate: TuiSortableDirective, descendants: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSortByDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'table[tuiTable][tuiSortBy]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiTableDirective, decorators: [{
                    type: Inject,
                    args: [TuiTableDirective]
                }] }]; }, propDecorators: { sortables: [{
                type: ContentChildren,
                args: [TuiSortableDirective, { descendants: true }]
            }], sortBy: [{
                type: Input,
                args: ['tuiSortBy']
            }], tuiSortByChange: [{
                type: Output
            }] } });
