// @taiga-ui/addon-tablebars/esm2015/components/table-bars-host/table-bars-host.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule, TuiModeModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTableBarsHostComponent } from './table-bars-host.component';
import * as i0 from "@angular/core";
export class TuiTableBarsHostModule {
}
TuiTableBarsHostModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTableBarsHostModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostModule, declarations: [TuiTableBarsHostComponent], imports: [CommonModule, PolymorpheusModule, TuiModeModule, TuiButtonModule], exports: [TuiTableBarsHostComponent] });
TuiTableBarsHostModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostModule, imports: [[CommonModule, PolymorpheusModule, TuiModeModule, TuiButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiModeModule, TuiButtonModule],
                    declarations: [TuiTableBarsHostComponent],
                    exports: [TuiTableBarsHostComponent],
                }]
        }] });
