// @taiga-ui/addon-tablebars/esm2015/components/table-bars-host/table-bars-host.component.js
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk';
import { TUI_ANIMATION_OPTIONS, TUI_CLOSE_WORD, TUI_COMMON_ICONS, TUI_MEDIA, tuiIsMobile, tuiSlideInTop, } from '@taiga-ui/core';
import { TuiTableBarsService } from '../../services/table-bars.service';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
import * as i2 from "@angular/common";
import * as i3 from "@tinkoff/ng-polymorpheus";
import * as i4 from "../../services/table-bars.service";
import * as i5 from "rxjs";
// TODO: Accessibility
export class TuiTableBarsHostComponent {
    constructor(service, closeWord$, icons, animation, media, win) {
        this.service = service;
        this.closeWord$ = closeWord$;
        this.icons = icons;
        this.animation = animation;
        this.media = media;
        this.win = win;
    }
    get isMobile() {
        return tuiIsMobile(this.win, this.media);
    }
    getMode(mode) {
        return mode === 'onLight' ? 'onDark' : null;
    }
    onCloseClick(itemToRemove) {
        itemToRemove.close();
    }
    getItemContext(item) {
        return {
            $implicit: () => item.close(),
        };
    }
}
TuiTableBarsHostComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostComponent, deps: [{ token: TuiTableBarsService }, { token: TUI_CLOSE_WORD }, { token: TUI_COMMON_ICONS }, { token: TUI_ANIMATION_OPTIONS }, { token: TUI_MEDIA }, { token: WINDOW }], target: i0.ɵɵFactoryTarget.Component });
TuiTableBarsHostComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTableBarsHostComponent, selector: "tui-table-bars-host", ngImport: i0, template: "<div\n    *ngIf=\"service.bar$ | async as item\"\n    class=\"t-wrapper\"\n    @tuiParentAnimation\n>\n    <div\n        automation-id=\"tui-table-bar__bar\"\n        class=\"t-bar\"\n        [@tuiSlideInTop]=\"animation\"\n        [class.t-bar_light]=\"item.mode === 'onDark'\"\n    >\n        <div\n            class=\"t-container\"\n            [class.t-container_adaptive]=\"item.adaptive\"\n        >\n            <div class=\"t-inner\">\n                <div class=\"t-content\">\n                    <!-- TODO: Polymorpheus fix type -->\n                    <ng-container *polymorpheusOutlet=\"$any(item.content) as text; context: getItemContext(item)\">\n                        {{ text }}\n                    </ng-container>\n                </div>\n\n                <button\n                    *ngIf=\"item.hasCloseButton\"\n                    appearance=\"icon\"\n                    automation-id=\"tui-table-bar__close-button\"\n                    size=\"xs\"\n                    tuiIconButton\n                    type=\"button\"\n                    class=\"t-close-button\"\n                    [icon]=\"icons.close\"\n                    [title]=\"closeWord$ | async\"\n                    [tuiMode]=\"getMode(item.mode)\"\n                    (click)=\"onCloseClick(item)\"\n                ></button>\n            </div>\n        </div>\n    </div>\n</div>\n", styles: [":host{position:fixed;bottom:0;left:0;display:flex;align-items:flex-end;width:100%;height:0}.t-wrapper{display:flex;width:100%}.t-bar{box-shadow:var(--tui-shadow);flex:0 0 100%;background:var(--tui-base-07);color:var(--tui-base-01)}.t-bar_light{background:var(--tui-base-01);color:var(--tui-text-01)}.t-container{position:relative;margin:0 auto}@media screen and (min-width: 80em){.t-container.t-container_adaptive{width:69rem}}@media screen and (max-width: 79.9625em){.t-container.t-container_adaptive{width:51.5rem}}@media screen and (max-width: 47.9625em){.t-container.t-container_adaptive{width:100%;padding:0 1rem;box-sizing:border-box}}@media screen and (min-width: 80em){.t-container{width:69rem}}@media screen and (max-width: 79.9625em){.t-container{width:51.5rem;padding:0 3rem}}.t-inner{display:flex;align-items:center;height:4.5rem;width:100%;overflow:hidden}:host-context(tui-root._mobile) .t-inner{height:5.75rem}.t-close-button{margin-left:1.5rem}.t-content{width:100%;flex:1 1 0}\n"], components: [{ type: i1.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i1.TuiModeDirective, selector: "[tuiMode]", inputs: ["tuiMode"] }], pipes: { "async": i2.AsyncPipe }, animations: [tuiSlideInTop, TUI_PARENT_ANIMATION], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsHostComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-table-bars-host',
                    templateUrl: './table-bars-host.template.html',
                    styleUrls: ['./table-bars-host.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiSlideInTop, TUI_PARENT_ANIMATION],
                }]
        }], ctorParameters: function () { return [{ type: i4.TuiTableBarsService, decorators: [{
                    type: Inject,
                    args: [TuiTableBarsService]
                }] }, { type: i5.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_CLOSE_WORD]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_COMMON_ICONS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATION_OPTIONS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_MEDIA]
                }] }, { type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }]; } });
