// @taiga-ui/addon-tablebars/esm2015/services/table-bars.service.js
import { Injectable } from '@angular/core';
import { tuiAssert, tuiIsObserved } from '@taiga-ui/cdk';
import { Observable, ReplaySubject } from 'rxjs';
import { TuiTableBar } from '../classes/table-bar';
import * as i0 from "@angular/core";
const NO_HOST = 'Table bars are disabled, enable support by adding TuiTableBarsHostModule to your main app module ' +
    'and tui-table-bars-host component to app template';
export class TuiTableBarsService {
    constructor() {
        this.bar$ = new ReplaySubject(1);
    }
    open(content, options) {
        return new Observable(observer => {
            ngDevMode && tuiAssert.assert(tuiIsObserved(this.bar$), NO_HOST);
            const tableBar = new TuiTableBar(observer, content, options);
            this.bar$.next(tableBar);
            return () => {
                this.bar$.next(null);
            };
        });
    }
}
TuiTableBarsService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
TuiTableBarsService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableBarsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
