// @taiga-ui/addon-doc/esm2015/components/navigation/navigation.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TuiScrollIntoViewLinkModule } from '@taiga-ui/addon-doc/directives';
import { TuiActiveZoneModule, TuiAutoFocusModule, TuiLetModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiDataListModule, TuiDropdownModule, TuiExpandModule, TuiHostedDropdownModule, TuiLinkModule, TuiModeModule, TuiScrollbarModule, TuiSvgModule, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiAccordionModule, TuiInputModule } from '@taiga-ui/kit';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDocNavigationComponent } from './navigation.component';
import * as i0 from "@angular/core";
export class TuiDocNavigationModule {
}
TuiDocNavigationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocNavigationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocNavigationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocNavigationModule, declarations: [TuiDocNavigationComponent], imports: [CommonModule,
        ReactiveFormsModule,
        RouterModule,
        PolymorpheusModule,
        TuiScrollIntoViewLinkModule,
        TuiButtonModule,
        TuiInputModule,
        TuiTextfieldControllerModule,
        TuiActiveZoneModule,
        TuiLetModule,
        TuiModeModule,
        TuiLinkModule,
        TuiExpandModule,
        TuiHostedDropdownModule,
        TuiDropdownModule,
        TuiAccordionModule,
        TuiScrollbarModule,
        TuiSvgModule,
        TuiDataListModule,
        TuiAutoFocusModule], exports: [TuiDocNavigationComponent] });
TuiDocNavigationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocNavigationModule, imports: [[
            CommonModule,
            ReactiveFormsModule,
            RouterModule,
            PolymorpheusModule,
            TuiScrollIntoViewLinkModule,
            TuiButtonModule,
            TuiInputModule,
            TuiTextfieldControllerModule,
            TuiActiveZoneModule,
            TuiLetModule,
            TuiModeModule,
            TuiLinkModule,
            TuiExpandModule,
            TuiHostedDropdownModule,
            TuiDropdownModule,
            TuiAccordionModule,
            TuiScrollbarModule,
            TuiSvgModule,
            TuiDataListModule,
            TuiAutoFocusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocNavigationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        RouterModule,
                        PolymorpheusModule,
                        TuiScrollIntoViewLinkModule,
                        TuiButtonModule,
                        TuiInputModule,
                        TuiTextfieldControllerModule,
                        TuiActiveZoneModule,
                        TuiLetModule,
                        TuiModeModule,
                        TuiLinkModule,
                        TuiExpandModule,
                        TuiHostedDropdownModule,
                        TuiDropdownModule,
                        TuiAccordionModule,
                        TuiScrollbarModule,
                        TuiSvgModule,
                        TuiDataListModule,
                        TuiAutoFocusModule,
                    ],
                    declarations: [TuiDocNavigationComponent],
                    exports: [TuiDocNavigationComponent],
                }]
        }] });
