// @taiga-ui/cdk/esm2015/directives/autofilled/autofilled-style.component.js
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiAutofilledStyleComponent {
}
TuiAutofilledStyleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutofilledStyleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAutofilledStyleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAutofilledStyleComponent, selector: "ng-component", host: { classAttribute: "tui-autofilled-styles" }, ngImport: i0, template: '', isInline: true, styles: [".tui-autofill input,input.tui-autofill{transition:box-shadow .01s}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAutofilledStyleComponent, decorators: [{
            type: Component,
            args: [{
                    host: {
                        class: 'tui-autofilled-styles',
                    },
                    template: '',
                    styleUrls: ['./autofilled.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
