// @taiga-ui/addon-table/esm2015/directives/table-filters/generic-filter.directive.js
import { Directive, Input } from '@angular/core';
import { ALWAYS_TRUE_HANDLER } from '@taiga-ui/cdk';
import { AbstractTuiTableFilter } from './abstract-table-filter';
import * as i0 from "@angular/core";
export class TuiGenericFilterDirective extends AbstractTuiTableFilter {
    constructor() {
        super(...arguments);
        this.filter = ALWAYS_TRUE_HANDLER;
    }
}
TuiGenericFilterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGenericFilterDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiGenericFilterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiGenericFilterDirective, selector: "[tuiGenericFilter]", inputs: { filter: ["tuiGenericFilter", "filter"] }, providers: [
        {
            provide: AbstractTuiTableFilter,
            useExisting: TuiGenericFilterDirective,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiGenericFilterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiGenericFilter]',
                    providers: [
                        {
                            provide: AbstractTuiTableFilter,
                            useExisting: TuiGenericFilterDirective,
                        },
                    ],
                }]
        }], propDecorators: { filter: [{
                type: Input,
                args: ['tuiGenericFilter']
            }] } });
