// @taiga-ui/addon-table/esm2015/directives/table-filters/table-filters.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TuiTableFiltersDirective } from './table-filters.directive';
import * as i0 from "@angular/core";
import * as i1 from "./table-filters.directive";
export class TuiTableFiltersPipe {
    constructor(filters) {
        this.filters = filters;
    }
    transform(items) {
        return this.filters.filter(items);
    }
}
TuiTableFiltersPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFiltersPipe, deps: [{ token: TuiTableFiltersDirective }], target: i0.ɵɵFactoryTarget.Pipe });
TuiTableFiltersPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFiltersPipe, name: "tuiTableFilters" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFiltersPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiTableFilters',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiTableFiltersDirective, decorators: [{
                    type: Inject,
                    args: [TuiTableFiltersDirective]
                }] }]; } });
