// @taiga-ui/addon-table/esm2015/directives/table-filters/table-filter.directive.js
import { Directive, Inject, Input, Optional } from '@angular/core';
import { NgControl } from '@angular/forms';
import { TuiHeadDirective } from '@taiga-ui/addon-table/components';
import { defer, EMPTY, merge } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
import { AbstractTuiTableFilter } from './abstract-table-filter';
import { TuiTableFiltersDirective } from './table-filters.directive';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/addon-table/components";
import * as i2 from "./abstract-table-filter";
import * as i3 from "@angular/forms";
import * as i4 from "./table-filters.directive";
export class TuiTableFilterDirective {
    constructor(head, delegate, control, filters) {
        this.head = head;
        this.delegate = delegate;
        this.control = control;
        this.filters = filters;
        this.refresh$ = defer(() => {
            var _a;
            return merge(this.control.valueChanges || EMPTY, ((_a = this.control.statusChanges) === null || _a === void 0 ? void 0 : _a.pipe(distinctUntilChanged())) || EMPTY);
        });
    }
    ngOnInit() {
        this.filters.register(this);
    }
    ngOnDestroy() {
        this.filters.unregister(this);
    }
    filter(item) {
        const { disabled, value } = this.control;
        return (!!disabled ||
            !this.key ||
            this.delegate.filter(item[this.key], value));
    }
    get key() {
        var _a;
        return this.tuiTableFilter || ((_a = this.head) === null || _a === void 0 ? void 0 : _a.tuiHead);
    }
}
TuiTableFilterDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFilterDirective, deps: [{ token: TuiHeadDirective, optional: true }, { token: AbstractTuiTableFilter }, { token: NgControl }, { token: TuiTableFiltersDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiTableFilterDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTableFilterDirective, selector: "[tuiTableFilter]", inputs: { tuiTableFilter: "tuiTableFilter" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFilterDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTableFilter]',
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiHeadDirective, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TuiHeadDirective]
                }] }, { type: i2.AbstractTuiTableFilter, decorators: [{
                    type: Inject,
                    args: [AbstractTuiTableFilter]
                }] }, { type: i3.NgControl, decorators: [{
                    type: Inject,
                    args: [NgControl]
                }] }, { type: i4.TuiTableFiltersDirective, decorators: [{
                    type: Inject,
                    args: [TuiTableFiltersDirective]
                }] }]; }, propDecorators: { tuiTableFilter: [{
                type: Input
            }] } });
