// @taiga-ui/addon-table/esm2015/directives/table-filters/table-filters.directive.js
import { Directive } from '@angular/core';
import { identity, merge, ReplaySubject } from 'rxjs';
import { map, startWith, switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiTableFiltersDirective {
    constructor() {
        this.refresh$ = new ReplaySubject(1);
        this.filters = [];
    }
    register(filter) {
        this.filters = this.filters.concat(filter);
        this.update();
    }
    unregister(filter) {
        this.filters = this.filters.filter(item => item !== filter);
        this.update();
    }
    filter(items) {
        return this.refresh$.pipe(switchMap(identity), startWith(null), map(() => items.filter(item => this.check(item))));
    }
    check(item) {
        return this.filters.every(filter => filter.filter(item));
    }
    update() {
        this.refresh$.next(merge(...this.filters.map(({ refresh$ }) => refresh$)));
    }
}
TuiTableFiltersDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFiltersDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiTableFiltersDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTableFiltersDirective, selector: "[tuiTableFilters]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableFiltersDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTableFilters]',
                }]
        }] });
