// @taiga-ui/kit/esm2015/components/input-date-range/input-date-range.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiActiveZoneModule, TuiLetModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiCalendarRangeModule } from '@taiga-ui/kit/components/calendar-range';
import { TuiValueAccessorModule } from '@taiga-ui/kit/directives';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputDateRangeComponent } from './input-date-range.component';
import { TuiInputDateRangeDirective } from './input-date-range.directive';
import * as i0 from "@angular/core";
export class TuiInputDateRangeModule {
}
TuiInputDateRangeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateRangeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputDateRangeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateRangeModule, declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective], imports: [CommonModule,
        MaskitoModule,
        TuiActiveZoneModule,
        TuiLetModule,
        PolymorpheusModule,
        TuiWrapperModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiHostedDropdownModule,
        TuiSvgModule,
        TuiCalendarRangeModule,
        TuiValueAccessorModule], exports: [TuiInputDateRangeComponent,
        TuiInputDateRangeDirective,
        TuiTextfieldComponent] });
TuiInputDateRangeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateRangeModule, imports: [[
            CommonModule,
            MaskitoModule,
            TuiActiveZoneModule,
            TuiLetModule,
            PolymorpheusModule,
            TuiWrapperModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiHostedDropdownModule,
            TuiSvgModule,
            TuiCalendarRangeModule,
            TuiValueAccessorModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateRangeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoModule,
                        TuiActiveZoneModule,
                        TuiLetModule,
                        PolymorpheusModule,
                        TuiWrapperModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiHostedDropdownModule,
                        TuiSvgModule,
                        TuiCalendarRangeModule,
                        TuiValueAccessorModule,
                    ],
                    declarations: [TuiInputDateRangeComponent, TuiInputDateRangeDirective],
                    exports: [
                        TuiInputDateRangeComponent,
                        TuiInputDateRangeDirective,
                        TuiTextfieldComponent,
                    ],
                }]
        }] });
