// @taiga-ui/addon-mobile/esm2015/components/sheet/sheet.directive.js
import { ChangeDetectorRef, Directive, Inject, Input, Output, TemplateRef, } from '@angular/core';
import { tuiIfMap } from '@taiga-ui/cdk';
import { PolymorpheusTemplate } from '@tinkoff/ng-polymorpheus';
import { Subject } from 'rxjs';
import { endWith, ignoreElements, share } from 'rxjs/operators';
import { TuiSheetService } from './sheet.service';
import * as i0 from "@angular/core";
import * as i1 from "./sheet.service";
/**
 * @deprecated: use {@link TuiSheetDialogDirective}
 * drop in v4.0
 */
export class TuiSheetDirective extends PolymorpheusTemplate {
    constructor(cdr, templateRef, service) {
        super(templateRef, cdr);
        this.service = service;
        this.open$ = new Subject();
        this.options = {};
        this.tuiSheetChange = this.open$.pipe(tuiIfMap(() => this.service.open(this, this.options).pipe(ignoreElements(), endWith(false))), share());
    }
    set tuiSheet(open) {
        this.open$.next(open);
    }
}
TuiSheetDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDirective, deps: [{ token: ChangeDetectorRef }, { token: TemplateRef }, { token: TuiSheetService }], target: i0.ɵɵFactoryTarget.Directive });
TuiSheetDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDirective, selector: "ng-template[tuiSheet]", inputs: { options: ["tuiSheetOptions", "options"], tuiSheet: "tuiSheet" }, outputs: { tuiSheetChange: "tuiSheetChange" }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiSheet]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef, decorators: [{
                    type: Inject,
                    args: [ChangeDetectorRef]
                }] }, { type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }, { type: i1.TuiSheetService, decorators: [{
                    type: Inject,
                    args: [TuiSheetService]
                }] }]; }, propDecorators: { options: [{
                type: Input,
                args: ['tuiSheetOptions']
            }], tuiSheet: [{
                type: Input
            }], tuiSheetChange: [{
                type: Output
            }] } });
