// @taiga-ui/addon-mobile/esm2015/components/sheet/sheet.module.js
import { CommonModule } from '@angular/common';
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { TuiLetModule, TuiOverscrollModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiScrollbarModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiSheetComponent } from './components/sheet/sheet.component';
import { TuiSheetBarComponent } from './components/sheet-bar/sheet-bar.component';
import { TuiSheetHeadingComponent } from './components/sheet-heading/sheet-heading.component';
import { TuiSheetsHostComponent } from './components/sheets-host/sheets-host.component';
import { TuiSheetCloseDirective } from './directives/sheet-close/sheet-close.directive';
import { TuiSheetStopDirective } from './directives/sheet-stop/sheet-stop.directive';
import { TuiSheetTopDirective } from './directives/sheet-top/sheet-top.directive';
import { TuiSheetWrapperDirective } from './directives/sheet-wrapper/sheet-wrapper.directive';
import { TuiSheetDirective } from './sheet.directive';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiSheetDialogModule}
 * drop in v4.0
 */
export class TuiSheetModule {
}
TuiSheetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSheetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetModule, declarations: [TuiSheetsHostComponent,
        TuiSheetComponent,
        TuiSheetBarComponent,
        TuiSheetHeadingComponent,
        TuiSheetTopDirective,
        TuiSheetDirective,
        TuiSheetWrapperDirective,
        TuiSheetCloseDirective,
        TuiSheetStopDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiLetModule,
        TuiButtonModule,
        TuiOverscrollModule,
        TuiScrollbarModule], exports: [TuiSheetsHostComponent,
        TuiSheetComponent,
        TuiSheetHeadingComponent,
        TuiSheetDirective] });
TuiSheetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiLetModule,
            TuiButtonModule,
            TuiOverscrollModule,
            TuiScrollbarModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiLetModule,
                        TuiButtonModule,
                        TuiOverscrollModule,
                        TuiScrollbarModule,
                    ],
                    declarations: [
                        TuiSheetsHostComponent,
                        TuiSheetComponent,
                        TuiSheetBarComponent,
                        TuiSheetHeadingComponent,
                        TuiSheetTopDirective,
                        TuiSheetDirective,
                        TuiSheetWrapperDirective,
                        TuiSheetCloseDirective,
                        TuiSheetStopDirective,
                    ],
                    exports: [
                        TuiSheetsHostComponent,
                        TuiSheetComponent,
                        TuiSheetHeadingComponent,
                        TuiSheetDirective,
                    ],
                    schemas: [NO_ERRORS_SCHEMA],
                }]
        }] });
