// @taiga-ui/addon-mobile/esm2015/components/sheet/sheet.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { tuiGetNativeFocused, tuiIsHTMLElement } from '@taiga-ui/cdk';
import { BehaviorSubject, EMPTY, Observable } from 'rxjs';
import { TUI_SHEET_OPTIONS } from './sheet-options';
import * as i0 from "@angular/core";
/**
 * @deprecated: use {@link TuiSheetDialogService}
 * drop in v4.0
 */
export class TuiSheetService {
    constructor(doc, options) {
        this.doc = doc;
        this.options = options;
        this.sheets$ = new BehaviorSubject([]);
    }
    open(content, options = {}) {
        return new Observable($implicit => {
            const completeWith = (result) => {
                $implicit.next(result);
                $implicit.complete();
            };
            const sheet = Object.assign(Object.assign(Object.assign({}, this.options), options), { content,
                completeWith,
                $implicit, scroll$: EMPTY });
            const focused = tuiGetNativeFocused(this.doc);
            if (tuiIsHTMLElement(focused) && sheet.overlay) {
                focused.blur();
            }
            this.sheets$.next([...this.sheets$.value, sheet]);
            return () => {
                this.sheets$.next(this.sheets$.value.filter(item => item !== sheet));
            };
        });
    }
}
TuiSheetService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetService, deps: [{ token: DOCUMENT }, { token: TUI_SHEET_OPTIONS }], target: i0.ɵɵFactoryTarget.Injectable });
TuiSheetService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_SHEET_OPTIONS]
                }] }]; } });
