// @taiga-ui/addon-doc/bundles/taiga-ui-addon-doc-utils.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('markdown-it')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-doc/utils', ['exports', 'markdown-it'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-doc"] = global["taiga-ui"]["addon-doc"] || {}, global["taiga-ui"]["addon-doc"].utils = {}), global.MarkdownIt));
})(this, (function (exports, MarkdownIt) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var MarkdownIt__default = /*#__PURE__*/_interopDefaultLegacy(MarkdownIt);

    function tuiCoerceValueIsTrue(value) {
        return (value === null || value === void 0 ? void 0 : value.toString()) === 'true';
    }

    function isEmptyParamValue(value) {
        return ['undefined', 'null', 'NaN'].includes(value);
    }
    function isBooleanParamValue(value) {
        return value === 'true' || value === 'false';
    }
    function isNumberParamValue(value) {
        return !!value.trim() && !Number.isNaN(Number(value)) && !value.startsWith('+');
    }
    function isPossibleArray(value) {
        return value.startsWith('[') && value.endsWith(']');
    }
    function isPossibleObject(value) {
        return value.startsWith('{') && value.endsWith('}');
    }
    function tuiCoerceValue(value) {
        var prepared = String(value).trim();
        if (isEmptyParamValue(prepared)) {
            return null;
        }
        if (isBooleanParamValue(prepared)) {
            return String(prepared) === 'true';
        }
        if (isNumberParamValue(prepared)) {
            return Number(prepared);
        }
        var decodedValue = decodeURIComponent(prepared);
        try {
            return isPossibleArray(decodedValue) || isPossibleObject(decodedValue)
                ? JSON.parse(decodedValue)
                : decodedValue;
        }
        catch (_a) {
            return decodedValue;
        }
    }

    function tuiGenerateRoutes(type) {
        return [
            {
                path: '',
                component: type,
                children: [
                    {
                        path: ':tab',
                        component: type,
                    },
                ],
            },
        ];
    }

    function inspectArray(array, depth) {
        if (depth === 0) {
            return '[…]';
        }
        var result = '';
        var first = true;
        for (var index = 0; index < array.length; index++) {
            if (first) {
                first = false;
            }
            else {
                result += ', ';
            }
            result += index in array ? tuiInspectAny(array[index], depth - 1) : 'empty';
        }
        return "[" + result + "]";
    }
    function inspectObject(object, depth) {
        if (depth === 0) {
            return '{…}';
        }
        var result = '';
        var first = true;
        for (var key in object) {
            if (!object.hasOwnProperty(key)) {
                continue;
            }
            if (first) {
                first = false;
            }
            else {
                result += ', ';
            }
            result += key + ": " + tuiInspectAny(object[key], depth - 1);
        }
        return "{" + result + "}";
    }
    /**
     * Returns readable JS entity
     * @param data
     * @param depth
     * @return readable JS entity
     */
    function tuiInspectAny(data, depth) {
        if (data === null) {
            return 'null';
        }
        switch (typeof data) {
            case 'string':
                return "'" + data + "'";
            case 'undefined':
            case 'number':
            case 'boolean':
            case 'function':
                return String(data);
            default:
                break;
        }
        if (data instanceof RegExp) {
            return String(data);
        }
        if (Array.isArray(data)) {
            return inspectArray(data, depth);
        }
        return inspectObject(data, depth);
    }

    function tuiIsPageGroup(page) {
        return 'subPages' in page;
    }

    function tuiTryParseMarkdownCodeBlock(text) {
        if (text === void 0) { text = ''; }
        var tokens = new MarkdownIt__default["default"]().parse(text, {});
        var result = tokens
            .filter(function (_a) {
            var tag = _a.tag, type = _a.type;
            return tag === 'code' && type === 'fence';
        })
            .map(function (_a) {
            var content = _a.content;
            return content.trim();
        });
        return result.length ? result : [text];
    }

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function tuiRawLoad(content) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(content instanceof Promise)) return [3 /*break*/, 2];
                        return [4 /*yield*/, content];
                    case 1:
                        _a = (_b.sent()).default;
                        return [3 /*break*/, 3];
                    case 2:
                        _a = content;
                        _b.label = 3;
                    case 3: return [2 /*return*/, _a];
                }
            });
        });
    }

    function tuiRawLoadRecord(example) {
        return __awaiter(this, void 0, void 0, function () {
            var processedContent, _a, _b, _c, key, content, _d, _e, e_1_1;
            var e_1, _f;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        processedContent = {};
                        _g.label = 1;
                    case 1:
                        _g.trys.push([1, 6, 7, 8]);
                        _a = __values(Object.entries(example)), _b = _a.next();
                        _g.label = 2;
                    case 2:
                        if (!!_b.done) return [3 /*break*/, 5];
                        _c = __read(_b.value, 2), key = _c[0], content = _c[1];
                        if (!content) return [3 /*break*/, 4];
                        _d = processedContent;
                        _e = key;
                        return [4 /*yield*/, tuiRawLoad(content)];
                    case 3:
                        _d[_e] = _g.sent();
                        _g.label = 4;
                    case 4:
                        _b = _a.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_1_1 = _g.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (_b && !_b.done && (_f = _a.return)) _f.call(_a);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/, processedContent];
                }
            });
        });
    }

    function tuiSortPages(pages, excludeSections) {
        if (excludeSections === void 0) { excludeSections = new Set(); }
        var sections = Array.from(new Set(pages.map(function (page) { return page.section; })));
        var sortedPages = pages.slice().sort(function (a, b) {
            var _a, _b;
            if (excludeSections.has((_a = a.section) !== null && _a !== void 0 ? _a : '') ||
                excludeSections.has((_b = b.section) !== null && _b !== void 0 ? _b : '')) {
                return 0;
            }
            var aSectionIndex = sections.indexOf(a.section);
            var bSectionIndex = sections.indexOf(b.section);
            if (aSectionIndex !== bSectionIndex) {
                return aSectionIndex - bSectionIndex;
            }
            if (a.title > b.title) {
                return 1;
            }
            return a.title.localeCompare(b.title);
        });
        return sortedPages.map(function (page) { return tuiIsPageGroup(page)
            ? Object.assign(Object.assign({}, page), { subPages: tuiSortPages(page.subPages, excludeSections) }) : page; });
    }

    function assertTitle(page, map) {
        var _a;
        if (map.has(page.title) && ((_a = map.get(page.title)) === null || _a === void 0 ? void 0 : _a.route) !== page.route) {
            console.error('Title for page should be unique for prevent inconsistent page names', page, '<== Collisions between ==>', map.get(page.title));
        }
    }
    function tuiToFlatMapPages(pages) {
        var map = new Map();
        pages.forEach(function (page) {
            if ('subPages' in page) {
                page.subPages.forEach(function (subPage) {
                    ngDevMode && assertTitle(subPage, map);
                    map.set(subPage.title, subPage);
                });
            }
            else {
                ngDevMode && assertTitle(page, map);
                map.set(page.title, page);
            }
        });
        return map;
    }

    var MAP = {
        а: 'f',
        б: ',',
        в: 'd',
        г: 'u',
        д: 'l',
        е: 't',
        ё: '`',
        ж: ';',
        з: 'p',
        и: 'b',
        й: 'q',
        к: 'r',
        л: 'k',
        м: 'v',
        н: 'y',
        о: 'j',
        п: 'g',
        р: 'h',
        с: 'c',
        т: 'n',
        у: 'e',
        ф: 'a',
        х: '[',
        ц: 'w',
        ч: 'x',
        ш: 'i',
        щ: 'o',
        ъ: ']',
        ы: 's',
        ь: 'm',
        ю: '.',
        я: 'z',
    };
    /**
     * Translates text mistakenly typed in the Russian layout into English
     * @param string string with Russian layout characters
     * @return string with English layout characters
     */
    function tuiTransliterateKeyboardLayout(string) {
        var newStr = '';
        for (var i = 0; i < string.length; i++) {
            newStr += string.charAt(i) in MAP ? MAP[string.charAt(i)] : string.charAt(i);
        }
        return newStr;
    }

    function tuiTypeReferenceParser(types) {
        var generics = types.match(/<([^>]+)>/g);
        var escaped = generics
            ? generics
                .reduce(function (result, current) { return result.replace(current, current.replace(/\|/g, '&')); }, types)
                .split('|')
                .map(function (item) { return item.trim(); })
            : types.split('|').map(function (item) { return item.trim(); });
        return escaped.reduce(function (result, type) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            var extracted = type
                .trim()
                .replace(/readonly /g, '')
                .replace(/\[\]/g, '');
            extracted =
                (_d = (_c = (_b = (_a = extracted.match(/ReadonlyArray<([^>]+)>/)) === null || _a === void 0 ? void 0 : _a[1]) === null || _b === void 0 ? void 0 : _b.split('&')) === null || _c === void 0 ? void 0 : _c[0]) !== null && _d !== void 0 ? _d : extracted;
            extracted = (_h = (_g = (_f = (_e = extracted.match(/\[([^\]]+)\]/)) === null || _e === void 0 ? void 0 : _e[1]) === null || _f === void 0 ? void 0 : _f.split(',')) === null || _g === void 0 ? void 0 : _g[0]) !== null && _h !== void 0 ? _h : extracted;
            extracted = (_m = (_l = ((_k = (_j = extracted.split('<')) === null || _j === void 0 ? void 0 : _j[0]) !== null && _k !== void 0 ? _k : extracted)) === null || _l === void 0 ? void 0 : _l.trim()) !== null && _m !== void 0 ? _m : '';
            extracted = Number.isNaN(parseFloat(extracted)) ? extracted : 'number';
            extracted = /^'(.+)'$|^"(.+)"$|^`(.+)`$/.test(extracted) ? 'string' : extracted;
            extracted = extracted.length === 1 ? 'unknown' : extracted;
            return result.concat({ type: type.replace(/&/g, '|'), extracted: extracted });
        }, []);
    }

    /**
     * Generated bundle index. Do not edit.
     */

    exports.tuiCoerceValue = tuiCoerceValue;
    exports.tuiCoerceValueIsTrue = tuiCoerceValueIsTrue;
    exports.tuiGenerateRoutes = tuiGenerateRoutes;
    exports.tuiInspectAny = tuiInspectAny;
    exports.tuiIsPageGroup = tuiIsPageGroup;
    exports.tuiRawLoad = tuiRawLoad;
    exports.tuiRawLoadRecord = tuiRawLoadRecord;
    exports.tuiSortPages = tuiSortPages;
    exports.tuiToFlatMapPages = tuiToFlatMapPages;
    exports.tuiTransliterateKeyboardLayout = tuiTransliterateKeyboardLayout;
    exports.tuiTryParseMarkdownCodeBlock = tuiTryParseMarkdownCodeBlock;
    exports.tuiTypeReferenceParser = tuiTypeReferenceParser;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
