// @taiga-ui/addon-doc/bundles/taiga-ui-addon-doc-directives.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-doc/tokens'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-doc/directives', ['exports', '@angular/core', '@taiga-ui/addon-doc/tokens', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-doc"] = global["taiga-ui"]["addon-doc"] || {}, global["taiga-ui"]["addon-doc"].directives = {}), global.ng.core, global["taiga-ui"]["addon-doc"].tokens, global.i1, global.rxjs, global.rxjs.operators));
})(this, (function (exports, i0, tokens, i1, i2, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    var TuiScrollIntoViewLinkDirective = /** @class */ (function () {
        function TuiScrollIntoViewLinkDirective(destroy$, _a, readyToScroll$) {
            var _this = this;
            var nativeElement = _a.nativeElement;
            this.readyToScroll$ = readyToScroll$;
            this.scroll$ = new i2.ReplaySubject(1);
            this.readyToScroll$
                .pipe(operators.filter(Boolean), operators.switchMap(function () { return _this.scroll$; }), operators.debounceTime(750), operators.filter(function (shallWe) { return shallWe && !!i1.tuiGetElementObscures(nativeElement); }), operators.takeUntil(destroy$))
                .subscribe(function () {
                nativeElement.scrollIntoView();
            });
        }
        Object.defineProperty(TuiScrollIntoViewLinkDirective.prototype, "tuiScrollIntoViewLink", {
            set: function (shallWe) {
                this.scroll$.next(shallWe);
            },
            enumerable: false,
            configurable: true
        });
        return TuiScrollIntoViewLinkDirective;
    }());
    TuiScrollIntoViewLinkDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkDirective, deps: [{ token: i1.TuiDestroyService, self: true }, { token: i0.ElementRef }, { token: tokens.TUI_DOC_PAGE_LOADED }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiScrollIntoViewLinkDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiScrollIntoViewLinkDirective, selector: "[tuiScrollIntoViewLink]", inputs: { tuiScrollIntoViewLink: "tuiScrollIntoViewLink" }, providers: [i1.TuiDestroyService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiScrollIntoViewLink]',
                        providers: [i1.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace.TuiDestroyService, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i1.TuiDestroyService]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i2__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_DOC_PAGE_LOADED]
                        }] }];
        }, propDecorators: { tuiScrollIntoViewLink: [{
                    type: i0.Input
                }] } });

    var TuiScrollIntoViewLinkModule = /** @class */ (function () {
        function TuiScrollIntoViewLinkModule() {
        }
        return TuiScrollIntoViewLinkModule;
    }());
    TuiScrollIntoViewLinkModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiScrollIntoViewLinkModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkModule, declarations: [TuiScrollIntoViewLinkDirective], exports: [TuiScrollIntoViewLinkDirective] });
    TuiScrollIntoViewLinkModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiScrollIntoViewLinkModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiScrollIntoViewLinkDirective],
                        exports: [TuiScrollIntoViewLinkDirective],
                    }]
            }] });

    var TuiTextCodeDirective = /** @class */ (function () {
        function TuiTextCodeDirective() {
            this.code = '';
        }
        return TuiTextCodeDirective;
    }());
    TuiTextCodeDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiTextCodeDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTextCodeDirective, selector: "code[tuiText]", inputs: { code: ["tuiText", "code"] }, host: { properties: { "textContent": "this.code" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'code[tuiText]',
                    }]
            }], propDecorators: { code: [{
                    type: i0.Input,
                    args: ['tuiText']
                }, {
                    type: i0.HostBinding,
                    args: ['textContent']
                }] } });

    var TuiTextCodeModule = /** @class */ (function () {
        function TuiTextCodeModule() {
        }
        return TuiTextCodeModule;
    }());
    TuiTextCodeModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiTextCodeModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeModule, declarations: [TuiTextCodeDirective], exports: [TuiTextCodeDirective] });
    TuiTextCodeModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiTextCodeModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiTextCodeDirective],
                        exports: [TuiTextCodeDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiScrollIntoViewLinkDirective = TuiScrollIntoViewLinkDirective;
    exports.TuiScrollIntoViewLinkModule = TuiScrollIntoViewLinkModule;
    exports.TuiTextCodeDirective = TuiTextCodeDirective;
    exports.TuiTextCodeModule = TuiTextCodeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
