// @taiga-ui/kit/esm2015/components/carousel/carousel-buttons.directive.js
import { Directive } from '@angular/core';
import { TuiAppearance, tuiButtonOptionsProvider } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiCarouselButtonsDirective {
}
TuiCarouselButtonsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselButtonsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiCarouselButtonsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCarouselButtonsDirective, selector: "[tuiCarouselButtons]", providers: [
        tuiButtonOptionsProvider({
            appearance: TuiAppearance.Secondary,
            shape: 'rounded',
            size: 'm',
        }),
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCarouselButtonsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCarouselButtons]',
                    providers: [
                        tuiButtonOptionsProvider({
                            appearance: TuiAppearance.Secondary,
                            shape: 'rounded',
                            size: 'm',
                        }),
                    ],
                }]
        }] });
