// @taiga-ui/layout/esm2015/components/block-status/block-status.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBlockStatusComponent } from './block-status.component';
import { TuiBlockStatusDirective } from './block-status.directive';
import * as i0 from "@angular/core";
export class TuiBlockStatusModule {
}
TuiBlockStatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBlockStatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective], imports: [CommonModule], exports: [TuiBlockStatusComponent, TuiBlockStatusDirective] });
TuiBlockStatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBlockStatusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                    exports: [TuiBlockStatusComponent, TuiBlockStatusDirective],
                }]
        }] });
