// @taiga-ui/addon-charts/esm2015/components/pie-chart/pie-chart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiHoveredModule, TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiHintModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPieChartComponent } from './pie-chart.component';
import { TuiPieChartDirective } from './pie-chart.directive';
import * as i0 from "@angular/core";
export class TuiPieChartModule {
}
TuiPieChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPieChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPieChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPieChartModule, declarations: [TuiPieChartComponent, TuiPieChartDirective], imports: [CommonModule,
        TuiRepeatTimesModule,
        TuiHoveredModule,
        PolymorpheusModule,
        TuiHintModule], exports: [TuiPieChartComponent] });
TuiPieChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPieChartModule, imports: [[
            CommonModule,
            TuiRepeatTimesModule,
            TuiHoveredModule,
            PolymorpheusModule,
            TuiHintModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPieChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiRepeatTimesModule,
                        TuiHoveredModule,
                        PolymorpheusModule,
                        TuiHintModule,
                    ],
                    declarations: [TuiPieChartComponent, TuiPieChartDirective],
                    exports: [TuiPieChartComponent],
                }]
        }] });
