// @taiga-ui/experimental/esm2015/components/tooltip/tooltip.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiHintModule } from '@taiga-ui/core';
import { TuiButtonModule } from '@taiga-ui/experimental/components/button';
import { TuiIconModule } from '@taiga-ui/experimental/components/icon';
import { TuiAppearanceModule } from '@taiga-ui/experimental/directives/appearance';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTooltipComponent } from './tooltip.component';
import * as i0 from "@angular/core";
export class TuiTooltipModule {
}
TuiTooltipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTooltipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTooltipModule, declarations: [TuiTooltipComponent], imports: [CommonModule,
        TuiButtonModule,
        TuiAppearanceModule,
        TuiHintModule,
        TuiIconModule,
        PolymorpheusModule], exports: [TuiTooltipComponent] });
TuiTooltipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTooltipModule, imports: [[
            CommonModule,
            TuiButtonModule,
            TuiAppearanceModule,
            TuiHintModule,
            TuiIconModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiButtonModule,
                        TuiAppearanceModule,
                        TuiHintModule,
                        TuiIconModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiTooltipComponent],
                    exports: [TuiTooltipComponent],
                }]
        }] });
