// @taiga-ui/kit/esm2015/components/textarea/textarea.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiFocusableModule } from '@taiga-ui/cdk';
import { TuiPrimitiveTextfieldModule, TuiScrollbarModule, TuiSvgModule, TuiTextfieldComponent, TuiTooltipModule, TuiWrapperModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTextareaComponent } from './textarea.component';
import { TuiTextareaDirective } from './textarea.directive';
import * as i0 from "@angular/core";
export class TuiTextareaModule {
}
TuiTextareaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextareaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTextareaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextareaModule, declarations: [TuiTextareaComponent, TuiTextareaDirective], imports: [CommonModule,
        FormsModule,
        TuiFocusableModule,
        TuiScrollbarModule,
        TuiTooltipModule,
        TuiWrapperModule,
        TuiSvgModule,
        TuiPrimitiveTextfieldModule,
        PolymorpheusModule], exports: [TuiTextareaComponent, TuiTextareaDirective, TuiTextfieldComponent] });
TuiTextareaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextareaModule, imports: [[
            CommonModule,
            FormsModule,
            TuiFocusableModule,
            TuiScrollbarModule,
            TuiTooltipModule,
            TuiWrapperModule,
            TuiSvgModule,
            TuiPrimitiveTextfieldModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTextareaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiFocusableModule,
                        TuiScrollbarModule,
                        TuiTooltipModule,
                        TuiWrapperModule,
                        TuiSvgModule,
                        TuiPrimitiveTextfieldModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiTextareaComponent, TuiTextareaDirective],
                    exports: [TuiTextareaComponent, TuiTextareaDirective, TuiTextfieldComponent],
                }]
        }] });
