// @taiga-ui/addon-doc/esm2015/components/internal/header/header.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSidebarModule } from '@taiga-ui/addon-mobile';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDocNavigationModule } from '../../navigation/navigation.module';
import { TuiDocHeaderComponent } from './header.component';
import * as i0 from "@angular/core";
export class TuiDocHeaderModule {
}
TuiDocHeaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocHeaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderModule, declarations: [TuiDocHeaderComponent], imports: [CommonModule,
        PolymorpheusModule,
        TuiButtonModule,
        TuiSidebarModule,
        TuiActiveZoneModule,
        TuiDocNavigationModule], exports: [TuiDocHeaderComponent] });
TuiDocHeaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiButtonModule,
            TuiSidebarModule,
            TuiActiveZoneModule,
            TuiDocNavigationModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiButtonModule,
                        TuiSidebarModule,
                        TuiActiveZoneModule,
                        TuiDocNavigationModule,
                    ],
                    declarations: [TuiDocHeaderComponent],
                    exports: [TuiDocHeaderComponent],
                }]
        }] });
