// @taiga-ui/addon-doc/esm2015/components/internal/header/header.component.js
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { Router } from '@angular/router';
import { TUI_DOC_ICONS, TUI_DOC_LOGO, TUI_DOC_MENU_TEXT, } from '@taiga-ui/addon-doc/tokens';
import { ALWAYS_FALSE_HANDLER } from '@taiga-ui/cdk';
import { merge, Subject } from 'rxjs';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
import * as i2 from "../../navigation/navigation.component";
import * as i3 from "@taiga-ui/cdk";
import * as i4 from "@taiga-ui/addon-mobile";
import * as i5 from "@tinkoff/ng-polymorpheus";
import * as i6 from "@angular/common";
import * as i7 from "@angular/router";
export class TuiDocHeaderComponent {
    constructor(icons, logo, menu, router) {
        this.icons = icons;
        this.logo = logo;
        this.menu = menu;
        this.router = router;
        this.stream$ = new Subject();
        this.open$ = merge(this.router.events.pipe(map(ALWAYS_FALSE_HANDLER)), this.stream$).pipe(startWith(false), distinctUntilChanged());
    }
    onClick() {
        this.stream$.next(true);
    }
    onActiveZone(active) {
        if (!active) {
            this.stream$.next(false);
        }
    }
}
TuiDocHeaderComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderComponent, deps: [{ token: TUI_DOC_ICONS }, { token: TUI_DOC_LOGO }, { token: TUI_DOC_MENU_TEXT }, { token: Router }], target: i0.ɵɵFactoryTarget.Component });
TuiDocHeaderComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiDocHeaderComponent, selector: "header[tuiDocHeader]", ngImport: i0, template: "<button\n    appearance=\"icon\"\n    tuiIconButton\n    type=\"button\"\n    class=\"t-menu\"\n    [icon]=\"icons.menu\"\n    [title]=\"menu\"\n    (click)=\"onClick()\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-doc-navigation\n        *tuiSidebar=\"!!(open$ | async)\"\n        class=\"t-navigation\"\n    ></tui-doc-navigation>\n</button>\n<div class=\"t-logo\">\n    <img\n        *polymorpheusOutlet=\"logo as src\"\n        alt=\"Logo\"\n        class=\"t-img\"\n        [src]=\"src\"\n    />\n</div>\n<ng-content></ng-content>\n", styles: [":host{position:fixed;top:0;left:0;right:0;z-index:1;display:flex;box-shadow:var(--tui-shadow-navigation);height:3.9375rem;align-items:center;padding:0 1.25rem;background:var(--tui-base-01);border-bottom:1px solid var(--tui-base-03)}@supports (background: color-mix(in srgb,var(--tui-base-01),transparent 50%)){:host{background:color-mix(in srgb,var(--tui-base-01),transparent 50%);-webkit-backdrop-filter:blur(1.25rem);backdrop-filter:blur(1.25rem)}}.t-logo{margin-right:auto;font:var(--tui-font-heading-5)}.t-img{display:block}.t-menu{margin-left:-1.25rem}@media screen and (min-width: 48em){.t-menu{display:none}}.t-navigation{min-height:100%}\n"], components: [{ type: i1.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }, { type: i2.TuiDocNavigationComponent, selector: "tui-doc-navigation" }], directives: [{ type: i3.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }, { type: i4.TuiSidebarDirective, selector: "[tuiSidebar]", inputs: ["tuiSidebarDirection", "tuiSidebarAutoWidth", "tuiSidebar"] }, { type: i5.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i6.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocHeaderComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'header[tuiDocHeader]',
                    templateUrl: './header.template.html',
                    styleUrls: ['./header.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_ICONS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_LOGO]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_MENU_TEXT]
                }] }, { type: i7.Router, decorators: [{
                    type: Inject,
                    args: [Router]
                }] }]; } });
