// @taiga-ui/experimental/esm2015/directives/sensitive/sensitive.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDestroyService, TuiDirectiveStylesService, TuiResizeService, } from '@taiga-ui/cdk';
import { map } from 'rxjs/operators';
import { TuiSensitiveComponent } from './sensitive.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
const rowsInSvg = 3;
export class TuiSensitiveDirective {
    constructor(directiveStyles, resize$) {
        this.resize$ = resize$;
        this.tuiSensitive = false;
        this.offset = Math.round(Math.random() * 10) * 10;
        this.height$ = this.resize$.pipe(map(([{ contentRect }]) => [
            Math.max(2, Math.floor(contentRect.height / 16) + 1),
            contentRect.height,
        ]), map(([rows, height]) => height * (rowsInSvg / rows)));
        directiveStyles.addComponent(TuiSensitiveComponent);
    }
}
TuiSensitiveDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSensitiveDirective, deps: [{ token: TuiDirectiveStylesService }, { token: TuiResizeService }], target: i0.ɵɵFactoryTarget.Directive });
TuiSensitiveDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSensitiveDirective, selector: "[tuiSensitive]", inputs: { tuiSensitive: "tuiSensitive" }, host: { listeners: { "$.style.--t-mask-height.px": "height$" }, properties: { "style.--t-offset.px": "offset", "$.style.--t-mask-height.px": "height$", "class.tui-sensitive": "tuiSensitive" } }, providers: [TuiResizeService, TuiDestroyService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSensitiveDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSensitive]',
                    providers: [TuiResizeService, TuiDestroyService],
                    host: {
                        '[style.--t-offset.px]': 'offset',
                        '[$.style.--t-mask-height.px]': 'height$',
                        '($.style.--t-mask-height.px)': 'height$',
                        '[class.tui-sensitive]': 'tuiSensitive',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }, { type: i1.TuiResizeService, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }]; }, propDecorators: { tuiSensitive: [{
                type: Input
            }] } });
