// @taiga-ui/kit/esm2015/components/input-date/input-date.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiLetModule, TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiCalendarModule, TuiHostedDropdownModule, TuiLinkModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiValueAccessorModule } from '@taiga-ui/kit/directives';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputDateComponent } from './input-date.component';
import { TuiInputDateDirective } from './input-date.directive';
import { TuiNativeDateDirective } from './native-date/native-date.component';
import * as i0 from "@angular/core";
export class TuiInputDateModule {
}
TuiInputDateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputDateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateModule, declarations: [TuiInputDateComponent, TuiInputDateDirective, TuiNativeDateDirective], imports: [CommonModule,
        MaskitoModule,
        PolymorpheusModule,
        TuiWrapperModule,
        TuiPreventDefaultModule,
        TuiCalendarModule,
        TuiSvgModule,
        TuiLinkModule,
        TuiHostedDropdownModule,
        TuiPrimitiveTextfieldModule,
        TuiValueAccessorModule,
        TuiLetModule,
        TuiTextfieldControllerModule], exports: [TuiInputDateComponent, TuiInputDateDirective, TuiTextfieldComponent] });
TuiInputDateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateModule, imports: [[
            CommonModule,
            MaskitoModule,
            PolymorpheusModule,
            TuiWrapperModule,
            TuiPreventDefaultModule,
            TuiCalendarModule,
            TuiSvgModule,
            TuiLinkModule,
            TuiHostedDropdownModule,
            TuiPrimitiveTextfieldModule,
            TuiValueAccessorModule,
            TuiLetModule,
            TuiTextfieldControllerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoModule,
                        PolymorpheusModule,
                        TuiWrapperModule,
                        TuiPreventDefaultModule,
                        TuiCalendarModule,
                        TuiSvgModule,
                        TuiLinkModule,
                        TuiHostedDropdownModule,
                        TuiPrimitiveTextfieldModule,
                        TuiValueAccessorModule,
                        TuiLetModule,
                        TuiTextfieldControllerModule,
                    ],
                    declarations: [TuiInputDateComponent, TuiInputDateDirective, TuiNativeDateDirective],
                    exports: [TuiInputDateComponent, TuiInputDateDirective, TuiTextfieldComponent],
                }]
        }] });
