// @taiga-ui/kit/esm2015/components/push/push.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TUI_ALERTS } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiFormatDatePipeModule, TuiLinkModule, TuiSvgModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPushComponent } from './push.component';
import { TuiPushDirective } from './push.directive';
import { TuiPushService } from './push.service';
import { TuiPushAlertComponent } from './push-alert.component';
import { TuiPushAlertDirective } from './push-alert.directive';
import * as i0 from "@angular/core";
export class TuiPushModule {
}
TuiPushModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPushModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushModule, declarations: [TuiPushComponent,
        TuiPushDirective,
        TuiPushAlertComponent,
        TuiPushAlertDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiButtonModule,
        TuiLinkModule,
        TuiSvgModule,
        TuiFormatDatePipeModule], exports: [TuiPushComponent,
        TuiPushDirective,
        TuiPushAlertComponent,
        TuiPushAlertDirective] });
TuiPushModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushModule, providers: [
        {
            provide: TUI_ALERTS,
            useExisting: TuiPushService,
            multi: true,
        },
    ], imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiButtonModule,
            TuiLinkModule,
            TuiSvgModule,
            TuiFormatDatePipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPushModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiButtonModule,
                        TuiLinkModule,
                        TuiSvgModule,
                        TuiFormatDatePipeModule,
                    ],
                    declarations: [
                        TuiPushComponent,
                        TuiPushDirective,
                        TuiPushAlertComponent,
                        TuiPushAlertDirective,
                    ],
                    providers: [
                        {
                            provide: TUI_ALERTS,
                            useExisting: TuiPushService,
                            multi: true,
                        },
                    ],
                    exports: [
                        TuiPushComponent,
                        TuiPushDirective,
                        TuiPushAlertComponent,
                        TuiPushAlertDirective,
                    ],
                }]
        }] });
