// @taiga-ui/experimental/esm2015/components/button/button.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLoaderModule, TuiSvgModule } from '@taiga-ui/core';
import { TuiIconsDirective, TuiIconsModule } from '@taiga-ui/experimental/directives/icons';
import { TuiButtonComponent } from './button.component';
import { TuiButtonDirective, TuiButtonStylesComponent } from './button.directive';
import * as i0 from "@angular/core";
export class TuiButtonModule {
}
TuiButtonModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiButtonModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonModule, declarations: [TuiButtonComponent, TuiButtonDirective, TuiButtonStylesComponent], imports: [CommonModule, TuiSvgModule, TuiLoaderModule, TuiIconsModule], exports: [TuiButtonComponent, TuiButtonDirective, TuiIconsDirective] });
TuiButtonModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonModule, imports: [[CommonModule, TuiSvgModule, TuiLoaderModule, TuiIconsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiLoaderModule, TuiIconsModule],
                    declarations: [TuiButtonComponent, TuiButtonDirective, TuiButtonStylesComponent],
                    exports: [TuiButtonComponent, TuiButtonDirective, TuiIconsDirective],
                }]
        }] });
