// @taiga-ui/experimental/esm2015/components/button/button.component.js
import { ChangeDetectionStrategy, Component, HostListener, Inject, Input, } from '@angular/core';
import { tuiIsString } from '@taiga-ui/cdk';
import { tuiSizeBigger } from '@taiga-ui/core';
import { TUI_BUTTON_OPTIONS } from './button.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
export class TuiButtonComponent {
    constructor(options) {
        this.options = options;
        this.size = this.options.size;
        this.loading = false;
    }
    get loaderSize() {
        return tuiSizeBigger(this.size) ? 'm' : 's';
    }
    get label() {
        return tuiIsString(this.loading) ? this.loading : '';
    }
    onClick(event) {
        if (this.loading) {
            event.stopPropagation();
        }
    }
}
TuiButtonComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonComponent, deps: [{ token: TUI_BUTTON_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiButtonComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonComponent, selector: "[tuiButton][loading],[tuiIconButton][loading]", inputs: { size: "size", loading: "loading" }, host: { listeners: { "click.capture": "onClick($event)" }, properties: { "attr.aria-disabled": "loading", "class._loading": "loading" } }, ngImport: i0, template: "<ng-content></ng-content>\n<tui-loader\n    aria-live=\"polite\"\n    role=\"status\"\n    class=\"t-loader\"\n    [inheritColor]=\"true\"\n    [showLoader]=\"!!loading\"\n    [size]=\"loaderSize\"\n    [textContent]=\"label\"\n></tui-loader>\n", components: [{ type: i1.TuiLoaderComponent, selector: "tui-loader", inputs: ["size", "inheritColor", "overlay", "textContent", "showLoader"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[tuiButton][loading],[tuiIconButton][loading]',
                    templateUrl: './button.template.html',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[attr.aria-disabled]': 'loading',
                        '[class._loading]': 'loading',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_BUTTON_OPTIONS]
                }] }]; }, propDecorators: { size: [{
                type: Input
            }], loading: [{
                type: Input
            }], onClick: [{
                type: HostListener,
                args: ['click.capture', ['$event']]
            }] } });
