// @taiga-ui/addon-charts/esm2015/components/line-chart/line-chart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusableModule, TuiLetModule } from '@taiga-ui/cdk';
import { TuiHintModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiLineChartComponent } from './line-chart.component';
import { TuiLineChartHintDirective } from './line-chart-hint.directive';
import * as i0 from "@angular/core";
export class TuiLineChartModule {
}
TuiLineChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLineChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineChartModule, declarations: [TuiLineChartComponent, TuiLineChartHintDirective], imports: [CommonModule,
        PolymorpheusModule,
        TuiHintModule,
        TuiFocusableModule,
        TuiLetModule], exports: [TuiLineChartComponent, TuiLineChartHintDirective] });
TuiLineChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineChartModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiHintModule,
            TuiFocusableModule,
            TuiLetModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiHintModule,
                        TuiFocusableModule,
                        TuiLetModule,
                    ],
                    declarations: [TuiLineChartComponent, TuiLineChartHintDirective],
                    exports: [TuiLineChartComponent, TuiLineChartHintDirective],
                }]
        }] });
