// @taiga-ui/experimental/esm2015/pipes/fallback-src/fallback-src.pipe.js
import { ElementRef, Inject, Pipe } from '@angular/core';
import { TUI_ICON_ERROR } from '@taiga-ui/core';
import { fromEvent, merge } from 'rxjs';
import { map, startWith } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiFallbackSrcPipe {
    constructor(el) {
        this.el = el;
    }
    transform(src, fallback) {
        return merge(fromEvent(this.el.nativeElement, TUI_ICON_ERROR), fromEvent(this.el.nativeElement, 'error', { capture: true })).pipe(map(() => fallback), startWith(src));
    }
}
TuiFallbackSrcPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Pipe });
TuiFallbackSrcPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, name: "tuiFallbackSrc" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFallbackSrcPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'tuiFallbackSrc',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; } });
