// @taiga-ui/kit/esm2015/components/multi-select/multi-select.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost, tuiAsTextfieldHost } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiMultiSelectDirective extends AbstractTuiTextfieldHost {
    constructor() {
        super(...arguments);
        this.disableItemHandler = item => this.host.disabledItemHandler(item);
    }
    get readOnly() {
        return true;
    }
    onValueChange() { }
    get stringify() {
        return this.host.stringify;
    }
    onSelectionChange(value) {
        this.host.onValueChange(value);
    }
}
TuiMultiSelectDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiMultiSelectDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMultiSelectDirective, selector: "tui-multi-select", providers: [tuiAsTextfieldHost(TuiMultiSelectDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMultiSelectDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-multi-select',
                    providers: [tuiAsTextfieldHost(TuiMultiSelectDirective)],
                }]
        }] });
