// @taiga-ui/addon-commerce/esm2015/components/thumbnail-card/thumbnail-card.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiThumbnailCardComponent } from './thumbnail-card.component';
import * as i0 from "@angular/core";
export class TuiThumbnailCardModule {
}
TuiThumbnailCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiThumbnailCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, declarations: [TuiThumbnailCardComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiThumbnailCardComponent] });
TuiThumbnailCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiThumbnailCardComponent],
                    exports: [TuiThumbnailCardComponent],
                }]
        }] });
