// @taiga-ui/addon-mobile/esm2015/directives/mobile-tabs/mobile-tabs.directive.js
import { Directive, HostBinding, Inject } from '@angular/core';
import { TUI_IS_ANDROID, TUI_IS_IOS, TuiDirectiveStylesService, } from '@taiga-ui/cdk';
import { TUI_TAB_MARGIN } from '@taiga-ui/kit';
import { TuiMobileTabsComponent } from './mobile-tabs.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiMobileTabsDirective {
    constructor(isIos, isAndroid, directiveStyles) {
        this.isIos = isIos;
        this.isAndroid = isAndroid;
        directiveStyles.addComponent(TuiMobileTabsComponent);
    }
    // TODO: Refactor with TUI_PLATFORM
    get platform() {
        if (this.isIos) {
            return 'ios';
        }
        if (this.isAndroid) {
            return 'android';
        }
        return null;
    }
}
TuiMobileTabsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsDirective, deps: [{ token: TUI_IS_IOS }, { token: TUI_IS_ANDROID }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiMobileTabsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileTabsDirective, selector: "[tuiMobileTabs]", host: { properties: { "attr.data-platform": "this.platform" } }, providers: [
        {
            provide: TUI_TAB_MARGIN,
            useValue: 0,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiMobileTabs]',
                    providers: [
                        {
                            provide: TUI_TAB_MARGIN,
                            useValue: 0,
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_IOS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_ANDROID]
                }] }, { type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { platform: [{
                type: HostBinding,
                args: ['attr.data-platform']
            }] } });
