// @taiga-ui/addon-doc/esm2015/components/documentation/pipes/inspect.pipe.js
import { Inject, Pipe } from '@angular/core';
import { tuiInspectAny } from '@taiga-ui/addon-doc/utils';
import { TUI_IS_E2E } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiInspectPipe {
    constructor(isE2E) {
        this.isE2E = isE2E;
    }
    transform(value, depth = 2) {
        if (this.isE2E && typeof value === 'function') {
            /**
             * @description:
             * When developing in production mode the webpack bundler minify
             * functions in different ways, then due to which the string content
             * of the function may differ from build to build, which can be to
             * various problems when screenshot testing on e2e.
             */
            return 'λ(x) => y';
        }
        return tuiInspectAny(value, depth);
    }
}
TuiInspectPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInspectPipe, deps: [{ token: TUI_IS_E2E }], target: i0.ɵɵFactoryTarget.Pipe });
TuiInspectPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInspectPipe, name: "tuiInspectAny" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInspectPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'tuiInspectAny' }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_E2E]
                }] }]; } });
