// @taiga-ui/addon-doc/esm2015/components/documentation/pipes/type-reference.pipe.js
import { Inject, Pipe } from '@angular/core';
import { TUI_DOC_TYPE_REFERENCE_HANDLER, TUI_DOC_TYPE_REFERENCE_PARSER, } from '@taiga-ui/addon-doc/tokens';
import * as i0 from "@angular/core";
export class TuiDocTypeReferencePipe {
    constructor(parser, linkHandler) {
        this.parser = parser;
        this.linkHandler = linkHandler;
    }
    transform(original) {
        return this.parser(original)
            .map(({ type, extracted }) => {
            var _a, _b;
            return ({
                type,
                extracted,
                reference: (_b = (_a = this.linkHandler) === null || _a === void 0 ? void 0 : _a.call(this, extracted)) !== null && _b !== void 0 ? _b : null,
            });
        })
            .sort((a, b) => { var _a, _b, _c; return (_c = (_a = b.reference) === null || _a === void 0 ? void 0 : _a.localeCompare((_b = a.reference) !== null && _b !== void 0 ? _b : '')) !== null && _c !== void 0 ? _c : -1; });
    }
}
TuiDocTypeReferencePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTypeReferencePipe, deps: [{ token: TUI_DOC_TYPE_REFERENCE_PARSER }, { token: TUI_DOC_TYPE_REFERENCE_HANDLER }], target: i0.ɵɵFactoryTarget.Pipe });
TuiDocTypeReferencePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTypeReferencePipe, name: "tuiDocTypeReference" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocTypeReferencePipe, decorators: [{
            type: Pipe,
            args: [{ name: 'tuiDocTypeReference' }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_TYPE_REFERENCE_PARSER]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_TYPE_REFERENCE_HANDLER]
                }] }]; } });
