// @taiga-ui/experimental/esm2015/components/rating/rating.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiIconModule } from '@taiga-ui/experimental/components/icon';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiRatingComponent } from './rating.component';
import * as i0 from "@angular/core";
export class TuiRatingModule {
}
TuiRatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, declarations: [TuiRatingComponent], imports: [CommonModule,
        FormsModule,
        TuiIconModule,
        TuiRepeatTimesModule,
        PolymorpheusModule], exports: [TuiRatingComponent] });
TuiRatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, imports: [[
            CommonModule,
            FormsModule,
            TuiIconModule,
            TuiRepeatTimesModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        TuiIconModule,
                        TuiRepeatTimesModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiRatingComponent],
                    exports: [TuiRatingComponent],
                }]
        }] });
