// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-sheet.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@ng-web-apis/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/sheet', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit', 'rxjs', 'rxjs/operators', '@angular/common', '@ng-web-apis/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components.sheet = {}), global.ng.core, global.i3, global.i1, global.kit, global.rxjs, global.rxjs.operators, global.ng.common, global.common, global.i5));
})(this, (function (exports, i0, i3, i1, kit, i1$1, operators, i2, common, i5) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /**
     * Sheet main component
     */
    var TUI_SHEET = new i0.InjectionToken('[TUI_SHEET]');
    /**
     * Current scrollTop of a sheet
     */
    var TUI_SHEET_SCROLL = new i0.InjectionToken('[TUI_SHEET_SCROLL]');
    /**
     * The sheet is being dragged
     */
    var TUI_SHEET_DRAGGED = new i0.InjectionToken('[TUI_SHEET_DRAGGED]');

    var TUI_SHEET_CLOSE = 'tui-sheet-close';
    var TUI_SHEET_ID = 'tui-sheet-id';
    var TuiSheetHeadingComponent = /** @class */ (function () {
        function TuiSheetHeadingComponent(idService, el, closeWord$, icons) {
            this.idService = idService;
            this.el = el;
            this.closeWord$ = closeWord$;
            this.icons = icons;
            this.id = this.idService.generate();
        }
        TuiSheetHeadingComponent.prototype.ngAfterViewInit = function () {
            this.el.nativeElement.dispatchEvent(new CustomEvent(TUI_SHEET_ID, { bubbles: true, detail: this.id }));
        };
        TuiSheetHeadingComponent.prototype.onClick = function () {
            this.el.nativeElement.dispatchEvent(new CustomEvent(TUI_SHEET_CLOSE, { bubbles: true }));
        };
        return TuiSheetHeadingComponent;
    }());
    TuiSheetHeadingComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetHeadingComponent, deps: [{ token: i3.TuiIdService }, { token: i0.ElementRef }, { token: i1.TUI_CLOSE_WORD }, { token: i1.TUI_COMMON_ICONS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSheetHeadingComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetHeadingComponent, selector: "[tuiSheetHeading]", ngImport: i0__namespace, template: "<button\n    appearance=\"icon\"\n    size=\"xs\"\n    tuiIconButton\n    type=\"button\"\n    class=\"t-close\"\n    [icon]=\"icons.close\"\n    [title]=\"closeWord$ | async\"\n    (click)=\"onClick()\"\n></button>\n<span [id]=\"id\">\n    <ng-content></ng-content>\n</span>\n", styles: [":host{position:-webkit-sticky;position:sticky;top:0;z-index:1;min-height:2rem;box-sizing:border-box;margin:-1.25rem -1rem 0;padding:1.5rem 1rem 1rem;background:var(--tui-elevation-01);border-radius:.75rem .75rem 0 0;font:var(--tui-font-heading-6);box-shadow:inset 0 1px #ffffff40}:host:after{content:\"\";position:absolute;top:100%;left:0;right:0;height:1px;background:var(--tui-base-03);opacity:0}:host :host-context(._stuck):after{opacity:1}.t-close{display:none;float:right;margin-right:-.25rem}:host-context(.t-wrapper_closeable) .t-close{display:block}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetHeadingComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: '[tuiSheetHeading]',
                        templateUrl: './sheet-heading.template.html',
                        styleUrls: ['./sheet-heading.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: i3__namespace.TuiIdService, decorators: [{
                            type: i0.Inject,
                            args: [i3.TuiIdService]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_CLOSE_WORD]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_COMMON_ICONS]
                        }] }];
        } });

    // eslint-disable-next-line @typescript-eslint/naming-convention
    function iosScrollFactory(element, doc, zone) {
        var load$ = i3.tuiTypedFromEvent(element, 'load', { capture: true });
        var touchstart$ = i3.tuiTypedFromEvent(element, 'touchstart', { passive: true });
        var touchmove$ = i3.tuiTypedFromEvent(doc, 'touchmove', { passive: true });
        var touchend$ = i3.tuiTypedFromEvent(doc, 'touchend');
        var scroll$ = i3.tuiTypedFromEvent(element, 'scroll').pipe(operators.map(function () { return element.scrollTop; }));
        var result$ = i1$1.merge(load$.pipe(operators.delay(0), operators.map(function () { return element.scrollTop; })), touchstart$.pipe(operators.switchMap(function (_a) {
            var touches = _a.touches;
            var screenY = touches[0].screenY;
            var scrollTop = element.scrollTop;
            return i1$1.concat(
            // Sometimes touch is triggered without scroll in iOS, filter that
            i1$1.zip(touchmove$, scroll$).pipe(operators.map(function (_a) {
                var _b = __read(_a, 1), touches = _b[0].touches;
                return scrollTop + screenY - touches[0].screenY;
            }), operators.takeUntil(touchend$)), scroll$);
        })));
        return i1$1.concat(scroll$.pipe(operators.take(1)), result$).pipe(i3.tuiZonefree(zone), operators.share());
    }

    var TUI_SHEET_PROVIDERS = [
        {
            provide: TUI_SHEET_DRAGGED,
            deps: [i0.ElementRef],
            useFactory: function (_a) {
                var nativeElement = _a.nativeElement;
                return i1$1.merge(i3.tuiTypedFromEvent(nativeElement, 'touchstart', { passive: true }).pipe(operators.map(i3.ALWAYS_TRUE_HANDLER)), i3.tuiTypedFromEvent(nativeElement, 'touchend').pipe(operators.map(i3.ALWAYS_FALSE_HANDLER)));
            },
        },
        {
            provide: TUI_SHEET_SCROLL,
            deps: [i0.ElementRef, i0.NgZone, i2.DOCUMENT, i3.TUI_IS_IOS],
            useFactory: function (_a, zone, doc, isIos) {
                var nativeElement = _a.nativeElement;
                return isIos
                    ? iosScrollFactory(nativeElement, doc, zone)
                    : i1$1.merge(i3.tuiTypedFromEvent(nativeElement, 'scroll'), i3.tuiTypedFromEvent(nativeElement, 'load', { capture: true })).pipe(operators.map(function () { return nativeElement.scrollTop; }), i3.tuiZonefree(zone), operators.share());
            },
        },
        {
            provide: i3.TUI_SCROLL_REF,
            useExisting: i0.ElementRef,
        },
        {
            provide: TUI_SHEET,
            useExisting: i0.forwardRef(function () { return TuiSheetComponent; }),
        },
    ];

    var TuiSheetBarComponent = /** @class */ (function () {
        function TuiSheetBarComponent(sheet, scroll$, zone) {
            var _this = this;
            this.sheet = sheet;
            this.scroll$ = scroll$;
            this.zone = zone;
            this.rotate$ = this.scroll$.pipe(i1.tuiZonefulMap(function (y) { return i3.tuiClamp(10 - (y - _this.stop) / 5, 0, 10); }, this.zone));
        }
        Object.defineProperty(TuiSheetBarComponent.prototype, "stop", {
            get: function () {
                return this.sheet.stops[0] || 0;
            },
            enumerable: false,
            configurable: true
        });
        return TuiSheetBarComponent;
    }());
    TuiSheetBarComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetBarComponent, deps: [{ token: TUI_SHEET }, { token: TUI_SHEET_SCROLL }, { token: i0.NgZone }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSheetBarComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetBarComponent, selector: "tui-sheet-bar", ngImport: i0__namespace, template: "<ng-container *tuiLet=\"rotate$ | async as deg\">\n    <div\n        class=\"t-bar\"\n        [ngStyle]=\"{transform: 'rotate(-' + deg + 'deg)'}\"\n    ></div>\n    <ng-content *ngIf=\"deg\"></ng-content>\n    <div\n        class=\"t-bar\"\n        [ngStyle]=\"{transform: 'rotate(' + deg + 'deg)'}\"\n    ></div>\n</ng-container>\n", styles: [":host{transition-property:border-radius;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:-webkit-sticky;position:sticky;top:0;z-index:2;display:flex;height:1.25rem;align-items:center;justify-content:center;opacity:.3}.t-bar{content:\"\";height:.25rem;width:1rem;background:var(--tui-base-09);pointer-events:none}:host-context(._ios:not(._dragged)) .t-bar{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out}.t-bar:first-child{border-radius:1rem 0 0 1rem;transform-origin:top right}.t-bar:last-child{border-radius:0 1rem 1rem 0;transform-origin:top left}\n"], directives: [{ type: i3__namespace.TuiLetDirective, selector: "[tuiLet]", inputs: ["tuiLet"] }, { type: i2__namespace.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetBarComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-sheet-bar',
                        templateUrl: './sheet-bar.template.html',
                        styleUrls: ['./sheet-bar.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_SCROLL]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }];
        } });

    // So that borders get rounded when image is visible for at least 10px
    var OFFSET$1 = 10;
    var TuiSheetTopDirective = /** @class */ (function () {
        function TuiSheetTopDirective(scroll$, component, win) {
            var _this = this;
            this.scroll$ = scroll$;
            this.component = component;
            this.win = win;
            this.stop = 0;
            this.transform$ = this.scroll$.pipe(operators.map(function (y) { return "translateY(" + _this.getY(y) + "%) scaleX(-1)"; }));
            this.rounded$ = this.scroll$.pipe(operators.map(function (y) { return y < _this.stop + OFFSET$1; }));
            this.clickthrough$ = this.scroll$.pipe(operators.map(function (y) { return !!Math.round(_this.getY(y)); }));
        }
        TuiSheetTopDirective.prototype.getY = function (scrollTop) {
            var value = scrollTop - this.stop;
            var total = this.win.innerHeight - this.component.item.offset - this.stop;
            return this.stop && i3.tuiClamp(100 - (value / total) * 100, 0, 100);
        };
        return TuiSheetTopDirective;
    }());
    TuiSheetTopDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetTopDirective, deps: [{ token: TUI_SHEET_SCROLL }, { token: TUI_SHEET }, { token: common.WINDOW }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetTopDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetTopDirective, selector: "[tuiSheetTop]", inputs: { stop: ["tuiSheetTop", "stop"] }, host: { listeners: { "$.style.transform": "transform$", "$.class._rounded": "rounded$", "$.class._clickthrough": "clickthrough$" }, properties: { "$.style.transform": "transform$", "$.class._rounded": "rounded$", "$.class._clickthrough": "clickthrough$" } }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetTopDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiSheetTop]',
                        host: {
                            '[$.style.transform]': 'transform$',
                            '($.style.transform)': 'transform$',
                            '[$.class._rounded]': 'rounded$',
                            '($.class._rounded)': 'rounded$',
                            '[$.class._clickthrough]': 'clickthrough$',
                            '($.class._clickthrough)': 'clickthrough$',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_SCROLL]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET]
                        }] }, { type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }];
        }, propDecorators: { stop: [{
                    type: i0.Input,
                    args: ['tuiSheetTop']
                }] } });

    var TuiSheetStopDirective = /** @class */ (function () {
        function TuiSheetStopDirective(el, destroy$, dragged$, scroll$, _a) {
            var nativeElement = _a.nativeElement;
            scroll$
                .pipe(operators.map(function (y) { return Math.floor(y) > el.nativeElement.offsetTop; }), operators.distinctUntilChanged(), operators.withLatestFrom(dragged$), operators.map(function (_a) {
                var _b = __read(_a, 2), above = _b[0], dragged = _b[1];
                return !above && !dragged;
            }), operators.filter(Boolean), operators.throttleTime(100), operators.takeUntil(destroy$))
                .subscribe(function () {
                nativeElement.style.overflow = 'hidden';
                nativeElement.classList.remove('_stuck'); // iOS
                nativeElement.scrollTop = el.nativeElement.offsetTop;
                i1$1.timer(100)
                    .pipe(operators.takeUntil(destroy$))
                    // eslint-disable-next-line rxjs/no-nested-subscribe
                    .subscribe(function () {
                    nativeElement.style.overflow = '';
                });
            });
        }
        return TuiSheetStopDirective;
    }());
    TuiSheetStopDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetStopDirective, deps: [{ token: i0.ElementRef }, { token: i3.TuiDestroyService, self: true }, { token: TUI_SHEET_DRAGGED }, { token: TUI_SHEET_SCROLL }, { token: i3.TUI_SCROLL_REF }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetStopDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetStopDirective, selector: "[tuiSheetStop]", providers: [i3.TuiDestroyService], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetStopDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiSheetStop]',
                        providers: [i3.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_DRAGGED]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_SCROLL]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_SCROLL_REF]
                        }] }];
        } });

    var TuiSheetComponent = /** @class */ (function () {
        function TuiSheetComponent(scroll$, el, zone, isIos, moreWord$, destroy$) {
            var _this = this;
            this.scroll$ = scroll$;
            this.el = el;
            this.zone = zone;
            this.isIos = isIos;
            this.moreWord$ = moreWord$;
            this.destroy$ = destroy$;
            this.stopsRefs = i3.EMPTY_QUERY;
            this.id = '';
            this.stuck$ = this.scroll$.pipe(operators.map(function (y) { return Math.floor(y) > _this.contentTop; }));
        }
        Object.defineProperty(TuiSheetComponent.prototype, "stops", {
            get: function () {
                return this.getStops(this.stopsRefs);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetComponent.prototype, "imageStop", {
            get: function () {
                return (this.item.imageSlide && this.stops[this.stops.length - 1]) || 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetComponent.prototype, "imageHeight", {
            get: function () {
                return this.contentTop - this.sheetTop;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetComponent.prototype, "context", {
            get: function () {
                return Object.assign(Object.assign({}, this.item), { scroll$: this.scroll$.pipe(i3.tuiZonefull(this.zone)) });
            },
            enumerable: false,
            configurable: true
        });
        TuiSheetComponent.prototype.onId = function (id) {
            this.id = id;
        };
        TuiSheetComponent.prototype.ngAfterViewInit = function () {
            this.el.nativeElement.scrollTop = __spreadArray(__spreadArray([], __read(this.stops)), [this.sheetTop, this.contentTop])[this.item.initial];
        };
        TuiSheetComponent.prototype.scrollTo = function (top) {
            if (top === void 0) { top = this.sheetTop; }
            var nativeElement = this.el.nativeElement;
            if (this.isIos) {
                var offset = top - nativeElement.scrollTop - 16;
                nativeElement.style.transition = 'none';
                nativeElement.style.transform = "scaleX(-1) translate3d(0, " + offset + "px, 0)";
                i1$1.timer(0)
                    .pipe(operators.takeUntil(this.destroy$))
                    .subscribe(function () {
                    nativeElement.style.transition = '';
                    nativeElement.style.transform = '';
                });
            }
            nativeElement.scrollTo({ top: top, behavior: 'smooth' });
        };
        TuiSheetComponent.prototype.close = function () {
            if (this.context.closeable) {
                this.context.$implicit.complete();
            }
        };
        Object.defineProperty(TuiSheetComponent.prototype, "contentTop", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.content) === null || _a === void 0 ? void 0 : _a.nativeElement.offsetTop) !== null && _b !== void 0 ? _b : Infinity;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetComponent.prototype, "sheetTop", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.sheet) === null || _a === void 0 ? void 0 : _a.nativeElement.offsetTop) !== null && _b !== void 0 ? _b : Infinity;
            },
            enumerable: false,
            configurable: true
        });
        TuiSheetComponent.prototype.getStops = function (stops) {
            return stops.map(function (_c) {
                var nativeElement = _c.nativeElement;
                return nativeElement.offsetTop + nativeElement.clientHeight;
            });
        };
        return TuiSheetComponent;
    }());
    TuiSheetComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetComponent, deps: [{ token: TUI_SHEET_SCROLL }, { token: i0.ElementRef }, { token: i0.NgZone }, { token: i3.TUI_IS_IOS }, { token: kit.TUI_MORE_WORD }, { token: i3.TuiDestroyService, self: true }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSheetComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetComponent, selector: "tui-sheet", inputs: { item: "item" }, host: { attributes: { "role": "dialog" }, listeners: { "$.class._stuck": "stuck$", "tui-sheet-id": "onId($event.detail)" }, properties: { "attr.aria-labelledby": "id", "class._ios": "isIos", "$.class._stuck": "stuck$" } }, providers: TUI_SHEET_PROVIDERS, viewQueries: [{ propertyName: "sheet", first: true, predicate: ["sheet"], descendants: true }, { propertyName: "content", first: true, predicate: ["content"], descendants: true }, { propertyName: "stopsRefs", predicate: ["stops"], descendants: true }], ngImport: i0__namespace, template: "<div\n    class=\"t-bumpers\"\n    (click)=\"close()\"\n>\n    <div\n        *ngFor=\"let stop of item.stops\"\n        #stops\n        class=\"t-bumper\"\n        [style.marginTop]=\"stop\"\n    ></div>\n</div>\n<div\n    #sheet\n    class=\"t-wrapper\"\n    [class.t-wrapper_shadow]=\"!item.image\"\n>\n    <header\n        *ngIf=\"item.image\"\n        class=\"t-top\"\n        [tuiSheetTop]=\"imageStop\"\n    >\n        <img\n            *polymorpheusOutlet=\"item.image as src; context: context\"\n            alt=\"\"\n            class=\"t-image\"\n            [src]=\"src\"\n        />\n    </header>\n    <section\n        #content\n        tuiSheetStop\n        class=\"t-sheet\"\n    >\n        <div class=\"t-bar\"></div>\n        <tui-sheet-bar>\n            <button\n                type=\"button\"\n                class=\"t-button\"\n                [title]=\"moreWord$ | async\"\n                (click)=\"scrollTo(stops[1])\"\n            ></button>\n        </tui-sheet-bar>\n        <div class=\"t-content\">\n            <h2\n                *polymorpheusOutlet=\"item.content as text; context: context\"\n                tuiSheetHeading\n                class=\"t-heading\"\n            >\n                {{ text }}\n            </h2>\n        </div>\n    </section>\n</div>\n", styles: [":host{scrollbar-width:none;-ms-overflow-style:none;position:absolute;top:0;left:0;bottom:0;right:0;transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;top:auto;border-radius:.75rem .75rem 0 0;overflow-y:auto;overflow-x:hidden;scroll-snap-type:y mandatory;box-shadow:0 50vh var(--tui-elevation-01);padding-right:1rem;margin-left:-1rem;transform:scaleX(-1);-webkit-clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0);clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0)}:host::-webkit-scrollbar,:host::-webkit-scrollbar-thumb{display:none}:host._stuck{scroll-snap-type:none}@supports (-moz-appearance: none){:host{scroll-snap-type:none}}.t-bumpers{display:flex;height:100%}:host-context(.t-wrapper_closeable) .t-bumpers{scroll-snap-stop:always;scroll-snap-align:start;scroll-margin:-1px}.t-bumper{scroll-snap-stop:always;scroll-snap-align:start;height:1rem;width:1rem}.t-wrapper{border-radius:inherit;scroll-snap-stop:always;scroll-snap-align:start}.t-wrapper_shadow{box-shadow:var(--tui-shadow)}.t-top{position:-webkit-sticky;position:sticky;top:0;border-radius:.8rem .8rem 0 0;box-shadow:var(--tui-shadow);transform:scaleX(-1);overflow:hidden}.t-top._clickthrough{pointer-events:none}:host-context(._overlay:not(._visible)) .t-top{transition-property:transform;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;transform:scaleX(-1)!important}.t-image{display:block;width:100%}.t-sheet{position:relative;border-radius:inherit;scroll-snap-stop:always;scroll-snap-align:start;transform:scaleX(-1)}.t-top:not(._rounded)~.t-sheet .t-bar{border-radius:0}.t-bar{transition-property:border-radius;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:-webkit-sticky;position:sticky;top:0;z-index:1;height:1.5rem;margin-bottom:-1.5rem;border-radius:inherit;background:var(--tui-elevation-01);box-shadow:inset 0 1px #ffffff40}.t-button{position:absolute;top:0;height:1.5rem;width:3rem;padding:0;border:0;opacity:0}.t-content{padding:1rem;margin-top:-1rem;border-radius:inherit;background:var(--tui-elevation-01)}.t-heading{padding-bottom:.5rem;background:var(--tui-elevation-01)}\n"], components: [{ type: TuiSheetBarComponent, selector: "tui-sheet-bar" }, { type: TuiSheetHeadingComponent, selector: "[tuiSheetHeading]" }], directives: [{ type: i2__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: TuiSheetTopDirective, selector: "[tuiSheetTop]", inputs: ["tuiSheetTop"] }, { type: i5__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: TuiSheetStopDirective, selector: "[tuiSheetStop]" }], pipes: { "async": i2__namespace.AsyncPipe }, animations: [i1.tuiSlideInTop], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        i3.tuiPure
    ], TuiSheetComponent.prototype, "context", null);
    __decorate([
        i3.tuiPure
    ], TuiSheetComponent.prototype, "getStops", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-sheet',
                        templateUrl: './sheet.template.html',
                        styleUrls: ['./sheet.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: TUI_SHEET_PROVIDERS,
                        host: {
                            role: 'dialog',
                            '[attr.aria-labelledby]': 'id',
                            '[class._ios]': 'isIos',
                            // '[class._stuck]': 'true', // Initially disable snapping for Firefox
                            '[$.class._stuck]': 'stuck$',
                            '($.class._stuck)': 'stuck$',
                        },
                        animations: [i1.tuiSlideInTop],
                    }]
            }], ctorParameters: function () {
            return [{ type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_SCROLL]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_IS_IOS]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [kit.TUI_MORE_WORD]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }];
        }, propDecorators: { sheet: [{
                    type: i0.ViewChild,
                    args: ['sheet']
                }], content: [{
                    type: i0.ViewChild,
                    args: ['content']
                }], stopsRefs: [{
                    type: i0.ViewChildren,
                    args: ['stops']
                }], item: [{
                    type: i0.Input
                }], context: [], onId: [{
                    type: i0.HostListener,
                    args: [TUI_SHEET_ID, ['$event.detail']]
                }], getStops: [] } });

    /**
     * @deprecated: drop in v4.0
     */
    var TUI_SHEET_DEFAULT_OPTIONS = {
        image: '',
        imageSlide: true,
        stops: [],
        initial: 0,
        offset: 16,
        closeable: true,
        overlay: false,
    };
    /**
     * @deprecated: drop in v4.0
     * Default parameters for sheet component
     */
    var TUI_SHEET_OPTIONS = i3.tuiCreateToken(Object.assign({}, TUI_SHEET_DEFAULT_OPTIONS));

    /**
     * @deprecated: use {@link TuiSheetDialogService}
     * drop in v4.0
     */
    var TuiSheetService = /** @class */ (function () {
        function TuiSheetService(doc, options) {
            this.doc = doc;
            this.options = options;
            this.sheets$ = new i1$1.BehaviorSubject([]);
        }
        TuiSheetService.prototype.open = function (content, options) {
            var _this = this;
            if (options === void 0) { options = {}; }
            return new i1$1.Observable(function ($implicit) {
                var completeWith = function (result) {
                    $implicit.next(result);
                    $implicit.complete();
                };
                var sheet = Object.assign(Object.assign(Object.assign({}, _this.options), options), { content: content, completeWith: completeWith, $implicit: $implicit, scroll$: i1$1.EMPTY });
                var focused = i3.tuiGetNativeFocused(_this.doc);
                if (i3.tuiIsHTMLElement(focused) && sheet.overlay) {
                    focused.blur();
                }
                _this.sheets$.next(__spreadArray(__spreadArray([], __read(_this.sheets$.value)), [sheet]));
                return function () {
                    _this.sheets$.next(_this.sheets$.value.filter(function (item) { return item !== sheet; }));
                };
            });
        };
        return TuiSheetService;
    }());
    TuiSheetService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetService, deps: [{ token: i2.DOCUMENT }, { token: TUI_SHEET_OPTIONS }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiSheetService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetService, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }], ctorParameters: function () {
            return [{ type: Document, decorators: [{
                            type: i0.Inject,
                            args: [i2.DOCUMENT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_OPTIONS]
                        }] }];
        } });

    // Safety offset for shadow
    var OFFSET = 16;
    function processDragged(dragged$, scroll$) {
        var touchstart$ = dragged$.pipe(operators.filter(Boolean));
        var touchend$ = dragged$.pipe(operators.filter(i3.tuiIsFalsy));
        var race$ = i1$1.race(scroll$, i1$1.timer(100)).pipe(operators.debounceTime(200), operators.take(1), operators.map(i3.ALWAYS_FALSE_HANDLER));
        return touchstart$.pipe(operators.switchMap(function () { return touchend$.pipe(operators.switchMap(function () { return race$; }), operators.startWith(true)); }), operators.startWith(false));
    }
    var TuiSheetWrapperDirective = /** @class */ (function () {
        function TuiSheetWrapperDirective(zone, win) {
            this.zone = zone;
            this.win = win;
            this.tuiSheetWrapper = 16;
        }
        Object.defineProperty(TuiSheetWrapperDirective.prototype, "overlay$", {
            get: function () {
                var _this = this;
                return this.scroll$.pipe(operators.map(function (y) { return y + 16 > _this.win.innerHeight - _this.tuiSheetWrapper; }), operators.distinctUntilChanged(), i3.tuiZonefull(this.zone));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetWrapperDirective.prototype, "visible$", {
            get: function () {
                return processDragged(this.dragged$, this.scroll$).pipe(operators.distinctUntilChanged(), i3.tuiZonefull(this.zone));
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetWrapperDirective.prototype, "height$", {
            get: function () {
                return this.scroll$.pipe(operators.map(this.getHeight.bind(this)));
            },
            enumerable: false,
            configurable: true
        });
        TuiSheetWrapperDirective.prototype.getHeight = function (value) {
            var _a;
            return ((_a = this.sheet) === null || _a === void 0 ? void 0 : _a.context.overlay)
                ? null
                : i3.tuiClamp(this.withImage(value) + OFFSET, OFFSET, this.win.innerHeight);
        };
        TuiSheetWrapperDirective.prototype.withImage = function (value) {
            var _a;
            return !((_a = this.sheet) === null || _a === void 0 ? void 0 : _a.imageStop) || Math.floor(value) > this.sheet.imageStop
                ? value
                : value - this.sheet.imageHeight;
        };
        return TuiSheetWrapperDirective;
    }());
    TuiSheetWrapperDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetWrapperDirective, deps: [{ token: i0.NgZone }, { token: common.WINDOW }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetWrapperDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetWrapperDirective, selector: "[tuiSheetWrapper]", inputs: { tuiSheetWrapper: "tuiSheetWrapper" }, queries: [{ propertyName: "sheet", first: true, predicate: TuiSheetComponent, descendants: true }, { propertyName: "dragged$", first: true, predicate: TuiSheetComponent, descendants: true, read: TUI_SHEET_DRAGGED, static: true }, { propertyName: "scroll$", first: true, predicate: TuiSheetComponent, descendants: true, read: TUI_SHEET_SCROLL, static: true }], exportAs: ["tuiSheetWrapper"], ngImport: i0__namespace });
    __decorate([
        i3.tuiPure
    ], TuiSheetWrapperDirective.prototype, "overlay$", null);
    __decorate([
        i3.tuiPure
    ], TuiSheetWrapperDirective.prototype, "visible$", null);
    __decorate([
        i3.tuiPure
    ], TuiSheetWrapperDirective.prototype, "height$", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetWrapperDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiSheetWrapper]',
                        exportAs: 'tuiSheetWrapper',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }];
        }, propDecorators: { sheet: [{
                    type: i0.ContentChild,
                    args: [TuiSheetComponent]
                }], dragged$: [{
                    type: i0.ContentChild,
                    args: [TuiSheetComponent, { read: TUI_SHEET_DRAGGED, static: true }]
                }], scroll$: [{
                    type: i0.ContentChild,
                    args: [TuiSheetComponent, { read: TUI_SHEET_SCROLL, static: true }]
                }], tuiSheetWrapper: [{
                    type: i0.Input
                }], overlay$: [], visible$: [], height$: [] } });

    var TuiSheetCloseDirective = /** @class */ (function () {
        function TuiSheetCloseDirective(zone, dragged$, scroll$, win, el, sheet) {
            var _this = this;
            this.zone = zone;
            this.dragged$ = dragged$;
            this.scroll$ = scroll$;
            this.win = win;
            this.el = el;
            this.sheet = sheet;
            this.close = i1$1.merge(i3.tuiTypedFromEvent(this.el.nativeElement, TUI_SHEET_CLOSE), this.dragged$.pipe(i3.tuiIfMap(function () { return _this.scroll$.pipe(operators.startWith(_this.el.nativeElement.scrollTop)); }, i3.tuiIsFalsy), operators.filter(function (y) { var _a; return ((_a = _this.sheet.item) === null || _a === void 0 ? void 0 : _a.closeable) && _this.shouldClose(y); }), operators.distinctUntilChanged(), i3.tuiZonefull(this.zone)));
        }
        TuiSheetCloseDirective.prototype.shouldClose = function (scrollTop) {
            var height = Math.min(this.win.innerHeight, this.el.nativeElement.scrollHeight - this.win.innerHeight);
            var min = Math.min(height, this.sheet.stops[0] || Infinity);
            return scrollTop < min / 2;
        };
        return TuiSheetCloseDirective;
    }());
    TuiSheetCloseDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetCloseDirective, deps: [{ token: i0.NgZone }, { token: TUI_SHEET_DRAGGED }, { token: TUI_SHEET_SCROLL }, { token: common.WINDOW }, { token: i0.ElementRef }, { token: TuiSheetComponent }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetCloseDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetCloseDirective, selector: "tui-sheet[close]", outputs: { close: "close" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetCloseDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'tui-sheet[close]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_DRAGGED]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_SHEET_SCROLL]
                        }] }, { type: Window, decorators: [{
                            type: i0.Inject,
                            args: [common.WINDOW]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: TuiSheetComponent, decorators: [{
                            type: i0.Inject,
                            args: [TuiSheetComponent]
                        }] }];
        }, propDecorators: { close: [{
                    type: i0.Output
                }] } });

    var TuiSheetsHostComponent = /** @class */ (function () {
        function TuiSheetsHostComponent(options, service, height$, destroy$, cdr) {
            this.options = options;
            this.service = service;
            this.height$ = height$;
            this.destroy$ = destroy$;
            this.cdr = cdr;
            this.sheets = [];
        }
        TuiSheetsHostComponent.prototype.ngOnInit = function () {
            var _this = this;
            // Due to this view being parallel to app content, `markForCheck` from `async` pipe
            // can happen after view was checked, so calling `detectChanges` instead
            this.service.sheets$.pipe(operators.takeUntil(this.destroy$)).subscribe(function (sheets) {
                _this.sheets = sheets;
                _this.cdr.detectChanges();
            });
        };
        TuiSheetsHostComponent.prototype.close = function (_a) {
            var closeable = _a.closeable, $implicit = _a.$implicit;
            if (closeable) {
                $implicit.complete();
            }
        };
        return TuiSheetsHostComponent;
    }());
    TuiSheetsHostComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetsHostComponent, deps: [{ token: i1.TUI_ANIMATION_OPTIONS }, { token: TuiSheetService }, { token: i3.TUI_WINDOW_HEIGHT }, { token: i3.TuiDestroyService, self: true }, { token: i0.ChangeDetectorRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSheetsHostComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetsHostComponent, selector: "tui-sheets-host", providers: [i3.TuiDestroyService], ngImport: i0__namespace, template: "<div\n    *ngFor=\"let item of sheets\"\n    #wrapper=\"tuiSheetWrapper\"\n    class=\"t-wrapper\"\n    [@tuiFadeIn]=\"options\"\n    [@tuiSlideInTop]=\"options\"\n    [class.t-wrapper_closeable]=\"item.closeable\"\n    [class.t-wrapper_overlay]=\"item.overlay || (wrapper.overlay$ | async)\"\n    [class.t-wrapper_visible]=\"wrapper.visible$ | async\"\n    [style.height.px]=\"wrapper.height$ | async\"\n    [tuiSheetWrapper]=\"item.offset\"\n    (click.self)=\"close(item)\"\n>\n    <tui-sheet\n        tuiOverscroll=\"all\"\n        tuiScrollRef\n        [item]=\"item\"\n        [style.height.px]=\"((height$ | async) || 0) - item.offset\"\n        (close)=\"close(item)\"\n    ></tui-sheet>\n</div>\n<div\n    class=\"t-overlay\"\n    (click)=\"close(sheets[0])\"\n></div>\n", styles: [":host{position:-webkit-sticky;position:sticky;bottom:0;display:block;height:0}.t-overlay{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:fixed;top:0;left:0;width:100%;z-index:-1;height:100%;background:#000;opacity:.001;pointer-events:none}.t-wrapper_overlay~.t-overlay{opacity:.8;pointer-events:auto;touch-action:none}.t-wrapper{position:absolute;bottom:0;z-index:0;width:100%;height:100vh;overflow:hidden}.t-wrapper_visible{overflow:visible}\n"], components: [{ type: TuiSheetComponent, selector: "tui-sheet", inputs: ["item"] }], directives: [{ type: i2__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: TuiSheetWrapperDirective, selector: "[tuiSheetWrapper]", inputs: ["tuiSheetWrapper"], exportAs: ["tuiSheetWrapper"] }, { type: TuiSheetCloseDirective, selector: "tui-sheet[close]", outputs: ["close"] }, { type: i3__namespace.TuiOverscrollDirective, selector: "[tuiOverscroll]", inputs: ["tuiOverscroll"] }, { type: i1__namespace.TuiScrollRefDirective, selector: "[tuiScrollRef]" }], pipes: { "async": i2__namespace.AsyncPipe }, animations: [i1.tuiSlideInTop, i1.tuiFadeIn], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetsHostComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-sheets-host',
                        templateUrl: './sheets-host.template.html',
                        styleUrls: ['./sheets-host.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [i3.TuiDestroyService],
                        animations: [i1.tuiSlideInTop, i1.tuiFadeIn],
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_ANIMATION_OPTIONS]
                        }] }, { type: TuiSheetService, decorators: [{
                            type: i0.Inject,
                            args: [TuiSheetService]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i3.TUI_WINDOW_HEIGHT]
                        }] }, { type: i1__namespace$1.Observable, decorators: [{
                            type: i0.Self
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiDestroyService]
                        }] }, { type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }];
        } });

    /**
     * @deprecated: use {@link TuiSheetDialogDirective}
     * drop in v4.0
     */
    var TuiSheetDirective = /** @class */ (function (_super) {
        __extends(TuiSheetDirective, _super);
        function TuiSheetDirective(cdr, templateRef, service) {
            var _this = _super.call(this, templateRef, cdr) || this;
            _this.service = service;
            _this.open$ = new i1$1.Subject();
            _this.options = {};
            _this.tuiSheetChange = _this.open$.pipe(i3.tuiIfMap(function () { return _this.service.open(_this, _this.options).pipe(operators.ignoreElements(), operators.endWith(false)); }), operators.share());
            return _this;
        }
        Object.defineProperty(TuiSheetDirective.prototype, "tuiSheet", {
            set: function (open) {
                this.open$.next(open);
            },
            enumerable: false,
            configurable: true
        });
        return TuiSheetDirective;
    }(i5.PolymorpheusTemplate));
    TuiSheetDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDirective, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.TemplateRef }, { token: TuiSheetService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDirective, selector: "ng-template[tuiSheet]", inputs: { options: ["tuiSheetOptions", "options"], tuiSheet: "tuiSheet" }, outputs: { tuiSheetChange: "tuiSheetChange" }, usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'ng-template[tuiSheet]',
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ChangeDetectorRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ChangeDetectorRef]
                        }] }, { type: i0__namespace.TemplateRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.TemplateRef]
                        }] }, { type: TuiSheetService, decorators: [{
                            type: i0.Inject,
                            args: [TuiSheetService]
                        }] }];
        }, propDecorators: { options: [{
                    type: i0.Input,
                    args: ['tuiSheetOptions']
                }], tuiSheet: [{
                    type: i0.Input
                }], tuiSheetChange: [{
                    type: i0.Output
                }] } });

    /**
     * @deprecated: use {@link TuiSheetDialogModule}
     * drop in v4.0
     */
    var TuiSheetModule = /** @class */ (function () {
        function TuiSheetModule() {
        }
        return TuiSheetModule;
    }());
    TuiSheetModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSheetModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetModule, declarations: [TuiSheetsHostComponent,
            TuiSheetComponent,
            TuiSheetBarComponent,
            TuiSheetHeadingComponent,
            TuiSheetTopDirective,
            TuiSheetDirective,
            TuiSheetWrapperDirective,
            TuiSheetCloseDirective,
            TuiSheetStopDirective], imports: [i2.CommonModule,
            i5.PolymorpheusModule,
            i3.TuiLetModule,
            i1.TuiButtonModule,
            i3.TuiOverscrollModule,
            i1.TuiScrollbarModule], exports: [TuiSheetsHostComponent,
            TuiSheetComponent,
            TuiSheetHeadingComponent,
            TuiSheetDirective] });
    TuiSheetModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetModule, imports: [[
                i2.CommonModule,
                i5.PolymorpheusModule,
                i3.TuiLetModule,
                i1.TuiButtonModule,
                i3.TuiOverscrollModule,
                i1.TuiScrollbarModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i2.CommonModule,
                            i5.PolymorpheusModule,
                            i3.TuiLetModule,
                            i1.TuiButtonModule,
                            i3.TuiOverscrollModule,
                            i1.TuiScrollbarModule,
                        ],
                        declarations: [
                            TuiSheetsHostComponent,
                            TuiSheetComponent,
                            TuiSheetBarComponent,
                            TuiSheetHeadingComponent,
                            TuiSheetTopDirective,
                            TuiSheetDirective,
                            TuiSheetWrapperDirective,
                            TuiSheetCloseDirective,
                            TuiSheetStopDirective,
                        ],
                        exports: [
                            TuiSheetsHostComponent,
                            TuiSheetComponent,
                            TuiSheetHeadingComponent,
                            TuiSheetDirective,
                        ],
                        schemas: [i0.NO_ERRORS_SCHEMA],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_SHEET_CLOSE = TUI_SHEET_CLOSE;
    exports.TUI_SHEET_DEFAULT_OPTIONS = TUI_SHEET_DEFAULT_OPTIONS;
    exports.TUI_SHEET_ID = TUI_SHEET_ID;
    exports.TUI_SHEET_OPTIONS = TUI_SHEET_OPTIONS;
    exports.TUI_SHEET_PROVIDERS = TUI_SHEET_PROVIDERS;
    exports.TuiSheetBarComponent = TuiSheetBarComponent;
    exports.TuiSheetCloseDirective = TuiSheetCloseDirective;
    exports.TuiSheetComponent = TuiSheetComponent;
    exports.TuiSheetDirective = TuiSheetDirective;
    exports.TuiSheetHeadingComponent = TuiSheetHeadingComponent;
    exports.TuiSheetModule = TuiSheetModule;
    exports.TuiSheetService = TuiSheetService;
    exports.TuiSheetStopDirective = TuiSheetStopDirective;
    exports.TuiSheetTopDirective = TuiSheetTopDirective;
    exports.TuiSheetWrapperDirective = TuiSheetWrapperDirective;
    exports.TuiSheetsHostComponent = TuiSheetsHostComponent;
    exports.iosScrollFactory = iosScrollFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
