// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-directives-mobile-tabs.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/kit')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/directives/mobile-tabs', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/kit'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].directives = global["taiga-ui"]["addon-mobile"].directives || {}, global["taiga-ui"]["addon-mobile"].directives["mobile-tabs"] = {}), global.ng.core, global.i1, global.kit));
})(this, (function (exports, i0, i1, kit) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiMobileTabsComponent = /** @class */ (function () {
            function TuiMobileTabsComponent() {
            }
            return TuiMobileTabsComponent;
        }());
        TuiMobileTabsComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiMobileTabsComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileTabsComponent, selector: "ng-component", host: { classAttribute: "tui-mobile-tabs-styles" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["tui-tabs[tuiMobileTabs][data-platform=android],[tuiTabs][tuiMobileTabs][data-platform=android],[tuiMobileTabs] tui-tabs[data-platform=android],[tuiMobileTabs] [tuiTabs][data-platform=android]{height:auto}tui-tabs[tuiMobileTabs][data-platform=android] tui-underline,[tuiTabs][tuiMobileTabs][data-platform=android] tui-underline,[tuiMobileTabs] tui-tabs[data-platform=android] tui-underline,[tuiMobileTabs] [tuiTabs][data-platform=android] tui-underline{box-shadow:inset 0 -.1875rem var(--tui-link)}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab],[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab],[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab],[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;height:3rem;flex:1;flex-shrink:0;justify-content:center;text-transform:uppercase;color:var(--tui-base-09);opacity:.5;font-size:.875rem;font-weight:500;letter-spacing:-.008125rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab]:hover,[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab]:hover,[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab]:hover,[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]:hover{box-shadow:none}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab]._active,[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab]._active,[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab]._active,[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]._active{opacity:1;color:var(--tui-base-09)}tui-tabs[tuiMobileTabs][data-platform=ios],[tuiTabs][tuiMobileTabs][data-platform=ios],[tuiMobileTabs] tui-tabs[data-platform=ios],[tuiMobileTabs] [tuiTabs][data-platform=ios]{height:auto;border:2px solid transparent;border-radius:.5625rem;background:var(--tui-clear);box-shadow:none;overflow:visible}tui-tabs[tuiMobileTabs][data-platform=ios] tui-underline,[tuiTabs][tuiMobileTabs][data-platform=ios] tui-underline,[tuiMobileTabs] tui-tabs[data-platform=ios] tui-underline,[tuiMobileTabs] [tuiTabs][data-platform=ios] tui-underline{border-radius:.4375rem;background:var(--tui-base-01);box-shadow:0 .1875rem .0625rem #0000000a,0 .1875rem .5rem #0000001f}tui-tabs[tuiMobileTabs][data-platform=ios] [tuiTab],[tuiTabs][tuiMobileTabs][data-platform=ios] [tuiTab],[tuiMobileTabs] tui-tabs[data-platform=ios] [tuiTab],[tuiMobileTabs] [tuiTabs][data-platform=ios] [tuiTab]{height:1.75rem;flex:1;flex-shrink:0;justify-content:center;color:var(--tui-text-01);opacity:1;font-size:.8125rem;font-weight:600;letter-spacing:-.005rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}tui-tabs[tuiMobileTabs][data-platform=ios] [tuiTab]:hover,[tuiTabs][tuiMobileTabs][data-platform=ios] [tuiTab]:hover,[tuiMobileTabs] tui-tabs[data-platform=ios] [tuiTab]:hover,[tuiMobileTabs] [tuiTabs][data-platform=ios] [tuiTab]:hover{box-shadow:none}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            host: {
                                class: 'tui-mobile-tabs-styles',
                            },
                            template: '',
                            styleUrls: ['./mobile-tabs.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        }]
                }] });

        var TuiMobileTabsDirective = /** @class */ (function () {
            function TuiMobileTabsDirective(isIos, isAndroid, directiveStyles) {
                this.isIos = isIos;
                this.isAndroid = isAndroid;
                directiveStyles.addComponent(TuiMobileTabsComponent);
            }
            Object.defineProperty(TuiMobileTabsDirective.prototype, "platform", {
                // TODO: Refactor with TUI_PLATFORM
                get: function () {
                    if (this.isIos) {
                        return 'ios';
                    }
                    if (this.isAndroid) {
                        return 'android';
                    }
                    return null;
                },
                enumerable: false,
                configurable: true
            });
            return TuiMobileTabsDirective;
        }());
        TuiMobileTabsDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsDirective, deps: [{ token: i1.TUI_IS_IOS }, { token: i1.TUI_IS_ANDROID }, { token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiMobileTabsDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileTabsDirective, selector: "[tuiMobileTabs]", host: { properties: { "attr.data-platform": "this.platform" } }, providers: [
                {
                    provide: kit.TUI_TAB_MARGIN,
                    useValue: 0,
                },
            ], ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiMobileTabs]',
                            providers: [
                                {
                                    provide: kit.TUI_TAB_MARGIN,
                                    useValue: 0,
                                },
                            ],
                        }]
                }], ctorParameters: function () {
                return [{ type: undefined, decorators: [{
                                type: i0.Inject,
                                args: [i1.TUI_IS_IOS]
                            }] }, { type: undefined, decorators: [{
                                type: i0.Inject,
                                args: [i1.TUI_IS_ANDROID]
                            }] }, { type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { platform: [{
                        type: i0.HostBinding,
                        args: ['attr.data-platform']
                    }] } });

        var TuiMobileTabsModule = /** @class */ (function () {
            function TuiMobileTabsModule() {
            }
            return TuiMobileTabsModule;
        }());
        TuiMobileTabsModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiMobileTabsModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsModule, declarations: [TuiMobileTabsDirective, TuiMobileTabsComponent], exports: [TuiMobileTabsDirective, TuiMobileTabsComponent] });
        TuiMobileTabsModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileTabsModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiMobileTabsDirective, TuiMobileTabsComponent],
                            exports: [TuiMobileTabsDirective, TuiMobileTabsComponent],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiMobileTabsComponent = TuiMobileTabsComponent;
        exports.TuiMobileTabsDirective = TuiMobileTabsDirective;
        exports.TuiMobileTabsModule = TuiMobileTabsModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
