// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-sheet-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@tinkoff/ng-event-plugins'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/sheet-dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@tinkoff/ng-event-plugins', '@tinkoff/ng-polymorpheus', 'rxjs', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components["sheet-dialog"] = {}), global.ng.core, global.i3, global.i1, global.ngEventPlugins, global.i4, global.rxjs, global.ng.common));
})(this, (function (exports, i0, i3, i1, ngEventPlugins, i4, i5, i2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    // So we re-enter ngZone and trigger change detection
    function isCloseable() {
        return this.context.closeable;
    }
    var TuiSheetDialogComponent = /** @class */ (function () {
        function TuiSheetDialogComponent(el, duration, icons, closeWord$, context) {
            this.el = el;
            this.duration = duration;
            this.icons = icons;
            this.closeWord$ = closeWord$;
            this.context = context;
            this.stopsRefs = i3.EMPTY_QUERY;
            this.pointers = 0;
            this.slideInTop = {
                value: '',
                params: {
                    start: '100vh',
                    duration: this.duration,
                },
            };
            this.stuck$ = new i5.BehaviorSubject(false);
        }
        Object.defineProperty(TuiSheetDialogComponent.prototype, "offset", {
            get: function () {
                return this.context.offset;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetDialogComponent.prototype, "closeable", {
            get: function () {
                return this.context.closeable;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiSheetDialogComponent.prototype, "isSmall", {
            get: function () {
                var _a;
                return this.sheetTop > (((_a = this.sheet) === null || _a === void 0 ? void 0 : _a.nativeElement.clientHeight) || Infinity);
            },
            enumerable: false,
            configurable: true
        });
        TuiSheetDialogComponent.prototype.onPointerChange = function (delta) {
            this.pointers += delta;
            if (!delta) {
                var stuck = this.el.nativeElement.scrollTop > this.sheetTop;
                this.stuck$.value !== stuck && this.stuck$.next(stuck);
            }
            if (this.context.closeable &&
                !this.pointers &&
                !this.el.nativeElement.scrollTop) {
                this.close();
            }
        };
        TuiSheetDialogComponent.prototype.close = function () {
            // TODO: Refactor focus visible on mobile
            this.el.nativeElement.dispatchEvent(new Event('mousedown', { bubbles: true }));
            this.context.$implicit.complete();
        };
        TuiSheetDialogComponent.prototype.ngAfterViewInit = function () {
            this.el.nativeElement.scrollTop = __spreadArray(__spreadArray([], __read(this.getStops(this.stopsRefs))), [
                this.sheetTop,
            ])[this.context.initial];
        };
        Object.defineProperty(TuiSheetDialogComponent.prototype, "sheetTop", {
            get: function () {
                var _a, _b;
                return (_b = (_a = this.sheet) === null || _a === void 0 ? void 0 : _a.nativeElement.offsetTop) !== null && _b !== void 0 ? _b : Infinity;
            },
            enumerable: false,
            configurable: true
        });
        TuiSheetDialogComponent.prototype.getStops = function (stops) {
            return stops.map(function (_c) {
                var nativeElement = _c.nativeElement;
                return nativeElement.offsetTop + nativeElement.clientHeight;
            });
        };
        return TuiSheetDialogComponent;
    }());
    TuiSheetDialogComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogComponent, deps: [{ token: i0.ElementRef }, { token: i1.TUI_ANIMATIONS_DURATION }, { token: i1.TUI_COMMON_ICONS }, { token: i1.TUI_CLOSE_WORD }, { token: i4.POLYMORPHEUS_CONTEXT }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiSheetDialogComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDialogComponent, selector: "tui-sheet-dialog", host: { listeners: { "$.class._stuck": "stuck$", "document:touchstart.passive.silent": "onPointerChange(1)", "document:touchend.silent": "onPointerChange(-1)", "document:touchcancel.silent": "onPointerChange(-1)", "scroll.silent": "onPointerChange(0)" }, properties: { "$.class._stuck": "stuck$", "@tuiSlideInTop": "this.slideInTop", "style.top.px": "this.offset", "class._closeable": "this.closeable" } }, viewQueries: [{ propertyName: "sheet", first: true, predicate: ["sheet"], descendants: true }, { propertyName: "stopsRefs", predicate: ["stops"], descendants: true }], ngImport: i0__namespace, template: "<div class=\"t-stops\">\n    <div\n        *ngFor=\"let stop of context.stops\"\n        #stops\n        class=\"t-stop\"\n        [style.margin-top]=\"stop\"\n    ></div>\n</div>\n<div\n    #sheet\n    class=\"t-sheet\"\n    [class.t-sheet_small]=\"isSmall\"\n    (tuiClickOutside)=\"close()\"\n>\n    <div class=\"t-top\"></div>\n    <h2\n        *ngIf=\"context.label\"\n        class=\"t-heading\"\n        [id]=\"context.id\"\n    >\n        <ng-container *polymorpheusOutlet=\"context.label as label\">\n            {{ label }}\n        </ng-container>\n        <button\n            *ngIf=\"context.closeable\"\n            appearance=\"icon\"\n            size=\"xs\"\n            tuiIconButton\n            class=\"t-close\"\n            [icon]=\"icons.close\"\n            (click)=\"close()\"\n        >\n            {{ closeWord$ | async }}\n        </button>\n    </h2>\n    <div\n        *ngIf=\"context.content\"\n        class=\"t-content\"\n    >\n        <ng-container *polymorpheusOutlet=\"context.content as text; context: context\">\n            {{ text }}\n        </ng-container>\n    </div>\n</div>\n<div class=\"t-footer\"></div>\n", styles: [":host{position:fixed;top:0;left:0;bottom:0;right:0;right:-1rem;display:flex;flex-direction:column;-webkit-clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0);clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0);font:var(--tui-font-text-m);overflow-y:scroll;scroll-snap-type:y mandatory;overscroll-behavior:none;box-shadow:0 20rem var(--tui-elevation-01)}:host._closeable{display:block}:host._closeable .t-stops{display:flex}.t-stops{display:none;height:100%;scroll-snap-stop:always;scroll-snap-align:start}.t-stop{scroll-snap-stop:normal;scroll-snap-align:start;height:1rem;width:1rem}.t-sheet{box-shadow:var(--tui-shadow);width:calc(100% - 1rem);border-radius:.75rem .75rem 0 0;padding:0 1rem;margin-top:auto;background:var(--tui-elevation-01);box-sizing:border-box;scroll-snap-stop:always;scroll-snap-align:start}@supports (-moz-appearance: none){.t-sheet_small{scroll-snap-align:end}}.t-top{position:-webkit-sticky;position:sticky;top:0;z-index:1;height:1.5rem;background:var(--tui-elevation-01)}.t-top:after{content:\"\";position:absolute;top:.5rem;left:50%;width:2rem;height:.25rem;transform:translate(-50%);background:var(--tui-base-09);opacity:.3;border-radius:1rem}.t-heading{position:-webkit-sticky;position:sticky;top:1.5rem;z-index:1;display:flex;margin:0;padding-bottom:1rem;font:var(--tui-font-heading-6);background:var(--tui-elevation-01)}.t-heading:last-child{margin-bottom:-.75rem}.t-heading:after{content:\"\";position:absolute;top:100%;left:0;right:0;height:1px;background:var(--tui-base-03);opacity:0}:host._stuck .t-heading:after{opacity:1}.t-close{right:-.25rem;flex-shrink:0;margin-left:auto}.t-content{position:relative;isolation:isolate}.t-content:nth-child(3){margin-top:1rem}.t-footer{height:1rem;scroll-snap-stop:always;scroll-snap-align:end;background:var(--tui-elevation-01)}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i2__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i3__namespace.TuiClickOutsideDirective, selector: "[tuiClickOutside]", outputs: ["tuiClickOutside"] }, { type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], pipes: { "async": i2__namespace.AsyncPipe }, animations: [i1.tuiSlideInTop], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    __decorate([
        ngEventPlugins.shouldCall(isCloseable)
    ], TuiSheetDialogComponent.prototype, "close", null);
    __decorate([
        i3.tuiPure
    ], TuiSheetDialogComponent.prototype, "getStops", null);
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-sheet-dialog',
                        templateUrl: './sheet-dialog.template.html',
                        styleUrls: ['./sheet-dialog.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        animations: [i1.tuiSlideInTop],
                        host: {
                            '[$.class._stuck]': 'stuck$',
                            '($.class._stuck)': 'stuck$',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_ANIMATIONS_DURATION]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_COMMON_ICONS]
                        }] }, { type: i5__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1.TUI_CLOSE_WORD]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i4.POLYMORPHEUS_CONTEXT]
                        }] }];
        }, propDecorators: { sheet: [{
                    type: i0.ViewChild,
                    args: ['sheet']
                }], stopsRefs: [{
                    type: i0.ViewChildren,
                    args: ['stops']
                }], slideInTop: [{
                    type: i0.HostBinding,
                    args: ['@tuiSlideInTop']
                }], offset: [{
                    type: i0.HostBinding,
                    args: ['style.top.px']
                }], closeable: [{
                    type: i0.HostBinding,
                    args: ['class._closeable']
                }], onPointerChange: [{
                    type: i0.HostListener,
                    args: ['document:touchstart.passive.silent', ['1']]
                }, {
                    type: i0.HostListener,
                    args: ['document:touchend.silent', ['-1']]
                }, {
                    type: i0.HostListener,
                    args: ['document:touchcancel.silent', ['-1']]
                }, {
                    type: i0.HostListener,
                    args: ['scroll.silent', ['0']]
                }], close: [], getStops: [] } });

    var TUI_SHEET_DIALOG_DEFAULT_OPTIONS = {
        label: '',
        stops: [],
        initial: 0,
        offset: 16,
        closeable: true,
    };
    /**
     * Default parameters for mobile dialog component
     */
    var TUI_SHEET_DIALOG_OPTIONS = i3.tuiCreateToken(TUI_SHEET_DIALOG_DEFAULT_OPTIONS);
    function tuiSheetDialogOptionsProvider(options) {
        return i3.tuiProvideOptions(TUI_SHEET_DIALOG_OPTIONS, options, TUI_SHEET_DIALOG_DEFAULT_OPTIONS);
    }

    var DIALOG = new i4.PolymorpheusComponent(TuiSheetDialogComponent);
    var TuiSheetDialogService = /** @class */ (function (_super) {
        __extends(TuiSheetDialogService, _super);
        function TuiSheetDialogService() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.component = DIALOG;
            _this.defaultOptions = Object.assign(Object.assign({}, i0.inject(TUI_SHEET_DIALOG_OPTIONS)), { data: undefined });
            return _this;
        }
        return TuiSheetDialogService;
    }(i3.AbstractTuiDialogService));
    TuiSheetDialogService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogService, deps: null, target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiSheetDialogService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogService, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }] });

    var TuiSheetDialogDirective = /** @class */ (function (_super) {
        __extends(TuiSheetDialogDirective, _super);
        function TuiSheetDialogDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return TuiSheetDialogDirective;
    }(i3.AbstractTuiDialogDirective));
    TuiSheetDialogDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogDirective, deps: null, target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiSheetDialogDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDialogDirective, selector: "ng-template[tuiSheetDialog]", inputs: { options: ["tuiSheetDialogOptions", "options"], open: ["tuiSheetDialog", "open"] }, outputs: { openChange: "tuiSheetDialogChange" }, providers: [
            {
                provide: i3.AbstractTuiDialogService,
                useExisting: TuiSheetDialogService,
            },
        ], usesInheritance: true, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: 'ng-template[tuiSheetDialog]',
                        inputs: ['options: tuiSheetDialogOptions', 'open: tuiSheetDialog'],
                        outputs: ['openChange: tuiSheetDialogChange'],
                        providers: [
                            {
                                provide: i3.AbstractTuiDialogService,
                                useExisting: TuiSheetDialogService,
                            },
                        ],
                    }]
            }] });

    var TuiSheetDialogModule = /** @class */ (function () {
        function TuiSheetDialogModule() {
        }
        return TuiSheetDialogModule;
    }());
    TuiSheetDialogModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiSheetDialogModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogModule, declarations: [TuiSheetDialogComponent, TuiSheetDialogDirective], imports: [i2.CommonModule, i4.PolymorpheusModule, i3.TuiClickOutsideModule, i1.TuiButtonModule], exports: [TuiSheetDialogComponent, TuiSheetDialogDirective] });
    TuiSheetDialogModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogModule, providers: [i3.tuiAsDialog(TuiSheetDialogService)], imports: [[i2.CommonModule, i4.PolymorpheusModule, i3.TuiClickOutsideModule, i1.TuiButtonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiSheetDialogModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i2.CommonModule, i4.PolymorpheusModule, i3.TuiClickOutsideModule, i1.TuiButtonModule],
                        declarations: [TuiSheetDialogComponent, TuiSheetDialogDirective],
                        providers: [i3.tuiAsDialog(TuiSheetDialogService)],
                        exports: [TuiSheetDialogComponent, TuiSheetDialogDirective],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_SHEET_DIALOG_DEFAULT_OPTIONS = TUI_SHEET_DIALOG_DEFAULT_OPTIONS;
    exports.TUI_SHEET_DIALOG_OPTIONS = TUI_SHEET_DIALOG_OPTIONS;
    exports.TuiSheetDialogComponent = TuiSheetDialogComponent;
    exports.TuiSheetDialogDirective = TuiSheetDialogDirective;
    exports.TuiSheetDialogModule = TuiSheetDialogModule;
    exports.TuiSheetDialogService = TuiSheetDialogService;
    exports.tuiSheetDialogOptionsProvider = tuiSheetDialogOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
