// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-pull-to-refresh.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/cdk'), require('rxjs'), require('@angular/common'), require('rxjs/operators'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/pull-to-refresh', ['exports', '@angular/core', '@tinkoff/ng-polymorpheus', '@taiga-ui/cdk', 'rxjs', '@angular/common', 'rxjs/operators', '@taiga-ui/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components["pull-to-refresh"] = {}), global.ng.core, global.i1, global.i1$1, global.rxjs, global.ng.common, global.rxjs.operators, global.i1$2));
})(this, (function (exports, i0, i1, i1$1, i3, i2, operators, i1$2) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace$2 = /*#__PURE__*/_interopNamespace(i1);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1$1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace$1 = /*#__PURE__*/_interopNamespace(i1$2);

    var LOADED_STEP = 8;
    var ROTATE_X_STEP = 30;
    var TuiMobileLoaderIOSComponent = /** @class */ (function () {
        function TuiMobileLoaderIOSComponent(context, threshold) {
            this.context = context;
            this.threshold = threshold;
            this.steps = 12;
        }
        Object.defineProperty(TuiMobileLoaderIOSComponent.prototype, "finished", {
            get: function () {
                return this.percent >= 100;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileLoaderIOSComponent.prototype, "percent", {
            get: function () {
                return (this.context.$implicit * 100) / this.threshold;
            },
            enumerable: false,
            configurable: true
        });
        TuiMobileLoaderIOSComponent.prototype.isShown = function (index) {
            return this.percent > (index + 1) * LOADED_STEP;
        };
        TuiMobileLoaderIOSComponent.prototype.calculateTransform = function (index) {
            return "rotate(" + index * ROTATE_X_STEP + " 50 50)";
        };
        TuiMobileLoaderIOSComponent.prototype.calculateAnimationBegin = function (index) {
            return (index * LOADED_STEP) / 100 + "s";
        };
        return TuiMobileLoaderIOSComponent;
    }());
    TuiMobileLoaderIOSComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileLoaderIOSComponent, deps: [{ token: i1.POLYMORPHEUS_CONTEXT }, { token: TUI_PULL_TO_REFRESH_THRESHOLD }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiMobileLoaderIOSComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileLoaderIOSComponent, selector: "tui-mobile-ios-loader", ngImport: i0__namespace, template: "<svg\n    height=\"36\"\n    preserveAspectRatio=\"xMidYMid\"\n    viewBox=\"0 0 100 100\"\n    width=\"36\"\n    xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n>\n    <ng-container *tuiRepeatTimes=\"let index of steps\">\n        <g\n            *ngIf=\"isShown(index)\"\n            [attr.transform]=\"calculateTransform(index)\"\n        >\n            <rect\n                fill=\"#c7c9cc\"\n                height=\"16\"\n                rx=\"7.05\"\n                ry=\"3.3\"\n                width=\"6\"\n                x=\"47\"\n                y=\"22\"\n            >\n                <animate\n                    *ngIf=\"finished\"\n                    attributeName=\"opacity\"\n                    dur=\"1s\"\n                    keyTimes=\"0;1\"\n                    repeatCount=\"indefinite\"\n                    values=\"1;0\"\n                    [attr.begin]=\"calculateAnimationBegin(index)\"\n                ></animate>\n            </rect>\n        </g>\n    </ng-container>\n</svg>\n", styles: [":host{position:absolute;top:-.5rem;left:50%;margin-left:-1.125rem}\n"], directives: [{ type: i1__namespace.TuiRepeatTimesDirective, selector: "[tuiRepeatTimes][tuiRepeatTimesOf]", inputs: ["tuiRepeatTimesOf"] }, { type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileLoaderIOSComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-mobile-ios-loader',
                        templateUrl: './loader-ios.template.html',
                        styleUrls: ['./loader-ios.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.POLYMORPHEUS_CONTEXT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_THRESHOLD]
                        }] }];
        } });
    var TUI_IOS_LOADER = new i1.PolymorpheusComponent(TuiMobileLoaderIOSComponent);

    /**
     * Stream that emits when loading is over
     */
    var TUI_PULL_TO_REFRESH_LOADED = i1$1.tuiCreateToken(i3.EMPTY);
    /**
     * Pull threshold in pixels until loading starts
     */
    var TUI_PULL_TO_REFRESH_THRESHOLD = i1$1.tuiCreateToken(50);
    /**
     * Loading indicator component that gets current pull distance in pixels as context
     */
    var TUI_PULL_TO_REFRESH_COMPONENT = i1$1.tuiCreateTokenFromFactory(function () { return (i0.inject(i1$1.TUI_IS_IOS) ? TUI_IOS_LOADER : TUI_ANDROID_LOADER); });
    /**
     * @deprecated renamed to {@link TUI_PULL_TO_REFRESH_LOADED}
     */
    var TUI_LOADED = TUI_PULL_TO_REFRESH_LOADED;

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var MICRO_OFFSET = Math.pow(10, -6);
    var TuiPullToRefreshService = /** @class */ (function (_super) {
        __extends(TuiPullToRefreshService, _super);
        function TuiPullToRefreshService(el, scrollRef, loaded$, threshold, component) {
            var _this = _super.call(this, function (subscriber) { return (component ? _this.pulling$ : i3.EMPTY).subscribe(subscriber); }) || this;
            _this.el = el;
            _this.scrollRef = scrollRef;
            _this.loaded$ = loaded$;
            _this.threshold = threshold;
            // Hack for iOS to determine if pulling stopped due to scroll
            // because Safari does not support `touch-action: pan-down`
            _this.touched = false;
            _this.pulling$ = _this.loaded$.pipe(operators.startWith(null), operators.switchMap(function () { return i1$1.tuiTypedFromEvent(_this.element, 'touchstart', { passive: true }).pipe(operators.filter(function () { return !_this.scrollTop; }), operators.map(function (_a) {
                var touches = _a.touches;
                return touches[0].clientY;
            }), operators.switchMap(function (start) { return i1$1.tuiTypedFromEvent(_this.element, 'touchmove').pipe(operators.tap(function () {
                _this.touched = true;
            }), operators.map(function (_a) {
                var touches = _a.touches;
                return touches[0].clientY - start;
            }), operators.filter(function (distance) { return distance > 0; }), operators.takeUntil(i1$1.tuiTypedFromEvent(_this.element, 'touchend').pipe(operators.tap(function () {
                _this.touched = false;
            }))), operators.takeUntil(i1$1.tuiScrollFrom(_this.scrollRef.nativeElement)), operators.endWith(0)); }), operators.scan(function (prev, current) { return !current && !_this.touched && prev > _this.threshold
                ? _this.threshold
                : current + current * MICRO_OFFSET; }, 0), operators.takeWhile(function (distance) { return distance !== _this.threshold; }, true), operators.startWith(0)); }), operators.distinctUntilChanged(), operators.share());
            return _this;
        }
        Object.defineProperty(TuiPullToRefreshService.prototype, "element", {
            get: function () {
                return this.el.nativeElement;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPullToRefreshService.prototype, "scrollTop", {
            get: function () {
                return this.scrollRef.nativeElement.scrollTop;
            },
            enumerable: false,
            configurable: true
        });
        return TuiPullToRefreshService;
    }(i3.Observable));
    TuiPullToRefreshService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshService, deps: [{ token: i0.ElementRef }, { token: i1$1.TUI_SCROLL_REF }, { token: TUI_PULL_TO_REFRESH_LOADED }, { token: TUI_PULL_TO_REFRESH_THRESHOLD }, { token: TUI_PULL_TO_REFRESH_COMPONENT }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiPullToRefreshService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshService });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshService, decorators: [{
                type: i0.Injectable
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TUI_SCROLL_REF]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_LOADED]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_THRESHOLD]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_COMPONENT]
                        }] }];
        } });

    var ROTATE_X_DEFAULT = 180;
    var ROTATE_X_MAX = 500;
    var ROTATE_X_MULTIPLIER = 2.3;
    var TuiMobileLoaderAndroidComponent = /** @class */ (function () {
        function TuiMobileLoaderAndroidComponent(context, threshold) {
            this.context = context;
            this.threshold = threshold;
        }
        Object.defineProperty(TuiMobileLoaderAndroidComponent.prototype, "transform", {
            get: function () {
                var rotateX = Math.min(ROTATE_X_DEFAULT + this.percent * ROTATE_X_MULTIPLIER, ROTATE_X_MAX);
                return "rotate(" + rotateX + " 0 0)";
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileLoaderAndroidComponent.prototype, "percent", {
            get: function () {
                return (this.context.$implicit * 100) / this.threshold;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileLoaderAndroidComponent.prototype, "opacity", {
            get: function () {
                return this.context.$implicit / (this.threshold * 1.5);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileLoaderAndroidComponent.prototype, "dropped", {
            get: function () {
                return (this.context.$implicit <= MICRO_OFFSET ||
                    this.context.$implicit === this.threshold);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileLoaderAndroidComponent.prototype, "hostTransform", {
            get: function () {
                return "translateY(" + Math.min(this.context.$implicit, this.threshold * 1.5) + "px)";
            },
            enumerable: false,
            configurable: true
        });
        return TuiMobileLoaderAndroidComponent;
    }());
    TuiMobileLoaderAndroidComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileLoaderAndroidComponent, deps: [{ token: i1.POLYMORPHEUS_CONTEXT }, { token: TUI_PULL_TO_REFRESH_THRESHOLD }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiMobileLoaderAndroidComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileLoaderAndroidComponent, selector: "tui-mobile-android-loader", host: { properties: { "class._visible": "this.percent", "class._dropped": "this.dropped", "style.transform": "this.hostTransform" } }, ngImport: i0__namespace, template: "<div class=\"t-wrapper\">\n    <svg\n        *ngIf=\"percent !== 100; else loading\"\n        fill=\"none\"\n        height=\"24\"\n        viewBox=\"0 0 24 24\"\n        width=\"24\"\n        xmlns=\"http://www.w3.org/2000/svg\"\n        [attr.transform]=\"transform\"\n        [style.opacity]=\"opacity\"\n    >\n        <defs>\n            <mask id=\"mask-1\">\n                <path\n                    clip-rule=\"evenodd\"\n                    d=\"M21 12C21 7.03 16.97 3 12 3C7.03 3 3 7.03 3 12C3 16.97 7.03 21 12 21C14.06 21 15.96 20.3 17.48 19.14L16.06 17.7C14.91 18.51 13.51 19 12 19C8.13 19 5 15.87 5 12C5 8.13 8.13 5 12 5C15.87 5 19 8.13 19 12H16L20 16L24 12H21Z\"\n                    fill=\"white\"\n                    fill-rule=\"evenodd\"\n                />\n            </mask>\n        </defs>\n        <g mask=\"url(#mask-1)\">\n            <path\n                clip-rule=\"evenodd\"\n                d=\"M21 12C21 7.03 16.97 3 12 3C7.03 3 3 7.03 3 12C3 16.97 7.03 21 12 21C14.06 21 15.96 20.3 17.48 19.14L16.06 17.7C14.91 18.51 13.51 19 12 19C8.13 19 5 15.87 5 12C5 8.13 8.13 5 12 5C15.87 5 19 8.13 19 12H16L20 16L24 12H21Z\"\n                fill=\"black\"\n                fill-rule=\"evenodd\"\n            />\n        </g>\n    </svg>\n</div>\n<ng-template #loading>\n    <tui-loader\n        size=\"s\"\n        class=\"t-loader\"\n        [inheritColor]=\"true\"\n    ></tui-loader>\n</ng-template>\n", styles: [":host{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;position:-webkit-sticky;position:sticky;top:0;z-index:1;display:block;height:0;opacity:0}:host._visible{opacity:1}:host._dropped{transition:transform var(--tui-duration) cubic-bezier(.4,0,.2,1),opacity var(--tui-duration) var(--tui-duration)}.t-wrapper{position:absolute;top:-2.5rem;left:50%;display:flex;box-shadow:var(--tui-shadow-dropdown);height:2.25rem;width:2.25rem;background-color:var(--tui-secondary);border-radius:6.25rem;align-items:center;justify-content:center;margin-left:-1.125rem}.t-loader{color:var(--tui-text-01)}\n"], components: [{ type: i1__namespace$1.TuiLoaderComponent, selector: "tui-loader", inputs: ["size", "inheritColor", "overlay", "textContent", "showLoader"] }], directives: [{ type: i2__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileLoaderAndroidComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-mobile-android-loader',
                        templateUrl: './loader-android.template.html',
                        styleUrls: ['./loader-android.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1.POLYMORPHEUS_CONTEXT]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_THRESHOLD]
                        }] }];
        }, propDecorators: { percent: [{
                    type: i0.HostBinding,
                    args: ['class._visible']
                }], dropped: [{
                    type: i0.HostBinding,
                    args: ['class._dropped']
                }], hostTransform: [{
                    type: i0.HostBinding,
                    args: ['style.transform']
                }] } });
    var TUI_ANDROID_LOADER = new i1.PolymorpheusComponent(TuiMobileLoaderAndroidComponent);

    var TuiPullToRefreshComponent = /** @class */ (function () {
        function TuiPullToRefreshComponent(zone, destroy$, _a, isIOS, threshold, component, pulling$) {
            var _this = this;
            var nativeElement = _a.nativeElement;
            this.isIOS = isIOS;
            this.threshold = threshold;
            this.component = component;
            this.pulling$ = pulling$;
            this.styleHandler = this.isIOS
                ? function (distance) { return ({ top: i1$1.tuiPx(distance / 2) }); }
                : function () { return null; };
            this.pulled = this.pulling$.pipe(operators.filter(function (distance) { return distance === _this.threshold; }));
            this.dropped$ = this.pulling$.pipe(operators.map(function (distance) { return distance <= MICRO_OFFSET || distance === _this.threshold; }), operators.distinctUntilChanged());
            // Ensure scrolling down is impossible while pulling
            if (this.component) {
                i1$1.tuiScrollFrom(nativeElement)
                    .pipe(operators.startWith(null), i1$1.tuiZonefree(zone), operators.takeUntil(destroy$))
                    .subscribe(function () {
                    if (nativeElement.scrollTop) {
                        nativeElement.style.touchAction = '';
                    }
                    else {
                        nativeElement.style.touchAction = 'pan-down';
                    }
                });
            }
        }
        return TuiPullToRefreshComponent;
    }());
    TuiPullToRefreshComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshComponent, deps: [{ token: i0.NgZone }, { token: i1$1.TuiDestroyService, self: true }, { token: i1$1.TUI_SCROLL_REF }, { token: i1$1.TUI_IS_IOS }, { token: TUI_PULL_TO_REFRESH_THRESHOLD }, { token: TUI_PULL_TO_REFRESH_COMPONENT }, { token: TuiPullToRefreshService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPullToRefreshComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPullToRefreshComponent, selector: "tui-pull-to-refresh", inputs: { styleHandler: "styleHandler" }, outputs: { pulled: "pulled" }, providers: [TuiPullToRefreshService, i1$1.TuiDestroyService], ngImport: i0__namespace, template: "<ng-container *polymorpheusOutlet=\"component; context: {$implicit: (pulling$ | async) || 0}\"></ng-container>\n\n<div\n    [class.t-drop]=\"dropped$ | async\"\n    [ngStyle]=\"styleHandler((pulling$ | async) || 0)\"\n    [style.position]=\"'relative'\"\n>\n    <ng-content></ng-content>\n</div>\n", styles: [".t-drop{transition:all var(--tui-duration) cubic-bezier(.4,0,.2,1)}\n"], directives: [{ type: i1__namespace$2.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }, { type: i2__namespace.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], pipes: { "async": i2__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-pull-to-refresh',
                        templateUrl: './pull-to-refresh.template.html',
                        styleUrls: ['./pull-to-refresh.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        providers: [TuiPullToRefreshService, i1$1.TuiDestroyService],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.NgZone, decorators: [{
                            type: i0.Inject,
                            args: [i0.NgZone]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TuiDestroyService]
                        }, {
                            type: i0.Self
                        }] }, { type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TUI_SCROLL_REF]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i1$1.TUI_IS_IOS]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_THRESHOLD]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [TUI_PULL_TO_REFRESH_COMPONENT]
                        }] }, { type: i3__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [TuiPullToRefreshService]
                        }] }];
        }, propDecorators: { styleHandler: [{
                    type: i0.Input
                }], pulled: [{
                    type: i0.Output
                }] } });

    var TuiPullToRefreshModule = /** @class */ (function () {
        function TuiPullToRefreshModule() {
        }
        return TuiPullToRefreshModule;
    }());
    TuiPullToRefreshModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPullToRefreshModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshModule, declarations: [TuiPullToRefreshComponent,
            TuiMobileLoaderAndroidComponent,
            TuiMobileLoaderIOSComponent], imports: [i2.CommonModule,
            i1$2.TuiLoaderModule,
            i1$2.TuiSvgModule,
            i1$1.TuiRepeatTimesModule,
            i1.PolymorpheusModule], exports: [TuiPullToRefreshComponent] });
    TuiPullToRefreshModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshModule, imports: [[
                i2.CommonModule,
                i1$2.TuiLoaderModule,
                i1$2.TuiSvgModule,
                i1$1.TuiRepeatTimesModule,
                i1.PolymorpheusModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPullToRefreshModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i2.CommonModule,
                            i1$2.TuiLoaderModule,
                            i1$2.TuiSvgModule,
                            i1$1.TuiRepeatTimesModule,
                            i1.PolymorpheusModule,
                        ],
                        declarations: [
                            TuiPullToRefreshComponent,
                            TuiMobileLoaderAndroidComponent,
                            TuiMobileLoaderIOSComponent,
                        ],
                        exports: [TuiPullToRefreshComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.MICRO_OFFSET = MICRO_OFFSET;
    exports.TUI_ANDROID_LOADER = TUI_ANDROID_LOADER;
    exports.TUI_IOS_LOADER = TUI_IOS_LOADER;
    exports.TUI_LOADED = TUI_LOADED;
    exports.TUI_PULL_TO_REFRESH_COMPONENT = TUI_PULL_TO_REFRESH_COMPONENT;
    exports.TUI_PULL_TO_REFRESH_LOADED = TUI_PULL_TO_REFRESH_LOADED;
    exports.TUI_PULL_TO_REFRESH_THRESHOLD = TUI_PULL_TO_REFRESH_THRESHOLD;
    exports.TuiMobileLoaderAndroidComponent = TuiMobileLoaderAndroidComponent;
    exports.TuiMobileLoaderIOSComponent = TuiMobileLoaderIOSComponent;
    exports.TuiPullToRefreshComponent = TuiPullToRefreshComponent;
    exports.TuiPullToRefreshModule = TuiPullToRefreshModule;
    exports.TuiPullToRefreshService = TuiPullToRefreshService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
