// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-mobile-calendar-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/addon-mobile/components/mobile-calendar'), require('@taiga-ui/kit')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/mobile-calendar-dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@tinkoff/ng-polymorpheus', '@taiga-ui/addon-mobile/components/mobile-calendar', '@taiga-ui/kit'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components["mobile-calendar-dialog"] = {}), global.ng.core, global.cdk, global.ngPolymorpheus, global["taiga-ui"]["addon-mobile"].components["mobile-calendar"], global.kit));
})(this, (function (exports, i0, cdk, ngPolymorpheus, i1, kit) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var TuiMobileCalendarDialogComponent = /** @class */ (function () {
        function TuiMobileCalendarDialogComponent(context) {
            this.context = context;
        }
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "single", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.single) === true;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "multi", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.multi) === true;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "min", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.min) || cdk.TUI_FIRST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "max", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.max) || cdk.TUI_LAST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "disabledItemHandler", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.disabledItemHandler) || cdk.ALWAYS_FALSE_HANDLER;
            },
            enumerable: false,
            configurable: true
        });
        return TuiMobileCalendarDialogComponent;
    }());
    TuiMobileCalendarDialogComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogComponent, deps: [{ token: ngPolymorpheus.POLYMORPHEUS_CONTEXT }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiMobileCalendarDialogComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileCalendarDialogComponent, selector: "tui-mobile-calendar-dialog", ngImport: i0__namespace, template: "<tui-mobile-calendar\n    [disabledItemHandler]=\"disabledItemHandler\"\n    [max]=\"max\"\n    [min]=\"min\"\n    [multi]=\"multi\"\n    [single]=\"single\"\n    (cancel)=\"context.$implicit.complete()\"\n    (confirm)=\"context.completeWith($event)\"\n></tui-mobile-calendar>\n", styles: [":host{display:block;height:100vh;margin:-1.5rem -1.5rem -2rem}\n"], components: [{ type: i1__namespace.TuiMobileCalendarComponent, selector: "tui-mobile-calendar", inputs: ["single", "multi", "min", "max", "disabledItemHandler"], outputs: ["cancel", "confirm"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-mobile-calendar-dialog',
                        templateUrl: './mobile-calendar-dialog.template.html',
                        styleUrls: ['./mobile-calendar-dialog.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [ngPolymorpheus.POLYMORPHEUS_CONTEXT]
                        }] }];
        } });

    var TuiMobileCalendarDialogModule = /** @class */ (function () {
        function TuiMobileCalendarDialogModule() {
        }
        return TuiMobileCalendarDialogModule;
    }());
    TuiMobileCalendarDialogModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiMobileCalendarDialogModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, declarations: [TuiMobileCalendarDialogComponent], imports: [i1.TuiMobileCalendarModule], exports: [TuiMobileCalendarDialogComponent] });
    TuiMobileCalendarDialogModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, providers: [
            {
                provide: kit.TUI_MOBILE_CALENDAR,
                useValue: TuiMobileCalendarDialogComponent,
            },
        ], imports: [[i1.TuiMobileCalendarModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.TuiMobileCalendarModule],
                        declarations: [TuiMobileCalendarDialogComponent],
                        providers: [
                            {
                                provide: kit.TUI_MOBILE_CALENDAR,
                                useValue: TuiMobileCalendarDialogComponent,
                            },
                        ],
                        exports: [TuiMobileCalendarDialogComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiMobileCalendarDialogComponent = TuiMobileCalendarDialogComponent;
    exports.TuiMobileCalendarDialogModule = TuiMobileCalendarDialogModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
