// @taiga-ui/experimental/esm2015/directives/skeleton/skeleton.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { CHAR_NO_BREAK_SPACE, TuiDirectiveStylesService, tuiPure } from '@taiga-ui/cdk';
import { TUI_ANIMATIONS_DURATION } from '@taiga-ui/core';
import { TuiSkeletonComponent } from './skeleton.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
const FADE = [{ opacity: 0.06 }, { opacity: 1 }];
export class TuiSkeletonDirective {
    constructor(el, duration, directiveStyles) {
        this.el = el;
        this.duration = duration;
        this.tuiSkeleton = false;
        directiveStyles.addComponent(TuiSkeletonComponent);
    }
    getPlaceholder(value) {
        switch (typeof value) {
            case 'number':
                return Array.from({ length: value })
                    .map(() => CHAR_NO_BREAK_SPACE.repeat(getLength()))
                    .join(' ');
            case 'string':
                return value;
            default:
                return null;
        }
    }
    ngOnChanges({ tuiSkeleton }) {
        var _a;
        (_a = this.animation) === null || _a === void 0 ? void 0 : _a.cancel();
        if (!tuiSkeleton.currentValue && !tuiSkeleton.firstChange) {
            this.animation = this.el.nativeElement.animate(FADE, this.duration * 2);
        }
    }
}
TuiSkeletonDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSkeletonDirective, deps: [{ token: ElementRef }, { token: TUI_ANIMATIONS_DURATION }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiSkeletonDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSkeletonDirective, selector: "[tuiSkeleton]", inputs: { tuiSkeleton: "tuiSkeleton" }, host: { attributes: { "tuiSkeleton": "" }, properties: { "class._skeleton": "tuiSkeleton", "attr.data-tui-skeleton": "getPlaceholder(tuiSkeleton)" } }, usesOnChanges: true, ngImport: i0 });
__decorate([
    tuiPure
], TuiSkeletonDirective.prototype, "getPlaceholder", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSkeletonDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSkeleton]',
                    host: {
                        tuiSkeleton: '',
                        '[class._skeleton]': 'tuiSkeleton',
                        '[attr.data-tui-skeleton]': 'getPlaceholder(tuiSkeleton)',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATIONS_DURATION]
                }] }, { type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiSkeleton: [{
                type: Input
            }], getPlaceholder: [] } });
function getLength() {
    return Math.floor(Math.random() * (15 - 5 + 1)) + 5;
}
