// @taiga-ui/experimental/esm2015/components/swipe-action/swipe-actions-auto-close.directive.js
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { tuiGetActualTarget } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export class TuiSwipeActionsAutoCloseDirective {
    constructor(el) {
        this.el = el;
        this.autoClose = true;
    }
    handleEvent(event) {
        const target = tuiGetActualTarget(event);
        if (this.autoClose !== false && !this.el.nativeElement.contains(target)) {
            this.close();
        }
    }
    close() {
        this.el.nativeElement.scrollTo({
            left: 0,
            behavior: 'smooth',
        });
    }
}
TuiSwipeActionsAutoCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsAutoCloseDirective, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiSwipeActionsAutoCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSwipeActionsAutoCloseDirective, selector: "tui-swipe-actions[autoClose]", inputs: { autoClose: "autoClose" }, host: { listeners: { "document:pointerdown.silent": "handleEvent($event)", "document:focusin.silent": "handleEvent($event)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsAutoCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-swipe-actions[autoClose]',
                    host: {
                        '(document:pointerdown.silent)': 'handleEvent($event)',
                        '(document:focusin.silent)': 'handleEvent($event)',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { autoClose: [{
                type: Input
            }] } });
