// @taiga-ui/experimental/esm2015/components/swipe-action/swipe-actions.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiResizeModule } from '@taiga-ui/cdk';
import { TuiSwipeActionsComponent } from './swipe-actions.component';
import { TuiSwipeActionsAutoCloseDirective } from './swipe-actions-auto-close.directive';
import * as i0 from "@angular/core";
export class TuiSwipeActionsModule {
}
TuiSwipeActionsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSwipeActionsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsModule, declarations: [TuiSwipeActionsComponent, TuiSwipeActionsAutoCloseDirective], imports: [CommonModule, TuiResizeModule], exports: [TuiSwipeActionsComponent, TuiSwipeActionsAutoCloseDirective] });
TuiSwipeActionsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsModule, imports: [[CommonModule, TuiResizeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSwipeActionsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiResizeModule],
                    declarations: [TuiSwipeActionsComponent, TuiSwipeActionsAutoCloseDirective],
                    exports: [TuiSwipeActionsComponent, TuiSwipeActionsAutoCloseDirective],
                }]
        }] });
