// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-components-theme-ios.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { AbstractTuiThemeSwitcher } from '@taiga-ui/cdk';

class TuiThemeIosComponent extends AbstractTuiThemeSwitcher {
}
TuiThemeIosComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiThemeIosComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiThemeIosComponent, selector: "tui-theme-ios", usesInheritance: true, ngImport: i0, template: '', isInline: true, styles: ["html{touch-action:manipulation}body{overscroll-behavior:none}*{outline:none;scrollbar-width:none;-webkit-tap-highlight-color:transparent}input,textarea,[contenteditable] *{-webkit-user-select:auto!important;-moz-user-select:auto!important;user-select:auto!important}::selection{background-color:transparent}.tui-island{padding:1rem;border-radius:.75rem;box-shadow:0 .625rem 1.875rem #0003}.tui-island:before{display:none}a[tuiButton],a[tuiIconButton],button[tuiButton],button[tuiIconButton]{border-radius:.5rem;font-size:1.0625rem!important;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-theme-ios',
                    template: '',
                    styleUrls: ['./theme-ios.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiThemeIosModule {
}
TuiThemeIosModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiThemeIosModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosModule, declarations: [TuiThemeIosComponent], exports: [TuiThemeIosComponent] });
TuiThemeIosModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeIosModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiThemeIosComponent],
                    exports: [TuiThemeIosComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiThemeIosComponent, TuiThemeIosModule };
