// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-components-app-bar.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Inject, ViewChildren, Directive, Input, NgModule } from '@angular/core';
import { MutationObserverService, MUTATION_OBSERVER_INIT } from '@ng-web-apis/mutation-observer';
import { TuiDestroyService, TuiResizeService, EMPTY_QUERY } from '@taiga-ui/cdk';
import * as i2 from 'rxjs';
import { merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { tuiButtonOptionsProvider, TuiSvgModule } from '@taiga-ui/core';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

const TUI_APP_BAR_PROVIDERS = [
    TuiDestroyService,
    TuiResizeService,
    MutationObserverService,
    {
        provide: MUTATION_OBSERVER_INIT,
        useValue: {
            characterData: true,
            childList: true,
            subtree: true,
        },
    },
    tuiButtonOptionsProvider({
        size: 'm',
        appearance: '',
    }),
];

class TuiAppBarComponent {
    constructor(resize$, mutation$) {
        this.resize$ = resize$;
        this.mutation$ = mutation$;
        this.side = EMPTY_QUERY;
        this.width$ = merge(this.resize$, this.mutation$).pipe(map(() => {
            var _a, _b;
            return 2 *
                Math.max((_a = this.side.first) === null || _a === void 0 ? void 0 : _a.nativeElement.clientWidth, (_b = this.side.last) === null || _b === void 0 ? void 0 : _b.nativeElement.clientWidth);
        }));
    }
}
TuiAppBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, deps: [{ token: TuiResizeService }, { token: MutationObserverService }], target: i0.ɵɵFactoryTarget.Component });
TuiAppBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarComponent, selector: "tui-app-bar", providers: TUI_APP_BAR_PROVIDERS, viewQueries: [{ propertyName: "side", predicate: ["side"], descendants: true }], ngImport: i0, template: "<div\n    #side\n    class=\"t-left\"\n>\n    <ng-content select=\"[tuiSlot='left']\"></ng-content>\n</div>\n<div\n    class=\"t-content\"\n    [style.--sides.px]=\"width$ | async\"\n>\n    <ng-content></ng-content>\n</div>\n<div\n    #side\n    class=\"t-right\"\n>\n    <ng-content select=\"[tuiSlot='right']\"></ng-content>\n</div>\n", styles: [":host{--tui-padding: .625rem;position:relative;display:flex;height:3rem;align-items:center;justify-content:space-between;border:solid transparent;border-width:0 .375rem;box-sizing:border-box;font:var(--tui-font-text-m);font-weight:bold;color:var(--tui-text-01);background:var(--tui-base-01)}:host ::ng-deep [tuiButton][data-size=m]{font-weight:normal}.t-content{top:50%;left:50%;transform:translate(-50%,-50%);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;position:absolute;max-width:calc(100% - var(--sides, 0px));flex:1;padding:0 .375rem;box-sizing:border-box;text-align:center;color:var(--tui-text-01)}\n"], pipes: { "async": i1.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-app-bar',
                    templateUrl: './app-bar.template.html',
                    styleUrls: ['./app-bar.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: TUI_APP_BAR_PROVIDERS,
                }]
        }], ctorParameters: function () { return [{ type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }, { type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [MutationObserverService]
                }] }]; }, propDecorators: { side: [{
                type: ViewChildren,
                args: ['side']
            }] } });

class TuiAppBarDirective {
    constructor() {
        this.tuiSlot = 'left';
    }
}
TuiAppBarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiAppBarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarDirective, selector: "[tuiSlot]", inputs: { tuiSlot: "tuiSlot" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSlot]',
                }]
        }], propDecorators: { tuiSlot: [{
                type: Input
            }] } });

class TuiAppBarBackComponent {
}
TuiAppBarBackComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarBackComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAppBarBackComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppBarBackComponent, selector: "button[tuiAppBarBack], a[tuiAppBarBack]", ngImport: i0, template: "<svg\n    height=\"24\"\n    viewBox=\"0 0 24 24\"\n    width=\"24\"\n    xmlns=\"http://www.w3.org/2000/svg\"\n>\n    <path\n        d=\"M7.44025 12L14.9826 4.43872C15.5391 3.88083 15.5391 2.97631 14.9826 2.41842C14.4261 1.86053 13.5239 1.86053 12.9674 2.41842L4.41737 10.9898C3.86088 11.5477 3.86088 12.4523 4.41737 13.0102L12.9674 21.5816C13.5239 22.1395 14.4261 22.1395 14.9826 21.5816C15.5391 21.0237 15.5391 20.1192 14.9826 19.5613L7.44025 12Z\"\n        fill=\"currentColor\"\n    />\n</svg>\n<ng-content></ng-content>\n", styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;border:0;background:none;font-size:inherit;line-height:inherit;text-decoration:none;display:flex;align-items:center;padding:0 .625rem 0 .125rem}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarBackComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'button[tuiAppBarBack], a[tuiAppBarBack]',
                    templateUrl: './app-bar-back.template.html',
                    styleUrls: ['./app-bar-back.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiAppBarModule {
}
TuiAppBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAppBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent] });
TuiAppBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                    exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_APP_BAR_PROVIDERS, TuiAppBarBackComponent, TuiAppBarComponent, TuiAppBarDirective, TuiAppBarModule };
