// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-components-theme-android.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { AbstractTuiThemeSwitcher } from '@taiga-ui/cdk';

class TuiThemeAndroidComponent extends AbstractTuiThemeSwitcher {
}
TuiThemeAndroidComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiThemeAndroidComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiThemeAndroidComponent, selector: "tui-theme-android", usesInheritance: true, ngImport: i0, template: '', isInline: true, styles: ["body{overscroll-behavior:none}*{outline:none;scrollbar-width:none;-webkit-tap-highlight-color:transparent}input,textarea,[contenteditable] *{-webkit-user-select:auto!important;-moz-user-select:auto!important;user-select:auto!important}::selection{background-color:transparent}.tui-island{padding:1rem;border-radius:.125rem;box-shadow:0 0 .125rem #0003,0 .0625rem .125rem #0003}.tui-island:before{display:none}a[tuiButton],a[tuiIconButton],button[tuiButton],button[tuiIconButton]{font-size:.875rem!important;font-weight:500;text-transform:uppercase;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-theme-android',
                    template: '',
                    styleUrls: ['./theme-android.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiThemeAndroidModule {
}
TuiThemeAndroidModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiThemeAndroidModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidModule, declarations: [TuiThemeAndroidComponent], exports: [TuiThemeAndroidComponent] });
TuiThemeAndroidModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThemeAndroidModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiThemeAndroidComponent],
                    exports: [TuiThemeAndroidComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiThemeAndroidComponent, TuiThemeAndroidModule };
