// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-directives-mobile-tabs.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, HostBinding, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TUI_IS_IOS, TUI_IS_ANDROID, TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TUI_TAB_MARGIN } from '@taiga-ui/kit';

class TuiMobileTabsComponent {
}
TuiMobileTabsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiMobileTabsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileTabsComponent, selector: "ng-component", host: { classAttribute: "tui-mobile-tabs-styles" }, ngImport: i0, template: '', isInline: true, styles: ["tui-tabs[tuiMobileTabs][data-platform=android],[tuiTabs][tuiMobileTabs][data-platform=android],[tuiMobileTabs] tui-tabs[data-platform=android],[tuiMobileTabs] [tuiTabs][data-platform=android]{height:auto}tui-tabs[tuiMobileTabs][data-platform=android] tui-underline,[tuiTabs][tuiMobileTabs][data-platform=android] tui-underline,[tuiMobileTabs] tui-tabs[data-platform=android] tui-underline,[tuiMobileTabs] [tuiTabs][data-platform=android] tui-underline{box-shadow:inset 0 -.1875rem var(--tui-link)}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab],[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab],[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab],[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]{transition-property:opacity;transition-duration:var(--tui-duration, .3s);transition-timing-function:ease-in-out;height:3rem;flex:1;flex-shrink:0;justify-content:center;text-transform:uppercase;color:var(--tui-base-09);opacity:.5;font-size:.875rem;font-weight:500;letter-spacing:-.008125rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab]:hover,[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab]:hover,[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab]:hover,[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]:hover{box-shadow:none}tui-tabs[tuiMobileTabs][data-platform=android] [tuiTab]._active,[tuiTabs][tuiMobileTabs][data-platform=android] [tuiTab]._active,[tuiMobileTabs] tui-tabs[data-platform=android] [tuiTab]._active,[tuiMobileTabs] [tuiTabs][data-platform=android] [tuiTab]._active{opacity:1;color:var(--tui-base-09)}tui-tabs[tuiMobileTabs][data-platform=ios],[tuiTabs][tuiMobileTabs][data-platform=ios],[tuiMobileTabs] tui-tabs[data-platform=ios],[tuiMobileTabs] [tuiTabs][data-platform=ios]{height:auto;border:2px solid transparent;border-radius:.5625rem;background:var(--tui-clear);box-shadow:none;overflow:visible}tui-tabs[tuiMobileTabs][data-platform=ios] tui-underline,[tuiTabs][tuiMobileTabs][data-platform=ios] tui-underline,[tuiMobileTabs] tui-tabs[data-platform=ios] tui-underline,[tuiMobileTabs] [tuiTabs][data-platform=ios] tui-underline{border-radius:.4375rem;background:var(--tui-base-01);box-shadow:0 .1875rem .0625rem #0000000a,0 .1875rem .5rem #0000001f}tui-tabs[tuiMobileTabs][data-platform=ios] [tuiTab],[tuiTabs][tuiMobileTabs][data-platform=ios] [tuiTab],[tuiMobileTabs] tui-tabs[data-platform=ios] [tuiTab],[tuiMobileTabs] [tuiTabs][data-platform=ios] [tuiTab]{height:1.75rem;flex:1;flex-shrink:0;justify-content:center;color:var(--tui-text-01);opacity:1;font-size:.8125rem;font-weight:600;letter-spacing:-.005rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,\"Helvetica Neue\",sans-serif}tui-tabs[tuiMobileTabs][data-platform=ios] [tuiTab]:hover,[tuiTabs][tuiMobileTabs][data-platform=ios] [tuiTab]:hover,[tuiMobileTabs] tui-tabs[data-platform=ios] [tuiTab]:hover,[tuiMobileTabs] [tuiTabs][data-platform=ios] [tuiTab]:hover{box-shadow:none}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsComponent, decorators: [{
            type: Component,
            args: [{
                    host: {
                        class: 'tui-mobile-tabs-styles',
                    },
                    template: '',
                    styleUrls: ['./mobile-tabs.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });

class TuiMobileTabsDirective {
    constructor(isIos, isAndroid, directiveStyles) {
        this.isIos = isIos;
        this.isAndroid = isAndroid;
        directiveStyles.addComponent(TuiMobileTabsComponent);
    }
    // TODO: Refactor with TUI_PLATFORM
    get platform() {
        if (this.isIos) {
            return 'ios';
        }
        if (this.isAndroid) {
            return 'android';
        }
        return null;
    }
}
TuiMobileTabsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsDirective, deps: [{ token: TUI_IS_IOS }, { token: TUI_IS_ANDROID }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiMobileTabsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileTabsDirective, selector: "[tuiMobileTabs]", host: { properties: { "attr.data-platform": "this.platform" } }, providers: [
        {
            provide: TUI_TAB_MARGIN,
            useValue: 0,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiMobileTabs]',
                    providers: [
                        {
                            provide: TUI_TAB_MARGIN,
                            useValue: 0,
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_IOS]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_ANDROID]
                }] }, { type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { platform: [{
                type: HostBinding,
                args: ['attr.data-platform']
            }] } });

class TuiMobileTabsModule {
}
TuiMobileTabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMobileTabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsModule, declarations: [TuiMobileTabsDirective, TuiMobileTabsComponent], exports: [TuiMobileTabsDirective, TuiMobileTabsComponent] });
TuiMobileTabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileTabsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiMobileTabsDirective, TuiMobileTabsComponent],
                    exports: [TuiMobileTabsDirective, TuiMobileTabsComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMobileTabsComponent, TuiMobileTabsDirective, TuiMobileTabsModule };
