// @taiga-ui/core/fesm2015/taiga-ui-core-directives-number-format.js
import * as i0 from '@angular/core';
import { forwardRef, Directive, Inject, Input, NgModule } from '@angular/core';
import { TUI_NUMBER_FORMAT, TUI_NUMBER_FORMAT_OBSERVABLE } from '@taiga-ui/core/tokens';
import { BehaviorSubject } from 'rxjs';

class TuiNumberFormatDirective extends BehaviorSubject {
    constructor(settings) {
        super(Object.assign(Object.assign({}, settings), { decimalLimit: NaN }));
        this.settings = settings;
    }
    set tuiNumberFormat(format) {
        this.next(Object.assign(Object.assign(Object.assign({}, this.settings), { decimalLimit: NaN }), format));
    }
}
TuiNumberFormatDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatDirective, deps: [{ token: TUI_NUMBER_FORMAT }], target: i0.ɵɵFactoryTarget.Directive });
TuiNumberFormatDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiNumberFormatDirective, selector: "[tuiNumberFormat]", inputs: { tuiNumberFormat: "tuiNumberFormat" }, providers: [
        {
            provide: TUI_NUMBER_FORMAT_OBSERVABLE,
            useExisting: forwardRef(() => TuiNumberFormatDirective),
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiNumberFormat]',
                    providers: [
                        {
                            provide: TUI_NUMBER_FORMAT_OBSERVABLE,
                            useExisting: forwardRef(() => TuiNumberFormatDirective),
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_NUMBER_FORMAT]
                }] }]; }, propDecorators: { tuiNumberFormat: [{
                type: Input
            }] } });

class TuiNumberFormatModule {
}
TuiNumberFormatModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiNumberFormatModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatModule, declarations: [TuiNumberFormatDirective], exports: [TuiNumberFormatDirective] });
TuiNumberFormatModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNumberFormatModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiNumberFormatDirective],
                    exports: [TuiNumberFormatDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiNumberFormatDirective, TuiNumberFormatModule };
