// @taiga-ui/kit/esm2015/components/routable-dialog/routable-dialog.component.js
import { ChangeDetectionStrategy, Component, Inject, Injector, Self } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { TuiDialogService } from '@taiga-ui/core';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@taiga-ui/core";
import * as i3 from "@taiga-ui/cdk";
export class TuiRoutableDialogComponent {
    constructor(route, router, dialogs, injector, destroy$) {
        this.route = route;
        this.router = router;
        this.initialUrl = this.router.url;
        dialogs
            .open(new PolymorpheusComponent(this.route.snapshot.data['dialog'], injector), this.route.snapshot.data['dialogOptions'])
            .pipe(takeUntil(destroy$))
            .subscribe({ complete: () => this.onDialogClosing() });
    }
    get lazyLoadedBackUrl() {
        var _a;
        return (((_a = this.route.parent) === null || _a === void 0 ? void 0 : _a.snapshot.url) || []).map(() => '..').join('/');
    }
    onDialogClosing() {
        if (this.initialUrl === this.router.url) {
            this.navigateToParent();
        }
    }
    navigateToParent() {
        const backUrl = this.route.snapshot.data['isLazy']
            ? this.lazyLoadedBackUrl
            : this.route.snapshot.data['backUrl'];
        void this.router.navigate([backUrl], {
            relativeTo: this.route,
        });
    }
}
TuiRoutableDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogComponent, deps: [{ token: ActivatedRoute }, { token: Router }, { token: TuiDialogService }, { token: Injector }, { token: TuiDestroyService, self: true }], target: i0.ɵɵFactoryTarget.Component });
TuiRoutableDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiRoutableDialogComponent, selector: "tui-routable-dialog", providers: [TuiDestroyService], ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRoutableDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-routable-dialog',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i1.ActivatedRoute, decorators: [{
                    type: Inject,
                    args: [ActivatedRoute]
                }] }, { type: i1.Router, decorators: [{
                    type: Inject,
                    args: [Router]
                }] }, { type: i2.TuiDialogService, decorators: [{
                    type: Inject,
                    args: [TuiDialogService]
                }] }, { type: i0.Injector, decorators: [{
                    type: Inject,
                    args: [Injector]
                }] }, { type: i3.TuiDestroyService, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }]; } });
