// @taiga-ui/addon-mobile/esm2015/internal/primitive-calendar-mobile/primitive-calendar-mobile.module.js
import { NgModule } from '@angular/core';
import { TuiRippleModule } from '@taiga-ui/addon-mobile/directives/ripple';
import { TuiLetModule, TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiCalendarSheetPipeModule } from '@taiga-ui/core';
import { TuiPrimitiveCalendarMobileComponent } from './primitive-calendar-mobile.component';
import * as i0 from "@angular/core";
/**
 * @internal
 */
export class TuiPrimitiveCalendarMobileModule {
}
TuiPrimitiveCalendarMobileModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarMobileModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPrimitiveCalendarMobileModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarMobileModule, declarations: [TuiPrimitiveCalendarMobileComponent], imports: [TuiRepeatTimesModule,
        TuiCalendarSheetPipeModule,
        TuiRippleModule,
        TuiLetModule], exports: [TuiPrimitiveCalendarMobileComponent] });
TuiPrimitiveCalendarMobileModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarMobileModule, imports: [[
            TuiRepeatTimesModule,
            TuiCalendarSheetPipeModule,
            TuiRippleModule,
            TuiLetModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPrimitiveCalendarMobileModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        TuiRepeatTimesModule,
                        TuiCalendarSheetPipeModule,
                        TuiRippleModule,
                        TuiLetModule,
                    ],
                    declarations: [TuiPrimitiveCalendarMobileComponent],
                    exports: [TuiPrimitiveCalendarMobileComponent],
                }]
        }] });
