// @taiga-ui/kit/esm2015/components/radio-block/radio-block.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiWrapperModule } from '@taiga-ui/core';
import { TuiRadioModule } from '@taiga-ui/kit/components/radio';
import { TuiRadioBlockComponent } from './radio-block.component';
import * as i0 from "@angular/core";
export class TuiRadioBlockModule {
}
TuiRadioBlockModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioBlockModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiRadioBlockModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioBlockModule, declarations: [TuiRadioBlockComponent], imports: [CommonModule, FormsModule, TuiRadioModule, TuiWrapperModule], exports: [TuiRadioBlockComponent] });
TuiRadioBlockModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioBlockModule, imports: [[CommonModule, FormsModule, TuiRadioModule, TuiWrapperModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiRadioBlockModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, TuiRadioModule, TuiWrapperModule],
                    declarations: [TuiRadioBlockComponent],
                    exports: [TuiRadioBlockComponent],
                }]
        }] });
