// @taiga-ui/experimental/esm2015/components/app-bar/app-bar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiFadeModule } from '@taiga-ui/experimental/directives/fade';
import { TuiAppBarComponent } from './app-bar.component';
import { TuiAppBarDirective } from './app-bar.directive';
import { TuiAppBarBackComponent } from './app-bar-back.component';
import * as i0 from "@angular/core";
export class TuiAppBarModule {
}
TuiAppBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAppBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent], imports: [CommonModule, TuiSvgModule, TuiFadeModule], exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent] });
TuiAppBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, imports: [[CommonModule, TuiSvgModule, TuiFadeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiFadeModule],
                    declarations: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                    exports: [TuiAppBarComponent, TuiAppBarDirective, TuiAppBarBackComponent],
                }]
        }] });
