// @taiga-ui/experimental/esm2015/directives/surface/surface.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiSurfaceComponent } from './surface.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiSurfaceDirective {
    constructor(directiveStyles) {
        this.tuiSurface = '';
        directiveStyles.addComponent(TuiSurfaceComponent);
    }
}
TuiSurfaceDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSurfaceDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiSurfaceDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSurfaceDirective, selector: "[tuiSurface]", inputs: { tuiSurface: "tuiSurface" }, host: { attributes: { "tuiSurface": "" }, properties: { "attr.data-surface": "tuiSurface" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSurfaceDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSurface]',
                    host: {
                        tuiSurface: '',
                        '[attr.data-surface]': 'tuiSurface',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiSurface: [{
                type: Input
            }] } });
