// @taiga-ui/addon-doc/esm2015/components/demo/demo.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TuiResizerModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiDataListModule, TuiExpandModule, TuiGroupModule, TuiHintModule, TuiLinkModule, TuiModeModule, TuiSvgModule, TuiTextfieldControllerModule, TuiTooltipModule, } from '@taiga-ui/core';
import { TuiCheckboxLabeledModule, TuiDataListWrapperModule, TuiRadioBlockModule, TuiSelectModule, } from '@taiga-ui/kit';
import { TuiDocDemoComponent } from './demo.component';
import * as i0 from "@angular/core";
export class TuiDocDemoModule {
}
TuiDocDemoModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDemoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocDemoModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDemoModule, declarations: [TuiDocDemoComponent], imports: [CommonModule,
        ReactiveFormsModule,
        FormsModule,
        RouterModule,
        TuiLinkModule,
        TuiButtonModule,
        TuiSelectModule,
        TuiExpandModule,
        TuiGroupModule,
        TuiTooltipModule,
        TuiModeModule,
        TuiRadioBlockModule,
        TuiCheckboxLabeledModule,
        TuiSvgModule,
        TuiDataListWrapperModule,
        TuiDataListModule,
        TuiTextfieldControllerModule,
        TuiHintModule,
        TuiResizerModule], exports: [TuiDocDemoComponent] });
TuiDocDemoModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDemoModule, imports: [[
            CommonModule,
            ReactiveFormsModule,
            FormsModule,
            RouterModule,
            TuiLinkModule,
            TuiButtonModule,
            TuiSelectModule,
            TuiExpandModule,
            TuiGroupModule,
            TuiTooltipModule,
            TuiModeModule,
            TuiRadioBlockModule,
            TuiCheckboxLabeledModule,
            TuiSvgModule,
            TuiDataListWrapperModule,
            TuiDataListModule,
            TuiTextfieldControllerModule,
            TuiHintModule,
            TuiResizerModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocDemoModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        FormsModule,
                        RouterModule,
                        TuiLinkModule,
                        TuiButtonModule,
                        TuiSelectModule,
                        TuiExpandModule,
                        TuiGroupModule,
                        TuiTooltipModule,
                        TuiModeModule,
                        TuiRadioBlockModule,
                        TuiCheckboxLabeledModule,
                        TuiSvgModule,
                        TuiDataListWrapperModule,
                        TuiDataListModule,
                        TuiTextfieldControllerModule,
                        TuiHintModule,
                        TuiResizerModule,
                    ],
                    declarations: [TuiDocDemoComponent],
                    exports: [TuiDocDemoComponent],
                }]
        }] });
