// @taiga-ui/core/esm2015/services/breakpoint.service.js
import { Inject, Injectable, NgZone } from '@angular/core';
import { TUI_WINDOW_SIZE, tuiZoneOptimized } from '@taiga-ui/cdk';
import { TUI_MEDIA } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
import { distinctUntilChanged, map, shareReplay } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
/**
 * Service to provide the current breakpoint based on Taiga UI's media queries
 */
export class TuiBreakpointService extends Observable {
    constructor(media, ngZone, size$) {
        super(subscriber => this.stream$.subscribe(subscriber));
        this.media = media;
        this.ngZone = ngZone;
        this.size$ = size$;
        this.sorted = Object.values(this.media).sort((a, b) => a - b);
        this.invert = Object.keys(this.media).reduce((ret, key) => (Object.assign(Object.assign({}, ret), { [this.media[key]]: key })), {});
        this.stream$ = this.size$.pipe(map(({ width }) => this.sorted.find(size => size > width)), map(key => this.invert[key || this.sorted[this.sorted.length - 1]]), distinctUntilChanged(), tuiZoneOptimized(this.ngZone), shareReplay({ bufferSize: 1, refCount: true }));
    }
}
TuiBreakpointService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, deps: [{ token: TUI_MEDIA }, { token: NgZone }, { token: TUI_WINDOW_SIZE }], target: i0.ɵɵFactoryTarget.Injectable });
TuiBreakpointService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBreakpointService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_MEDIA]
                }] }, { type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: i1.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_WINDOW_SIZE]
                }] }]; } });
