// @taiga-ui/addon-commerce/esm2015/components/input-cvc/input-cvc.module.js
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiPrimitiveTextfieldModule, TuiTextfieldControllerModule } from '@taiga-ui/core';
import { TuiValueAccessorModule } from '@taiga-ui/kit';
import { TuiInputCVCComponent } from './input-cvc.component';
import * as i0 from "@angular/core";
export class TuiInputCVCModule {
}
TuiInputCVCModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCVCModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputCVCModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCVCModule, declarations: [TuiInputCVCComponent], imports: [MaskitoModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiValueAccessorModule], exports: [TuiInputCVCComponent] });
TuiInputCVCModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCVCModule, imports: [[
            MaskitoModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiValueAccessorModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCVCModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiValueAccessorModule,
                    ],
                    declarations: [TuiInputCVCComponent],
                    exports: [TuiInputCVCComponent],
                }]
        }] });
