// @taiga-ui/experimental/esm2015/components/chip/chip.module.js
import { NgModule } from '@angular/core';
import { TuiIconsDirective, TuiIconsModule } from '@taiga-ui/experimental/directives/icons';
import { TuiChipComponent } from './chip.component';
import { TuiChipDirective } from './chip.directive';
import * as i0 from "@angular/core";
export class TuiChipModule {
}
TuiChipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiChipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipModule, declarations: [TuiChipComponent, TuiChipDirective], imports: [TuiIconsModule], exports: [TuiChipDirective, TuiIconsDirective] });
TuiChipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipModule, imports: [[TuiIconsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiIconsModule],
                    declarations: [TuiChipComponent, TuiChipDirective],
                    exports: [TuiChipDirective, TuiIconsDirective],
                }]
        }] });
