// @taiga-ui/experimental/esm2015/components/chip/chip.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { tuiAvatarOptionsProvider } from '@taiga-ui/experimental/components/avatar';
import { tuiButtonOptionsProvider } from '@taiga-ui/experimental/components/button';
import { tuiCheckboxOptionsProvider } from '@taiga-ui/experimental/components/checkbox';
import { tuiToggleOptionsProvider } from '@taiga-ui/experimental/components/toggle';
import { TuiChipComponent } from './chip.component';
import { TUI_CHIP_OPTIONS } from './chip.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiChipDirective {
    constructor(options, directiveStyles) {
        this.options = options;
        this.size = this.options.size;
        this.appearance = this.options.appearance;
        directiveStyles.addComponent(TuiChipComponent);
    }
}
TuiChipDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipDirective, deps: [{ token: TUI_CHIP_OPTIONS }, { token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiChipDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiChipDirective, selector: "tui-chip,[tuiChip]", inputs: { size: "size", appearance: "appearance" }, host: { attributes: { "tuiAppearance": "" }, properties: { "attr.data-appearance": "appearance", "attr.data-size": "size" } }, providers: [
        tuiToggleOptionsProvider({ size: 's' }),
        tuiCheckboxOptionsProvider({ size: 's' }),
        tuiAvatarOptionsProvider({ size: 'xs' }),
        tuiButtonOptionsProvider({
            size: 'xs',
            appearance: 'icon',
        }),
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiChipDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-chip,[tuiChip]',
                    providers: [
                        tuiToggleOptionsProvider({ size: 's' }),
                        tuiCheckboxOptionsProvider({ size: 's' }),
                        tuiAvatarOptionsProvider({ size: 'xs' }),
                        tuiButtonOptionsProvider({
                            size: 'xs',
                            appearance: 'icon',
                        }),
                    ],
                    host: {
                        tuiAppearance: '',
                        '[attr.data-appearance]': 'appearance',
                        '[attr.data-size]': 'size',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_CHIP_OPTIONS]
                }] }, { type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { size: [{
                type: Input
            }], appearance: [{
                type: Input
            }] } });
