// @taiga-ui/cdk/esm2015/tokens/window-size.js
import { inject, InjectionToken } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { tuiTypedFromEvent } from '@taiga-ui/cdk/observables';
import { map, shareReplay, startWith } from 'rxjs/operators';
export const TUI_WINDOW_SIZE = new InjectionToken('[TUI_WINDOW_SIZE]', {
    factory: () => {
        const w = inject(WINDOW);
        return tuiTypedFromEvent(w, 'resize').pipe(startWith(null), map(() => {
            var _a, _b;
            const width = Math.max(w.document.documentElement.clientWidth || 0, w.innerWidth || 0, ((_a = w.visualViewport) === null || _a === void 0 ? void 0 : _a.width) || 0);
            const height = Math.max(w.document.documentElement.clientHeight || 0, w.innerHeight || 0, ((_b = w.visualViewport) === null || _b === void 0 ? void 0 : _b.height) || 0);
            return {
                width,
                height,
                top: 0,
                left: 0,
                right: width,
                bottom: height,
            };
        }), shareReplay({ bufferSize: 1, refCount: true }));
    },
});
