// @taiga-ui/kit/esm2015/components/files/files.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiFocusVisibleModule, TuiItemDirective, TuiItemModule, TuiPreventDefaultModule, } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiExpandModule, TuiGroupModule, TuiLoaderModule, TuiSvgModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiFileComponent } from './file/file.component';
import { TuiFilesComponent } from './files.component';
import * as i0 from "@angular/core";
export class TuiFilesModule {
}
TuiFilesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiFilesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilesModule, declarations: [TuiFilesComponent, TuiFileComponent], imports: [CommonModule,
        TuiGroupModule,
        TuiLoaderModule,
        PolymorpheusModule,
        TuiSvgModule,
        TuiFocusVisibleModule,
        TuiButtonModule,
        TuiItemModule,
        TuiExpandModule,
        TuiPreventDefaultModule], exports: [TuiFilesComponent, TuiFileComponent, TuiItemDirective] });
TuiFilesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilesModule, imports: [[
            CommonModule,
            TuiGroupModule,
            TuiLoaderModule,
            PolymorpheusModule,
            TuiSvgModule,
            TuiFocusVisibleModule,
            TuiButtonModule,
            TuiItemModule,
            TuiExpandModule,
            TuiPreventDefaultModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiFilesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiGroupModule,
                        TuiLoaderModule,
                        PolymorpheusModule,
                        TuiSvgModule,
                        TuiFocusVisibleModule,
                        TuiButtonModule,
                        TuiItemModule,
                        TuiExpandModule,
                        TuiPreventDefaultModule,
                    ],
                    declarations: [TuiFilesComponent, TuiFileComponent],
                    exports: [TuiFilesComponent, TuiFileComponent, TuiItemDirective],
                }]
        }] });
