// @taiga-ui/kit/esm2015/components/select/select.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { TuiDataListModule, TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiTextfieldComponent, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiArrowModule } from '@taiga-ui/kit/components/arrow';
import { TuiDataListWrapperModule } from '@taiga-ui/kit/components/data-list-wrapper';
import { TuiSelectOptionModule } from '@taiga-ui/kit/components/select-option';
import { TuiStringifyContentPipeModule } from '@taiga-ui/kit/pipes';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiNativeSelectComponent } from './native-select/native-select.component';
import { TuiNativeSelectGroupComponent } from './native-select/native-select-group.component';
import { TuiSelectComponent } from './select.component';
import { TuiSelectDirective } from './select.directive';
import * as i0 from "@angular/core";
export class TuiSelectModule {
}
TuiSelectModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiSelectModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectModule, declarations: [TuiSelectComponent,
        TuiSelectDirective,
        TuiNativeSelectComponent,
        TuiNativeSelectGroupComponent], imports: [CommonModule,
        PolymorpheusModule,
        TuiActiveZoneModule,
        TuiPrimitiveTextfieldModule,
        TuiHostedDropdownModule,
        TuiSelectOptionModule,
        TuiArrowModule,
        TuiWrapperModule,
        TuiTextfieldControllerModule,
        TuiDataListWrapperModule,
        TuiDataListModule,
        TuiStringifyContentPipeModule], exports: [TuiSelectComponent,
        TuiSelectDirective,
        TuiTextfieldComponent,
        TuiNativeSelectComponent,
        TuiNativeSelectGroupComponent] });
TuiSelectModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectModule, imports: [[
            CommonModule,
            PolymorpheusModule,
            TuiActiveZoneModule,
            TuiPrimitiveTextfieldModule,
            TuiHostedDropdownModule,
            TuiSelectOptionModule,
            TuiArrowModule,
            TuiWrapperModule,
            TuiTextfieldControllerModule,
            TuiDataListWrapperModule,
            TuiDataListModule,
            TuiStringifyContentPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PolymorpheusModule,
                        TuiActiveZoneModule,
                        TuiPrimitiveTextfieldModule,
                        TuiHostedDropdownModule,
                        TuiSelectOptionModule,
                        TuiArrowModule,
                        TuiWrapperModule,
                        TuiTextfieldControllerModule,
                        TuiDataListWrapperModule,
                        TuiDataListModule,
                        TuiStringifyContentPipeModule,
                    ],
                    declarations: [
                        TuiSelectComponent,
                        TuiSelectDirective,
                        TuiNativeSelectComponent,
                        TuiNativeSelectGroupComponent,
                    ],
                    exports: [
                        TuiSelectComponent,
                        TuiSelectDirective,
                        TuiTextfieldComponent,
                        TuiNativeSelectComponent,
                        TuiNativeSelectGroupComponent,
                    ],
                }]
        }] });
