// @taiga-ui/kit/esm2015/components/select/select.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost, tuiAsTextfieldHost } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiSelectDirective extends AbstractTuiTextfieldHost {
    get readOnly() {
        return true;
    }
    get value() {
        return this.host.computedValue;
    }
    get stringify() {
        return this.host.stringify;
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
}
TuiSelectDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiSelectDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSelectDirective, selector: "tui-select", providers: [tuiAsTextfieldHost(TuiSelectDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSelectDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-select',
                    providers: [tuiAsTextfieldHost(TuiSelectDirective)],
                }]
        }] });
