// @taiga-ui/addon-mobile/esm2015/components/mobile-calendar/mobile-calendar.providers.js
import { VIRTUAL_SCROLL_STRATEGY } from '@angular/cdk/scrolling';
import { ChangeDetectorRef, InjectionToken, Optional } from '@angular/core';
import { TUI_IS_IOS, TuiDestroyService, TuiScrollService, tuiWatch, } from '@taiga-ui/cdk';
import { TUI_CALENDAR_DATE_STREAM } from '@taiga-ui/kit';
import { EMPTY } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TuiMobileCalendarStrategy } from './mobile-calendar.strategy';
/**
 * Stream for updating value
 */
export const TUI_VALUE_STREAM = new InjectionToken('[TUI_VALUE_STREAM]');
export const TUI_MOBILE_CALENDAR_PROVIDERS = [
    TuiDestroyService,
    TuiScrollService,
    {
        provide: VIRTUAL_SCROLL_STRATEGY,
        deps: [TUI_IS_IOS, TuiScrollService],
        useClass: TuiMobileCalendarStrategy,
    },
    {
        provide: TUI_VALUE_STREAM,
        deps: [
            [new Optional(), TUI_CALENDAR_DATE_STREAM],
            TuiDestroyService,
            ChangeDetectorRef,
        ],
        useFactory: (value$, destroy$, cdr) => (value$ || EMPTY).pipe(tuiWatch(cdr), takeUntil(destroy$)),
    },
];
