// @taiga-ui/addon-mobile/esm2015/components/mobile-calendar/mobile-calendar.module.js
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRippleModule } from '@taiga-ui/addon-mobile/directives/ripple';
import { TuiTouchableModule } from '@taiga-ui/addon-mobile/directives/touchable';
import { TuiPrimitiveCalendarMobileModule } from '@taiga-ui/addon-mobile/internal/primitive-calendar-mobile';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiButtonModule, TuiLinkModule, TuiMonthPipeModule, TuiOrderWeekDaysPipeModule, } from '@taiga-ui/core';
import { TuiMobileCalendarComponent } from './mobile-calendar.component';
import * as i0 from "@angular/core";
export class TuiMobileCalendarModule {
}
TuiMobileCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMobileCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarModule, declarations: [TuiMobileCalendarComponent], imports: [CommonModule,
        ScrollingModule,
        TuiTouchableModule,
        TuiRippleModule,
        TuiMapperPipeModule,
        TuiLinkModule,
        TuiPrimitiveCalendarMobileModule,
        TuiButtonModule,
        TuiMonthPipeModule,
        TuiOrderWeekDaysPipeModule], exports: [TuiMobileCalendarComponent] });
TuiMobileCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarModule, imports: [[
            CommonModule,
            ScrollingModule,
            TuiTouchableModule,
            TuiRippleModule,
            TuiMapperPipeModule,
            TuiLinkModule,
            TuiPrimitiveCalendarMobileModule,
            TuiButtonModule,
            TuiMonthPipeModule,
            TuiOrderWeekDaysPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ScrollingModule,
                        TuiTouchableModule,
                        TuiRippleModule,
                        TuiMapperPipeModule,
                        TuiLinkModule,
                        TuiPrimitiveCalendarMobileModule,
                        TuiButtonModule,
                        TuiMonthPipeModule,
                        TuiOrderWeekDaysPipeModule,
                    ],
                    declarations: [TuiMobileCalendarComponent],
                    exports: [TuiMobileCalendarComponent],
                }]
        }] });
