// @taiga-ui/cdk/esm2015/components/dropdown-host/dropdown-host.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { AbstractTuiPortalHostComponent, AbstractTuiPortalService, } from '@taiga-ui/cdk/abstract';
import { TuiDropdownPortalService } from './dropdown-portal.service';
import * as i0 from "@angular/core";
/**
 * Host element for dynamically created portals, for example using {@link TuiDropdownDirective}.
 */
export class TuiDropdownHostComponent extends AbstractTuiPortalHostComponent {
}
TuiDropdownHostComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHostComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiDropdownHostComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownHostComponent, selector: "tui-dropdown-host", providers: [
        { provide: AbstractTuiPortalService, useExisting: TuiDropdownPortalService },
        // TODO: Remove in 4.0
        { provide: AbstractTuiPortalHostComponent, useExisting: TuiDropdownHostComponent },
    ], usesInheritance: true, ngImport: i0, template: "<ng-content></ng-content>\n<ng-container #viewContainer></ng-container>\n", styles: [":host{position:relative;z-index:0;display:block;height:100%}:host:before{content:\"\";display:block;overflow:hidden}.t-position-fixed-offset{position:fixed;left:0;top:0;pointer-events:none;visibility:hidden;width:100%;height:100%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownHostComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-dropdown-host',
                    templateUrl: './dropdown-host.template.html',
                    styleUrls: ['./dropdown-host.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        { provide: AbstractTuiPortalService, useExisting: TuiDropdownPortalService },
                        // TODO: Remove in 4.0
                        { provide: AbstractTuiPortalHostComponent, useExisting: TuiDropdownHostComponent },
                    ],
                }]
        }] });
