// @taiga-ui/experimental/esm2015/directives/appearance/appearance.directive.js
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiAppearanceDirective {
    constructor() {
        this.tuiAppearance = '';
        this.tuiAppearanceState = null;
        this.tuiAppearanceFocus = null;
    }
}
TuiAppearanceDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppearanceDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiAppearanceDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAppearanceDirective, selector: "[tuiAppearance]", inputs: { tuiAppearance: "tuiAppearance", tuiAppearanceState: "tuiAppearanceState", tuiAppearanceFocus: "tuiAppearanceFocus" }, host: { attributes: { "tuiAppearance": "" }, properties: { "attr.data-appearance": "tuiAppearance", "attr.data-state": "tuiAppearanceState", "attr.data-focus": "tuiAppearanceFocus" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAppearanceDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAppearance]',
                    host: {
                        tuiAppearance: '',
                        '[attr.data-appearance]': 'tuiAppearance',
                        '[attr.data-state]': 'tuiAppearanceState',
                        '[attr.data-focus]': 'tuiAppearanceFocus',
                    },
                }]
        }], propDecorators: { tuiAppearance: [{
                type: Input
            }], tuiAppearanceState: [{
                type: Input
            }], tuiAppearanceFocus: [{
                type: Input
            }] } });
