// @taiga-ui/addon-commerce/esm2015/components/input-card-grouped/input-card-grouped.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MaskitoModule } from '@maskito/angular';
import { TuiFormatCardModule } from '@taiga-ui/addon-commerce/pipes';
import { TuiActiveZoneModule, TuiAutoFocusModule, TuiFocusableModule, TuiLetModule, TuiMapperPipeModule, TuiPreventDefaultModule, TuiResizeModule, } from '@taiga-ui/cdk';
import { TuiDropdownModule, TuiSvgModule, TuiWrapperModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputCardGroupedComponent } from './input-card-grouped.component';
import * as i0 from "@angular/core";
export class TuiInputCardGroupedModule {
}
TuiInputCardGroupedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardGroupedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputCardGroupedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardGroupedModule, declarations: [TuiInputCardGroupedComponent], imports: [CommonModule,
        FormsModule,
        MaskitoModule,
        TuiFocusableModule,
        TuiSvgModule,
        TuiWrapperModule,
        TuiActiveZoneModule,
        TuiMapperPipeModule,
        TuiDropdownModule,
        TuiPreventDefaultModule,
        PolymorpheusModule,
        TuiLetModule,
        TuiResizeModule,
        TuiFormatCardModule,
        TuiAutoFocusModule], exports: [TuiInputCardGroupedComponent] });
TuiInputCardGroupedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardGroupedModule, imports: [[
            CommonModule,
            FormsModule,
            MaskitoModule,
            TuiFocusableModule,
            TuiSvgModule,
            TuiWrapperModule,
            TuiActiveZoneModule,
            TuiMapperPipeModule,
            TuiDropdownModule,
            TuiPreventDefaultModule,
            PolymorpheusModule,
            TuiLetModule,
            TuiResizeModule,
            TuiFormatCardModule,
            TuiAutoFocusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardGroupedModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        MaskitoModule,
                        TuiFocusableModule,
                        TuiSvgModule,
                        TuiWrapperModule,
                        TuiActiveZoneModule,
                        TuiMapperPipeModule,
                        TuiDropdownModule,
                        TuiPreventDefaultModule,
                        PolymorpheusModule,
                        TuiLetModule,
                        TuiResizeModule,
                        TuiFormatCardModule,
                        TuiAutoFocusModule,
                    ],
                    declarations: [TuiInputCardGroupedComponent],
                    exports: [TuiInputCardGroupedComponent],
                }]
        }] });
