// @ng-web-apis/mutation-observer/esm2015/directives/mutation-observer.directive.js
import { Attribute, Directive, ElementRef, EventEmitter, Inject, Output, } from '@angular/core';
import { MUTATION_OBSERVER_INIT } from '../tokens/mutation-observer-init';
import { mutationObserverInitFactory } from '../utils/mutation-observer-init-factory';
import * as i0 from "@angular/core";
export const SafeObserver = typeof MutationObserver !== `undefined`
    ? MutationObserver
    : class {
        observe() { }
        disconnect() { }
        takeRecords() {
            return [];
        }
    };
export class MutationObserverDirective extends SafeObserver {
    constructor({ nativeElement }, config, _1, _2, _3, _4, _5, _6, _7) {
        super(records => {
            this.waMutationObserver.emit(records);
        });
        this.waMutationObserver = new EventEmitter();
        this.observe(nativeElement, config);
    }
    ngOnDestroy() {
        this.disconnect();
    }
}
MutationObserverDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: MutationObserverDirective, deps: [{ token: ElementRef }, { token: MUTATION_OBSERVER_INIT }, { token: 'attributeFilter', attribute: true }, { token: 'attributeOldValue', attribute: true }, { token: 'attributes', attribute: true }, { token: 'characterData', attribute: true }, { token: 'characterDataOldValue', attribute: true }, { token: 'childList', attribute: true }, { token: 'subtree', attribute: true }], target: i0.ɵɵFactoryTarget.Directive });
MutationObserverDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: MutationObserverDirective, selector: "[waMutationObserver]", outputs: { waMutationObserver: "waMutationObserver" }, providers: [
        {
            provide: MUTATION_OBSERVER_INIT,
            deps: [ElementRef],
            useFactory: mutationObserverInitFactory,
        },
    ], exportAs: ["MutationObserver"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: MutationObserverDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[waMutationObserver]',
                    exportAs: 'MutationObserver',
                    providers: [
                        {
                            provide: MUTATION_OBSERVER_INIT,
                            deps: [ElementRef],
                            useFactory: mutationObserverInitFactory,
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [MUTATION_OBSERVER_INIT]
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['attributeFilter']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['attributeOldValue']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['attributes']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['characterData']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['characterDataOldValue']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['childList']
                }] }, { type: undefined, decorators: [{
                    type: Attribute,
                    args: ['subtree']
                }] }]; }, propDecorators: { waMutationObserver: [{
                type: Output
            }] } });
