// @taiga-ui/cdk/esm2015/directives/auto-focus/handlers/abstract.handler.js
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class AbstractTuiAutofocusHandler {
    constructor(focusable, el) {
        this.focusable = focusable;
        this.el = el;
    }
    get element() {
        var _a;
        return ((_a = this.focusable) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || this.el.nativeElement;
    }
    get isTextFieldElement() {
        return this.element.matches('input, textarea, [contenteditable]');
    }
}
AbstractTuiAutofocusHandler.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiAutofocusHandler, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiAutofocusHandler.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiAutofocusHandler, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiAutofocusHandler, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }, { type: i0.ElementRef }]; } });
