// @taiga-ui/kit/esm2015/components/input-date-multi/input-date-multi.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MaskitoModule } from '@maskito/angular';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiCalendarModule, TuiHostedDropdownModule, TuiLinkModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldControllerModule, TuiWrapperModule, } from '@taiga-ui/core';
import { TuiInputTagModule } from '@taiga-ui/kit/components/input-tag';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiInputDateMultiComponent } from './input-date-multi.component';
import * as i0 from "@angular/core";
export class TuiInputDateMultiModule {
}
TuiInputDateMultiModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateMultiModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputDateMultiModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateMultiModule, declarations: [TuiInputDateMultiComponent], imports: [CommonModule,
        MaskitoModule,
        PolymorpheusModule,
        TuiWrapperModule,
        TuiCalendarModule,
        TuiSvgModule,
        TuiLinkModule,
        TuiInputTagModule,
        FormsModule,
        TuiMapperPipeModule,
        TuiHostedDropdownModule,
        TuiTextfieldControllerModule,
        TuiPrimitiveTextfieldModule], exports: [TuiInputDateMultiComponent] });
TuiInputDateMultiModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateMultiModule, imports: [[
            CommonModule,
            MaskitoModule,
            PolymorpheusModule,
            TuiWrapperModule,
            TuiCalendarModule,
            TuiSvgModule,
            TuiLinkModule,
            TuiInputTagModule,
            FormsModule,
            TuiMapperPipeModule,
            TuiHostedDropdownModule,
            TuiTextfieldControllerModule,
            TuiPrimitiveTextfieldModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputDateMultiModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MaskitoModule,
                        PolymorpheusModule,
                        TuiWrapperModule,
                        TuiCalendarModule,
                        TuiSvgModule,
                        TuiLinkModule,
                        TuiInputTagModule,
                        FormsModule,
                        TuiMapperPipeModule,
                        TuiHostedDropdownModule,
                        TuiTextfieldControllerModule,
                        TuiPrimitiveTextfieldModule,
                    ],
                    declarations: [TuiInputDateMultiComponent],
                    exports: [TuiInputDateMultiComponent],
                }]
        }] });
