// @taiga-ui/addon-charts/esm2015/components/line-days-chart/line-days-chart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLineChartModule } from '@taiga-ui/addon-charts/components/line-chart';
import { TuiHintModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiLineDaysChartComponent } from './line-days-chart.component';
import { TuiLineDaysChartHintDirective } from './line-days-chart-hint.directive';
import * as i0 from "@angular/core";
export class TuiLineDaysChartModule {
}
TuiLineDaysChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineDaysChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLineDaysChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineDaysChartModule, declarations: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective], imports: [CommonModule, TuiLineChartModule, PolymorpheusModule, TuiHintModule], exports: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective] });
TuiLineDaysChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineDaysChartModule, imports: [[CommonModule, TuiLineChartModule, PolymorpheusModule, TuiHintModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLineDaysChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiLineChartModule, PolymorpheusModule, TuiHintModule],
                    declarations: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective],
                    exports: [TuiLineDaysChartComponent, TuiLineDaysChartHintDirective],
                }]
        }] });
