// @taiga-ui/experimental/esm2015/components/badged-content/badged-content.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiResizeModule } from '@taiga-ui/cdk';
import { TuiBadgedContentComponent } from './badged-content.component';
import { TuiBadgedContentDirective } from './badged-content.directive';
import * as i0 from "@angular/core";
export class TuiBadgedContentModule {
}
TuiBadgedContentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBadgedContentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective], imports: [CommonModule, TuiResizeModule], exports: [TuiBadgedContentComponent, TuiBadgedContentDirective] });
TuiBadgedContentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, imports: [[CommonModule, TuiResizeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBadgedContentModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiResizeModule],
                    declarations: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                    exports: [TuiBadgedContentComponent, TuiBadgedContentDirective],
                }]
        }] });
