// @taiga-ui/core/esm2015/components/scrollbar/scroll-ref.directive.js
import { Directive, ElementRef } from '@angular/core';
import { TUI_SCROLL_REF } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
/** @deprecated import from `@taiga-ui/cdk` instead */
export const SCROLL_REF_SELECTOR = '[tuiScrollRef]';
/** @deprecated import from `@taiga-ui/cdk` instead */
export class TuiScrollRefDirective {
}
TuiScrollRefDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollRefDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TuiScrollRefDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiScrollRefDirective, selector: "[tuiScrollRef]", providers: [
        {
            provide: TUI_SCROLL_REF,
            useExisting: ElementRef,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollRefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: SCROLL_REF_SELECTOR,
                    providers: [
                        {
                            provide: TUI_SCROLL_REF,
                            useExisting: ElementRef,
                        },
                    ],
                }]
        }] });
