// @taiga-ui/addon-table/esm2015/components/table-pagination/table-pagination.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiButtonModule, TuiDataListModule, TuiHostedDropdownModule, TuiLinkModule, TuiSvgModule, } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiTablePaginationComponent } from './table-pagination.component';
import * as i0 from "@angular/core";
export class TuiTablePaginationModule {
}
TuiTablePaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTablePaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTablePaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTablePaginationModule, declarations: [TuiTablePaginationComponent], imports: [CommonModule,
        TuiButtonModule,
        TuiLinkModule,
        TuiHostedDropdownModule,
        TuiDataListModule,
        TuiSvgModule,
        PolymorpheusModule], exports: [TuiTablePaginationComponent] });
TuiTablePaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTablePaginationModule, imports: [[
            CommonModule,
            TuiButtonModule,
            TuiLinkModule,
            TuiHostedDropdownModule,
            TuiDataListModule,
            TuiSvgModule,
            PolymorpheusModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTablePaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiButtonModule,
                        TuiLinkModule,
                        TuiHostedDropdownModule,
                        TuiDataListModule,
                        TuiSvgModule,
                        PolymorpheusModule,
                    ],
                    declarations: [TuiTablePaginationComponent],
                    exports: [TuiTablePaginationComponent],
                }]
        }] });
