// @taiga-ui/addon-charts/esm2015/components/bar/bar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBarComponent } from './bar.component';
import * as i0 from "@angular/core";
export class TuiBarModule {
}
TuiBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, declarations: [TuiBarComponent], imports: [CommonModule], exports: [TuiBarComponent] });
TuiBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBarComponent],
                    exports: [TuiBarComponent],
                }]
        }] });
