// @taiga-ui/addon-preview/esm2015/components/preview-dialog/preview-dialog.module.js
import { NgModule } from '@angular/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPreviewDialogComponent } from './preview-dialog.component';
import * as i0 from "@angular/core";
export class TuiPreviewDialogModule {
}
TuiPreviewDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPreviewDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, declarations: [TuiPreviewDialogComponent], imports: [PolymorpheusModule], exports: [TuiPreviewDialogComponent] });
TuiPreviewDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, imports: [[PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PolymorpheusModule],
                    declarations: [TuiPreviewDialogComponent],
                    exports: [TuiPreviewDialogComponent],
                }]
        }] });
