// @taiga-ui/addon-doc/esm2015/directives/scroll-into-view/scroll-into-view.directive.js
import { Directive, ElementRef, Inject, Input, Self } from '@angular/core';
import { TUI_DOC_PAGE_LOADED } from '@taiga-ui/addon-doc/tokens';
import { TuiDestroyService, tuiGetElementObscures } from '@taiga-ui/cdk';
import { ReplaySubject } from 'rxjs';
import { debounceTime, filter, switchMap, takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
import * as i2 from "rxjs";
export class TuiScrollIntoViewLinkDirective {
    constructor(destroy$, { nativeElement }, readyToScroll$) {
        this.readyToScroll$ = readyToScroll$;
        this.scroll$ = new ReplaySubject(1);
        this.readyToScroll$
            .pipe(filter(Boolean), switchMap(() => this.scroll$), debounceTime(750), filter(shallWe => shallWe && !!tuiGetElementObscures(nativeElement)), takeUntil(destroy$))
            .subscribe(() => {
            nativeElement.scrollIntoView();
        });
    }
    set tuiScrollIntoViewLink(shallWe) {
        this.scroll$.next(shallWe);
    }
}
TuiScrollIntoViewLinkDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollIntoViewLinkDirective, deps: [{ token: TuiDestroyService, self: true }, { token: ElementRef }, { token: TUI_DOC_PAGE_LOADED }], target: i0.ɵɵFactoryTarget.Directive });
TuiScrollIntoViewLinkDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiScrollIntoViewLinkDirective, selector: "[tuiScrollIntoViewLink]", inputs: { tuiScrollIntoViewLink: "tuiScrollIntoViewLink" }, providers: [TuiDestroyService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollIntoViewLinkDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiScrollIntoViewLink]',
                    providers: [TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDestroyService, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_DOC_PAGE_LOADED]
                }] }]; }, propDecorators: { tuiScrollIntoViewLink: [{
                type: Input
            }] } });
