// @taiga-ui/experimental/esm2015/directives/progress-segmented/progress-segmented.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiProgressSegmentedComponent } from './progress-segmented.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiProgressSegmentedDirective {
    constructor(directiveStyles) {
        this.segments = 1;
        directiveStyles.addComponent(TuiProgressSegmentedComponent);
    }
}
TuiProgressSegmentedDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressSegmentedDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiProgressSegmentedDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiProgressSegmentedDirective, selector: "[tuiProgressBar][segments]", inputs: { segments: "segments" }, host: { properties: { "style.--t-segment-width": "1 / segments", "attr.new": "\"\"" }, classAttribute: "_segmented" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiProgressSegmentedDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiProgressBar][segments]',
                    host: {
                        class: '_segmented',
                        '[style.--t-segment-width]': '1 / segments',
                        '[attr.new]': '""', // TODO: drop in v4.0
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { segments: [{
                type: Input
            }] } });
