// @taiga-ui/experimental/esm2015/directives/button-close/button-close.directive.js
import { Directive, Inject } from '@angular/core';
import { tuiButtonOptionsProvider } from '@taiga-ui/experimental/components';
import { TUI_ICON_RESOLVER } from '@taiga-ui/experimental/tokens';
import { TUI_BUTTON_CLOSE_ICON } from './button-close.options';
import * as i0 from "@angular/core";
// TODO: 4.0 revisit after turning tuiIcon to host directive
export class TuiButtonCloseDirective {
    constructor(resolver, icon) {
        this.resolver = resolver;
        this.icon = icon;
    }
}
TuiButtonCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonCloseDirective, deps: [{ token: TUI_ICON_RESOLVER }, { token: TUI_BUTTON_CLOSE_ICON }], target: i0.ɵɵFactoryTarget.Directive });
TuiButtonCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiButtonCloseDirective, selector: "[tuiIconButton][tuiButtonClose]", host: { properties: { "style.--t-mask-left": "\"url(\" + resolver(icon) + \")\"", "style.--t-radius.%": "100", "style.--tui-height-s.rem": "1.875", "class._icon-left": "true" } }, providers: [tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' })], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiButtonCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiIconButton][tuiButtonClose]',
                    providers: [tuiButtonOptionsProvider({ appearance: 'neutral', size: 's' })],
                    host: {
                        '[style.--t-mask-left]': '"url(" + resolver(icon) + ")"',
                        '[style.--t-radius.%]': '100',
                        '[style.--tui-height-s.rem]': '1.875',
                        '[class._icon-left]': 'true',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ICON_RESOLVER]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_BUTTON_CLOSE_ICON]
                }] }]; } });
