// @taiga-ui/addon-doc/esm2015/components/page/page.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TuiReplacePipeModule } from '@taiga-ui/cdk';
import { TuiTabsModule, TuiTagModule } from '@taiga-ui/kit';
import { TuiDocSeeAlsoModule } from '../internal/see-also/see-also.module';
import { TuiDocSourceCodeModule } from '../internal/source-code/source-code.module';
import { TuiDocPageComponent } from './page.component';
import { TuiDocPageTabConnectorDirective } from './page-tab.directive';
import * as i0 from "@angular/core";
export class TuiDocPageModule {
}
TuiDocPageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocPageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDocPageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocPageModule, declarations: [TuiDocPageComponent, TuiDocPageTabConnectorDirective], imports: [CommonModule,
        RouterModule,
        TuiDocSeeAlsoModule,
        TuiTabsModule,
        TuiTagModule,
        TuiDocSourceCodeModule,
        TuiReplacePipeModule], exports: [TuiDocPageComponent, TuiDocPageTabConnectorDirective] });
TuiDocPageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocPageModule, imports: [[
            CommonModule,
            RouterModule,
            TuiDocSeeAlsoModule,
            TuiTabsModule,
            TuiTagModule,
            TuiDocSourceCodeModule,
            TuiReplacePipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDocPageModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        TuiDocSeeAlsoModule,
                        TuiTabsModule,
                        TuiTagModule,
                        TuiDocSourceCodeModule,
                        TuiReplacePipeModule,
                    ],
                    declarations: [TuiDocPageComponent, TuiDocPageTabConnectorDirective],
                    exports: [TuiDocPageComponent, TuiDocPageTabConnectorDirective],
                }]
        }] });
