// @taiga-ui/addon-table/esm2015/components/table/tr/tr.component.js
import { __awaiter } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, forwardRef, Inject, } from '@angular/core';
import { EMPTY_QUERY, tuiQueryListChanges } from '@taiga-ui/cdk';
import { ReplaySubject } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { TuiCellDirective } from '../directives/cell.directive';
import { TuiTableDirective } from '../directives/table.directive';
import { TUI_TABLE_PROVIDER } from '../providers/table.provider';
import { TuiTbodyComponent } from '../tbody/tbody.component';
import * as i0 from "@angular/core";
import * as i1 from "../td/td.component";
import * as i2 from "@angular/common";
import * as i3 from "../directives/table.directive";
import * as i4 from "../tbody/tbody.component";
export class TuiTrComponent {
    constructor(table, body) {
        this.table = table;
        this.body = body;
        this.cells = EMPTY_QUERY;
        this.contentReady$ = new ReplaySubject(1);
        this.cells$ = this.contentReady$.pipe(switchMap(() => tuiQueryListChanges(this.cells)), map(cells => cells.reduce((record, item) => (Object.assign(Object.assign({}, record), { [item.tuiCell]: item })), {})));
        this.item$ = this.contentReady$.pipe(switchMap(() => tuiQueryListChanges(this.body.rows)), map(rows => 
        /**
         * TODO v4.0 replace `this.body.sorted` with `this.body.data` (don't forget to drop `sorted`-getter).
         */
        this.body.sorted[rows.findIndex(row => row === this)]));
    }
    ngAfterContentInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield Promise.resolve();
            this.contentReady$.next(true);
        });
    }
}
TuiTrComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTrComponent, deps: [{ token: forwardRef(() => TuiTableDirective) }, { token: forwardRef(() => TuiTbodyComponent) }], target: i0.ɵɵFactoryTarget.Component });
TuiTrComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTrComponent, selector: "tr[tuiTr]", providers: [TUI_TABLE_PROVIDER], queries: [{ propertyName: "cells", predicate: i0.forwardRef(function () { return TuiCellDirective; }) }], ngImport: i0, template: "<ng-container *ngIf=\"cells$ | async as items\">\n    <ng-container\n        *ngFor=\"let key of table.columns\"\n        [ngTemplateOutlet]=\"items[key]?.template || plain\"\n    >\n        <ng-template #plain>\n            <td\n                *ngIf=\"item$ | async as item\"\n                tuiTd\n            >\n                {{ item[key] }}\n            </td>\n        </ng-template>\n    </ng-container>\n</ng-container>\n", components: [{ type: i1.TuiTdComponent, selector: "th[tuiTd], td[tuiTd]" }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet"] }], pipes: { "async": i2.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTrComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tr[tuiTr]',
                    templateUrl: './tr.template.html',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [TUI_TABLE_PROVIDER],
                }]
        }], ctorParameters: function () { return [{ type: i3.TuiTableDirective, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => TuiTableDirective)]
                }] }, { type: i4.TuiTbodyComponent, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => TuiTbodyComponent)]
                }] }]; }, propDecorators: { cells: [{
                type: ContentChildren,
                args: [forwardRef(() => TuiCellDirective)]
            }] } });
