// @taiga-ui/addon-preview/fesm2015/taiga-ui-addon-preview-components-preview-dialog.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Inject, HostListener, NgModule, Injectable } from '@angular/core';
import { tuiSlideInTop } from '@taiga-ui/core';
import * as i1 from '@tinkoff/ng-polymorpheus';
import { POLYMORPHEUS_CONTEXT, PolymorpheusModule, PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';

class TuiPreviewDialogComponent {
    constructor(context) {
        this.context = context;
    }
    onKeyDownEsc() {
        this.context.$implicit.complete();
    }
}
TuiPreviewDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogComponent, deps: [{ token: POLYMORPHEUS_CONTEXT }], target: i0.ɵɵFactoryTarget.Component });
TuiPreviewDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewDialogComponent, selector: "tui-preview-dialog", host: { listeners: { "document:keydown.esc": "onKeyDownEsc()" } }, ngImport: i0, template: "<div\n    class=\"t-dialog-content\"\n    @tuiSlideInTop\n>\n    <!-- TODO: Polymorpheus fix type -->\n    <ng-container *polymorpheusOutlet=\"$any(context.content) as text; context: context\">\n        {{ text }}\n    </ng-container>\n</div>\n", styles: [".t-dialog-content{position:fixed;width:100%;height:100%}[tuiWrapper][data-appearance=preview-action]{background:rgba(104,104,104,.96);color:var(--tui-text-01-night)}@media (hover: hover){[tuiWrapper][data-appearance=preview-action]:hover:not(._no-hover),[tuiWrapper][data-appearance=preview-action][data-state=hover]{background:rgba(159,159,159,.86)}}[tuiWrapper][data-appearance=preview-action]:active:not(._no-active),[tuiWrapper][data-appearance=preview-action][data-state=active],[tuiWrapper][data-appearance=preview-action][data-state=active]:hover{background:rgba(159,159,159,.75)}\n"], directives: [{ type: i1.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [tuiSlideInTop], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-preview-dialog',
                    templateUrl: './preview-dialog.template.html',
                    styleUrls: ['./preview-dialog.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiSlideInTop],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [POLYMORPHEUS_CONTEXT]
                }] }]; }, propDecorators: { onKeyDownEsc: [{
                type: HostListener,
                args: ['document:keydown.esc']
            }] } });

class TuiPreviewDialogModule {
}
TuiPreviewDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiPreviewDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, declarations: [TuiPreviewDialogComponent], imports: [PolymorpheusModule], exports: [TuiPreviewDialogComponent] });
TuiPreviewDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, imports: [[PolymorpheusModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PolymorpheusModule],
                    declarations: [TuiPreviewDialogComponent],
                    exports: [TuiPreviewDialogComponent],
                }]
        }] });

class TuiPreviewDialogService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.defaultOptions = {};
        this.component = new PolymorpheusComponent(TuiPreviewDialogComponent);
    }
}
TuiPreviewDialogService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogService, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
TuiPreviewDialogService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiPreviewDialogService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPreviewDialogComponent, TuiPreviewDialogModule, TuiPreviewDialogService };
