// @taiga-ui/addon-commerce/esm2015/components/money/pipes/integer-part.pipe.js
import { Inject, Pipe } from '@angular/core';
import { tuiRoundWith } from '@taiga-ui/cdk';
import { TUI_NUMBER_FORMAT, tuiFormatNumber, } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiIntegerPartPipe {
    constructor(numberFormat) {
        this.numberFormat = numberFormat;
    }
    transform(value, precision = 2) {
        const rounded = tuiRoundWith({
            value: Math.abs(Number(value)),
            precision,
            method: this.numberFormat.rounding,
        });
        return tuiFormatNumber(rounded, Object.assign(Object.assign({}, this.numberFormat), { decimalLimit: 0, rounding: 'truncate' }));
    }
}
TuiIntegerPartPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIntegerPartPipe, deps: [{ token: TUI_NUMBER_FORMAT }], target: i0.ɵɵFactoryTarget.Pipe });
TuiIntegerPartPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIntegerPartPipe, name: "tuiIntegerPart" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiIntegerPartPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'tuiIntegerPart' }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_NUMBER_FORMAT]
                }] }]; } });
