// @taiga-ui/addon-commerce/fesm2015/taiga-ui-addon-commerce-components-thumbnail-card.js
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Inject, Input, HostBinding, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/core';
import { TuiSvgModule } from '@taiga-ui/core';
import { TUI_INPUT_CARD_OPTIONS } from '@taiga-ui/addon-commerce/components/input-card';

class TuiThumbnailCardComponent {
    constructor(options) {
        this.options = options;
        this.active = false;
        this.brandLogo = '';
        this.cardNumber = '';
        this.paymentSystem = null;
        this.size = 'm';
    }
    get hasBrandLogo() {
        return !!this.brandLogo && this.size === 'm';
    }
    get paymentSystemLogo() {
        return this.paymentSystem ? this.options.icons[this.paymentSystem] : '';
    }
}
TuiThumbnailCardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardComponent, deps: [{ token: TUI_INPUT_CARD_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiThumbnailCardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiThumbnailCardComponent, selector: "tui-thumbnail-card", inputs: { active: "active", brandLogo: "brandLogo", cardNumber: "cardNumber", paymentSystem: "paymentSystem", size: "size" }, host: { properties: { "class._active": "this.active", "attr.data-size": "this.size" } }, ngImport: i0, template: "<div class=\"t-front\">\n    <tui-svg\n        *ngIf=\"hasBrandLogo\"\n        class=\"t-brand-logo\"\n        [src]=\"brandLogo\"\n    ></tui-svg>\n    <span class=\"t-number\">{{ cardNumber }}</span>\n    <tui-svg\n        *ngIf=\"!!paymentSystem\"\n        class=\"t-payment-system-logo\"\n        [src]=\"paymentSystemLogo\"\n    ></tui-svg>\n</div>\n<div class=\"t-back\"></div>\n", styles: [":host{position:relative;display:block;flex-shrink:0;color:var(--tui-base-01);transform-style:preserve-3d;cursor:default;border-radius:var(--tui-radius-xs);background-size:100%}:host[data-size=s]{width:2rem;height:1.5rem}:host[data-size=m]{width:3rem;height:2rem}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\";box-shadow:inset 0 0 0 1px;border-radius:inherit;color:var(--tui-clear)}:host._active{box-shadow:0 0 0 1px var(--tui-base-01)}.t-front{position:absolute;top:0;left:0;width:100%;height:100%;-webkit-backface-visibility:hidden;backface-visibility:hidden}.t-brand-logo{position:absolute;top:.1875rem;left:.125rem;height:.875rem;width:.875rem}.t-number{position:absolute;right:.25rem;display:flex;justify-content:center;align-items:center}:host[data-size=s] .t-number{top:.125rem;height:.625rem;width:1.1875rem;font-size:.5rem}:host[data-size=m] .t-number{top:.1875rem;height:.875rem;width:1.5rem;font-size:.625rem}.t-payment-system-logo{position:absolute;right:.25rem;bottom:-.5rem;width:2rem;height:2rem;transform:scale(.5);transform-origin:right}.t-back{position:absolute;top:0;left:0;width:100%;height:100%;-webkit-backface-visibility:hidden;backface-visibility:hidden;transform:rotateY(180deg) translateZ(1px);background-color:var(--tui-base-05);border-radius:var(--tui-radius-xs)}.t-back:after{content:\"\";position:absolute;top:20%;left:0;right:0;height:20%;background-color:var(--tui-base-06)}\n"], components: [{ type: i1.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-thumbnail-card',
                    templateUrl: './thumbnail-card.template.html',
                    styleUrls: ['./thumbnail-card.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_INPUT_CARD_OPTIONS]
                }] }]; }, propDecorators: { active: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['class._active']
            }], brandLogo: [{
                type: Input
            }], cardNumber: [{
                type: Input
            }], paymentSystem: [{
                type: Input
            }], size: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-size']
            }] } });

/**
 * @deprecated use {@link TuiThumbnailCardComponent}
 */
class TuiCardComponent extends TuiThumbnailCardComponent {
}
TuiCardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TuiCardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiCardComponent, selector: "tui-card", usesInheritance: true, ngImport: i0, template: "<div class=\"t-front\">\n    <tui-svg\n        *ngIf=\"hasBrandLogo\"\n        class=\"t-brand-logo\"\n        [src]=\"brandLogo\"\n    ></tui-svg>\n    <span class=\"t-number\">{{ cardNumber }}</span>\n    <tui-svg\n        *ngIf=\"!!paymentSystem\"\n        class=\"t-payment-system-logo\"\n        [src]=\"paymentSystemLogo\"\n    ></tui-svg>\n</div>\n<div class=\"t-back\"></div>\n", styles: [":host{position:relative;display:block;flex-shrink:0;color:var(--tui-base-01);transform-style:preserve-3d;cursor:default;border-radius:var(--tui-radius-xs);background-size:100%}:host[data-size=s]{width:2rem;height:1.5rem}:host[data-size=m]{width:3rem;height:2rem}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\";box-shadow:inset 0 0 0 1px;border-radius:inherit;color:var(--tui-clear)}:host._active{box-shadow:0 0 0 1px var(--tui-base-01)}.t-front{position:absolute;top:0;left:0;width:100%;height:100%;-webkit-backface-visibility:hidden;backface-visibility:hidden}.t-brand-logo{position:absolute;top:.1875rem;left:.125rem;height:.875rem;width:.875rem}.t-number{position:absolute;right:.25rem;display:flex;justify-content:center;align-items:center}:host[data-size=s] .t-number{top:.125rem;height:.625rem;width:1.1875rem;font-size:.5rem}:host[data-size=m] .t-number{top:.1875rem;height:.875rem;width:1.5rem;font-size:.625rem}.t-payment-system-logo{position:absolute;right:.25rem;bottom:-.5rem;width:2rem;height:2rem;transform:scale(.5);transform-origin:right}.t-back{position:absolute;top:0;left:0;width:100%;height:100%;-webkit-backface-visibility:hidden;backface-visibility:hidden;transform:rotateY(180deg) translateZ(1px);background-color:var(--tui-base-05);border-radius:var(--tui-radius-xs)}.t-back:after{content:\"\";position:absolute;top:20%;left:0;right:0;height:20%;background-color:var(--tui-base-06)}\n"], components: [{ type: i1.TuiSvgComponent, selector: "tui-svg", inputs: ["src"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-card',
                    templateUrl: './thumbnail-card.template.html',
                    styleUrls: ['./thumbnail-card.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
/**
 * @deprecated use {@link TuiThumbnailCardModule}
 */
class TuiCardModule {
}
TuiCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, declarations: [TuiCardComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiCardComponent] });
TuiCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiCardComponent],
                    exports: [TuiCardComponent],
                }]
        }] });

class TuiThumbnailCardModule {
}
TuiThumbnailCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiThumbnailCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, declarations: [TuiThumbnailCardComponent], imports: [CommonModule, TuiSvgModule], exports: [TuiThumbnailCardComponent] });
TuiThumbnailCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, imports: [[CommonModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule],
                    declarations: [TuiThumbnailCardComponent],
                    exports: [TuiThumbnailCardComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiCardComponent, TuiCardModule, TuiThumbnailCardComponent, TuiThumbnailCardModule };
