// @taiga-ui/kit/esm2015/components/input-phone/input-phone.module.js
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiTextfieldComponent, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiValueAccessorModule } from '@taiga-ui/kit/directives';
import { TuiInputPhoneComponent } from './input-phone.component';
import { TuiInputPhoneDirective } from './input-phone.directive';
import * as i0 from "@angular/core";
export class TuiInputPhoneModule {
}
TuiInputPhoneModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputPhoneModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneModule, declarations: [TuiInputPhoneComponent, TuiInputPhoneDirective], imports: [MaskitoModule,
        TuiPrimitiveTextfieldModule,
        TuiHostedDropdownModule,
        TuiTextfieldControllerModule,
        TuiActiveZoneModule,
        TuiValueAccessorModule], exports: [TuiInputPhoneComponent, TuiInputPhoneDirective, TuiTextfieldComponent] });
TuiInputPhoneModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneModule, imports: [[
            MaskitoModule,
            TuiPrimitiveTextfieldModule,
            TuiHostedDropdownModule,
            TuiTextfieldControllerModule,
            TuiActiveZoneModule,
            TuiValueAccessorModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputPhoneModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoModule,
                        TuiPrimitiveTextfieldModule,
                        TuiHostedDropdownModule,
                        TuiTextfieldControllerModule,
                        TuiActiveZoneModule,
                        TuiValueAccessorModule,
                    ],
                    declarations: [TuiInputPhoneComponent, TuiInputPhoneDirective],
                    exports: [TuiInputPhoneComponent, TuiInputPhoneDirective, TuiTextfieldComponent],
                }]
        }] });
