// ngx-highlightjs/esm2020/plus/highlight-plus.module.mjs
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { HighlightModule } from 'ngx-highlightjs';
// Uncomment the following line for development
// import { HighlightModule } from '../../src/public-api';
import { GistFilePipe, GistDirective } from './gist';
import { CodeFromUrlPipe } from './code-from-url';
import * as i0 from "@angular/core";
export class HighlightPlusModule {
}
HighlightPlusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: HighlightPlusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HighlightPlusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: HighlightPlusModule, declarations: [GistDirective,
        GistFilePipe,
        CodeFromUrlPipe], imports: [HighlightModule,
        HttpClientModule], exports: [HighlightModule,
        GistDirective,
        GistFilePipe,
        CodeFromUrlPipe] });
HighlightPlusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: HighlightPlusModule, imports: [[
            HighlightModule,
            HttpClientModule
        ], HighlightModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: HighlightPlusModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        HighlightModule,
                        HttpClientModule
                    ],
                    declarations: [
                        GistDirective,
                        GistFilePipe,
                        CodeFromUrlPipe
                    ],
                    exports: [
                        HighlightModule,
                        GistDirective,
                        GistFilePipe,
                        CodeFromUrlPipe
                    ]
                }]
        }] });
