// @taiga-ui/experimental/esm2015/directives/comment/comment.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiCommentComponent } from './comment.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiCommentDirective {
    constructor(directiveStyles) {
        this.tuiComment = 'top';
        directiveStyles.addComponent(TuiCommentComponent);
    }
}
TuiCommentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiCommentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCommentDirective, selector: "[tuiComment]", inputs: { tuiComment: "tuiComment" }, host: { properties: { "attr.data-direction": "tuiComment" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCommentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiComment]',
                    host: {
                        '[attr.data-direction]': 'tuiComment',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiComment: [{
                type: Input
            }] } });
