// @taiga-ui/addon-doc/esm2015/components/language-switcher/language-switcher.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { TuiDataListModule, TuiFlagPipeModule } from '@taiga-ui/core';
import { TuiSelectModule } from '@taiga-ui/kit';
import { TuiLanguageSwitcherComponent } from './language-switcher.component';
import * as i0 from "@angular/core";
export class TuiLanguageSwitcherModule {
}
TuiLanguageSwitcherModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLanguageSwitcherModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiLanguageSwitcherModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLanguageSwitcherModule, declarations: [TuiLanguageSwitcherComponent], imports: [CommonModule,
        TuiFlagPipeModule,
        TuiSelectModule,
        TuiDataListModule,
        ReactiveFormsModule], exports: [TuiLanguageSwitcherComponent] });
TuiLanguageSwitcherModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLanguageSwitcherModule, imports: [[
            CommonModule,
            TuiFlagPipeModule,
            TuiSelectModule,
            TuiDataListModule,
            ReactiveFormsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiLanguageSwitcherModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiFlagPipeModule,
                        TuiSelectModule,
                        TuiDataListModule,
                        ReactiveFormsModule,
                    ],
                    declarations: [TuiLanguageSwitcherComponent],
                    exports: [TuiLanguageSwitcherComponent],
                }]
        }] });
