// @taiga-ui/addon-mobile/esm2015/components/tab-bar/tab-bar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiRepeatTimesModule } from '@taiga-ui/cdk';
import { TuiSvgModule } from '@taiga-ui/core';
import { TuiTabBarComponent } from './tab-bar.component';
import { TuiTabBarItemComponent } from './tab-bar-item.component';
import { TuiTabBarItemDirective } from './tab-bar-item.directive';
import * as i0 from "@angular/core";
export class TuiTabBarModule {
}
TuiTabBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTabBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabBarModule, declarations: [TuiTabBarComponent, TuiTabBarItemComponent, TuiTabBarItemDirective], imports: [CommonModule, TuiSvgModule, TuiRepeatTimesModule], exports: [TuiTabBarComponent, TuiTabBarItemComponent, TuiTabBarItemDirective] });
TuiTabBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabBarModule, imports: [[CommonModule, TuiSvgModule, TuiRepeatTimesModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTabBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiSvgModule, TuiRepeatTimesModule],
                    declarations: [TuiTabBarComponent, TuiTabBarItemComponent, TuiTabBarItemDirective],
                    exports: [TuiTabBarComponent, TuiTabBarItemComponent, TuiTabBarItemDirective],
                }]
        }] });
