// @taiga-ui/addon-commerce/esm2015/components/input-card/abstract-input-card.js
import { __decorate } from "tslib";
import { ChangeDetectorRef, Directive, EventEmitter, Inject, Input, Optional, Output, Self, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiNullableControl, tuiPure, } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
export class AbstractTuiInputCard extends AbstractTuiNullableControl {
    constructor(control, cdr, options) {
        super(control, cdr);
        this.options = options;
        this.autocompleteEnabled = this.options.autocompleteEnabled;
        this.binChange = new EventEmitter();
    }
    get defaultIcon() {
        const paymentSystem = this.getPaymentSystem(this.card);
        return paymentSystem && this.options.icons[paymentSystem];
    }
    get paymentSystem() {
        return this.getPaymentSystem(this.card);
    }
    /** @deprecated remove in 4.0 */
    get icon() {
        return this.cardSrc || this.defaultIcon;
    }
    get autocomplete() {
        return this.autocompleteEnabled ? 'cc-number' : 'off';
    }
    /**
     * @deprecated: drop in v4.0
     * use {@link autocomplete}
     */
    get autocompleteCard() {
        return this.autocomplete;
    }
    get bin() {
        return this.card.length < 6 ? null : this.card.slice(0, 6);
    }
    getPaymentSystem(cardNumber) {
        return this.options.paymentSystemHandler(cardNumber);
    }
}
AbstractTuiInputCard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiInputCard, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiInputCard.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiInputCard, inputs: { cardSrc: "cardSrc", autocompleteEnabled: "autocompleteEnabled" }, outputs: { binChange: "binChange" }, usesInheritance: true, ngImport: i0 });
__decorate([
    tuiPure
], AbstractTuiInputCard.prototype, "getPaymentSystem", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiInputCard, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgControl, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }, {
                    type: Inject,
                    args: [NgControl]
                }] }, { type: i0.ChangeDetectorRef, decorators: [{
                    type: Inject,
                    args: [ChangeDetectorRef]
                }] }, { type: undefined }]; }, propDecorators: { cardSrc: [{
                type: Input
            }], autocompleteEnabled: [{
                type: Input
            }], binChange: [{
                type: Output
            }], getPaymentSystem: [] } });
