// @taiga-ui/addon-commerce/esm2015/components/input-card/input-card.module.js
import { NgModule } from '@angular/core';
import { MaskitoModule } from '@maskito/angular';
import { TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiValueAccessorModule } from '@taiga-ui/kit';
import { TuiInputCardComponent } from './input-card.component';
import { TuiInputCardDirective } from './input-card.directive';
import * as i0 from "@angular/core";
export class TuiInputCardModule {
}
TuiInputCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardModule, declarations: [TuiInputCardComponent, TuiInputCardDirective], imports: [MaskitoModule,
        TuiPrimitiveTextfieldModule,
        TuiTextfieldControllerModule,
        TuiValueAccessorModule,
        TuiSvgModule], exports: [TuiInputCardComponent, TuiInputCardDirective, TuiTextfieldComponent] });
TuiInputCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardModule, imports: [[
            MaskitoModule,
            TuiPrimitiveTextfieldModule,
            TuiTextfieldControllerModule,
            TuiValueAccessorModule,
            TuiSvgModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        MaskitoModule,
                        TuiPrimitiveTextfieldModule,
                        TuiTextfieldControllerModule,
                        TuiValueAccessorModule,
                        TuiSvgModule,
                    ],
                    declarations: [TuiInputCardComponent, TuiInputCardDirective],
                    exports: [TuiInputCardComponent, TuiInputCardDirective, TuiTextfieldComponent],
                }]
        }] });
