// @taiga-ui/addon-commerce/esm2015/components/input-card/input-card.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiTextfieldHost, tuiAsTextfieldHost } from '@taiga-ui/core';
import * as i0 from "@angular/core";
export class TuiInputCardDirective extends AbstractTuiTextfieldHost {
    onValueChange(value) {
        this.host.onValueChange(value);
    }
    ngAfterViewInit() {
        const { nativeFocusableElement } = this.host;
        if (!nativeFocusableElement) {
            return;
        }
        nativeFocusableElement.inputMode = 'numeric';
        nativeFocusableElement.placeholder =
            nativeFocusableElement.placeholder || '0000 0000 0000 0000';
    }
}
TuiInputCardDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiInputCardDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiInputCardDirective, selector: "tui-input-card", providers: [tuiAsTextfieldHost(TuiInputCardDirective)], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputCardDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'tui-input-card',
                    providers: [tuiAsTextfieldHost(TuiInputCardDirective)],
                }]
        }] });
