// @taiga-ui/addon-table/esm2015/components/table/table.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiSvgModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiCellDirective } from './directives/cell.directive';
import { TuiDirectionOrderDirective } from './directives/direction-order.directive';
import { TuiHeadDirective } from './directives/head.directive';
import { TuiResizedDirective } from './directives/resized.directive';
import { TuiRowDirective } from './directives/row.directive';
import { TuiSortByDirective } from './directives/sort-by.directive';
import { TuiSortableDirective } from './directives/sortable.directive';
import { TuiTableDirective } from './directives/table.directive';
import { TuiTheadDirective } from './directives/thead.directive';
import { TuiTableSortPipe } from './pipes/table-sort.pipe';
import { TuiTbodyComponent } from './tbody/tbody.component';
import { TuiTdComponent } from './td/td.component';
import { TuiThComponent } from './th/th.component';
import { TuiThGroupComponent } from './th-group/th-group.component';
import { TuiTrComponent } from './tr/tr.component';
import * as i0 from "@angular/core";
export class TuiTableModule {
}
TuiTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableModule, declarations: [TuiTableDirective,
        TuiTbodyComponent,
        TuiThGroupComponent,
        TuiThComponent,
        TuiTdComponent,
        TuiTrComponent,
        TuiCellDirective,
        TuiHeadDirective,
        TuiRowDirective,
        TuiSortByDirective,
        TuiSortableDirective,
        TuiTheadDirective,
        TuiResizedDirective,
        TuiTableSortPipe,
        TuiDirectionOrderDirective], imports: [CommonModule, PolymorpheusModule, TuiMapperPipeModule, TuiSvgModule], exports: [TuiTableDirective,
        TuiTbodyComponent,
        TuiThGroupComponent,
        TuiThComponent,
        TuiTdComponent,
        TuiTrComponent,
        TuiCellDirective,
        TuiHeadDirective,
        TuiRowDirective,
        TuiSortByDirective,
        TuiSortableDirective,
        TuiTheadDirective,
        TuiTableSortPipe,
        TuiDirectionOrderDirective] });
TuiTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableModule, imports: [[CommonModule, PolymorpheusModule, TuiMapperPipeModule, TuiSvgModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PolymorpheusModule, TuiMapperPipeModule, TuiSvgModule],
                    declarations: [
                        TuiTableDirective,
                        TuiTbodyComponent,
                        TuiThGroupComponent,
                        TuiThComponent,
                        TuiTdComponent,
                        TuiTrComponent,
                        TuiCellDirective,
                        TuiHeadDirective,
                        TuiRowDirective,
                        TuiSortByDirective,
                        TuiSortableDirective,
                        TuiTheadDirective,
                        TuiResizedDirective,
                        TuiTableSortPipe,
                        TuiDirectionOrderDirective,
                    ],
                    exports: [
                        TuiTableDirective,
                        TuiTbodyComponent,
                        TuiThGroupComponent,
                        TuiThComponent,
                        TuiTdComponent,
                        TuiTrComponent,
                        TuiCellDirective,
                        TuiHeadDirective,
                        TuiRowDirective,
                        TuiSortByDirective,
                        TuiSortableDirective,
                        TuiTheadDirective,
                        TuiTableSortPipe,
                        TuiDirectionOrderDirective,
                    ],
                }]
        }] });
