// @taiga-ui/cdk/esm2015/components/scroll-controls/scroll-controls.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiScrollControlsComponent } from './scroll-controls.component';
import { TuiScrollRefDirective } from './scroll-ref.directive';
import { TuiScrollbarDirective } from './scrollbar.directive';
import * as i0 from "@angular/core";
export class TuiScrollControlsModule {
}
TuiScrollControlsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiScrollControlsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, declarations: [TuiScrollbarDirective,
        TuiScrollControlsComponent,
        TuiScrollRefDirective], imports: [CommonModule], exports: [TuiScrollControlsComponent, TuiScrollRefDirective] });
TuiScrollControlsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiScrollControlsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        TuiScrollbarDirective,
                        TuiScrollControlsComponent,
                        TuiScrollRefDirective,
                    ],
                    exports: [TuiScrollControlsComponent, TuiScrollRefDirective],
                }]
        }] });
