// @taiga-ui/cdk/esm2015/services/static-request.service.js
import { __awaiter } from "tslib";
import { isPlatformServer } from '@angular/common';
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { defer, from } from 'rxjs';
import { fromFetch } from 'rxjs/fetch';
import { shareReplay, switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiStaticRequestService {
    constructor(win, platformId) {
        this.win = win;
        this.platformId = platformId;
        this.cache = new Map();
    }
    request(url) {
        const cache = this.cache.get(url);
        if (cache) {
            return cache;
        }
        const response$ = 'AbortController' in this.win || isPlatformServer(this.platformId)
            ? fromFetch(url)
            : /**
               * Fallback for Firefox 55 and 56
               * TODO: drop after browser support bump
               */
                defer(() => from(fetch(url)));
        const piped = response$.pipe(switchMap((response) => __awaiter(this, void 0, void 0, function* () {
            if (response.ok) {
                return response.text();
            }
            throw new Error(`Failed to load ${url} (${response.statusText})`);
        })), shareReplay({ bufferSize: 1, refCount: false }));
        this.cache.set(url, piped);
        return piped;
    }
}
TuiStaticRequestService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStaticRequestService, deps: [{ token: WINDOW }, { token: PLATFORM_ID }], target: i0.ɵɵFactoryTarget.Injectable });
TuiStaticRequestService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStaticRequestService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStaticRequestService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: Window, decorators: [{
                    type: Inject,
                    args: [WINDOW]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }]; } });
