// @taiga-ui/kit/esm2015/components/input-month/input-month.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TuiMapperPipeModule, TuiPreventDefaultModule } from '@taiga-ui/cdk';
import { TuiHostedDropdownModule, TuiPrimitiveTextfieldModule, TuiSvgModule, TuiTextfieldComponent, TuiTextfieldControllerModule, } from '@taiga-ui/core';
import { TuiCalendarMonthModule } from '@taiga-ui/kit/components/calendar-month';
import { TuiInputMonthComponent } from './input-month.component';
import { TuiInputMonthDirective } from './input-month.directive';
import * as i0 from "@angular/core";
export class TuiInputMonthModule {
}
TuiInputMonthModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiInputMonthModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthModule, declarations: [TuiInputMonthComponent, TuiInputMonthDirective], imports: [CommonModule,
        TuiCalendarMonthModule,
        TuiHostedDropdownModule,
        TuiPrimitiveTextfieldModule,
        TuiSvgModule,
        TuiPreventDefaultModule,
        TuiMapperPipeModule,
        TuiTextfieldControllerModule,
        FormsModule], exports: [TuiInputMonthComponent, TuiInputMonthDirective, TuiTextfieldComponent] });
TuiInputMonthModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthModule, imports: [[
            CommonModule,
            TuiCalendarMonthModule,
            TuiHostedDropdownModule,
            TuiPrimitiveTextfieldModule,
            TuiSvgModule,
            TuiPreventDefaultModule,
            TuiMapperPipeModule,
            TuiTextfieldControllerModule,
            FormsModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiInputMonthModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiCalendarMonthModule,
                        TuiHostedDropdownModule,
                        TuiPrimitiveTextfieldModule,
                        TuiSvgModule,
                        TuiPreventDefaultModule,
                        TuiMapperPipeModule,
                        TuiTextfieldControllerModule,
                        FormsModule,
                    ],
                    declarations: [TuiInputMonthComponent, TuiInputMonthDirective],
                    exports: [TuiInputMonthComponent, TuiInputMonthDirective, TuiTextfieldComponent],
                }]
        }] });
