// @taiga-ui/kit/esm2015/components/tiles/tile-handle.directive.js
import { __decorate } from "tslib";
import { Directive, HostListener, Inject } from '@angular/core';
import { tuiGetActualTarget, tuiIsElement } from '@taiga-ui/cdk';
import { shouldCall } from '@tinkoff/ng-event-plugins';
import { TuiTileComponent } from './tile.component';
import * as i0 from "@angular/core";
import * as i1 from "./tile.component";
function isInteracting(x = NaN) {
    return !Number.isNaN(x) || !Number.isNaN(this['x']);
}
function isDragging() {
    return !Number.isNaN(this['x']);
}
export class TuiTileHandleDirective {
    constructor(tile) {
        this.tile = tile;
        this.x = NaN;
        this.y = NaN;
    }
    onStart(event) {
        const target = tuiGetActualTarget(event);
        const { x, y, pointerId } = event;
        if (tuiIsElement(target)) {
            target.releasePointerCapture(pointerId);
        }
        this.onPointer(x, y);
    }
    onPointer(x = NaN, y = NaN) {
        const { left, top } = this.tile.element.getBoundingClientRect();
        this.x = x - left;
        this.y = y - top;
        this.tile.onDrag([NaN, NaN]);
    }
    onMove(x, y) {
        this.tile.onDrag([x - this.x, y - this.y]);
    }
}
TuiTileHandleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileHandleDirective, deps: [{ token: TuiTileComponent }], target: i0.ɵɵFactoryTarget.Directive });
TuiTileHandleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiTileHandleDirective, selector: "[tuiTileHandle]", host: { listeners: { "pointerdown.silent": "onStart($event)", "document:pointerup.silent": "onPointer()", "document:pointermove.silent": "onMove($event.x,$event.y)" }, properties: { "style.touchAction": "\"none\"", "style.userSelect": "\"none\"" } }, ngImport: i0 });
__decorate([
    shouldCall(isInteracting)
], TuiTileHandleDirective.prototype, "onPointer", null);
__decorate([
    shouldCall(isDragging)
], TuiTileHandleDirective.prototype, "onMove", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileHandleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiTileHandle]',
                    host: {
                        '[style.touchAction]': '"none"',
                        '[style.userSelect]': '"none"',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiTileComponent, decorators: [{
                    type: Inject,
                    args: [TuiTileComponent]
                }] }]; }, propDecorators: { onStart: [{
                type: HostListener,
                args: ['pointerdown.silent', ['$event']]
            }], onPointer: [{
                type: HostListener,
                args: ['document:pointerup.silent']
            }], onMove: [{
                type: HostListener,
                args: ['document:pointermove.silent', ['$event.x', '$event.y']]
            }] } });
