// @taiga-ui/kit/esm2015/components/tiles/tile.service.js
import { ElementRef, Inject, Injectable } from '@angular/core';
import { MutationObserverService } from '@ng-web-apis/mutation-observer';
import { tuiArrayShallowEquals, tuiPx, TuiResizeService } from '@taiga-ui/cdk';
import { BehaviorSubject, combineLatest, Subscription } from 'rxjs';
import { debounceTime, distinctUntilChanged, map, startWith } from 'rxjs/operators';
import { TuiTilesComponent } from './tiles.component';
import * as i0 from "@angular/core";
import * as i1 from "./tiles.component";
import * as i2 from "rxjs";
export class TuiTileService {
    constructor(el, tiles, resize$, mutation$) {
        this.el = el;
        this.tiles = tiles;
        this.resize$ = resize$;
        this.mutation$ = mutation$;
        this.sub = new Subscription();
        this.offset$ = new BehaviorSubject([NaN, NaN]);
        this.position$ = combineLatest([
            this.offset$.pipe(distinctUntilChanged(tuiArrayShallowEquals)),
            this.resize$.pipe(startWith(null)),
            this.mutation$.pipe(startWith(null)),
            this.tiles.order$.pipe(debounceTime(0)),
        ]).pipe(map(([offset]) => offset));
    }
    init(element) {
        this.sub.add(this.position$.subscribe(offset => {
            this.setPosition(element, offset);
            this.setRect(element, offset);
        }));
    }
    setOffset(offset) {
        this.offset$.next(offset);
    }
    ngOnDestroy() {
        this.sub.unsubscribe();
    }
    getRect([left, top]) {
        return {
            top: Number.isNaN(top) ? this.el.nativeElement.offsetTop : top,
            left: Number.isNaN(left) ? this.el.nativeElement.offsetLeft : left,
            width: this.el.nativeElement.clientWidth,
            height: this.el.nativeElement.clientHeight,
            right: NaN,
            bottom: NaN,
        };
    }
    setRect({ style }, offset) {
        const { top, left, width, height } = this.getRect(offset);
        style.top = tuiPx(top);
        style.left = tuiPx(left);
        style.width = tuiPx(width);
        style.height = tuiPx(height);
    }
    setPosition(element, [left]) {
        if (!Number.isNaN(left)) {
            element.style.setProperty('position', 'fixed');
            return;
        }
        const { style } = element;
        const rect = element.getBoundingClientRect();
        const host = this.el.nativeElement.getBoundingClientRect();
        style.removeProperty('position');
        style.top = tuiPx(rect.top - host.top + this.el.nativeElement.offsetTop);
        style.left = tuiPx(rect.left - host.left + this.el.nativeElement.offsetLeft);
    }
}
TuiTileService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileService, deps: [{ token: ElementRef }, { token: TuiTilesComponent }, { token: TuiResizeService }, { token: MutationObserverService }], target: i0.ɵɵFactoryTarget.Injectable });
TuiTileService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i1.TuiTilesComponent, decorators: [{
                    type: Inject,
                    args: [TuiTilesComponent]
                }] }, { type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [TuiResizeService]
                }] }, { type: i2.Observable, decorators: [{
                    type: Inject,
                    args: [MutationObserverService]
                }] }]; } });
