// @taiga-ui/kit/esm2015/components/tiles/tile.component.js
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, HostListener, Inject, Input, ViewChild, } from '@angular/core';
import { TuiTileService } from './tile.service';
import { TuiTilesComponent } from './tiles.component';
import * as i0 from "@angular/core";
import * as i1 from "./tile.service";
import * as i2 from "./tiles.component";
export class TuiTileComponent {
    constructor(service, el, tiles) {
        this.service = service;
        this.el = el;
        this.tiles = tiles;
        this.width = 1;
        this.height = 1;
        this.dragged = false;
    }
    get column() {
        return `span var(--tui-width, ${this.width})`;
    }
    get row() {
        return `span var(--tui-height, ${this.height})`;
    }
    get element() {
        return this.el.nativeElement;
    }
    onEnter() {
        this.tiles.rearrange(this.element);
    }
    onDrag(offset) {
        const dragged = !Number.isNaN(offset[0]);
        this.dragged = this.dragged || dragged;
        this.tiles.element = dragged ? this.element : null;
        this.service.setOffset(offset);
    }
    onTransitionEnd() {
        this.dragged = false;
    }
    ngAfterViewInit() {
        if (this.wrapper) {
            this.service.init(this.wrapper.nativeElement);
        }
    }
    ngOnDestroy() {
        if (this.tiles.element === this.element) {
            this.tiles.element = null;
        }
    }
}
TuiTileComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileComponent, deps: [{ token: TuiTileService }, { token: ElementRef }, { token: TuiTilesComponent }], target: i0.ɵɵFactoryTarget.Component });
TuiTileComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiTileComponent, selector: "tui-tile", inputs: { width: "width", height: "height" }, host: { listeners: { "pointerenter": "onEnter()" }, properties: { "class._dragged": "this.dragged", "style.gridColumn": "this.column", "style.gridRow": "this.row" } }, providers: [TuiTileService], viewQueries: [{ propertyName: "wrapper", first: true, predicate: ["wrapper"], descendants: true }], ngImport: i0, template: "<div\n    #wrapper\n    class=\"t-wrapper\"\n    (transitionend.self)=\"onTransitionEnd()\"\n>\n    <ng-content></ng-content>\n</div>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiTileComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-tile',
                    templateUrl: './tile.template.html',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [TuiTileService],
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiTileService, decorators: [{
                    type: Inject,
                    args: [TuiTileService]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i2.TuiTilesComponent, decorators: [{
                    type: Inject,
                    args: [TuiTilesComponent]
                }] }]; }, propDecorators: { wrapper: [{
                type: ViewChild,
                args: ['wrapper']
            }], width: [{
                type: Input
            }], height: [{
                type: Input
            }], dragged: [{
                type: HostBinding,
                args: ['class._dragged']
            }], column: [{
                type: HostBinding,
                args: ['style.gridColumn']
            }], row: [{
                type: HostBinding,
                args: ['style.gridRow']
            }], onEnter: [{
                type: HostListener,
                args: ['pointerenter']
            }] } });
