// @taiga-ui/addon-preview/bundles/taiga-ui-addon-preview-tokens.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@taiga-ui/cdk'), require('@taiga-ui/i18n')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-preview/tokens', ['exports', '@taiga-ui/cdk', '@taiga-ui/i18n'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-preview"] = global["taiga-ui"]["addon-preview"] || {}, global["taiga-ui"]["addon-preview"].tokens = {}), global.cdk, global.i18n));
})(this, (function (exports, cdk, i18n) { 'use strict';

    /**
     * tui-preview i18n
     */
    var TUI_PREVIEW_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('previewTexts'));
    /**
     * tui-preview zoom i18n
     */
    var TUI_PREVIEW_ZOOM_TEXTS = cdk.tuiCreateTokenFromFactory(i18n.tuiExtractI18n('zoomTexts'));

    var TUI_PREVIEW_ICONS_DEFAULT = {
        rotate: 'tuiIconRotate',
        prev: 'tuiIconArrowLeft',
        next: 'tuiIconArrowRight',
        zoomIn: 'tuiIconPlus',
        zoomOut: 'tuiIconMinus',
        zoomReset: 'tuiIconMinimize',
    };
    var TUI_PREVIEW_ICONS = cdk.tuiCreateToken(TUI_PREVIEW_ICONS_DEFAULT);
    function tuiPreviewIconsProvider(icons) {
        return cdk.tuiProvideOptions(TUI_PREVIEW_ICONS, icons, TUI_PREVIEW_ICONS_DEFAULT);
    }

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TUI_PREVIEW_ICONS = TUI_PREVIEW_ICONS;
    exports.TUI_PREVIEW_ICONS_DEFAULT = TUI_PREVIEW_ICONS_DEFAULT;
    exports.TUI_PREVIEW_TEXTS = TUI_PREVIEW_TEXTS;
    exports.TUI_PREVIEW_ZOOM_TEXTS = TUI_PREVIEW_ZOOM_TEXTS;
    exports.tuiPreviewIconsProvider = tuiPreviewIconsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
