// @taiga-ui/addon-preview/bundles/taiga-ui-addon-preview-components-preview.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/addon-preview/tokens'), require('@taiga-ui/cdk'), require('@taiga-ui/kit'), require('@taiga-ui/core'), require('@angular/common'), require('rxjs'), require('rxjs/operators'), require('@angular/forms'), require('@ng-web-apis/mutation-observer'), require('@taiga-ui/addon-preview/components/preview-dialog'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-preview/components/preview', ['exports', '@angular/core', '@taiga-ui/addon-preview/tokens', '@taiga-ui/cdk', '@taiga-ui/kit', '@taiga-ui/core', '@angular/common', 'rxjs', 'rxjs/operators', '@angular/forms', '@ng-web-apis/mutation-observer', '@taiga-ui/addon-preview/components/preview-dialog', '@tinkoff/ng-polymorpheus'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-preview"] = global["taiga-ui"]["addon-preview"] || {}, global["taiga-ui"]["addon-preview"].components = global["taiga-ui"]["addon-preview"].components || {}, global["taiga-ui"]["addon-preview"].components.preview = {}), global.ng.core, global["taiga-ui"]["addon-preview"].tokens, global.i4, global.i2, global.i1, global.ng.common, global.rxjs, global.rxjs.operators, global.ng.forms, global.i5$1, global["taiga-ui"]["addon-preview"].components["preview-dialog"], global.ngPolymorpheus));
})(this, (function (exports, i0, tokens, i4, i2, i1, i3, i4$1, operators, i5, i5$1, previewDialog, ngPolymorpheus) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i4__namespace$1 = /*#__PURE__*/_interopNamespace(i4);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4$1);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i5__namespace$1 = /*#__PURE__*/_interopNamespace(i5$1);

    var TuiPreviewActionDirective = /** @class */ (function () {
        function TuiPreviewActionDirective() {
        }
        return TuiPreviewActionDirective;
    }());
    TuiPreviewActionDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    TuiPreviewActionDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewActionDirective, selector: "[tuiPreviewAction]", providers: [
            i1.tuiButtonOptionsProvider({
                appearance: 'preview-action',
                shape: 'rounded',
                size: 's',
            }),
        ], ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionDirective, decorators: [{
                type: i0.Directive,
                args: [{
                        selector: '[tuiPreviewAction]',
                        providers: [
                            i1.tuiButtonOptionsProvider({
                                appearance: 'preview-action',
                                shape: 'rounded',
                                size: 's',
                            }),
                        ],
                    }]
            }] });

    var TuiPreviewPaginationComponent = /** @class */ (function () {
        function TuiPreviewPaginationComponent(icons, texts$) {
            this.icons = icons;
            this.texts$ = texts$;
            this.length = 1;
            this.index = 0;
            this.indexChange = new i0.EventEmitter();
        }
        Object.defineProperty(TuiPreviewPaginationComponent.prototype, "leftButtonDisabled", {
            get: function () {
                return this.index === 0;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPreviewPaginationComponent.prototype, "rightButtonDisabled", {
            get: function () {
                return this.index === this.length - 1;
            },
            enumerable: false,
            configurable: true
        });
        TuiPreviewPaginationComponent.prototype.onArrowClick = function (step) {
            this.updateIndex(i4.tuiClamp(this.index + step, 0, this.length - 1));
        };
        TuiPreviewPaginationComponent.prototype.updateIndex = function (index) {
            if (this.index === index) {
                return;
            }
            this.index = index;
            this.indexChange.emit(index);
        };
        return TuiPreviewPaginationComponent;
    }());
    TuiPreviewPaginationComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewPaginationComponent, deps: [{ token: tokens.TUI_PREVIEW_ICONS }, { token: i2.TUI_PAGINATION_TEXTS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPreviewPaginationComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewPaginationComponent, selector: "tui-preview-pagination", inputs: { length: "length", index: "index" }, outputs: { indexChange: "indexChange" }, host: { listeners: { "document:keydown.arrowRight.prevent": "onArrowClick(1)", "document:keydown.arrowLeft.prevent": "onArrowClick(-1)" } }, ngImport: i0__namespace, template: "<ng-container *ngIf=\"texts$ | async as texts\">\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        type=\"button\"\n        class=\"t-arrow_left\"\n        [disabled]=\"leftButtonDisabled\"\n        [icon]=\"icons.prev\"\n        [title]=\"texts[0]\"\n        (click)=\"onArrowClick(-1)\"\n    ></button>\n    {{ index + 1 }}/{{ length }}\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        type=\"button\"\n        class=\"t-arrow_right\"\n        [disabled]=\"rightButtonDisabled\"\n        [icon]=\"icons.next\"\n        [title]=\"texts[1]\"\n        (click)=\"onArrowClick(1)\"\n    ></button>\n</ng-container>\n", styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:6.25rem}.t-arrow_left{border-top-right-radius:0;border-bottom-right-radius:0}.t-arrow_right{border-top-left-radius:0;border-bottom-left-radius:0}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: TuiPreviewActionDirective, selector: "[tuiPreviewAction]" }], pipes: { "async": i3__namespace.AsyncPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewPaginationComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-preview-pagination',
                        templateUrl: './preview-pagination.template.html',
                        styleUrls: ['./preview-pagination.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PREVIEW_ICONS]
                        }] }, { type: i4__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [i2.TUI_PAGINATION_TEXTS]
                        }] }];
        }, propDecorators: { length: [{
                    type: i0.Input
                }], index: [{
                    type: i0.Input
                }], indexChange: [{
                    type: i0.Output
                }], onArrowClick: [{
                    type: i0.HostListener,
                    args: ['document:keydown.arrowRight.prevent', ['1']]
                }, {
                    type: i0.HostListener,
                    args: ['document:keydown.arrowLeft.prevent', ['-1']]
                }] } });

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var STEP = 0.5;
    var TuiPreviewZoomComponent = /** @class */ (function () {
        function TuiPreviewZoomComponent(icons, zoomTexts$) {
            this.icons = icons;
            this.zoomTexts$ = zoomTexts$;
            this.min = 0.5;
            this.max = 2;
            this.value = 1;
            this.valueChange = new i0.EventEmitter();
            this.reset = new i0.EventEmitter();
            this.hintShow$ = this.valueChange.pipe(operators.switchMap(function () { return i4$1.merge(i4$1.of(true), i4$1.timer(1000).pipe(operators.map(i4.ALWAYS_FALSE_HANDLER))); }), operators.startWith(false));
        }
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "leftButtonDisabled", {
            get: function () {
                return this.value === this.min;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "rightButtonDisabled", {
            get: function () {
                return this.value === this.max;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "collapseVisible", {
            get: function () {
                return this.value > this.min;
            },
            enumerable: false,
            configurable: true
        });
        TuiPreviewZoomComponent.prototype.onModelChange = function (value) {
            var clamped = i4.tuiClamp(value, this.min, this.max);
            if (clamped === this.value) {
                return;
            }
            this.value = clamped;
            this.valueChange.emit(clamped);
        };
        TuiPreviewZoomComponent.prototype.onReset = function () {
            this.reset.emit();
        };
        TuiPreviewZoomComponent.prototype.onMinus = function () {
            this.onModelChange(this.value - STEP);
        };
        TuiPreviewZoomComponent.prototype.onPlus = function () {
            this.onModelChange(this.value < 1 ? 1 : this.value + STEP);
        };
        return TuiPreviewZoomComponent;
    }());
    TuiPreviewZoomComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewZoomComponent, deps: [{ token: tokens.TUI_PREVIEW_ICONS }, { token: tokens.TUI_PREVIEW_ZOOM_TEXTS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPreviewZoomComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewZoomComponent, selector: "tui-preview-zoom", inputs: { min: "min", max: "max", value: "value" }, outputs: { valueChange: "valueChange", reset: "reset" }, ngImport: i0__namespace, template: "<ng-container *ngIf=\"zoomTexts$ | async as texts\">\n    <section class=\"t-zoom\">\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            type=\"button\"\n            class=\"t-sign_minus\"\n            [disabled]=\"leftButtonDisabled\"\n            [icon]=\"icons.zoomOut\"\n            [title]=\"texts.zoomOut\"\n            (click)=\"onMinus()\"\n        ></button>\n        <label tuiSliderThumbLabel>\n            <div\n                tuiHintAppearance=\"onDark\"\n                tuiHintDirection=\"top-right\"\n                [tuiHint]=\"hint\"\n                [tuiHintManual]=\"!!(hintShow$ | async)\"\n            ></div>\n\n            <ng-template #hint>\n                {{ (valueChange | async) || value | percent }}\n            </ng-template>\n\n            <input\n                step=\"any\"\n                tuiSlider\n                type=\"range\"\n                class=\"t-slider\"\n                [max]=\"max\"\n                [min]=\"min\"\n                [ngModel]=\"value\"\n                (ngModelChange)=\"onModelChange($event)\"\n            />\n        </label>\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            type=\"button\"\n            class=\"t-sign_plus\"\n            [disabled]=\"rightButtonDisabled\"\n            [icon]=\"icons.zoomIn\"\n            [title]=\"texts.zoomIn\"\n            (click)=\"onPlus()\"\n        ></button>\n    </section>\n\n    <button\n        tuiHintAppearance=\"onDark\"\n        tuiHintDescribe\n        tuiHintDirection=\"top-right\"\n        tuiIconButton\n        tuiPreviewAction\n        type=\"button\"\n        class=\"t-reset-button\"\n        [class.t-invisible]=\"!collapseVisible\"\n        [icon]=\"icons.zoomReset\"\n        [tuiHint]=\"texts.reset\"\n        (click)=\"onReset()\"\n    ></button>\n</ng-container>\n", styles: [":host{position:relative;display:flex}.t-zoom{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:12rem}.t-slider{width:7.5rem}.t-sign_minus{border-top-right-radius:0;border-bottom-right-radius:0}.t-sign_plus{border-top-left-radius:0;border-bottom-left-radius:0}.t-invisible{visibility:hidden}.t-reset-button{margin-left:.3125rem}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }, { type: i2__namespace.TuiSliderThumbLabelComponent, selector: "[tuiSliderThumbLabel]" }, { type: i2__namespace.TuiSliderComponent, selector: "input[type=range][tuiSlider]", inputs: ["size", "segments"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: TuiPreviewActionDirective, selector: "[tuiPreviewAction]" }, { type: i1__namespace.TuiHintDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHint", "tuiHintContext", "tuiHintAppearance"] }, { type: i1__namespace.TuiHintDriverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)" }, { type: i1__namespace.TuiHintHoverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHintShowDelay", "tuiHintHideDelay"], exportAs: ["tuiHintHover"] }, { type: i1__namespace.TuiHintPositionDirective, selector: "[tuiHint]:not([tuiHintCustomPosition]):not(ng-container):not(ng-template)", inputs: ["tuiHintDirection"] }, { type: i1__namespace.TuiHintManualDirective, selector: "[tuiHint][tuiHintManual]", inputs: ["tuiHintManual"] }, { type: i5__namespace.RangeValueAccessor, selector: "input[type=range][formControlName],input[type=range][formControl],input[type=range][ngModel]" }, { type: i5__namespace.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i5__namespace.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i5__namespace.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i1__namespace.TuiHintDescribeDirective, selector: "[tuiHintDescribe]", inputs: ["tuiHintDescribe"] }], pipes: { "async": i3__namespace.AsyncPipe, "percent": i3__namespace.PercentPipe }, changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewZoomComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-preview-zoom',
                        templateUrl: './preview-zoom.template.html',
                        styleUrls: ['./preview-zoom.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PREVIEW_ICONS]
                        }] }, { type: i4__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PREVIEW_ZOOM_TEXTS]
                        }] }];
        }, propDecorators: { min: [{
                    type: i0.Input
                }], max: [{
                    type: i0.Input
                }], value: [{
                    type: i0.Input
                }], valueChange: [{
                    type: i0.Output
                }], reset: [{
                    type: i0.Output
                }] } });

    var INITIAL_SCALE_COEF = 0.8;
    var EMPTY_COORDINATES = [0, 0];
    var ROTATION_ANGLE = 90;
    var TuiPreviewComponent = /** @class */ (function () {
        function TuiPreviewComponent(el, icons, texts$) {
            this.el = el;
            this.icons = icons;
            this.texts$ = texts$;
            this.zoomable = true;
            this.rotatable = false;
            this.minZoom = 1;
            this.width = 0;
            this.height = 0;
            this.zoom$ = new i4$1.BehaviorSubject(this.minZoom);
            this.rotation$ = new i4$1.BehaviorSubject(0);
            this.coordinates$ = new i4$1.BehaviorSubject(EMPTY_COORDINATES);
            this.transitioned$ = i4$1.merge(i4.tuiDragAndDropFrom(this.el.nativeElement).pipe(operators.map(function (_a) {
                var stage = _a.stage;
                return stage !== i4.TuiDragStage.Continues;
            })), i4.tuiTypedFromEvent(this.el.nativeElement, 'touchmove', {
                passive: true,
            }).pipe(operators.map(i4.ALWAYS_FALSE_HANDLER)), i4.tuiTypedFromEvent(this.el.nativeElement, 'wheel', { passive: true }).pipe(operators.map(i4.ALWAYS_FALSE_HANDLER)));
            this.cursor$ = i4.tuiDragAndDropFrom(this.el.nativeElement).pipe(operators.map(function (_a) {
                var stage = _a.stage;
                return (stage === i4.TuiDragStage.Continues ? 'grabbing' : 'initial');
            }), operators.startWith('initial'));
            this.wrapperTransform$ = i4$1.combineLatest([
                this.coordinates$.pipe(operators.map(function (_a) {
                    var _b = __read(_a, 2), x = _b[0], y = _b[1];
                    return i4.tuiPx(x) + ", " + i4.tuiPx(y);
                })),
                this.zoom$,
                this.rotation$,
            ]).pipe(operators.map(function (_a) {
                var _b = __read(_a, 3), translate = _b[0], zoom = _b[1], rotation = _b[2];
                return "translate(" + translate + ") scale(" + zoom + ") rotate(" + rotation + "deg)";
            }));
        }
        TuiPreviewComponent.prototype.rotate = function () {
            this.rotation$.next(this.rotation$.value - ROTATION_ANGLE);
        };
        TuiPreviewComponent.prototype.onPan = function (delta) {
            this.coordinates$.next(this.getGuardedCoordinates(this.coordinates$.value[0] + delta[0], this.coordinates$.value[1] + delta[1]));
        };
        TuiPreviewComponent.prototype.onMutation = function (contentWrapper) {
            var clientWidth = contentWrapper.clientWidth, clientHeight = contentWrapper.clientHeight;
            this.refresh(clientWidth, clientHeight);
        };
        TuiPreviewComponent.prototype.onZoom = function (_a) {
            var clientX = _a.clientX, clientY = _a.clientY, delta = _a.delta;
            if (this.zoomable) {
                this.processZoom(clientX, clientY, delta);
            }
        };
        TuiPreviewComponent.prototype.onResize = function (contentResizeEntries) {
            if (contentResizeEntries.length === 0) {
                return;
            }
            var _a = contentResizeEntries[0].contentRect, width = _a.width, height = _a.height;
            this.refresh(width, height);
        };
        TuiPreviewComponent.prototype.reset = function () {
            this.zoom$.next(this.minZoom);
            this.coordinates$.next(EMPTY_COORDINATES);
        };
        TuiPreviewComponent.prototype.setZoom = function (zoom) {
            this.zoom$.next(zoom);
            var _a = __read(this.coordinates$.value, 2), x = _a[0], y = _a[1];
            this.coordinates$.next(this.getGuardedCoordinates(x, y));
        };
        Object.defineProperty(TuiPreviewComponent.prototype, "offsets", {
            get: function () {
                var offsetX = ((this.zoom$.value - this.minZoom) * this.width) / 2;
                var offsetY = ((this.zoom$.value - this.minZoom) * this.height) / 2;
                return { offsetX: offsetX, offsetY: offsetY };
            },
            enumerable: false,
            configurable: true
        });
        TuiPreviewComponent.prototype.calculateMinZoom = function (contentHeight, contentWidth, boxHeight, boxWidth) {
            var bigSize = contentHeight > boxHeight * INITIAL_SCALE_COEF ||
                contentWidth > boxWidth * INITIAL_SCALE_COEF;
            var _a = this.el.nativeElement, clientHeight = _a.clientHeight, clientWidth = _a.clientWidth;
            return bigSize
                ? i4.tuiRound(Math.min((clientHeight * INITIAL_SCALE_COEF) / contentHeight, (clientWidth * INITIAL_SCALE_COEF) / contentWidth), 2)
                : 1;
        };
        TuiPreviewComponent.prototype.refresh = function (width, height) {
            this.width = width;
            this.height = height;
            this.minZoom = this.calculateMinZoom(height, width, this.el.nativeElement.clientHeight, this.el.nativeElement.clientWidth);
            this.zoom$.next(this.minZoom);
            this.coordinates$.next(EMPTY_COORDINATES);
            this.rotation$.next(0);
        };
        TuiPreviewComponent.prototype.processZoom = function (clientX, clientY, delta) {
            var oldScale = this.zoom$.value;
            var newScale = i4.tuiClamp(oldScale + delta, this.minZoom, 2);
            var center = this.getScaleCenter({ clientX: clientX, clientY: clientY }, this.coordinates$.value, this.zoom$.value);
            var moveX = center[0] * oldScale - center[0] * newScale;
            var moveY = center[1] * oldScale - center[1] * newScale;
            this.zoom$.next(newScale);
            this.coordinates$.next(this.getGuardedCoordinates(this.coordinates$.value[0] + moveX, this.coordinates$.value[1] + moveY));
        };
        TuiPreviewComponent.prototype.getGuardedCoordinates = function (x, y) {
            var _a = this.offsets, offsetX = _a.offsetX, offsetY = _a.offsetY;
            return [i4.tuiClamp(x, -offsetX, offsetX), i4.tuiClamp(y, -offsetY, offsetY)];
        };
        TuiPreviewComponent.prototype.getScaleCenter = function (_a, _b, scale) {
            var clientX = _a.clientX, clientY = _a.clientY;
            var _c = __read(_b, 2), x = _c[0], y = _c[1];
            return [
                (clientX - x - this.el.nativeElement.offsetWidth / 2) / scale,
                (clientY - y - this.el.nativeElement.offsetHeight / 2) / scale,
            ];
        };
        return TuiPreviewComponent;
    }());
    TuiPreviewComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewComponent, deps: [{ token: i0.ElementRef }, { token: tokens.TUI_PREVIEW_ICONS }, { token: tokens.TUI_PREVIEW_TEXTS }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPreviewComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewComponent, selector: "tui-preview", inputs: { zoomable: "zoomable", rotatable: "rotatable" }, ngImport: i0__namespace, template: "<ng-container *ngIf=\"texts$ | async as texts\">\n    <section\n        #contentWrapper\n        attributeFilter=\"src\"\n        characterData\n        childList\n        subtree\n        class=\"t-wrapper\"\n        [class.t-not-interactive-content]=\"zoomable\"\n        [class.t-transitive]=\"transitioned$ | async\"\n        [style.cursor]=\"cursor$ | async\"\n        [style.transform]=\"wrapperTransform$ | async\"\n        (tuiPan)=\"onPan($event)\"\n        (tuiResize)=\"onResize($event)\"\n        (tuiZoom)=\"onZoom($event)\"\n        (waMutationObserver)=\"onMutation(contentWrapper)\"\n    >\n        <ng-content></ng-content>\n    </section>\n\n    <header class=\"t-header\">\n        <div class=\"t-title\">\n            <ng-content select=\"tui-preview-title\"></ng-content>\n        </div>\n\n        <ng-content select=\"tui-preview-pagination\"></ng-content>\n\n        <div class=\"t-actions\">\n            <ng-content select=\"[tuiPreviewAction]\"></ng-content>\n        </div>\n    </header>\n\n    <footer class=\"t-footer\">\n        <button\n            *ngIf=\"rotatable\"\n            tuiHintAppearance=\"onDark\"\n            tuiHintDescribe\n            tuiHintDirection=\"top-right\"\n            tuiIconButton\n            tuiPreviewAction\n            type=\"button\"\n            class=\"t-rotate-button\"\n            [icon]=\"icons.rotate\"\n            [tuiHint]=\"texts.rotate\"\n            (click)=\"rotate()\"\n        ></button>\n\n        <tui-preview-zoom\n            *ngIf=\"zoomable\"\n            [min]=\"minZoom\"\n            [value]=\"(zoom$ | async) || 1\"\n            (reset)=\"reset()\"\n            (valueChange)=\"setZoom($event)\"\n        ></tui-preview-zoom>\n    </footer>\n</ng-container>\n", styles: [":host{position:relative;display:flex;justify-content:center;align-items:center;width:100%;height:100%;-webkit-user-select:none;-moz-user-select:none;user-select:none}.t-header{position:fixed;top:1rem;display:flex;width:100%;padding:0 1rem;box-sizing:border-box}.t-footer{position:absolute;bottom:1rem;display:flex;width:100%;padding:0 1rem;box-sizing:border-box;justify-content:center}.t-actions{display:flex;flex:1;justify-content:flex-end}.t-actions ::ng-deep>*{margin-left:.625rem}.t-rotate-button{margin-right:.3125rem}.t-title{flex:1}:host-context(tui-root._mobile) .t-title{display:none}.t-not-interactive-content ::ng-deep>*{pointer-events:none}.t-wrapper{will-change:transform}.t-transitive{transition-duration:.3s}\n"], components: [{ type: i1__namespace.TuiButtonComponent, selector: "button[tuiButton], button[tuiIconButton], a[tuiButton], a[tuiIconButton]", inputs: ["appearance", "disabled", "icon", "iconRight", "shape", "showLoader", "size"] }, { type: TuiPreviewZoomComponent, selector: "tui-preview-zoom", inputs: ["min", "max", "value"], outputs: ["valueChange", "reset"] }], directives: [{ type: i3__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4__namespace$1.TuiPanDirective, selector: "[tuiPan]", outputs: ["tuiPan"] }, { type: i4__namespace$1.TuiResizeDirective, selector: "[tuiResize]", outputs: ["tuiResize"] }, { type: i4__namespace$1.TuiZoomDirective, selector: "[tuiZoom]", outputs: ["tuiZoom"] }, { type: i5__namespace$1.MutationObserverDirective, selector: "[waMutationObserver]", outputs: ["waMutationObserver"], exportAs: ["MutationObserver"] }, { type: i1__namespace.TuiHintDescribeDirective, selector: "[tuiHintDescribe]", inputs: ["tuiHintDescribe"] }, { type: TuiPreviewActionDirective, selector: "[tuiPreviewAction]" }, { type: i1__namespace.TuiHintDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHint", "tuiHintContext", "tuiHintAppearance"] }, { type: i1__namespace.TuiHintDriverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)" }, { type: i1__namespace.TuiHintHoverDirective, selector: "[tuiHint]:not(ng-container):not(ng-template)", inputs: ["tuiHintShowDelay", "tuiHintHideDelay"], exportAs: ["tuiHintHover"] }, { type: i1__namespace.TuiHintPositionDirective, selector: "[tuiHint]:not([tuiHintCustomPosition]):not(ng-container):not(ng-template)", inputs: ["tuiHintDirection"] }], pipes: { "async": i3__namespace.AsyncPipe }, animations: [i1.tuiSlideInTop], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-preview',
                        templateUrl: './preview.template.html',
                        styleUrls: ['./preview.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        animations: [i1.tuiSlideInTop],
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PREVIEW_ICONS]
                        }] }, { type: i4__namespace.Observable, decorators: [{
                            type: i0.Inject,
                            args: [tokens.TUI_PREVIEW_TEXTS]
                        }] }];
        }, propDecorators: { zoomable: [{
                    type: i0.Input
                }], rotatable: [{
                    type: i0.Input
                }] } });

    var TuiPreviewActionModule = /** @class */ (function () {
        function TuiPreviewActionModule() {
        }
        return TuiPreviewActionModule;
    }());
    TuiPreviewActionModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPreviewActionModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionModule, declarations: [TuiPreviewActionDirective], exports: [TuiPreviewActionDirective] });
    TuiPreviewActionModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewActionModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [TuiPreviewActionDirective],
                        exports: [TuiPreviewActionDirective],
                    }]
            }] });

    var TuiPreviewTitleComponent = /** @class */ (function () {
        function TuiPreviewTitleComponent() {
        }
        return TuiPreviewTitleComponent;
    }());
    TuiPreviewTitleComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewTitleComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiPreviewTitleComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiPreviewTitleComponent, selector: "tui-preview-title", ngImport: i0__namespace, template: "\n        <ng-content></ng-content>\n    ", isInline: true, styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:inline-block;padding:.375rem .75rem}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewTitleComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-preview-title',
                        template: "\n        <ng-content></ng-content>\n    ",
                        styleUrls: ['./preview-title.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }] });

    var TuiPreviewModule = /** @class */ (function () {
        function TuiPreviewModule() {
        }
        return TuiPreviewModule;
    }());
    TuiPreviewModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiPreviewModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewModule, declarations: [TuiPreviewComponent,
            TuiPreviewTitleComponent,
            TuiPreviewPaginationComponent,
            TuiPreviewZoomComponent], imports: [i3.CommonModule,
            ngPolymorpheus.PolymorpheusModule,
            i5$1.MutationObserverModule,
            i1.TuiButtonModule,
            i5.FormsModule,
            i2.TuiSliderModule,
            i1.TuiHintModule,
            i4.TuiDragModule,
            i4.TuiResizeModule,
            i1.TuiSvgModule,
            i1.TuiLoaderModule,
            i4.TuiPanModule,
            i4.TuiZoomModule,
            TuiPreviewActionModule], exports: [TuiPreviewComponent,
            TuiPreviewActionDirective,
            TuiPreviewTitleComponent,
            TuiPreviewPaginationComponent,
            TuiPreviewZoomComponent] });
    TuiPreviewModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewModule, providers: [i4.tuiAsDialog(previewDialog.TuiPreviewDialogService)], imports: [[
                i3.CommonModule,
                ngPolymorpheus.PolymorpheusModule,
                i5$1.MutationObserverModule,
                i1.TuiButtonModule,
                i5.FormsModule,
                i2.TuiSliderModule,
                i1.TuiHintModule,
                i4.TuiDragModule,
                i4.TuiResizeModule,
                i1.TuiSvgModule,
                i1.TuiLoaderModule,
                i4.TuiPanModule,
                i4.TuiZoomModule,
                TuiPreviewActionModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiPreviewModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            i3.CommonModule,
                            ngPolymorpheus.PolymorpheusModule,
                            i5$1.MutationObserverModule,
                            i1.TuiButtonModule,
                            i5.FormsModule,
                            i2.TuiSliderModule,
                            i1.TuiHintModule,
                            i4.TuiDragModule,
                            i4.TuiResizeModule,
                            i1.TuiSvgModule,
                            i1.TuiLoaderModule,
                            i4.TuiPanModule,
                            i4.TuiZoomModule,
                            TuiPreviewActionModule,
                        ],
                        declarations: [
                            TuiPreviewComponent,
                            TuiPreviewTitleComponent,
                            TuiPreviewPaginationComponent,
                            TuiPreviewZoomComponent,
                        ],
                        providers: [i4.tuiAsDialog(previewDialog.TuiPreviewDialogService)],
                        exports: [
                            TuiPreviewComponent,
                            TuiPreviewActionDirective,
                            TuiPreviewTitleComponent,
                            TuiPreviewPaginationComponent,
                            TuiPreviewZoomComponent,
                        ],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiPreviewActionDirective = TuiPreviewActionDirective;
    exports.TuiPreviewActionModule = TuiPreviewActionModule;
    exports.TuiPreviewComponent = TuiPreviewComponent;
    exports.TuiPreviewModule = TuiPreviewModule;
    exports.TuiPreviewPaginationComponent = TuiPreviewPaginationComponent;
    exports.TuiPreviewTitleComponent = TuiPreviewTitleComponent;
    exports.TuiPreviewZoomComponent = TuiPreviewZoomComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
