// @taiga-ui/addon-charts/esm2015/components/bar-chart/bar-chart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBarSetModule } from '@taiga-ui/addon-charts/components/bar-set';
import { TuiFocusableModule, TuiMapperPipeModule } from '@taiga-ui/cdk';
import { TuiHintModule } from '@taiga-ui/core';
import { TuiBarChartComponent } from './bar-chart.component';
import * as i0 from "@angular/core";
export class TuiBarChartModule {
}
TuiBarChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBarChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarChartModule, declarations: [TuiBarChartComponent], imports: [CommonModule,
        TuiBarSetModule,
        TuiFocusableModule,
        TuiHintModule,
        TuiMapperPipeModule], exports: [TuiBarChartComponent] });
TuiBarChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarChartModule, imports: [[
            CommonModule,
            TuiBarSetModule,
            TuiFocusableModule,
            TuiHintModule,
            TuiMapperPipeModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBarChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        TuiBarSetModule,
                        TuiFocusableModule,
                        TuiHintModule,
                        TuiMapperPipeModule,
                    ],
                    declarations: [TuiBarChartComponent],
                    exports: [TuiBarChartComponent],
                }]
        }] });
